{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "44830610-d070-4882-8273-da4c5c22dd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import bs4\n",
    "import json\n",
    "import re\n",
    "import tqdm\n",
    "\n",
    "import scipy.stats as st\n",
    "import numpy as np\n",
    "\n",
    "import bokeh\n",
    "from bokeh.plotting import show as show_interactive\n",
    "from bokeh.plotting import output_file, output_notebook\n",
    "from bokeh.layouts import column, row\n",
    "from bokeh.models import CustomJS, TextInput, LassoSelectTool, Select, MultiSelect, ColorBar, Legend, LegendItem\n",
    "from bokeh.models.widgets import DataTable, DateFormatter, TableColumn, Button, HTMLTemplateFormatter\n",
    "from bokeh.events import SelectionGeometry\n",
    "from bokeh.transform import linear_cmap, jitter\n",
    "from bokeh.models import FactorRange\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "5ca5c7a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annotation = pd.read_csv('../eggnog/complete_eggnog_annotation.csv')\n",
    "go_df = pd.read_csv('./go_annotations.csv')\n",
    "kegg_df = pd.read_csv('./kegg_annotations.csv')\n",
    "ec_df = pd.read_csv('./ec_annotations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "d996023a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_01528530</td>\n",
       "      <td>5911.EAS00195</td>\n",
       "      <td>2.2e-14</td>\n",
       "      <td>49.9</td>\n",
       "      <td>KOG0594@1|root,KOG0594@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>G</td>\n",
       "      <td>cyclin-dependent protein serine/threonine kina...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_01528510</td>\n",
       "      <td>5911.EAR81750</td>\n",
       "      <td>1.1e-194</td>\n",
       "      <td>633.4</td>\n",
       "      <td>2EI75@1|root,2SNPE@2759|Eukaryota,3ZBUW@5878|C...</td>\n",
       "      <td>5878|Ciliophora</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_01528500</td>\n",
       "      <td>5911.EAR81749</td>\n",
       "      <td>4e-141</td>\n",
       "      <td>462.0</td>\n",
       "      <td>COG1100@1|root,KOG0074@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>KLT</td>\n",
       "      <td>GTP binding</td>\n",
       "      <td>ARL13B</td>\n",
       "      <td>GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko04031</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ADP-ribosylation factor family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_0015284992</td>\n",
       "      <td>5911.EAR97791</td>\n",
       "      <td>8.4e-64</td>\n",
       "      <td>205.1</td>\n",
       "      <td>2ERUB@1|root,2SUIJ@2759|Eukaryota</td>\n",
       "      <td>5911.EAR97791|-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Leucine-rich repeat-containing protein 74A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00897120</td>\n",
       "      <td>5911.EAR86071</td>\n",
       "      <td>5.2e-77</td>\n",
       "      <td>248.8</td>\n",
       "      <td>2ERUB@1|root,2SUIJ@2759|Eukaryota</td>\n",
       "      <td>5911.EAR86071|-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>kinase domain protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TTHERM_ID  seed_ortholog    evalue  score  \\\n",
       "0    TTHERM_01528530  5911.EAS00195   2.2e-14   49.9   \n",
       "1    TTHERM_01528510  5911.EAR81750  1.1e-194  633.4   \n",
       "2    TTHERM_01528500  5911.EAR81749    4e-141  462.0   \n",
       "3  TTHERM_0015284992  5911.EAR97791   8.4e-64  205.1   \n",
       "4    TTHERM_00897120  5911.EAR86071   5.2e-77  248.8   \n",
       "\n",
       "                                          eggNOG_OGs    max_annot_lvl  \\\n",
       "0              KOG0594@1|root,KOG0594@2759|Eukaryota   2759|Eukaryota   \n",
       "1  2EI75@1|root,2SNPE@2759|Eukaryota,3ZBUW@5878|C...  5878|Ciliophora   \n",
       "2              COG1100@1|root,KOG0074@2759|Eukaryota   2759|Eukaryota   \n",
       "3                  2ERUB@1|root,2SUIJ@2759|Eukaryota  5911.EAR97791|-   \n",
       "4                  2ERUB@1|root,2SUIJ@2759|Eukaryota  5911.EAR86071|-   \n",
       "\n",
       "  COG_category                                        Description  \\\n",
       "0            G  cyclin-dependent protein serine/threonine kina...   \n",
       "1            -                                                  -   \n",
       "2          KLT                                        GTP binding   \n",
       "3            -                                                  -   \n",
       "4            -                                                  -   \n",
       "\n",
       "  Preferred_name                                                GOs  ...  \\\n",
       "0              -                                                  -  ...   \n",
       "1              -                                                  -  ...   \n",
       "2         ARL13B  GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...  ...   \n",
       "3              -                                                  -  ...   \n",
       "4              -                                                  -  ...   \n",
       "\n",
       "  KEGG_Pathway KEGG_Module KEGG_Reaction KEGG_rclass            BRITE KEGG_TC  \\\n",
       "0            -           -             -           -                -       -   \n",
       "1            -           -             -           -                -       -   \n",
       "2            -           -             -           -  ko00000,ko04031       -   \n",
       "3            -           -             -           -                -       -   \n",
       "4            -           -             -           -                -       -   \n",
       "\n",
       "  CAZy BiGG_Reaction PFAMs                         TGD2021_description  \n",
       "0    -             -     -                              protein kinase  \n",
       "1    -             -     -                        hypothetical protein  \n",
       "2    -             -     -      ADP-ribosylation factor family protein  \n",
       "3    -             -     -  Leucine-rich repeat-containing protein 74A  \n",
       "4    -             -     -                       kinase domain protein  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "9bdcf8af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26258, 22)"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annotation.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "96ba81e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_cols = ['TTHERM_ID', 'GOs', 'KEGG_ko', 'EC']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "4fb1da66",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annotated = complete_annotation.loc[\n",
    "    (complete_annotation['GOs'] != '-') |\n",
    "    (complete_annotation['KEGG_ko'] != '-') |\n",
    "    (complete_annotation['EC'] != '-')\n",
    "].loc[:, annotation_cols]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "36ba3cba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>GOs</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>EC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_01528500</td>\n",
       "      <td>GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...</td>\n",
       "      <td>ko:K07962</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TTHERM_00898250</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K12833</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TTHERM_00898300</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K05643</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TTHERM_01074550</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K01373</td>\n",
       "      <td>3.4.22.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TTHERM_01071490</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K08794,ko:K13412</td>\n",
       "      <td>2.7.11.1,2.7.11.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17269</th>\n",
       "      <td>TTHERM_01076950</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K09391</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17270</th>\n",
       "      <td>TTHERM_01076960</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K12869</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17271</th>\n",
       "      <td>TTHERM_01076970</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K04958,ko:K04960</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17272</th>\n",
       "      <td>TTHERM_01076980</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K06631</td>\n",
       "      <td>2.7.11.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17273</th>\n",
       "      <td>TTHERM_01077000</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K14713</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7210 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID                                                GOs  \\\n",
       "2      TTHERM_01528500  GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...   \n",
       "13     TTHERM_00898250                                                  -   \n",
       "17     TTHERM_00898300                                                  -   \n",
       "22     TTHERM_01074550                                                  -   \n",
       "25     TTHERM_01071490                                                  -   \n",
       "...                ...                                                ...   \n",
       "17269  TTHERM_01076950                                                  -   \n",
       "17270  TTHERM_01076960                                                  -   \n",
       "17271  TTHERM_01076970                                                  -   \n",
       "17272  TTHERM_01076980                                                  -   \n",
       "17273  TTHERM_01077000                                                  -   \n",
       "\n",
       "                   KEGG_ko                  EC  \n",
       "2                ko:K07962                   -  \n",
       "13               ko:K12833                   -  \n",
       "17               ko:K05643                   -  \n",
       "22               ko:K01373           3.4.22.41  \n",
       "25     ko:K08794,ko:K13412  2.7.11.1,2.7.11.17  \n",
       "...                    ...                 ...  \n",
       "17269            ko:K09391                   -  \n",
       "17270            ko:K12869                   -  \n",
       "17271  ko:K04958,ko:K04960                   -  \n",
       "17272            ko:K06631           2.7.11.21  \n",
       "17273            ko:K14713                   -  \n",
       "\n",
       "[7210 rows x 4 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "6755d52a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.2745829842333765"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "7210/26258"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "09c0d482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20326, 48)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/allgood_filt_agg_tidy_2021aligned_qc_rma_expression_full.csv')\n",
    "full_filtered_df = full_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "full_filtered_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "30dae79e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['TTHERM_000000042', 'TTHERM_000000045', 'TTHERM_00000010', ...,\n",
       "       'TTHERM_02555200', 'TTHERM_02607240', 'TTHERM_02653470'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df['TTHERM_ID'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "e44ce228",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>GOs</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>EC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_01528500</td>\n",
       "      <td>GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...</td>\n",
       "      <td>ko:K07962</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TTHERM_00898250</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K12833</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TTHERM_00898300</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K05643</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TTHERM_01074550</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K01373</td>\n",
       "      <td>3.4.22.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TTHERM_01071490</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K08794,ko:K13412</td>\n",
       "      <td>2.7.11.1,2.7.11.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17269</th>\n",
       "      <td>TTHERM_01076950</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K09391</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17270</th>\n",
       "      <td>TTHERM_01076960</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K12869</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17271</th>\n",
       "      <td>TTHERM_01076970</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K04958,ko:K04960</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17272</th>\n",
       "      <td>TTHERM_01076980</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K06631</td>\n",
       "      <td>2.7.11.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17273</th>\n",
       "      <td>TTHERM_01077000</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K14713</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6625 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID                                                GOs  \\\n",
       "2      TTHERM_01528500  GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...   \n",
       "13     TTHERM_00898250                                                  -   \n",
       "17     TTHERM_00898300                                                  -   \n",
       "22     TTHERM_01074550                                                  -   \n",
       "25     TTHERM_01071490                                                  -   \n",
       "...                ...                                                ...   \n",
       "17269  TTHERM_01076950                                                  -   \n",
       "17270  TTHERM_01076960                                                  -   \n",
       "17271  TTHERM_01076970                                                  -   \n",
       "17272  TTHERM_01076980                                                  -   \n",
       "17273  TTHERM_01077000                                                  -   \n",
       "\n",
       "                   KEGG_ko                  EC  \n",
       "2                ko:K07962                   -  \n",
       "13               ko:K12833                   -  \n",
       "17               ko:K05643                   -  \n",
       "22               ko:K01373           3.4.22.41  \n",
       "25     ko:K08794,ko:K13412  2.7.11.1,2.7.11.17  \n",
       "...                    ...                 ...  \n",
       "17269            ko:K09391                   -  \n",
       "17270            ko:K12869                   -  \n",
       "17271  ko:K04958,ko:K04960                   -  \n",
       "17272            ko:K06631           2.7.11.21  \n",
       "17273            ko:K14713                   -  \n",
       "\n",
       "[6625 rows x 4 columns]"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annotated_filtered = complete_annotated.loc[complete_annotated['TTHERM_ID'].isin(full_filtered_df['TTHERM_ID'].values)]\n",
    "complete_annotated_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "18770c2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.3259372232608482"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annotated_filtered.shape[0]/full_filtered_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "a8f20b8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GOs</th>\n",
       "      <th>GOs_description</th>\n",
       "      <th>GOs_definition</th>\n",
       "      <th>GOs_obsolete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>The distribution of mitochondria, including th...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>The maintenance of the structure and integrity...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>The production of new individuals that contain...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000011</td>\n",
       "      <td>vacuole inheritance</td>\n",
       "      <td>The distribution of vacuoles into daughter cel...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GOs                   GOs_description  \\\n",
       "0           -                                 -   \n",
       "1  GO:0000001         mitochondrion inheritance   \n",
       "2  GO:0000002  mitochondrial genome maintenance   \n",
       "3  GO:0000003                      reproduction   \n",
       "4  GO:0000011               vacuole inheritance   \n",
       "\n",
       "                                      GOs_definition GOs_obsolete  \n",
       "0                                                  -            -  \n",
       "1  The distribution of mitochondria, including th...        False  \n",
       "2  The maintenance of the structure and integrity...        False  \n",
       "3  The production of new individuals that contain...        False  \n",
       "4  The distribution of vacuoles into daughter cel...        False  "
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "bff38f9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mitochondrion inheritance'"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df['GOs_description'].loc[go_df['GOs'] == 'GO:0000001'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "223b06f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_ko_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ko:K00002</td>\n",
       "      <td>alcohol dehydrogenase (NADP+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ko:K00006</td>\n",
       "      <td>glycerol-3-phosphate dehydrogenase (NAD+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ko:K00011</td>\n",
       "      <td>aldehyde reductase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ko:K00012</td>\n",
       "      <td>UDPglucose 6-dehydrogenase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     KEGG_ko                        KEGG_ko_description\n",
       "0          -                                          -\n",
       "1  ko:K00002              alcohol dehydrogenase (NADP+)\n",
       "2  ko:K00006  glycerol-3-phosphate dehydrogenase (NAD+)\n",
       "3  ko:K00011                         aldehyde reductase\n",
       "4  ko:K00012                 UDPglucose 6-dehydrogenase"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "2414d9c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC</th>\n",
       "      <th>EC_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>alcohol dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.1.1.105</td>\n",
       "      <td>all-trans-retinol dehydrogenase (NAD(+))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.1.1.141</td>\n",
       "      <td>15-hydroxyprostaglandin dehydrogenase (NAD(+))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.1.1.178</td>\n",
       "      <td>3-hydroxy-2-methylbutyryl-CoA dehydrogenase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          EC                                  EC_description\n",
       "0          -                                               -\n",
       "1    1.1.1.1                           alcohol dehydrogenase\n",
       "2  1.1.1.105        all-trans-retinol dehydrogenase (NAD(+))\n",
       "3  1.1.1.141  15-hydroxyprostaglandin dehydrogenase (NAD(+))\n",
       "4  1.1.1.178     3-hydroxy-2-methylbutyryl-CoA dehydrogenase"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ec_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "108df2b3-df6f-4028-a879-e9034f79309a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# As of 2020 https://www.ncbi.nlm.nih.gov/research/cog/\n",
    "COG_dict = {\n",
    "    \"A\" : \"RNA processing and modification\",\n",
    "    \"B\" : \"Chromatin structure and dynamics\",\n",
    "    \"C\" : \"Energy production and conversion\",\n",
    "    \"D\" : \"Cell cycle control, cell division, chromosome partitioning\",\n",
    "    \"E\" : \"Amino acid transport and metabolism\",\n",
    "    \"F\" : \"Nucleotide transport and metabolism\",\n",
    "    \"G\" : \"Carbohydrate transport and metabolism\",\n",
    "    \"H\" : \"Coenzyme transport and metabolism\",\n",
    "    \"I\" : \"Lipid transport and metabolism\",\n",
    "    \"J\" : \"Translation, ribosomal structure and biogenesis\",\n",
    "    \"K\" : \"Transcription\",\n",
    "    \"L\" : \"Replication, recombination, and repair\",\n",
    "    \"M\" : \"Cell wall/membrane/envelope biogenesis\",\n",
    "    \"N\" : \"Cell motility\",\n",
    "    \"O\" : \"Posttranslational modification, protein turnover, chaperones\",\n",
    "    \"P\" : \"Inorganic ion transport and metabolism\",\n",
    "    \"Q\" : \"Secondary metabolites biosynthesis, transport and catabolism\",\n",
    "    \"T\" : \"Signal transduction mechanisms\",\n",
    "    \"U\" : \"Intracellular trafficking, secretion, and vesicular transport\",\n",
    "    \"V\" : \"Defense mechanisms\",\n",
    "    \"W\" : \"Extracellular structures\",\n",
    "    \"X\" : \"Mobilome: prophages, transposons\",\n",
    "    \"Y\" : \"Nuclear structure\",\n",
    "    \"Z\" : \"Cytoskeleton\",\n",
    "    \"R\" : \"General function prediction only\",\n",
    "    \"S\" : \"Function unknown\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "b0251f64-8d43-4f21-9a88-772f44b71f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "def term_count_dict_from_annotation_df(annot_df, term_column):\n",
    "    \n",
    "    column = annot_df[term_column].values\n",
    "    \n",
    "    funct_terms = []\n",
    "    for entry in column:\n",
    "        if entry != '-':\n",
    "            if term_column == 'COG_category':\n",
    "                # terms = [f'{e}: {COG_dict[e]}' for e in entry]\n",
    "                terms = list(entry)\n",
    "            else:\n",
    "                terms = entry.split(',')\n",
    "            for t in terms:\n",
    "                funct_terms.append(t)\n",
    "\n",
    "#     len(terms)\n",
    "    \n",
    "    term_count_dict = {}\n",
    "    \n",
    "    for t in funct_terms:\n",
    "        count = term_count_dict.get(t, 0)\n",
    "        count += 1\n",
    "        term_count_dict[t] = count\n",
    "        \n",
    "    return term_count_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "0fddca3d-22c7-42cf-b4d1-f663bfa84cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrichment_analysis(module, leiden_label_df, phases, background_annotation, term_column):\n",
    "    \n",
    "    module_ttids = leiden_label_df.loc[leiden_label_df[f'leiden_label_{phases}'] == module]['TTHERM_ID'].values\n",
    "    \n",
    "    module_annotation = background_annotation.loc[background_annotation['TTHERM_ID'].isin(module_ttids)]\n",
    "    \n",
    "    background_term_dict = term_count_dict_from_annotation_df(background_annotation, term_column)\n",
    "    module_term_dict = term_count_dict_from_annotation_df(module_annotation, term_column)\n",
    "    \n",
    "    bs = []\n",
    "    ps = []\n",
    "    folds = []\n",
    "    terms = []\n",
    "    \n",
    "    for t, module_count in module_term_dict.items():\n",
    "        \n",
    "        background_count = background_term_dict[t]\n",
    "        module_size = len(module_annotation)\n",
    "        background_size = len(background_annotation)\n",
    "        \n",
    "        standard_contingency_table = [\n",
    "                                [module_count, background_count - module_count], \n",
    "                                [module_size - module_count, background_size - module_size - (background_count - module_count)]\n",
    "                            ]\n",
    "        \n",
    "        # The -1 and +1 make this more conservative (see explanation from the DAVID database: \n",
    "        # https://david.ncifcrf.gov/helps/functional_annotation.html#geneenrich)\n",
    "        conservative_contingency_table = [\n",
    "                                [module_count - 1, background_count - module_count + 1], \n",
    "                                [module_size - module_count, background_size - module_size - (background_count - module_count)]\n",
    "                            ]\n",
    "        \n",
    "        \n",
    "        odds, p_standard = st.fisher_exact(standard_contingency_table, 'greater')\n",
    "        odds, p_conservative = st.fisher_exact(conservative_contingency_table, 'greater')\n",
    "        \n",
    "        p_reasonable = np.mean([p_standard, p_conservative])\n",
    "        \n",
    "        bonferroni  = p_reasonable * len(module_term_dict)\n",
    "\n",
    "        fold_enrichment = (module_count/module_size) / (background_count/background_size)\n",
    "\n",
    "        if bonferroni <= 0.05:\n",
    "            \n",
    "            ps.append(p_reasonable)\n",
    "            bs.append(bonferroni)\n",
    "            folds.append(fold_enrichment)\n",
    "            terms.append(t)\n",
    "            \n",
    "#         else:\n",
    "#             ps.append('')\n",
    "#             bs.append('')\n",
    "#             folds.append('')\n",
    "#             terms.append('')\n",
    "            \n",
    "    return ps, bs, folds, terms\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "6482b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_GO_info(go_term):\n",
    "    \n",
    "    name = go_df['GOs_description'].loc[go_df['GOs'] == go_term].values[0]\n",
    "    \n",
    "    definition = go_df['GOs_definition'].loc[go_df['GOs'] == go_term].values[0]\n",
    "    \n",
    "    obsolete = go_df['GOs_obsolete'].loc[go_df['GOs'] == go_term].values[0]\n",
    "    \n",
    "    return name, definition, obsolete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "c9dad6d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('mitochondrion inheritance',\n",
       " 'The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.',\n",
       " 'False')"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_GO_info('GO:0000001')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "e47679cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_KEGG_info(term):\n",
    "    return kegg_df['KEGG_ko_description'].loc[kegg_df['KEGG_ko'] == term].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "0f53543b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'alcohol dehydrogenase (NADP+)'"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_KEGG_info('ko:K00002')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "22b083ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_EC_info(term):\n",
    "    return ec_df['EC_description'].loc[ec_df['EC'] == term].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "39579c7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'15-hydroxyprostaglandin dehydrogenase (NAD(+))'"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_EC_info('1.1.1.141')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "14cb6323-a773-46b7-85e3-f9d5bd609e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_enrichment(lldf, phases, out_prefix, term_column):\n",
    "    \n",
    "    with open(f'../enrichment/{out_prefix}_enrichment.txt', 'w') as fl:\n",
    "    \n",
    "        for m in tqdm.tqdm(sorted(lldf[f'leiden_label_{phases}'].unique())):\n",
    "\n",
    "            # print(f'Module {m}')\n",
    "            ps, bs, folds, terms = enrichment_analysis(m, lldf, phases, complete_annot, term_column)\n",
    "\n",
    "            fl.write(f'Module {m}\\n')\n",
    "            for b, fold, t in zip(bs, folds, terms):\n",
    "                \n",
    "                if term_column == 'COG_category':\n",
    "                    \n",
    "                    definition = COG_dict[t]\n",
    "                    t = f'{t}: {definition}'\n",
    "                \n",
    "                elif term_column == 'GOs':\n",
    "                    name, definition, obsolete = get_GO_info(t)\n",
    "                    if obsolete:\n",
    "                        t = f'{name.capitalize()} ({t}): {definition} (obsolete)'\n",
    "                    else:\n",
    "                        t = f'{name.capitalize()} ({t}): {definition}'\n",
    "                        \n",
    "                elif term_column.split('_')[0] == 'KEGG':\n",
    "                    name = get_KEGG_info(t)\n",
    "                    \n",
    "                    t = f'{t}: {name}'\n",
    "                    \n",
    "                elif term_column == 'EC':\n",
    "                    \n",
    "                    name = get_EC_info(t)\n",
    "                    t = f'{t}: {name}'\n",
    "                    \n",
    "                        \n",
    "                # print(f\"{t}\\nFold-enrichment: {fold:.02f}\\nBonferroni-corrected p-value: {b:.02e}\\n\\n\")\n",
    "\n",
    "                fl.write(f'{t}\\nFold-enrichment: {fold:.02f}\\nBonferroni-corrected p-value: {b:.02e}\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "dd5d7697-d044-45e7-a4d4-5945fa23775b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enrichment_df(lldf, phases, background_annotation, term_columns=['COG_category', 'GOs', 'KEGG_ko', 'EC'], outfile=None):\n",
    "    \n",
    "    module_dfs = []\n",
    "    \n",
    "    for m in tqdm.tqdm(sorted(lldf[f'leiden_label_{phases}'].unique())):\n",
    "        \n",
    "        term_dfs = []\n",
    "\n",
    "        for tc in term_columns:\n",
    "        \n",
    "            ps, bs, folds, terms = enrichment_analysis(m, lldf, phases, background_annotation, tc)\n",
    "\n",
    "            # fl.write(f'Module {m}\\n')\n",
    "            \n",
    "            info = []\n",
    "\n",
    "            if tc == 'GOs':\n",
    "\n",
    "                for t in terms:\n",
    "                    name, definition, obsolete = get_GO_info(t)\n",
    "                    if obsolete:\n",
    "                        info.append(f'{name.capitalize()}: {definition} (obsolete)')\n",
    "                    else:\n",
    "                        info.append(f'{name.capitalize()}: {definition}')\n",
    "                        \n",
    "            elif tc == 'COG_category':\n",
    "                for t in terms:\n",
    "                    info.append(COG_dict[t])\n",
    "                                \n",
    "            elif tc == 'KEGG_ko':\n",
    "                for t in terms:\n",
    "                    info.append(get_KEGG_info(t))\n",
    "                    \n",
    "            elif tc == 'EC':\n",
    "                for t in terms:\n",
    "                    info.append(get_EC_info(t))\n",
    "                    \n",
    "            term_df = pd.DataFrame({'module': [m]*len(terms),\n",
    "                                    'term': terms,\n",
    "                                    'info': info,\n",
    "                                    'fold_change': folds,\n",
    "                                    'bonferroni': bs})\n",
    "            \n",
    "            term_dfs.append(term_df)\n",
    "            \n",
    "        module_df = pd.concat(term_dfs)\n",
    "        \n",
    "        module_dfs.append(module_df)\n",
    "        \n",
    "    all_enrichment_df = pd.concat(module_dfs)\n",
    "    \n",
    "    if outfile:\n",
    "        all_enrichment_df.to_csv(outfile, index=False)\n",
    "    \n",
    "    return all_enrichment_df\n",
    "        \n",
    "        \n",
    "                        \n",
    "            \n",
    "                    \n",
    "            \n",
    "        \n",
    "#         for b, fold, t in zip(bs, folds, terms):\n",
    "\n",
    "#             if term_column == 'GOs':\n",
    "#                 name, definition, obsolete = get_GO_info(t)\n",
    "#                 if obsolete:\n",
    "#                     t = f'{name.capitalize()} ({t}): {definition} (obsolete)'\n",
    "#                 else:\n",
    "#                     t = f'{name.capitalize()} ({t}): {definition}'\n",
    "\n",
    "#             if term_column.split('_')[0] == 'KEGG':\n",
    "#                 dfs = pd.read_html(f'https://www.genome.jp/entry/{t}')\n",
    "\n",
    "#                 data = dfs[4]\n",
    "#                 entry = data.loc[data[0] == 'Entry'][1].values[0]\n",
    "#                 name = data.loc[data[0] == 'Name'][1].values[0]\n",
    "\n",
    "#                 t = f'{t}: {name}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "3e001ab1-173b-4ed8-85d5-091142480820",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annotation = pd.read_csv('../eggnog/complete_eggnog_annotation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "7f839c0b-7ed1-4a84-b058-dafccbfdad88",
   "metadata": {},
   "outputs": [],
   "source": [
    "lldf_nn3 = pd.read_csv('../embedding/test_nn3_leiden_label_df_round_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "c4b49a02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "838\n",
      "837\n",
      "837\n",
      "837\n",
      "838\n",
      "838\n",
      "838\n",
      "838\n",
      "838\n",
      "838\n",
      "839\n",
      "839\n",
      "838\n",
      "838\n",
      "838\n",
      "839\n",
      "839\n",
      "837\n",
      "837\n",
      "839\n",
      "837\n",
      "837\n",
      "837\n"
     ]
    }
   ],
   "source": [
    "gene_list_test = ['TTHERM_000013409', 'TTHERM_01321550', 'TTHERM_00011710', 'TTHERM_00321680', 'TTHERM_00355700', 'TTHERM_00938950', 'TTHERM_01372820', 'TTHERM_00013410', 'TTHERM_00390080', 'TTHERM_00516380', 'TTHERM_00038880', 'TTHERM_00059370', 'TTHERM_00473020', 'TTHERM_00497590', 'TTHERM_00558350', 'TTHERM_00052190', 'TTHERM_00392790', 'TTHERM_00410180', 'TTHERM_00685980', 'TTHERM_00445920', 'TTHERM_00471040', 'TTHERM_00140780', 'TTHERM_00145480', 'TTHERM_00321720', 'TTHERM_00628650', 'TTHERM_00526730', 'TTHERM_01156770', 'TTHERM_00312200', 'TTHERM_01332070', 'TTHERM_00318900', 'TTHERM_00340180', 'TTHERM_00592740', 'TTHERM_00440600', 'TTHERM_01321570', 'TTHERM_00537380', 'TTHERM_00585170', 'TTHERM_01197130', 'TTHERM_00554390', 'TTHERM_00649180', 'TTHERM_00691410', ]\n",
    "for gene in gene_list_test:\n",
    "    print((lldf_nn3.loc[lldf_nn3['TTHERM_ID'] == gene])['leiden_label_full'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "0a18a134-a62a-4166-9764-a05d90f45361",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1283, 1282, 1281, ...,    2,    1,    0])"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lldf_nn3['leiden_label_full'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "93dc3cc2-1fc1-4018-80cb-21f998c0ed74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "fb55c9e2-7d38-44d2-a944-51bf9519ab39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 2/1284 [00:00<02:35,  8.27it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1284/1284 [02:03<00:00, 10.36it/s]\n"
     ]
    }
   ],
   "source": [
    "enrich_full_nn3 = get_enrichment_df(lldf_nn3, 'full', complete_annotation, outfile='./test_nn3_full_enrichment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "2c4fefa0-73c9-422e-8090-61fa18d1faf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrich_full_nn3 = pd.read_csv('./test_nn3_full_enrichment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "6b2f9dee-8c14-40eb-a3b8-bc3865d6f367",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>term</th>\n",
       "      <th>info</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>bonferroni</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>U</td>\n",
       "      <td>Intracellular trafficking, secretion, and vesi...</td>\n",
       "      <td>32.699875</td>\n",
       "      <td>4.473578e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0005575</td>\n",
       "      <td>Cellular_component: A location, relative to ce...</td>\n",
       "      <td>13.756647</td>\n",
       "      <td>1.892595e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0005622</td>\n",
       "      <td>Intracellular anatomical structure: A componen...</td>\n",
       "      <td>14.877054</td>\n",
       "      <td>1.493668e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0005623</td>\n",
       "      <td>Obsolete cell: OBSOLETE. The basic structural ...</td>\n",
       "      <td>14.309537</td>\n",
       "      <td>1.680029e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>Cytoplasm: The contents of a cell excluding th...</td>\n",
       "      <td>18.124590</td>\n",
       "      <td>8.225290e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0005794</td>\n",
       "      <td>Golgi apparatus: A membrane-bound cytoplasmic ...</td>\n",
       "      <td>140.042667</td>\n",
       "      <td>1.722394e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0006464</td>\n",
       "      <td>Protein modification process: The covalent alt...</td>\n",
       "      <td>42.609331</td>\n",
       "      <td>6.239795e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0006471</td>\n",
       "      <td>Obsolete protein adp-ribosylation: OBSOLETE. T...</td>\n",
       "      <td>2100.640000</td>\n",
       "      <td>3.739236e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0006807</td>\n",
       "      <td>Nitrogen compound metabolic process: The chemi...</td>\n",
       "      <td>22.957814</td>\n",
       "      <td>4.028969e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0006810</td>\n",
       "      <td>Transport: The directed movement of substances...</td>\n",
       "      <td>51.739901</td>\n",
       "      <td>3.475154e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0006886</td>\n",
       "      <td>Intracellular protein transport: The directed ...</td>\n",
       "      <td>244.260465</td>\n",
       "      <td>3.193529e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0008104</td>\n",
       "      <td>Protein localization: Any process in which a p...</td>\n",
       "      <td>78.090706</td>\n",
       "      <td>1.004480e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>Biological_process: A biological process is th...</td>\n",
       "      <td>13.613999</td>\n",
       "      <td>1.953197e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0008152</td>\n",
       "      <td>Metabolic process: The chemical reactions and ...</td>\n",
       "      <td>19.968061</td>\n",
       "      <td>6.139191e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0009987</td>\n",
       "      <td>Cellular process: Any process that is carried ...</td>\n",
       "      <td>15.782419</td>\n",
       "      <td>1.249424e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0012505</td>\n",
       "      <td>Endomembrane system: A collection of membranou...</td>\n",
       "      <td>55.134908</td>\n",
       "      <td>2.869236e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0015031</td>\n",
       "      <td>Protein transport: The directed movement of pr...</td>\n",
       "      <td>125.786826</td>\n",
       "      <td>2.382531e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0015833</td>\n",
       "      <td>Peptide transport: The directed movement of pe...</td>\n",
       "      <td>121.424277</td>\n",
       "      <td>2.650629e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0016192</td>\n",
       "      <td>Vesicle-mediated transport: A cellular transpo...</td>\n",
       "      <td>114.165217</td>\n",
       "      <td>3.193058e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0019538</td>\n",
       "      <td>Protein metabolic process: The chemical reacti...</td>\n",
       "      <td>33.290650</td>\n",
       "      <td>1.313201e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0033036</td>\n",
       "      <td>Macromolecule localization: Any process in whi...</td>\n",
       "      <td>69.557616</td>\n",
       "      <td>1.423959e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0034613</td>\n",
       "      <td>Protein localization: Any process in which a p...</td>\n",
       "      <td>114.165217</td>\n",
       "      <td>3.193058e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0036211</td>\n",
       "      <td>Protein modification process: The covalent alt...</td>\n",
       "      <td>42.609331</td>\n",
       "      <td>6.239795e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0042886</td>\n",
       "      <td>Amide transport: The directed movement of an a...</td>\n",
       "      <td>120.036571</td>\n",
       "      <td>2.744275e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0043170</td>\n",
       "      <td>Macromolecule metabolic process: The chemical ...</td>\n",
       "      <td>25.157365</td>\n",
       "      <td>3.056950e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0043226</td>\n",
       "      <td>Organelle: Organized structure of distinctive ...</td>\n",
       "      <td>17.092270</td>\n",
       "      <td>9.819402e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0043227</td>\n",
       "      <td>Membrane-bounded organelle: Organized structur...</td>\n",
       "      <td>19.342910</td>\n",
       "      <td>6.758082e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0043229</td>\n",
       "      <td>Intracellular organelle: Organized structure o...</td>\n",
       "      <td>17.461679</td>\n",
       "      <td>9.205166e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0043231</td>\n",
       "      <td>Intracellular membrane-bounded organelle: Orga...</td>\n",
       "      <td>20.716371</td>\n",
       "      <td>5.493777e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0043412</td>\n",
       "      <td>Macromolecule modification: The covalent alter...</td>\n",
       "      <td>41.432742</td>\n",
       "      <td>6.789421e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044237</td>\n",
       "      <td>Cellular metabolic process: The chemical react...</td>\n",
       "      <td>21.090763</td>\n",
       "      <td>5.204607e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044238</td>\n",
       "      <td>Primary metabolic process: The chemical reacti...</td>\n",
       "      <td>21.500921</td>\n",
       "      <td>4.910625e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044260</td>\n",
       "      <td>Obsolete cellular macromolecule metabolic proc...</td>\n",
       "      <td>28.658117</td>\n",
       "      <td>2.063432e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044267</td>\n",
       "      <td>Protein metabolic process: The chemical reacti...</td>\n",
       "      <td>35.364310</td>\n",
       "      <td>1.094457e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044424</td>\n",
       "      <td>Obsolete intracellular part: OBSOLETE. Any con...</td>\n",
       "      <td>15.026037</td>\n",
       "      <td>1.449350e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044444</td>\n",
       "      <td>Obsolete cytoplasmic part: OBSOLETE. Any const...</td>\n",
       "      <td>22.323486</td>\n",
       "      <td>4.384506e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0044464</td>\n",
       "      <td>Obsolete cell part: OBSOLETE. Any constituent ...</td>\n",
       "      <td>14.309537</td>\n",
       "      <td>1.680029e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0045184</td>\n",
       "      <td>Establishment of protein localization: The dir...</td>\n",
       "      <td>114.165217</td>\n",
       "      <td>3.193058e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0046907</td>\n",
       "      <td>Intracellular transport: The directed movement...</td>\n",
       "      <td>135.525161</td>\n",
       "      <td>1.901855e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GO:0051179</td>\n",
       "      <td>Localization: Any process in which a cell, a s...</td>\n",
       "      <td>40.552896</td>\n",
       "      <td>7.243292e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    module        term                                               info  \\\n",
       "0      1.0           U  Intracellular trafficking, secretion, and vesi...   \n",
       "1      1.0  GO:0005575  Cellular_component: A location, relative to ce...   \n",
       "2      1.0  GO:0005622  Intracellular anatomical structure: A componen...   \n",
       "3      1.0  GO:0005623  Obsolete cell: OBSOLETE. The basic structural ...   \n",
       "4      1.0  GO:0005737  Cytoplasm: The contents of a cell excluding th...   \n",
       "5      1.0  GO:0005794  Golgi apparatus: A membrane-bound cytoplasmic ...   \n",
       "6      1.0  GO:0006464  Protein modification process: The covalent alt...   \n",
       "7      1.0  GO:0006471  Obsolete protein adp-ribosylation: OBSOLETE. T...   \n",
       "8      1.0  GO:0006807  Nitrogen compound metabolic process: The chemi...   \n",
       "9      1.0  GO:0006810  Transport: The directed movement of substances...   \n",
       "10     1.0  GO:0006886  Intracellular protein transport: The directed ...   \n",
       "11     1.0  GO:0008104  Protein localization: Any process in which a p...   \n",
       "12     1.0  GO:0008150  Biological_process: A biological process is th...   \n",
       "13     1.0  GO:0008152  Metabolic process: The chemical reactions and ...   \n",
       "14     1.0  GO:0009987  Cellular process: Any process that is carried ...   \n",
       "15     1.0  GO:0012505  Endomembrane system: A collection of membranou...   \n",
       "16     1.0  GO:0015031  Protein transport: The directed movement of pr...   \n",
       "17     1.0  GO:0015833  Peptide transport: The directed movement of pe...   \n",
       "18     1.0  GO:0016192  Vesicle-mediated transport: A cellular transpo...   \n",
       "19     1.0  GO:0019538  Protein metabolic process: The chemical reacti...   \n",
       "20     1.0  GO:0033036  Macromolecule localization: Any process in whi...   \n",
       "21     1.0  GO:0034613  Protein localization: Any process in which a p...   \n",
       "22     1.0  GO:0036211  Protein modification process: The covalent alt...   \n",
       "23     1.0  GO:0042886  Amide transport: The directed movement of an a...   \n",
       "24     1.0  GO:0043170  Macromolecule metabolic process: The chemical ...   \n",
       "25     1.0  GO:0043226  Organelle: Organized structure of distinctive ...   \n",
       "26     1.0  GO:0043227  Membrane-bounded organelle: Organized structur...   \n",
       "27     1.0  GO:0043229  Intracellular organelle: Organized structure o...   \n",
       "28     1.0  GO:0043231  Intracellular membrane-bounded organelle: Orga...   \n",
       "29     1.0  GO:0043412  Macromolecule modification: The covalent alter...   \n",
       "30     1.0  GO:0044237  Cellular metabolic process: The chemical react...   \n",
       "31     1.0  GO:0044238  Primary metabolic process: The chemical reacti...   \n",
       "32     1.0  GO:0044260  Obsolete cellular macromolecule metabolic proc...   \n",
       "33     1.0  GO:0044267  Protein metabolic process: The chemical reacti...   \n",
       "34     1.0  GO:0044424  Obsolete intracellular part: OBSOLETE. Any con...   \n",
       "35     1.0  GO:0044444  Obsolete cytoplasmic part: OBSOLETE. Any const...   \n",
       "36     1.0  GO:0044464  Obsolete cell part: OBSOLETE. Any constituent ...   \n",
       "37     1.0  GO:0045184  Establishment of protein localization: The dir...   \n",
       "38     1.0  GO:0046907  Intracellular transport: The directed movement...   \n",
       "39     1.0  GO:0051179  Localization: Any process in which a cell, a s...   \n",
       "\n",
       "    fold_change    bonferroni  \n",
       "0     32.699875  4.473578e-07  \n",
       "1     13.756647  1.892595e-02  \n",
       "2     14.877054  1.493668e-02  \n",
       "3     14.309537  1.680029e-02  \n",
       "4     18.124590  8.225290e-03  \n",
       "5    140.042667  1.722394e-05  \n",
       "6     42.609331  6.239795e-04  \n",
       "7   2100.640000  3.739236e-09  \n",
       "8     22.957814  4.028969e-03  \n",
       "9     51.739901  3.475154e-04  \n",
       "10   244.260465  3.193529e-06  \n",
       "11    78.090706  1.004480e-04  \n",
       "12    13.613999  1.953197e-02  \n",
       "13    19.968061  6.139191e-03  \n",
       "14    15.782419  1.249424e-02  \n",
       "15    55.134908  2.869236e-04  \n",
       "16   125.786826  2.382531e-05  \n",
       "17   121.424277  2.650629e-05  \n",
       "18   114.165217  3.193058e-05  \n",
       "19    33.290650  1.313201e-03  \n",
       "20    69.557616  1.423959e-04  \n",
       "21   114.165217  3.193058e-05  \n",
       "22    42.609331  6.239795e-04  \n",
       "23   120.036571  2.744275e-05  \n",
       "24    25.157365  3.056950e-03  \n",
       "25    17.092270  9.819402e-03  \n",
       "26    19.342910  6.758082e-03  \n",
       "27    17.461679  9.205166e-03  \n",
       "28    20.716371  5.493777e-03  \n",
       "29    41.432742  6.789421e-04  \n",
       "30    21.090763  5.204607e-03  \n",
       "31    21.500921  4.910625e-03  \n",
       "32    28.658117  2.063432e-03  \n",
       "33    35.364310  1.094457e-03  \n",
       "34    15.026037  1.449350e-02  \n",
       "35    22.323486  4.384506e-03  \n",
       "36    14.309537  1.680029e-02  \n",
       "37   114.165217  3.193058e-05  \n",
       "38   135.525161  1.901855e-05  \n",
       "39    40.552896  7.243292e-04  "
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrich_full_nn3.head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "10c2768f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "U\n",
      "G\n",
      "Q\n",
      "M\n",
      "O\n",
      "T\n",
      "B\n",
      "Y\n",
      "L\n",
      "Z\n",
      "D\n",
      "K\n",
      "J\n",
      "S\n",
      "P\n",
      "A\n",
      "W\n",
      "F\n",
      "C\n",
      "E\n",
      "H\n",
      "I\n"
     ]
    }
   ],
   "source": [
    "for term in enrich_full_nn3['term'].unique():\n",
    "    if 'ko:' not in term and 'GO:' not in term and '.' not in term:\n",
    "        print(term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "0dfb99e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1281.0\n",
      "1.0\n",
      "1282\n",
      "1281\n",
      "\n",
      "{0, 5, 10, 11, 14, 15, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 94, 95, 96, 99, 100, 101, 102, 103, 104, 105, 106, 108, 109, 110, 111, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 125, 128, 130, 134, 136, 137, 138, 140, 141, 142, 143, 145, 146, 147, 148, 149, 150, 151, 152, 153, 155, 156, 157, 159, 160, 162, 163, 164, 165, 172, 173, 174, 175, 176, 177, 178, 180, 182, 183, 184, 185, 187, 188, 190, 192, 193, 194, 196, 203, 205, 206, 207, 208, 209, 210, 214, 215, 217, 219, 220, 221, 222, 223, 224, 226, 227, 228, 229, 230, 231, 232, 234, 235, 238, 239, 240, 241, 242, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 259, 260, 261, 263, 264, 265, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 296, 297, 298, 299, 300, 301, 302, 303, 305, 306, 307, 308, 311, 312, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 329, 331, 332, 333, 334, 335, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 348, 349, 350, 351, 355, 356, 357, 359, 360, 361, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 391, 393, 394, 395, 397, 398, 400, 401, 403, 404, 405, 406, 407, 408, 409, 410, 412, 415, 416, 417, 418, 420, 421, 422, 423, 424, 425, 427, 428, 429, 430, 431, 433, 434, 435, 436, 437, 438, 440, 441, 442, 445, 446, 447, 448, 449, 451, 453, 455, 456, 458, 459, 460, 463, 465, 466, 467, 469, 470, 471, 472, 473, 474, 475, 476, 477, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 495, 497, 498, 499, 500, 501, 504, 505, 506, 511, 512, 513, 514, 515, 516, 517, 518, 520, 521, 522, 523, 524, 526, 527, 528, 529, 530, 531, 532, 533, 535, 536, 537, 538, 540, 542, 543, 544, 545, 546, 548, 549, 551, 552, 553, 555, 556, 559, 560, 561, 562, 564, 565, 566, 567, 569, 570, 571, 573, 574, 576, 577, 578, 579, 580, 582, 583, 584, 585, 586, 588, 590, 591, 592, 593, 594, 596, 597, 598, 599, 600, 601, 605, 606, 607, 608, 609, 610, 611, 612, 613, 614, 616, 617, 618, 619, 620, 621, 622, 628, 630, 631, 632, 633, 634, 635, 637, 638, 639, 640, 641, 642, 644, 645, 647, 648, 649, 650, 651, 652, 653, 654, 656, 657, 658, 659, 660, 661, 662, 663, 664, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 676, 677, 678, 679, 680, 681, 682, 683, 684, 685, 686, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 702, 703, 704, 705, 706, 707, 710, 711, 712, 713, 714, 715, 717, 718, 719, 720, 721, 724, 726, 727, 728, 729, 730, 734, 735, 736, 737, 738, 739, 740, 741, 742, 743, 744, 749, 750, 751, 752, 753, 754, 755, 756, 757, 759, 760, 761, 762, 764, 765, 766, 767, 768, 769, 770, 771, 772, 773, 775, 776, 777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787, 789, 790, 791, 792, 794, 795, 797, 798, 801, 803, 804, 805, 807, 808, 809, 810, 811, 812, 813, 814, 816, 817, 818, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 832, 833, 834, 835, 836, 837, 838, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857, 860, 861, 862, 863, 864, 865, 867, 868, 869, 870, 871, 872, 874, 875, 876, 877, 878, 879, 880, 881, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 895, 896, 898, 899, 901, 902, 903, 905, 906, 907, 908, 909, 910, 911, 912, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 934, 936, 937, 938, 940, 942, 944, 945, 946, 947, 948, 949, 950, 953, 956, 957, 958, 960, 961, 962, 963, 964, 966, 970, 974, 975, 979, 980, 981, 982, 983, 985, 987, 988, 989, 990, 991, 992, 993, 994, 997, 998, 999, 1002, 1005, 1006, 1007, 1008, 1009, 1011, 1014, 1015, 1017, 1018, 1019, 1020, 1021, 1022, 1023, 1025, 1026, 1027, 1028, 1029, 1030, 1031, 1032, 1033, 1034, 1035, 1036, 1037, 1038, 1039, 1040, 1041, 1042, 1043, 1044, 1045, 1046, 1047, 1048, 1052, 1053, 1054, 1055, 1058, 1059, 1061, 1062, 1063, 1065, 1066, 1067, 1068, 1069, 1070, 1071, 1072, 1073, 1074, 1075, 1076, 1077, 1079, 1080, 1081, 1082, 1083, 1084, 1086, 1088, 1089, 1092, 1093, 1096, 1097, 1098, 1099, 1100, 1101, 1102, 1103, 1105, 1106, 1107, 1108, 1109, 1112, 1113, 1114, 1115, 1116, 1117, 1119, 1120, 1121, 1122, 1123, 1125, 1126, 1127, 1128, 1129, 1130, 1131, 1132, 1133, 1134, 1135, 1137, 1138, 1139, 1140, 1142, 1144, 1145, 1146, 1147, 1148, 1149, 1150, 1153, 1154, 1156, 1157, 1159, 1160, 1162, 1164, 1166, 1169, 1170, 1171, 1172, 1173, 1175, 1176, 1177, 1178, 1179, 1180, 1182, 1183, 1185, 1186, 1187, 1189, 1190, 1191, 1192, 1193, 1194, 1195, 1196, 1197, 1198, 1199, 1200, 1201, 1202, 1203, 1204, 1205, 1206, 1207, 1208, 1209, 1210, 1211, 1212, 1213, 1214, 1215, 1216, 1217, 1218, 1219, 1221, 1222, 1223, 1224, 1226, 1227, 1228, 1229, 1232, 1234, 1236, 1238, 1239, 1240, 1243, 1244, 1247, 1249, 1251, 1253, 1254, 1255, 1256, 1257, 1258, 1260, 1264, 1265, 1266, 1269, 1271, 1276, 1279, 1280}\n",
      "285\n",
      "285\n",
      "997\n"
     ]
    }
   ],
   "source": [
    "print(max(enrich_full_nn3['module'].values))\n",
    "print(min(enrich_full_nn3['module'].values))\n",
    "\n",
    "not_enriched_mod_list = []\n",
    "enriched_mod_list = []\n",
    "\n",
    "for num in range(int(max(enrich_full_nn3['module'].values)) + 1):\n",
    "    if num not in enrich_full_nn3['module'].values:\n",
    "        not_enriched_mod_list.append(num)\n",
    "    if num in enrich_full_nn3['module'].values:\n",
    "        enriched_mod_list.append(num)\n",
    "\n",
    "print(len(not_enriched_mod_list) + len(enriched_mod_list))\n",
    "print(max(not_enriched_mod_list + enriched_mod_list))\n",
    "print()\n",
    "\n",
    "print(set(range(int(max(enrich_full_nn3['module'].values)) + 1)) - set(enrich_full_nn3['module'].values))\n",
    "print(len(set(enrich_full_nn3['module'].values)))\n",
    "print(len(enriched_mod_list))\n",
    "print(len(not_enriched_mod_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "323ec30f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 5, 10, 11, 14, 15, 17, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 94, 95, 96, 99, 100, 101, 102, 103, 104, 105, 106, 108, 109, 110, 111, 112, 114, 115, 116, 117, 118, 119, 120, 122, 123, 125, 128, 130, 134, 136, 137, 138, 140, 141, 142, 143, 145, 146, 147, 148, 149, 150, 151, 152, 153, 155, 156, 157, 159, 160, 162, 163, 164, 165, 172, 173, 174, 175, 176, 177, 178, 180, 182, 183, 184, 185, 187, 188, 190, 192, 193, 194, 196, 203, 205, 206, 207, 208, 209, 210, 214, 215, 217, 219, 220, 221, 222, 223, 224, 226, 227, 228, 229, 230, 231, 232, 234, 235, 238, 239, 240, 241, 242, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 259, 260, 261, 263, 264, 265, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 296, 297, 298, 299, 300, 301, 302, 303, 305, 306, 307, 308, 311, 312, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 329, 331, 332, 333, 334, 335, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 348, 349, 350, 351, 355, 356, 357, 359, 360, 361, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 391, 393, 394, 395, 397, 398, 400, 401, 403, 404, 405, 406, 407, 408, 409, 410, 412, 415, 416, 417, 418, 420, 421, 422, 423, 424, 425, 427, 428, 429, 430, 431, 433, 434, 435, 436, 437, 438, 440, 441, 442, 445, 446, 447, 448, 449, 451, 453, 455, 456, 458, 459, 460, 463, 465, 466, 467, 469, 470, 471, 472, 473, 474, 475, 476, 477, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 495, 497, 498, 499, 500, 501, 504, 505, 506, 511, 512, 513, 514, 515, 516, 517, 518, 520, 521, 522, 523, 524, 526, 527, 528, 529, 530, 531, 532, 533, 535, 536, 537, 538, 540, 542, 543, 544, 545, 546, 548, 549, 551, 552, 553, 555, 556, 559, 560, 561, 562, 564, 565, 566, 567, 569, 570, 571, 573, 574, 576, 577, 578, 579, 580, 582, 583, 584, 585, 586, 588, 590, 591, 592, 593, 594, 596, 597, 598, 599, 600, 601, 605, 606, 607, 608, 609, 610, 611, 612, 613, 614, 616, 617, 618, 619, 620, 621, 622, 628, 630, 631, 632, 633, 634, 635, 637, 638, 639, 640, 641, 642, 644, 645, 647, 648, 649, 650, 651, 652, 653, 654, 656, 657, 658, 659, 660, 661, 662, 663, 664, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 676, 677, 678, 679, 680, 681, 682, 683, 684, 685, 686, 688, 689, 690, 691, 692, 693, 694, 695, 696, 697, 698, 699, 702, 703, 704, 705, 706, 707, 710, 711, 712, 713, 714, 715, 717, 718, 719, 720, 721, 724, 726, 727, 728, 729, 730, 734, 735, 736, 737, 738, 739, 740, 741, 742, 743, 744, 749, 750, 751, 752, 753, 754, 755, 756, 757, 759, 760, 761, 762, 764, 765, 766, 767, 768, 769, 770, 771, 772, 773, 775, 776, 777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787, 789, 790, 791, 792, 794, 795, 797, 798, 801, 803, 804, 805, 807, 808, 809, 810, 811, 812, 813, 814, 816, 817, 818, 819, 820, 821, 822, 823, 824, 825, 826, 827, 828, 829, 830, 832, 833, 834, 835, 836, 837, 838, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856, 857, 860, 861, 862, 863, 864, 865, 867, 868, 869, 870, 871, 872, 874, 875, 876, 877, 878, 879, 880, 881, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 895, 896, 898, 899, 901, 902, 903, 905, 906, 907, 908, 909, 910, 911, 912, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924, 925, 926, 927, 928, 929, 930, 931, 934, 936, 937, 938, 940, 942, 944, 945, 946, 947, 948, 949, 950, 953, 956, 957, 958, 960, 961, 962, 963, 964, 966, 970, 974, 975, 979, 980, 981, 982, 983, 985, 987, 988, 989, 990, 991, 992, 993, 994, 997, 998, 999, 1002, 1005, 1006, 1007, 1008, 1009, 1011, 1014, 1015, 1017, 1018, 1019, 1020, 1021, 1022, 1023, 1025, 1026, 1027, 1028, 1029, 1030, 1031, 1032, 1033, 1034, 1035, 1036, 1037, 1038, 1039, 1040, 1041, 1042, 1043, 1044, 1045, 1046, 1047, 1048, 1052, 1053, 1054, 1055, 1058, 1059, 1061, 1062, 1063, 1065, 1066, 1067, 1068, 1069, 1070, 1071, 1072, 1073, 1074, 1075, 1076, 1077, 1079, 1080, 1081, 1082, 1083, 1084, 1086, 1088, 1089, 1092, 1093, 1096, 1097, 1098, 1099, 1100, 1101, 1102, 1103, 1105, 1106, 1107, 1108, 1109, 1112, 1113, 1114, 1115, 1116, 1117, 1119, 1120, 1121, 1122, 1123, 1125, 1126, 1127, 1128, 1129, 1130, 1131, 1132, 1133, 1134, 1135, 1137, 1138, 1139, 1140, 1142, 1144, 1145, 1146, 1147, 1148, 1149, 1150, 1153, 1154, 1156, 1157, 1159, 1160, 1162, 1164, 1166, 1169, 1170, 1171, 1172, 1173, 1175, 1176, 1177, 1178, 1179, 1180, 1182, 1183, 1185, 1186, 1187, 1189, 1190, 1191, 1192, 1193, 1194, 1195, 1196, 1197, 1198, 1199, 1200, 1201, 1202, 1203, 1204, 1205, 1206, 1207, 1208, 1209, 1210, 1211, 1212, 1213, 1214, 1215, 1216, 1217, 1218, 1219, 1221, 1222, 1223, 1224, 1226, 1227, 1228, 1229, 1232, 1234, 1236, 1238, 1239, 1240, 1243, 1244, 1247, 1249, 1251, 1253, 1254, 1255, 1256, 1257, 1258, 1260, 1264, 1265, 1266, 1269, 1271, 1276, 1279, 1280]\n"
     ]
    }
   ],
   "source": [
    "print(not_enriched_mod_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "b4b518af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>term</th>\n",
       "      <th>info</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>bonferroni</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [module, term, info, fold_change, bonferroni]\n",
       "Index: []"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrich_full_nn3.loc[enrich_full_nn3['module']==72]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "801013e7-b277-4b7f-a6b7-773fc734c1b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.000e+00, 2.000e+00, 3.000e+00, 4.000e+00, 6.000e+00, 7.000e+00,\n",
       "       8.000e+00, 9.000e+00, 1.200e+01, 1.300e+01, 1.600e+01, 1.800e+01,\n",
       "       3.000e+01, 5.400e+01, 9.300e+01, 9.700e+01, 9.800e+01, 1.070e+02,\n",
       "       1.130e+02, 1.210e+02, 1.240e+02, 1.260e+02, 1.270e+02, 1.290e+02,\n",
       "       1.310e+02, 1.320e+02, 1.330e+02, 1.350e+02, 1.390e+02, 1.440e+02,\n",
       "       1.540e+02, 1.580e+02, 1.610e+02, 1.660e+02, 1.670e+02, 1.680e+02,\n",
       "       1.690e+02, 1.700e+02, 1.710e+02, 1.790e+02, 1.810e+02, 1.860e+02,\n",
       "       1.890e+02, 1.910e+02, 1.950e+02, 1.970e+02, 1.980e+02, 1.990e+02,\n",
       "       2.000e+02, 2.010e+02, 2.020e+02, 2.040e+02, 2.110e+02, 2.120e+02,\n",
       "       2.130e+02, 2.160e+02, 2.180e+02, 2.250e+02, 2.330e+02, 2.360e+02,\n",
       "       2.370e+02, 2.430e+02, 2.580e+02, 2.620e+02, 2.660e+02, 2.840e+02,\n",
       "       2.950e+02, 3.040e+02, 3.090e+02, 3.100e+02, 3.130e+02, 3.140e+02,\n",
       "       3.280e+02, 3.300e+02, 3.360e+02, 3.470e+02, 3.520e+02, 3.530e+02,\n",
       "       3.540e+02, 3.580e+02, 3.620e+02, 3.900e+02, 3.920e+02, 3.960e+02,\n",
       "       3.990e+02, 4.020e+02, 4.110e+02, 4.130e+02, 4.140e+02, 4.190e+02,\n",
       "       4.260e+02, 4.320e+02, 4.390e+02, 4.430e+02, 4.440e+02, 4.500e+02,\n",
       "       4.520e+02, 4.540e+02, 4.570e+02, 4.610e+02, 4.620e+02, 4.640e+02,\n",
       "       4.680e+02, 4.780e+02, 4.790e+02, 4.930e+02, 4.940e+02, 4.960e+02,\n",
       "       5.020e+02, 5.030e+02, 5.070e+02, 5.080e+02, 5.090e+02, 5.100e+02,\n",
       "       5.190e+02, 5.250e+02, 5.340e+02, 5.390e+02, 5.410e+02, 5.470e+02,\n",
       "       5.500e+02, 5.540e+02, 5.570e+02, 5.580e+02, 5.630e+02, 5.680e+02,\n",
       "       5.720e+02, 5.750e+02, 5.810e+02, 5.870e+02, 5.890e+02, 5.950e+02,\n",
       "       6.020e+02, 6.030e+02, 6.040e+02, 6.150e+02, 6.230e+02, 6.240e+02,\n",
       "       6.250e+02, 6.260e+02, 6.270e+02, 6.290e+02, 6.360e+02, 6.430e+02,\n",
       "       6.460e+02, 6.550e+02, 6.650e+02, 6.870e+02, 7.000e+02, 7.010e+02,\n",
       "       7.080e+02, 7.090e+02, 7.160e+02, 7.220e+02, 7.230e+02, 7.250e+02,\n",
       "       7.310e+02, 7.320e+02, 7.330e+02, 7.450e+02, 7.460e+02, 7.470e+02,\n",
       "       7.480e+02, 7.580e+02, 7.630e+02, 7.740e+02, 7.880e+02, 7.930e+02,\n",
       "       7.960e+02, 7.990e+02, 8.000e+02, 8.020e+02, 8.060e+02, 8.150e+02,\n",
       "       8.310e+02, 8.390e+02, 8.580e+02, 8.590e+02, 8.660e+02, 8.730e+02,\n",
       "       8.820e+02, 8.830e+02, 8.940e+02, 8.970e+02, 9.000e+02, 9.040e+02,\n",
       "       9.130e+02, 9.140e+02, 9.320e+02, 9.330e+02, 9.350e+02, 9.390e+02,\n",
       "       9.410e+02, 9.430e+02, 9.510e+02, 9.520e+02, 9.540e+02, 9.550e+02,\n",
       "       9.590e+02, 9.650e+02, 9.670e+02, 9.680e+02, 9.690e+02, 9.710e+02,\n",
       "       9.720e+02, 9.730e+02, 9.760e+02, 9.770e+02, 9.780e+02, 9.840e+02,\n",
       "       9.860e+02, 9.950e+02, 9.960e+02, 1.000e+03, 1.001e+03, 1.003e+03,\n",
       "       1.004e+03, 1.010e+03, 1.012e+03, 1.013e+03, 1.016e+03, 1.024e+03,\n",
       "       1.049e+03, 1.050e+03, 1.051e+03, 1.056e+03, 1.057e+03, 1.060e+03,\n",
       "       1.064e+03, 1.078e+03, 1.085e+03, 1.087e+03, 1.090e+03, 1.091e+03,\n",
       "       1.094e+03, 1.095e+03, 1.104e+03, 1.110e+03, 1.111e+03, 1.118e+03,\n",
       "       1.124e+03, 1.136e+03, 1.141e+03, 1.143e+03, 1.151e+03, 1.152e+03,\n",
       "       1.155e+03, 1.158e+03, 1.161e+03, 1.163e+03, 1.165e+03, 1.167e+03,\n",
       "       1.168e+03, 1.174e+03, 1.181e+03, 1.184e+03, 1.188e+03, 1.220e+03,\n",
       "       1.225e+03, 1.230e+03, 1.231e+03, 1.233e+03, 1.235e+03, 1.237e+03,\n",
       "       1.241e+03, 1.242e+03, 1.245e+03, 1.246e+03, 1.248e+03, 1.250e+03,\n",
       "       1.252e+03, 1.259e+03, 1.261e+03, 1.262e+03, 1.263e+03, 1.267e+03,\n",
       "       1.268e+03, 1.270e+03, 1.272e+03, 1.273e+03, 1.274e+03, 1.275e+03,\n",
       "       1.277e+03, 1.278e+03, 1.281e+03])"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrich_full_nn3['module'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "1228547a-b5d6-4d2f-9534-5a5a10043c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1283, 1282, 1281, ...,    2,    1,    0])"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lldf_nn3['leiden_label_full'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "1a6ff685-b372-4979-969e-2ae278aa46cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8200</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>748</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            TTHERM_ID  leiden_label_full\n",
       "8200  TTHERM_01386050                748"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lldf_nn3.loc[lldf_nn3['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "5230794a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "839\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "837\n",
      "838\n",
      "837\n",
      "837\n",
      "837\n",
      "838\n",
      "838\n",
      "838\n",
      "838\n",
      "838\n",
      "838\n",
      "839\n",
      "839\n",
      "838\n",
      "838\n",
      "838\n",
      "839\n",
      "839\n",
      "837\n",
      "837\n",
      "839\n",
      "837\n",
      "837\n",
      "837\n"
     ]
    }
   ],
   "source": [
    "gene_list_test = ['TTHERM_000013409', 'TTHERM_01321550', 'TTHERM_00011710', 'TTHERM_00321680', 'TTHERM_00355700', 'TTHERM_00938950', 'TTHERM_01372820', 'TTHERM_00013410', 'TTHERM_00390080', 'TTHERM_00516380', 'TTHERM_00038880', 'TTHERM_00059370', 'TTHERM_00473020', 'TTHERM_00497590', 'TTHERM_00558350', 'TTHERM_00052190', 'TTHERM_00392790', 'TTHERM_00410180', 'TTHERM_00685980', 'TTHERM_00445920', 'TTHERM_00471040', 'TTHERM_00140780', 'TTHERM_00145480', 'TTHERM_00321720', 'TTHERM_00628650', 'TTHERM_00526730', 'TTHERM_01156770', 'TTHERM_00312200', 'TTHERM_01332070', 'TTHERM_00318900', 'TTHERM_00340180', 'TTHERM_00592740', 'TTHERM_00440600', 'TTHERM_01321570', 'TTHERM_00537380', 'TTHERM_00585170', 'TTHERM_01197130', 'TTHERM_00554390', 'TTHERM_00649180', 'TTHERM_00691410', ]\n",
    "for gene in gene_list_test:\n",
    "    print((lldf_nn3.loc[lldf_nn3['TTHERM_ID'] == gene])['leiden_label_full'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "3e1672f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "285"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(enrich_full_nn3['module'].values))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d35056e7-38f7-4cf6-a5df-f387d19e8d3a",
   "metadata": {},
   "source": [
    "end here. Return to embedding notebook!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
