{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "44830610-d070-4882-8273-da4c5c22dd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import bs4\n",
    "import json\n",
    "import re\n",
    "import tqdm\n",
    "\n",
    "import scipy.stats as st\n",
    "import numpy as np\n",
    "\n",
    "import bokeh\n",
    "from bokeh.plotting import show as show_interactive\n",
    "from bokeh.plotting import output_file, output_notebook\n",
    "from bokeh.layouts import column, row\n",
    "from bokeh.models import CustomJS, TextInput, LassoSelectTool, Select, MultiSelect, ColorBar, Legend, LegendItem\n",
    "from bokeh.models.widgets import DataTable, DateFormatter, TableColumn, Button, HTMLTemplateFormatter\n",
    "from bokeh.events import SelectionGeometry\n",
    "from bokeh.transform import linear_cmap, jitter\n",
    "from bokeh.models import FactorRange\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5ca5c7a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annotation = pd.read_csv('../eggnog/complete_eggnog_annotation.csv')\n",
    "go_df = pd.read_csv('./go_annotations.csv')\n",
    "kegg_df = pd.read_csv('./kegg_annotations.csv')\n",
    "ec_df = pd.read_csv('./ec_annotations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a8f20b8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GOs</th>\n",
       "      <th>GOs_description</th>\n",
       "      <th>GOs_definition</th>\n",
       "      <th>GOs_obsolete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>The distribution of mitochondria, including th...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>The maintenance of the structure and integrity...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>The production of new individuals that contain...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000011</td>\n",
       "      <td>vacuole inheritance</td>\n",
       "      <td>The distribution of vacuoles into daughter cel...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GOs                   GOs_description  \\\n",
       "0           -                                 -   \n",
       "1  GO:0000001         mitochondrion inheritance   \n",
       "2  GO:0000002  mitochondrial genome maintenance   \n",
       "3  GO:0000003                      reproduction   \n",
       "4  GO:0000011               vacuole inheritance   \n",
       "\n",
       "                                      GOs_definition GOs_obsolete  \n",
       "0                                                  -            -  \n",
       "1  The distribution of mitochondria, including th...        False  \n",
       "2  The maintenance of the structure and integrity...        False  \n",
       "3  The production of new individuals that contain...        False  \n",
       "4  The distribution of vacuoles into daughter cel...        False  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bff38f9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mitochondrion inheritance'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_df['GOs_description'].loc[go_df['GOs'] == 'GO:0000001'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "223b06f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_ko_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ko:K00002</td>\n",
       "      <td>alcohol dehydrogenase (NADP+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ko:K00006</td>\n",
       "      <td>glycerol-3-phosphate dehydrogenase (NAD+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ko:K00011</td>\n",
       "      <td>aldehyde reductase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ko:K00012</td>\n",
       "      <td>UDPglucose 6-dehydrogenase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     KEGG_ko                        KEGG_ko_description\n",
       "0          -                                          -\n",
       "1  ko:K00002              alcohol dehydrogenase (NADP+)\n",
       "2  ko:K00006  glycerol-3-phosphate dehydrogenase (NAD+)\n",
       "3  ko:K00011                         aldehyde reductase\n",
       "4  ko:K00012                 UDPglucose 6-dehydrogenase"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2414d9c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EC</th>\n",
       "      <th>EC_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.1.1.1</td>\n",
       "      <td>alcohol dehydrogenase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.1.1.105</td>\n",
       "      <td>all-trans-retinol dehydrogenase (NAD(+))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.1.1.141</td>\n",
       "      <td>15-hydroxyprostaglandin dehydrogenase (NAD(+))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.1.1.178</td>\n",
       "      <td>3-hydroxy-2-methylbutyryl-CoA dehydrogenase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          EC                                  EC_description\n",
       "0          -                                               -\n",
       "1    1.1.1.1                           alcohol dehydrogenase\n",
       "2  1.1.1.105        all-trans-retinol dehydrogenase (NAD(+))\n",
       "3  1.1.1.141  15-hydroxyprostaglandin dehydrogenase (NAD(+))\n",
       "4  1.1.1.178     3-hydroxy-2-methylbutyryl-CoA dehydrogenase"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ec_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "108df2b3-df6f-4028-a879-e9034f79309a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# As of 2020 https://www.ncbi.nlm.nih.gov/research/cog/\n",
    "COG_dict = {\n",
    "    \"A\" : \"RNA processing and modification\",\n",
    "    \"B\" : \"Chromatin structure and dynamics\",\n",
    "    \"C\" : \"Energy production and conversion\",\n",
    "    \"D\" : \"Cell cycle control, cell division, chromosome partitioning\",\n",
    "    \"E\" : \"Amino acid transport and metabolism\",\n",
    "    \"F\" : \"Nucleotide transport and metabolism\",\n",
    "    \"G\" : \"Carbohydrate transport and metabolism\",\n",
    "    \"H\" : \"Coenzyme transport and metabolism\",\n",
    "    \"I\" : \"Lipid transport and metabolism\",\n",
    "    \"J\" : \"Translation, ribosomal structure and biogenesis\",\n",
    "    \"K\" : \"Transcription\",\n",
    "    \"L\" : \"Replication, recombination, and repair\",\n",
    "    \"M\" : \"Cell wall/membrane/envelope biogenesis\",\n",
    "    \"N\" : \"Cell motility\",\n",
    "    \"O\" : \"Posttranslational modification, protein turnover, chaperones\",\n",
    "    \"P\" : \"Inorganic ion transport and metabolism\",\n",
    "    \"Q\" : \"Secondary metabolites biosynthesis, transport and catabolism\",\n",
    "    \"T\" : \"Signal transduction mechanisms\",\n",
    "    \"U\" : \"Intracellular trafficking, secretion, and vesicular transport\",\n",
    "    \"V\" : \"Defense mechanisms\",\n",
    "    \"W\" : \"Extracellular structures\",\n",
    "    \"X\" : \"Mobilome: prophages, transposons\",\n",
    "    \"Y\" : \"Nuclear structure\",\n",
    "    \"Z\" : \"Cytoskeleton\",\n",
    "    \"R\" : \"General function prediction only\",\n",
    "    \"S\" : \"Function unknown\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b0251f64-8d43-4f21-9a88-772f44b71f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "def term_count_dict_from_annotation_df(annot_df, term_column):\n",
    "    \n",
    "    column = annot_df[term_column].values\n",
    "    \n",
    "    funct_terms = []\n",
    "    for entry in column:\n",
    "        if entry != '-':\n",
    "            if term_column == 'COG_category':\n",
    "                # terms = [f'{e}: {COG_dict[e]}' for e in entry]\n",
    "                terms = list(entry)\n",
    "            else:\n",
    "                terms = entry.split(',')\n",
    "            for t in terms:\n",
    "                funct_terms.append(t)\n",
    "\n",
    "#     len(terms)\n",
    "    \n",
    "    term_count_dict = {}\n",
    "    \n",
    "    for t in funct_terms:\n",
    "        count = term_count_dict.get(t, 0)\n",
    "        count += 1\n",
    "        term_count_dict[t] = count\n",
    "        \n",
    "    return term_count_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0fddca3d-22c7-42cf-b4d1-f663bfa84cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrichment_analysis(module, leiden_label_df, phases, background_annotation, term_column):\n",
    "    \n",
    "    module_ttids = leiden_label_df.loc[leiden_label_df[f'leiden_label_{phases}'] == module]['TTHERM_ID'].values\n",
    "    \n",
    "    module_annotation = background_annotation.loc[background_annotation['TTHERM_ID'].isin(module_ttids)]\n",
    "    \n",
    "    background_term_dict = term_count_dict_from_annotation_df(background_annotation, term_column)\n",
    "    module_term_dict = term_count_dict_from_annotation_df(module_annotation, term_column)\n",
    "    \n",
    "    bs = []\n",
    "    ps = []\n",
    "    folds = []\n",
    "    terms = []\n",
    "    \n",
    "    for t, module_count in module_term_dict.items():\n",
    "        \n",
    "        background_count = background_term_dict[t]\n",
    "        module_size = len(module_annotation)\n",
    "        background_size = len(background_annotation)\n",
    "        \n",
    "        standard_contingency_table = [\n",
    "                                [module_count, background_count - module_count], \n",
    "                                [module_size - module_count, background_size - module_size - (background_count - module_count)]\n",
    "                            ]\n",
    "        \n",
    "        # The -1 and +1 make this more conservative (see explanation from the DAVID database: \n",
    "        # https://david.ncifcrf.gov/helps/functional_annotation.html#geneenrich)\n",
    "        conservative_contingency_table = [\n",
    "                                [module_count - 1, background_count - module_count + 1], \n",
    "                                [module_size - module_count, background_size - module_size - (background_count - module_count)]\n",
    "                            ]\n",
    "        \n",
    "        \n",
    "        odds, p_standard = st.fisher_exact(standard_contingency_table, 'greater')\n",
    "        odds, p_conservative = st.fisher_exact(conservative_contingency_table, 'greater')\n",
    "        \n",
    "        p_reasonable = np.mean([p_standard, p_conservative])\n",
    "        \n",
    "        bonferroni  = p_reasonable * len(module_term_dict)\n",
    "\n",
    "        fold_enrichment = (module_count/module_size) / (background_count/background_size)\n",
    "\n",
    "        if bonferroni <= 0.05:\n",
    "            \n",
    "            ps.append(p_reasonable)\n",
    "            bs.append(bonferroni)\n",
    "            folds.append(fold_enrichment)\n",
    "            terms.append(t)\n",
    "            \n",
    "#         else:\n",
    "#             ps.append('')\n",
    "#             bs.append('')\n",
    "#             folds.append('')\n",
    "#             terms.append('')\n",
    "            \n",
    "    return ps, bs, folds, terms\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6482b3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_GO_info(go_term):\n",
    "    \n",
    "    name = go_df['GOs_description'].loc[go_df['GOs'] == go_term].values[0]\n",
    "    \n",
    "    definition = go_df['GOs_definition'].loc[go_df['GOs'] == go_term].values[0]\n",
    "    \n",
    "    obsolete = go_df['GOs_obsolete'].loc[go_df['GOs'] == go_term].values[0]\n",
    "    \n",
    "    return name, definition, obsolete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c9dad6d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('mitochondrion inheritance',\n",
       " 'The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.',\n",
       " 'False')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_GO_info('GO:0000001')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e47679cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_KEGG_info(term):\n",
    "    return kegg_df['KEGG_ko_description'].loc[kegg_df['KEGG_ko'] == term].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0f53543b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'alcohol dehydrogenase (NADP+)'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_KEGG_info('ko:K00002')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "22b083ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_EC_info(term):\n",
    "    return ec_df['EC_description'].loc[ec_df['EC'] == term].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "39579c7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'15-hydroxyprostaglandin dehydrogenase (NAD(+))'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_EC_info('1.1.1.141')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "14cb6323-a773-46b7-85e3-f9d5bd609e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_enrichment(lldf, phases, out_prefix, term_column):\n",
    "    \n",
    "    with open(f'../enrichment/{out_prefix}_enrichment.txt', 'w') as fl:\n",
    "    \n",
    "        for m in tqdm.tqdm(sorted(lldf[f'leiden_label_{phases}'].unique())):\n",
    "\n",
    "            # print(f'Module {m}')\n",
    "            ps, bs, folds, terms = enrichment_analysis(m, lldf, phases, complete_annot, term_column)\n",
    "\n",
    "            fl.write(f'Module {m}\\n')\n",
    "            for b, fold, t in zip(bs, folds, terms):\n",
    "                \n",
    "                if term_column == 'COG_category':\n",
    "                    \n",
    "                    definition = COG_dict[t]\n",
    "                    t = f'{t}: {definition}'\n",
    "                \n",
    "                elif term_column == 'GOs':\n",
    "                    name, definition, obsolete = get_GO_info(t)\n",
    "                    if obsolete:\n",
    "                        t = f'{name.capitalize()} ({t}): {definition} (obsolete)'\n",
    "                    else:\n",
    "                        t = f'{name.capitalize()} ({t}): {definition}'\n",
    "                        \n",
    "                elif term_column.split('_')[0] == 'KEGG':\n",
    "                    name = get_KEGG_info(t)\n",
    "                    \n",
    "                    t = f'{t}: {name}'\n",
    "                    \n",
    "                elif term_column == 'EC':\n",
    "                    \n",
    "                    name = get_EC_info(t)\n",
    "                    t = f'{t}: {name}'\n",
    "                    \n",
    "                        \n",
    "                # print(f\"{t}\\nFold-enrichment: {fold:.02f}\\nBonferroni-corrected p-value: {b:.02e}\\n\\n\")\n",
    "\n",
    "                fl.write(f'{t}\\nFold-enrichment: {fold:.02f}\\nBonferroni-corrected p-value: {b:.02e}\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dd5d7697-d044-45e7-a4d4-5945fa23775b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enrichment_df(lldf, phases, background_annotation, term_columns=['COG_category', 'GOs', 'KEGG_ko', 'EC'], outfile=None):\n",
    "    \n",
    "    module_dfs = []\n",
    "    \n",
    "    for m in tqdm.tqdm(sorted(lldf[f'leiden_label_{phases}'].unique())):\n",
    "        \n",
    "        term_dfs = []\n",
    "\n",
    "        for tc in term_columns:\n",
    "        \n",
    "            ps, bs, folds, terms = enrichment_analysis(m, lldf, phases, background_annotation, tc)\n",
    "\n",
    "            # fl.write(f'Module {m}\\n')\n",
    "            \n",
    "            info = []\n",
    "\n",
    "            if tc == 'GOs':\n",
    "\n",
    "                for t in terms:\n",
    "                    name, definition, obsolete = get_GO_info(t)\n",
    "                    if obsolete:\n",
    "                        info.append(f'{name.capitalize()}: {definition} (obsolete)')\n",
    "                    else:\n",
    "                        info.append(f'{name.capitalize()}: {definition}')\n",
    "                        \n",
    "            elif tc == 'COG_category':\n",
    "                for t in terms:\n",
    "                    info.append(COG_dict[t])\n",
    "                                \n",
    "            elif tc == 'KEGG_ko':\n",
    "                for t in terms:\n",
    "                    info.append(get_KEGG_info(t))\n",
    "                    \n",
    "            elif tc == 'EC':\n",
    "                for t in terms:\n",
    "                    info.append(get_EC_info(t))\n",
    "                    \n",
    "            term_df = pd.DataFrame({'module': [m]*len(terms),\n",
    "                                    'term': terms,\n",
    "                                    'info': info,\n",
    "                                    'fold_change': folds,\n",
    "                                    'bonferroni': bs})\n",
    "            \n",
    "            term_dfs.append(term_df)\n",
    "            \n",
    "        module_df = pd.concat(term_dfs)\n",
    "        \n",
    "        module_dfs.append(module_df)\n",
    "        \n",
    "    all_enrichment_df = pd.concat(module_dfs)\n",
    "    \n",
    "    if outfile:\n",
    "        all_enrichment_df.to_csv(outfile, index=False)\n",
    "    \n",
    "    return all_enrichment_df\n",
    "        \n",
    "        \n",
    "                        \n",
    "            \n",
    "                    \n",
    "            \n",
    "        \n",
    "#         for b, fold, t in zip(bs, folds, terms):\n",
    "\n",
    "#             if term_column == 'GOs':\n",
    "#                 name, definition, obsolete = get_GO_info(t)\n",
    "#                 if obsolete:\n",
    "#                     t = f'{name.capitalize()} ({t}): {definition} (obsolete)'\n",
    "#                 else:\n",
    "#                     t = f'{name.capitalize()} ({t}): {definition}'\n",
    "\n",
    "#             if term_column.split('_')[0] == 'KEGG':\n",
    "#                 dfs = pd.read_html(f'https://www.genome.jp/entry/{t}')\n",
    "\n",
    "#                 data = dfs[4]\n",
    "#                 entry = data.loc[data[0] == 'Entry'][1].values[0]\n",
    "#                 name = data.loc[data[0] == 'Name'][1].values[0]\n",
    "\n",
    "#                 t = f'{t}: {name}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3e001ab1-173b-4ed8-85d5-091142480820",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annotation = pd.read_csv('../eggnog/complete_eggnog_annotation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7f839c0b-7ed1-4a84-b058-dafccbfdad88",
   "metadata": {},
   "outputs": [],
   "source": [
    "lldf_nn3 = pd.read_csv('../embedding/test_nn3_leiden_label_df_round_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c4b49a02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "24\n",
      "10\n",
      "24\n",
      "10\n",
      "10\n",
      "24\n",
      "10\n",
      "10\n",
      "24\n",
      "24\n",
      "24\n",
      "10\n",
      "10\n",
      "10\n",
      "10\n",
      "16\n",
      "24\n",
      "10\n",
      "24\n",
      "24\n",
      "10\n",
      "10\n",
      "10\n",
      "16\n",
      "10\n",
      "40\n",
      "24\n",
      "5\n",
      "24\n",
      "24\n",
      "24\n",
      "41\n",
      "23\n",
      "24\n",
      "24\n",
      "18\n",
      "10\n",
      "24\n",
      "24\n"
     ]
    }
   ],
   "source": [
    "gene_list_test = ['TTHERM_000013409', 'TTHERM_01321550', 'TTHERM_00011710', 'TTHERM_00321680', 'TTHERM_00355700', 'TTHERM_00938950', 'TTHERM_01372820', 'TTHERM_00013410', 'TTHERM_00390080', 'TTHERM_00516380', 'TTHERM_00038880', 'TTHERM_00059370', 'TTHERM_00473020', 'TTHERM_00497590', 'TTHERM_00558350', 'TTHERM_00052190', 'TTHERM_00392790', 'TTHERM_00410180', 'TTHERM_00685980', 'TTHERM_00445920', 'TTHERM_00471040', 'TTHERM_00140780', 'TTHERM_00145480', 'TTHERM_00321720', 'TTHERM_00628650', 'TTHERM_00526730', 'TTHERM_01156770', 'TTHERM_00312200', 'TTHERM_01332070', 'TTHERM_00318900', 'TTHERM_00340180', 'TTHERM_00592740', 'TTHERM_00440600', 'TTHERM_01321570', 'TTHERM_00537380', 'TTHERM_00585170', 'TTHERM_01197130', 'TTHERM_00554390', 'TTHERM_00649180', 'TTHERM_00691410', ]\n",
    "for gene in gene_list_test:\n",
    "    print((lldf_nn3.loc[lldf_nn3['TTHERM_ID'] == gene])['leiden_label_rna_seq'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0a18a134-a62a-4166-9764-a05d90f45361",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33,\n",
       "       32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,\n",
       "       15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lldf_nn3['leiden_label_rna_seq'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "93dc3cc2-1fc1-4018-80cb-21f998c0ed74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fb55c9e2-7d38-44d2-a944-51bf9519ab39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/50 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 50/50 [00:24<00:00,  2.03it/s]\n"
     ]
    }
   ],
   "source": [
    "enrich_full_nn3 = get_enrichment_df(lldf_nn3, 'rna_seq', complete_annotation, outfile='./test_nn3_full_enrichment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2c4fefa0-73c9-422e-8090-61fa18d1faf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrich_full_nn3 = pd.read_csv('./test_nn3_full_enrichment.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6b2f9dee-8c14-40eb-a3b8-bc3865d6f367",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>term</th>\n",
       "      <th>info</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>bonferroni</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>Z</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>4.247240</td>\n",
       "      <td>1.335030e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>T</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>1.395865</td>\n",
       "      <td>2.957352e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>S</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>1.234319</td>\n",
       "      <td>2.749970e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>D</td>\n",
       "      <td>Cell cycle control, cell division, chromosome ...</td>\n",
       "      <td>3.395502</td>\n",
       "      <td>4.816431e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0005856</td>\n",
       "      <td>Cytoskeleton: A cellular structure that forms ...</td>\n",
       "      <td>3.945512</td>\n",
       "      <td>2.432438e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0005929</td>\n",
       "      <td>Cilium: A specialized eukaryotic organelle tha...</td>\n",
       "      <td>5.765141</td>\n",
       "      <td>2.840020e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0006996</td>\n",
       "      <td>Organelle organization: A process that is carr...</td>\n",
       "      <td>2.452899</td>\n",
       "      <td>3.167779e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0016043</td>\n",
       "      <td>Cellular component organization: A process tha...</td>\n",
       "      <td>2.048829</td>\n",
       "      <td>6.066976e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0044430</td>\n",
       "      <td>Obsolete cytoskeletal part: OBSOLETE. Any cons...</td>\n",
       "      <td>4.027088</td>\n",
       "      <td>1.245529e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0071840</td>\n",
       "      <td>Cellular component organization or biogenesis:...</td>\n",
       "      <td>2.014741</td>\n",
       "      <td>7.663285e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0000910</td>\n",
       "      <td>Cytokinesis: The division of the cytoplasm and...</td>\n",
       "      <td>5.426702</td>\n",
       "      <td>7.692155e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007049</td>\n",
       "      <td>Cell cycle: The progression of biochemical and...</td>\n",
       "      <td>3.248388</td>\n",
       "      <td>2.692370e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0022402</td>\n",
       "      <td>Cell cycle process: The cellular process that ...</td>\n",
       "      <td>3.182325</td>\n",
       "      <td>2.032114e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0051301</td>\n",
       "      <td>Cell division: The process resulting in divisi...</td>\n",
       "      <td>4.083459</td>\n",
       "      <td>2.607360e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0000226</td>\n",
       "      <td>Microtubule cytoskeleton organization: A proce...</td>\n",
       "      <td>7.014104</td>\n",
       "      <td>1.036233e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0000278</td>\n",
       "      <td>Mitotic cell cycle: Progression through the ph...</td>\n",
       "      <td>3.360535</td>\n",
       "      <td>4.180151e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007010</td>\n",
       "      <td>Cytoskeleton organization: A process that is c...</td>\n",
       "      <td>5.575967</td>\n",
       "      <td>3.800951e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007017</td>\n",
       "      <td>Microtubule-based process: Any cellular proces...</td>\n",
       "      <td>5.458623</td>\n",
       "      <td>4.812967e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007051</td>\n",
       "      <td>Spindle organization: A process that is carrie...</td>\n",
       "      <td>6.270855</td>\n",
       "      <td>1.085222e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007052</td>\n",
       "      <td>Mitotic spindle organization: A process that i...</td>\n",
       "      <td>7.020074</td>\n",
       "      <td>9.977586e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0015630</td>\n",
       "      <td>Microtubule cytoskeleton: The part of the cyto...</td>\n",
       "      <td>4.230044</td>\n",
       "      <td>3.828275e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0030496</td>\n",
       "      <td>Midbody: A thin cytoplasmic bridge formed betw...</td>\n",
       "      <td>6.546497</td>\n",
       "      <td>2.245185e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:1902850</td>\n",
       "      <td>Microtubule cytoskeleton organization involved...</td>\n",
       "      <td>6.873822</td>\n",
       "      <td>1.255564e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:1903047</td>\n",
       "      <td>Mitotic cell cycle process: A process that is ...</td>\n",
       "      <td>3.306395</td>\n",
       "      <td>2.956888e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0061640</td>\n",
       "      <td>Cytoskeleton-dependent cytokinesis: A cytokine...</td>\n",
       "      <td>5.816311</td>\n",
       "      <td>2.145346e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0006928</td>\n",
       "      <td>Obsolete movement of cell or subcellular compo...</td>\n",
       "      <td>4.056682</td>\n",
       "      <td>3.140070e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0008017</td>\n",
       "      <td>Microtubule binding: Binding to a microtubule,...</td>\n",
       "      <td>9.819746</td>\n",
       "      <td>5.241274e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0008092</td>\n",
       "      <td>Cytoskeletal protein binding: Binding to a pro...</td>\n",
       "      <td>4.924529</td>\n",
       "      <td>4.003245e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0015631</td>\n",
       "      <td>Tubulin binding: Binding to monomeric or multi...</td>\n",
       "      <td>7.755081</td>\n",
       "      <td>1.272309e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0005819</td>\n",
       "      <td>Spindle: The array of microtubules and associa...</td>\n",
       "      <td>6.270855</td>\n",
       "      <td>1.085222e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0003774</td>\n",
       "      <td>Cytoskeletal motor activity: Generation of for...</td>\n",
       "      <td>12.497858</td>\n",
       "      <td>4.499222e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0003777</td>\n",
       "      <td>Microtubule motor activity: A motor activity t...</td>\n",
       "      <td>12.372880</td>\n",
       "      <td>3.286359e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0005871</td>\n",
       "      <td>Kinesin complex: Any complex that includes a d...</td>\n",
       "      <td>13.024084</td>\n",
       "      <td>2.028758e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0005875</td>\n",
       "      <td>Microtubule associated complex: Any multimeric...</td>\n",
       "      <td>9.165096</td>\n",
       "      <td>4.744623e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007018</td>\n",
       "      <td>Microtubule-based movement: A microtubule-base...</td>\n",
       "      <td>5.592262</td>\n",
       "      <td>1.112432e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0008574</td>\n",
       "      <td>Plus-end-directed microtubule motor activity: ...</td>\n",
       "      <td>14.664154</td>\n",
       "      <td>5.256285e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:1990939</td>\n",
       "      <td>Microtubule motor activity: A motor activity t...</td>\n",
       "      <td>14.664154</td>\n",
       "      <td>5.256285e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0005818</td>\n",
       "      <td>Aster: An array of microtubules emanating from...</td>\n",
       "      <td>10.692612</td>\n",
       "      <td>3.819943e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0007110</td>\n",
       "      <td>Meiosis i cytokinesis: A cell cycle process th...</td>\n",
       "      <td>16.497173</td>\n",
       "      <td>1.738208e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>0.0</td>\n",
       "      <td>GO:0018200</td>\n",
       "      <td>Peptidyl-glutamic acid modification: The modif...</td>\n",
       "      <td>11.576963</td>\n",
       "      <td>3.687373e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    module        term                                               info  \\\n",
       "0      0.0           Z                                       Cytoskeleton   \n",
       "1      0.0           T                     Signal transduction mechanisms   \n",
       "2      0.0           S                                   Function unknown   \n",
       "3      0.0           D  Cell cycle control, cell division, chromosome ...   \n",
       "4      0.0  GO:0005856  Cytoskeleton: A cellular structure that forms ...   \n",
       "5      0.0  GO:0005929  Cilium: A specialized eukaryotic organelle tha...   \n",
       "6      0.0  GO:0006996  Organelle organization: A process that is carr...   \n",
       "7      0.0  GO:0016043  Cellular component organization: A process tha...   \n",
       "8      0.0  GO:0044430  Obsolete cytoskeletal part: OBSOLETE. Any cons...   \n",
       "9      0.0  GO:0071840  Cellular component organization or biogenesis:...   \n",
       "10     0.0  GO:0000910  Cytokinesis: The division of the cytoplasm and...   \n",
       "11     0.0  GO:0007049  Cell cycle: The progression of biochemical and...   \n",
       "12     0.0  GO:0022402  Cell cycle process: The cellular process that ...   \n",
       "13     0.0  GO:0051301  Cell division: The process resulting in divisi...   \n",
       "14     0.0  GO:0000226  Microtubule cytoskeleton organization: A proce...   \n",
       "15     0.0  GO:0000278  Mitotic cell cycle: Progression through the ph...   \n",
       "16     0.0  GO:0007010  Cytoskeleton organization: A process that is c...   \n",
       "17     0.0  GO:0007017  Microtubule-based process: Any cellular proces...   \n",
       "18     0.0  GO:0007051  Spindle organization: A process that is carrie...   \n",
       "19     0.0  GO:0007052  Mitotic spindle organization: A process that i...   \n",
       "20     0.0  GO:0015630  Microtubule cytoskeleton: The part of the cyto...   \n",
       "21     0.0  GO:0030496  Midbody: A thin cytoplasmic bridge formed betw...   \n",
       "22     0.0  GO:1902850  Microtubule cytoskeleton organization involved...   \n",
       "23     0.0  GO:1903047  Mitotic cell cycle process: A process that is ...   \n",
       "24     0.0  GO:0061640  Cytoskeleton-dependent cytokinesis: A cytokine...   \n",
       "25     0.0  GO:0006928  Obsolete movement of cell or subcellular compo...   \n",
       "26     0.0  GO:0008017  Microtubule binding: Binding to a microtubule,...   \n",
       "27     0.0  GO:0008092  Cytoskeletal protein binding: Binding to a pro...   \n",
       "28     0.0  GO:0015631  Tubulin binding: Binding to monomeric or multi...   \n",
       "29     0.0  GO:0005819  Spindle: The array of microtubules and associa...   \n",
       "30     0.0  GO:0003774  Cytoskeletal motor activity: Generation of for...   \n",
       "31     0.0  GO:0003777  Microtubule motor activity: A motor activity t...   \n",
       "32     0.0  GO:0005871  Kinesin complex: Any complex that includes a d...   \n",
       "33     0.0  GO:0005875  Microtubule associated complex: Any multimeric...   \n",
       "34     0.0  GO:0007018  Microtubule-based movement: A microtubule-base...   \n",
       "35     0.0  GO:0008574  Plus-end-directed microtubule motor activity: ...   \n",
       "36     0.0  GO:1990939  Microtubule motor activity: A motor activity t...   \n",
       "37     0.0  GO:0005818  Aster: An array of microtubules emanating from...   \n",
       "38     0.0  GO:0007110  Meiosis i cytokinesis: A cell cycle process th...   \n",
       "39     0.0  GO:0018200  Peptidyl-glutamic acid modification: The modif...   \n",
       "\n",
       "    fold_change    bonferroni  \n",
       "0      4.247240  1.335030e-18  \n",
       "1      1.395865  2.957352e-02  \n",
       "2      1.234319  2.749970e-02  \n",
       "3      3.395502  4.816431e-10  \n",
       "4      3.945512  2.432438e-07  \n",
       "5      5.765141  2.840020e-03  \n",
       "6      2.452899  3.167779e-04  \n",
       "7      2.048829  6.066976e-03  \n",
       "8      4.027088  1.245529e-06  \n",
       "9      2.014741  7.663285e-03  \n",
       "10     5.426702  7.692155e-04  \n",
       "11     3.248388  2.692370e-04  \n",
       "12     3.182325  2.032114e-03  \n",
       "13     4.083459  2.607360e-02  \n",
       "14     7.014104  1.036233e-10  \n",
       "15     3.360535  4.180151e-03  \n",
       "16     5.575967  3.800951e-10  \n",
       "17     5.458623  4.812967e-09  \n",
       "18     6.270855  1.085222e-03  \n",
       "19     7.020074  9.977586e-04  \n",
       "20     4.230044  3.828275e-06  \n",
       "21     6.546497  2.245185e-02  \n",
       "22     6.873822  1.255564e-03  \n",
       "23     3.306395  2.956888e-02  \n",
       "24     5.816311  2.145346e-02  \n",
       "25     4.056682  3.140070e-03  \n",
       "26     9.819746  5.241274e-04  \n",
       "27     4.924529  4.003245e-02  \n",
       "28     7.755081  1.272309e-03  \n",
       "29     6.270855  1.085222e-03  \n",
       "30    12.497858  4.499222e-05  \n",
       "31    12.372880  3.286359e-04  \n",
       "32    13.024084  2.028758e-04  \n",
       "33     9.165096  4.744623e-03  \n",
       "34     5.592262  1.112432e-02  \n",
       "35    14.664154  5.256285e-04  \n",
       "36    14.664154  5.256285e-04  \n",
       "37    10.692612  3.819943e-02  \n",
       "38    16.497173  1.738208e-02  \n",
       "39    11.576963  3.687373e-03  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrich_full_nn3.head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "10c2768f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Z\n",
      "T\n",
      "S\n",
      "D\n",
      "Q\n",
      "P\n",
      "G\n",
      "J\n",
      "K\n",
      "L\n",
      "I\n",
      "C\n",
      "E\n",
      "B\n",
      "O\n",
      "U\n",
      "A\n",
      "Y\n"
     ]
    }
   ],
   "source": [
    "for term in enrich_full_nn3['term'].unique():\n",
    "    if 'ko:' not in term and 'GO:' not in term and '.' not in term:\n",
    "        print(term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0dfb99e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48.0\n",
      "0.0\n",
      "49\n",
      "48\n",
      "\n",
      "{1, 34, 35, 4, 5, 38, 40, 41, 10, 11, 14, 46, 20, 25, 26}\n",
      "34\n",
      "34\n",
      "15\n"
     ]
    }
   ],
   "source": [
    "print(max(enrich_full_nn3['module'].values))\n",
    "print(min(enrich_full_nn3['module'].values))\n",
    "\n",
    "not_enriched_mod_list = []\n",
    "enriched_mod_list = []\n",
    "\n",
    "for num in range(int(max(enrich_full_nn3['module'].values)) + 1):\n",
    "    if num not in enrich_full_nn3['module'].values:\n",
    "        not_enriched_mod_list.append(num)\n",
    "    if num in enrich_full_nn3['module'].values:\n",
    "        enriched_mod_list.append(num)\n",
    "\n",
    "print(len(not_enriched_mod_list) + len(enriched_mod_list))\n",
    "print(max(not_enriched_mod_list + enriched_mod_list))\n",
    "print()\n",
    "\n",
    "print(set(range(int(max(enrich_full_nn3['module'].values)) + 1)) - set(enrich_full_nn3['module'].values))\n",
    "print(len(set(enrich_full_nn3['module'].values)))\n",
    "print(len(enriched_mod_list))\n",
    "print(len(not_enriched_mod_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "323ec30f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 4, 5, 10, 11, 14, 20, 25, 26, 34, 35, 38, 40, 41, 46]\n"
     ]
    }
   ],
   "source": [
    "print(not_enriched_mod_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b4b518af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>term</th>\n",
       "      <th>info</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>bonferroni</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [module, term, info, fold_change, bonferroni]\n",
       "Index: []"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrich_full_nn3.loc[enrich_full_nn3['module']==72]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "801013e7-b277-4b7f-a6b7-773fc734c1b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.,  2.,  3.,  6.,  7.,  8.,  9., 12., 13., 15., 16., 17., 18.,\n",
       "       19., 21., 22., 23., 24., 27., 28., 29., 30., 31., 32., 33., 36.,\n",
       "       37., 39., 42., 43., 44., 45., 47., 48.])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enrich_full_nn3['module'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1228547a-b5d6-4d2f-9534-5a5a10043c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33,\n",
       "       32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16,\n",
       "       15, 14, 13, 12, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1,  0])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lldf_nn3['leiden_label_rna_seq'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1a6ff685-b372-4979-969e-2ae278aa46cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_rna_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5566</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            TTHERM_ID  leiden_label_rna_seq\n",
       "5566  TTHERM_01386050                    27"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lldf_nn3.loc[lldf_nn3['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5230794a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "24\n",
      "10\n",
      "24\n",
      "10\n",
      "10\n",
      "24\n",
      "10\n",
      "10\n",
      "24\n",
      "24\n",
      "24\n",
      "10\n",
      "10\n",
      "10\n",
      "10\n",
      "16\n",
      "24\n",
      "10\n",
      "24\n",
      "24\n",
      "10\n",
      "10\n",
      "10\n",
      "16\n",
      "10\n",
      "40\n",
      "24\n",
      "5\n",
      "24\n",
      "24\n",
      "24\n",
      "41\n",
      "23\n",
      "24\n",
      "24\n",
      "18\n",
      "10\n",
      "24\n",
      "24\n"
     ]
    }
   ],
   "source": [
    "gene_list_test = ['TTHERM_000013409', 'TTHERM_01321550', 'TTHERM_00011710', 'TTHERM_00321680', 'TTHERM_00355700', 'TTHERM_00938950', 'TTHERM_01372820', 'TTHERM_00013410', 'TTHERM_00390080', 'TTHERM_00516380', 'TTHERM_00038880', 'TTHERM_00059370', 'TTHERM_00473020', 'TTHERM_00497590', 'TTHERM_00558350', 'TTHERM_00052190', 'TTHERM_00392790', 'TTHERM_00410180', 'TTHERM_00685980', 'TTHERM_00445920', 'TTHERM_00471040', 'TTHERM_00140780', 'TTHERM_00145480', 'TTHERM_00321720', 'TTHERM_00628650', 'TTHERM_00526730', 'TTHERM_01156770', 'TTHERM_00312200', 'TTHERM_01332070', 'TTHERM_00318900', 'TTHERM_00340180', 'TTHERM_00592740', 'TTHERM_00440600', 'TTHERM_01321570', 'TTHERM_00537380', 'TTHERM_00585170', 'TTHERM_01197130', 'TTHERM_00554390', 'TTHERM_00649180', 'TTHERM_00691410', ]\n",
    "for gene in gene_list_test:\n",
    "    print((lldf_nn3.loc[lldf_nn3['TTHERM_ID'] == gene])['leiden_label_rna_seq'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3e1672f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(enrich_full_nn3['module'].values))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d35056e7-38f7-4cf6-a5df-f387d19e8d3a",
   "metadata": {},
   "source": [
    "end here. Return to embedding notebook!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
