{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e305761-a971-4232-8a21-5907cfeedfad",
   "metadata": {},
   "source": [
    "## Import all packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "id": "5d9163ea-8dc4-46e8-9542-ba27cc741b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pdb\n",
    "\n",
    "import bokeh\n",
    "from bokeh.plotting import show as show_interactive\n",
    "from bokeh.plotting import output_file, output_notebook\n",
    "from bokeh.layouts import column, row\n",
    "from bokeh.models import CustomJS, TextInput, LassoSelectTool, Select, MultiSelect, ColorBar, Legend, LegendItem\n",
    "from bokeh.models.widgets import DataTable, DateFormatter, TableColumn, Button, HTMLTemplateFormatter\n",
    "from bokeh.events import SelectionGeometry\n",
    "from bokeh.transform import linear_cmap, jitter\n",
    "from matplotlib.pyplot import show as show_static\n",
    "# from clustergrammer2 import net, Network, CGM2\n",
    "\n",
    "import igraph as ig\n",
    "import leidenalg as la\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import scipy.stats as st\n",
    "import scipy.spatial\n",
    "import scipy.cluster.hierarchy\n",
    "\n",
    "import glob\n",
    "import json\n",
    "import re\n",
    "import copy\n",
    "\n",
    "import requests\n",
    "import bs4\n",
    "import tqdm\n",
    "import os\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "import umap\n",
    "import pymde\n",
    "\n",
    "import torch\n",
    "\n",
    "# bokeh.io.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "id": "20cf4cee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20326"
      ]
     },
     "execution_count": 449,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/allgood_filt_agg_tidy_2021aligned_qc_rma_expression_full.csv')\n",
    "num_genes = full_filtered_df.shape[0]\n",
    "num_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7ace6be-7982-4461-bc6e-f49a60b1e56c",
   "metadata": {},
   "source": [
    "## Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "id": "674c6371-69ed-44cd-891b-f9bdcb1ee41b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_geom_mean_expression(expression_df):\n",
    "    \"\"\"\n",
    "    \n",
    "    Function to take an expression dataframe from the microarrays and collapse it into the means of\n",
    "    all replicate chips.\n",
    "    \"\"\"\n",
    "    # C2 and S12 got removed during quality control\n",
    "    x = [\n",
    "        'Ll', \n",
    "        'Lm', \n",
    "        'Lh', \n",
    "        'S0', \n",
    "        'S3', \n",
    "        'S6', \n",
    "        'S9', \n",
    "        # 'S12', \n",
    "        'S15', \n",
    "        'S24', \n",
    "        'C0', \n",
    "        # 'C2', \n",
    "        'C4', \n",
    "        'C6', \n",
    "        'C8', \n",
    "        'C10', \n",
    "        'C12', \n",
    "        'C14', \n",
    "        'C16', \n",
    "        'C18']\n",
    "    \n",
    "    # cols = expression_df.columns[1:]\n",
    "    # x = [c for c in x if c in cols]\n",
    "    \n",
    "    condition_expr_dict = {c.split(\"_\")[0]: [] for c in expression_df.columns[1:]}\n",
    "    \n",
    "    for c in list(expression_df.columns)[1:]:\n",
    "        \n",
    "        cond = c.split('_')[0]\n",
    "        if cond in condition_expr_dict.keys():\n",
    "            expr_list = condition_expr_dict.get(cond, [])\n",
    "\n",
    "            # Need to avoid true zeros\n",
    "            expr_list.append(expression_df[c].values)\n",
    "            condition_expr_dict[cond] = expr_list\n",
    "        \n",
    "    condition_mean_dict = {c: (st.mstats.gmean(np.array(condition_expr_dict[c]) + 1, 0) - 1) for c in condition_expr_dict.keys() if c in x}\n",
    "    \n",
    "    mean_expr_df = pd.DataFrame(condition_mean_dict)\n",
    "    mean_expr_df['TTHERM_ID'] = expression_df['TTHERM_ID'].values\n",
    "    cols = list(mean_expr_df.columns)\n",
    "    reorder = cols[-1:] + cols[:-1]\n",
    "    mean_expr_df = mean_expr_df[reorder]\n",
    "    \n",
    "    return mean_expr_df\n",
    "\n",
    "def normalizer(array):\n",
    "    \"\"\"\n",
    "    Normalizes the values of an array to range from zero to one\n",
    "    \"\"\"\n",
    "    \n",
    "    a = np.array(array)\n",
    "    \n",
    "    normalized = (array - np.min(array)) / (np.max(array) - np.min(array))\n",
    "    \n",
    "    return normalized\n",
    "\n",
    "def normalize_expression_per_gene(expression_df):\n",
    "    \"\"\"\n",
    "    Function to normalize all gene expression to range from zero to one.\n",
    "    \"\"\"\n",
    "    if 'TTHERM_ID' in expression_df.columns:\n",
    "        ttids = expression_df['TTHERM_ID'].values\n",
    "        data = expression_df[list(expression_df.columns)[1:]]\n",
    "        \n",
    "        norm_expression_df = data.apply(lambda row: normalizer(row), axis=1)\n",
    "        norm_expression_df['TTHERM_ID'] = ttids\n",
    "        \n",
    "        columns = norm_expression_df.columns.tolist()\n",
    "        \n",
    "        rearrangment = columns[-1:] + columns[:-1]\n",
    "        \n",
    "        norm_expression_df = norm_expression_df[rearrangment]\n",
    "        \n",
    "    else:\n",
    "        norm_expression_df = expression_df.apply(lambda row: normalizer(row), axis=1)\n",
    "    \n",
    "    return norm_expression_df\n",
    "    \n",
    "\n",
    "\n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in order of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', leiden_type='modularity', la_res_param=1.0, return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "    if leiden_type == 'modularity':\n",
    "        partition = la.find_partition(g, la.ModularityVertexPartition, seed=random_state, weights='weight')\n",
    "    elif leiden_type == 'cpm':\n",
    "        partition = la.find_partition(g, la.CPMVertexPartition, resolution_parameter = la_res_param, seed=random_state, weights='weight')\n",
    "    else:\n",
    "        raise ValueError('Invalid value for leiden_type parameter')\n",
    "    leiden_modules = np.array(partition.membership)\n",
    "    \n",
    "    return leiden_modules, dists\n",
    "\n",
    "def build_leiden_label_df(data_df, phases, random_state=42, n_neighbors=3, metric='manhattan', leiden_type='modularity', la_res_param=1.0, lldf=None):\n",
    "    \"\"\"\n",
    "    Function to build a dataframe of genes labeled according to their UMAP/Leiden modules\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    data_df : pandas DataFrame\n",
    "        The expression data\n",
    "    phases : str ('full', 'veg', or 'sex')\n",
    "        The physiological phases for which expression data is being provided\n",
    "    lldf : pandas DataFrame (default None)\n",
    "        Another leiden label df (lldf) to which to add a column\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    lldf : pandas DataFrame\n",
    "        Leiden Label DataFrame. Gene IDs and their corresponding UMAP/Leiden module\n",
    "        computed for a specific physiological regime (full set (full), vegetative only\n",
    "        (veg), or sexual only (sex))\n",
    "    \"\"\"\n",
    "    \n",
    "    if type(lldf) == type(None):\n",
    "        lldf = pd.DataFrame.from_dict({'TTHERM_ID': []})\n",
    "    \n",
    "    leiden_modules, dists = run_leiden(data_df, random_state=random_state, n_neighbors=n_neighbors, metric=metric, leiden_type=leiden_type, la_res_param=la_res_param)\n",
    "    lldf['TTHERM_ID'] = data_df['TTHERM_ID'].values\n",
    "    \n",
    "    lldf[f'leiden_label_{phases}'] = leiden_modules\n",
    "    \n",
    "    return lldf, dists\n",
    "\n",
    "\n",
    "# The two functions below are taken and adapted from the UMAP package\n",
    "def _get_umap_embedding(umap_object):\n",
    "    if hasattr(umap_object, \"embedding_\"):\n",
    "        return umap_object.embedding_\n",
    "    elif hasattr(umap_object, \"embedding\"):\n",
    "        return umap_object.embedding\n",
    "    else:\n",
    "        raise ValueError(\"Could not find embedding attribute of umap_object\")\n",
    "        \n",
    "def plot_enrichment(enrich_column_data_source):\n",
    "    \n",
    "    # pdb.set_trace()\n",
    "    \n",
    "    # y_range = FactorRange(factors=[str(y) for y in enrich_df['module'].unique()])\n",
    "    \n",
    "    # grouped = enrich_df.groupby('module')\n",
    "    \n",
    "    hover = [\n",
    "        ('module', '@module'),\n",
    "        ('term', '@term'),\n",
    "        ('info', '@info'),\n",
    "        ('fold-change', '@fold_change'),\n",
    "        ('bonferroni', '@bonferroni')\n",
    "    ]\n",
    "    \n",
    "    p = bokeh.plotting.figure(\n",
    "        height=1000,\n",
    "        width=400,\n",
    "        # y_range=y_range,\n",
    "        title='Functional term enrichment in modules',\n",
    "        x_axis_label='fold-change',\n",
    "        y_axis_label='module',\n",
    "        x_axis_type='log',\n",
    "        tooltips=hover,\n",
    "        # background_fill_color='black'\n",
    "    )\n",
    "    \n",
    "    # cds = bokeh.models.ColumnDataSource(enrich_df)\n",
    "    # print(enrich_df.head())\n",
    "    \n",
    "    p.circle(y=jitter('module', width=0.4), x='fold_change', source=enrich_column_data_source, alpha=0.3, size=7, color='color', line_color='black')\n",
    "    # p.xaxis.major_label_orientation = 45\n",
    "    p.ygrid.minor_grid_line_color = 'navy'\n",
    "    p.ygrid.minor_grid_line_alpha = 0.1\n",
    "    # p.xgrid.band_fill_alpha = 0.1\n",
    "    # p.xgrid.band_fill_color = \"navy\"\n",
    "    ticker = []\n",
    "    for m in enrich_column_data_source.data['module']:\n",
    "        if m not in ticker:\n",
    "            ticker.append(m)\n",
    "    p.yaxis.ticker = ticker\n",
    "    p.y_range.flipped = True\n",
    "    p.xaxis.major_label_text_font_size = '12pt'\n",
    "    p.yaxis.major_label_text_font_size = '12pt'\n",
    "    p.yaxis.axis_label_text_font_size = '12pt'\n",
    "    p.xaxis.axis_label_text_font_size = '12pt'\n",
    "    \n",
    "    return p\n",
    "\n",
    "def heatmap(column_data_source, ls_color_palette, r_low, r_high, x_axis_factors, y_axis_factors, s_z=\"normalized_expression\", index_name='TTHERM_ID', col_name='phase'):\n",
    "    # adapted from https://gitlab.com/biotransistor/bokehheat/-/blob/master/bokehheat/heat.py\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df_matrx: a dataframe in same xy orientation as the final heatmap.\n",
    "          the index should cary the y axis label.\n",
    "          the column should cary the x axis label.\n",
    "          the matrix as such should only cary the z axis values.\n",
    "\n",
    "        ls_color_palette: a list color strings to specify the color spectrum.\n",
    "            this variable is compatible with the ordinary bokeh palettes:\n",
    "            https://bokeh.pydata.org/en/latest/docs/reference/palettes.html\n",
    "\n",
    "        r_low: quantitative minimum value. the dataset can contain lower values,\n",
    "            but for color labeling they will be mapped to this minimum value.\n",
    "            e.g.: -8.\n",
    "\n",
    "        r_high: quantitative maximum value. the dataset can contain lower values,\n",
    "            but for color labeling they will be mapped to this maximum value.\n",
    "            e.g.: 8.\n",
    "\n",
    "        s_z: string. label that specifies what the values in the matrix actually\n",
    "            are. e.g.: 'gene expression [log2]'\n",
    "\n",
    "    output:\n",
    "        p: bokeh plot object.\n",
    "\n",
    "    description:\n",
    "        this function will return a bokeh based interactive heatmap plot.\n",
    "        the color are representing the z value.\n",
    "    \"\"\"\n",
    "    # index as string\n",
    "#     df_matrix.index = df_matrix.index.astype(str)\n",
    "#     df_matrix.columns = df_matrix.columns.astype(str)\n",
    "\n",
    "#     # handle y and x axis name\n",
    "#     if (df_matrix.index.name == None):\n",
    "#         df_matrix.index.name = \"y_axis\"\n",
    "#     if (df_matrix.columns.name == None):\n",
    "#         df_matrix.columns.name = \"x_axis\"\n",
    "    # pdb.set_trace()\n",
    "    s_y = index_name\n",
    "    \n",
    "    # df_matrix.columns.name = 'phase'\n",
    "    s_x = col_name\n",
    "    \n",
    "    \n",
    "    # print(df_matrix.head())\n",
    "    \n",
    "    # melt dataframe\n",
    "    # df_tidy = df_matrix.reset_index().melt(\n",
    "    #     id_vars=[df_matrix.index.name],\n",
    "    #     value_name=s_z\n",
    "    # )\n",
    "    # print(df_tidy.head())\n",
    "    # color declaration\n",
    "    d_zcolormapper = linear_cmap(\n",
    "        field_name=s_z,\n",
    "        palette=ls_color_palette,\n",
    "        low=r_low,\n",
    "        high=r_high\n",
    "    )\n",
    "    # tooltip declaration\n",
    "    lt_tooltip = [\n",
    "        (s_y, f\"@{s_y}\"),\n",
    "        (s_x, f\"@{s_x}\"),\n",
    "        (s_z, f\"@{s_z}\"),\n",
    "        ('module', f'@module')\n",
    "    ]\n",
    "    # generate figure\n",
    "    o_colorbar = ColorBar(color_mapper=d_zcolormapper['transform'])\n",
    "    p = bokeh.plotting.figure(\n",
    "        y_range=y_axis_factors,\n",
    "        x_range=x_axis_factors,\n",
    "        width=400,\n",
    "        height=1000,\n",
    "        tools = \"box_zoom,hover,pan,reset,wheel_zoom,save\",  # have to be set hardcoded\n",
    "        active_drag = \"box_zoom\",  # have to be set hardcoded\n",
    "        tooltips=lt_tooltip,\n",
    "        title=s_z,\n",
    "        toolbar_location='right',\n",
    "        \n",
    "    )\n",
    "    \n",
    "    p.rect(\n",
    "        source=column_data_source,\n",
    "        x=s_x,\n",
    "        y=s_y,\n",
    "        color=d_zcolormapper,\n",
    "        width=1,\n",
    "        height=1,\n",
    "        fill_alpha='fill_alpha',\n",
    "        line_alpha='line_alpha',\n",
    "        # line_color='white',\n",
    "        nonselection_fill_alpha=0.01,\n",
    "        nonselection_line_alpha=0.01,\n",
    "        # nonselection_line_color=\"white\"\n",
    "    )\n",
    "    p.add_layout(o_colorbar, place='left')\n",
    "    # p.yaxis.major_label_orientation = \"horizontal\"\n",
    "    p.xaxis.major_label_orientation = 45\n",
    "    # p.yaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.visible = False\n",
    "    p.xaxis.major_label_text_font_size = '12pt'\n",
    "\n",
    "    # out\n",
    "    return(p)\n",
    "        \n",
    "def interactive(\n",
    "    embedding_df,\n",
    "    x,\n",
    "    # mean_expression_df,\n",
    "    title=None,\n",
    "    labels=None,\n",
    "    values=None,\n",
    "    hover_data=None,\n",
    "    theme=None,\n",
    "    cmap=\"Blues\",\n",
    "    color_key=None,\n",
    "    color_key_cmap=\"Spectral\",\n",
    "    background=\"white\",\n",
    "#     width=800,\n",
    "#     height=800,\n",
    "    point_size=None,\n",
    "    radius=None, # My contribution\n",
    "#     subset_points=None,\n",
    "    interactive_text_search=False,\n",
    "    interactive_text_search_columns=None,\n",
    "    interactive_text_search_alpha_contrast=0.999,\n",
    "    alpha=None,\n",
    "    normalized=True\n",
    "):\n",
    "    \"\"\"Create an interactive bokeh plot of a UMAP embedding.\n",
    "    While static plots are useful, sometimes a plot that\n",
    "    supports interactive zooming, and hover tooltips for\n",
    "    individual points is much more desireable. This function\n",
    "    provides a simple interface for creating such plots. The\n",
    "    result is a bokeh plot that will be displayed in a notebook.\n",
    "    Note that more complex tooltips etc. will require custom\n",
    "    code -- this is merely meant to provide fast and easy\n",
    "    access to interactive plotting.\n",
    "    Parameters\n",
    "    ----------\n",
    "    embedding_df: pandas DataFrame\n",
    "        A expression dataframe with columns x and y, which are the\n",
    "        2D embedding of a model (e.g., UMAP or pyMDE) on the expression data, and all the\n",
    "        annotations, geometric means of expression, etc.\n",
    "    x: list\n",
    "        The categories for the x-axes of the heatmap and expression profiles\n",
    "    labels: array, shape (n_samples,) (optional, default None)\n",
    "        An array of labels (assumed integer or categorical),\n",
    "        one for each data sample.\n",
    "        This will be used for coloring the points in\n",
    "        the plot according to their label. Note that\n",
    "        this option is mutually exclusive to the ``values``\n",
    "        option.\n",
    "    values: array, shape (n_samples,) (optional, default None)\n",
    "        An array of values (assumed float or continuous),\n",
    "        one for each sample.\n",
    "        This will be used for coloring the points in\n",
    "        the plot according to a colorscale associated\n",
    "        to the total range of values. Note that this\n",
    "        option is mutually exclusive to the ``labels``\n",
    "        option.\n",
    "    hover_data: DataFrame, shape (n_samples, n_tooltip_features)\n",
    "    (optional, default None)\n",
    "        A dataframe of tooltip data. Each column of the dataframe\n",
    "        should be a Series of length ``n_samples`` providing a value\n",
    "        for each data point. Column names will be used for\n",
    "        identifying information within the tooltip.\n",
    "    theme: string (optional, default None)\n",
    "        A color theme to use for plotting. A small set of\n",
    "        predefined themes are provided which have relatively\n",
    "        good aesthetics. Available themes are:\n",
    "           * 'blue'\n",
    "           * 'red'\n",
    "           * 'green'\n",
    "           * 'inferno'\n",
    "           * 'fire'\n",
    "           * 'viridis'\n",
    "           * 'darkblue'\n",
    "           * 'darkred'\n",
    "           * 'darkgreen'\n",
    "    cmap: string (optional, default 'Blues')\n",
    "        The name of a matplotlib colormap to use for coloring\n",
    "        or shading points. If no labels or values are passed\n",
    "        this will be used for shading points according to\n",
    "        density (largely only of relevance for very large\n",
    "        datasets). If values are passed this will be used for\n",
    "        shading according the value. Note that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    color_key: dict or array, shape (n_categories) (optional, default None)\n",
    "        A way to assign colors to categoricals. This can either be\n",
    "        an explicit dict mapping labels to colors (as strings of form\n",
    "        '#RRGGBB'), or an array like object providing one color for\n",
    "        each distinct category being provided in ``labels``. Either\n",
    "        way this mapping will be used to color points according to\n",
    "        the label. Note that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    color_key_cmap: string (optional, default 'Spectral')\n",
    "        The name of a matplotlib colormap to use for categorical coloring.\n",
    "        If an explicit ``color_key`` is not given a color mapping for\n",
    "        categories can be generated from the label list and selecting\n",
    "        a matching list of colors from the given colormap. Note\n",
    "        that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    background: string (optional, default 'white')\n",
    "        The color of the background. Usually this will be either\n",
    "        'white' or 'black', but any color name will work. Ideally\n",
    "        one wants to match this appropriately to the colors being\n",
    "        used for points etc. This is one of the things that themes\n",
    "        handle for you. Note that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    width: int (optional, default 800)\n",
    "        The desired width of the plot in pixels.\n",
    "    height: int (optional, default 800)\n",
    "        The desired height of the plot in pixels\n",
    "    point_size: int (optional, default None)\n",
    "        The size of each point marker\n",
    "    radius: int (optional, default None)\n",
    "        The radius of each point marker (adjusts the point size while zooming)\n",
    "    subset_points: array, shape (n_samples,) (optional, default None)\n",
    "        A way to select a subset of points based on an array of boolean\n",
    "        values.\n",
    "    interactive_text_search: bool (optional, default False)\n",
    "        Whether to include a text search widget above the interactive plot\n",
    "    interactive_text_search_columns: list (optional, default None)\n",
    "        Columns of data source to search. Searches labels and hover_data by default.\n",
    "    interactive_text_search_alpha_contrast: float (optional, default 0.95)\n",
    "        Alpha value for points matching text search. Alpha value for points\n",
    "        not matching text search will be 1 - interactive_text_search_alpha_contrast\n",
    "    alpha: float (optional, default: None)\n",
    "        The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
    "    Returns\n",
    "    -------\n",
    "    \"\"\"\n",
    "    if theme is not None:\n",
    "        cmap = _themes[theme][\"cmap\"]\n",
    "        color_key_cmap = _themes[theme][\"color_key_cmap\"]\n",
    "        background = _themes[theme][\"background\"]\n",
    "\n",
    "    if labels is not None and values is not None:\n",
    "        raise ValueError(\n",
    "            \"Conflicting options; only one of labels or values should be set\"\n",
    "        )\n",
    "        \n",
    "    if alpha is not None:\n",
    "        if not 0.0 <= alpha <= 1.0:\n",
    "            raise ValueError(\"Alpha must be between 0 and 1 inclusive\")\n",
    "\n",
    "    if point_size is None and radius is None:\n",
    "        point_size = 100.0 / np.sqrt(points.shape[0])\n",
    "        \n",
    "    data = embedding_df\n",
    "    # data = data.set_index('TTHERM_ID')\n",
    "    # pdb.set_trace()\n",
    "    if radius is not None:\n",
    "        data['radius'] = radius\n",
    "\n",
    "    if labels is not None:\n",
    "        data[\"label\"] = labels\n",
    "\n",
    "        if color_key is None:\n",
    "            unique_labels = np.unique(labels)\n",
    "            num_labels = unique_labels.shape[0]\n",
    "            color_key = _to_hex(\n",
    "                plt.get_cmap(color_key_cmap)(np.linspace(0, 1, num_labels))\n",
    "            )\n",
    "\n",
    "        if isinstance(color_key, dict):\n",
    "            data[\"color\"] = pd.Series(labels).map(color_key)\n",
    "        else:\n",
    "            # print('here')\n",
    "            unique_labels = np.unique(labels)\n",
    "            if len(color_key) < unique_labels.shape[0]:\n",
    "                # raise ValueError(\n",
    "                #     \"Color key must have enough colors for the number of labels\"\n",
    "                # )\n",
    "                \n",
    "                print('Color key has fewer colors than labels. Making all white')\n",
    "                data['color'] = ['white']*len(labels)\n",
    "            else:\n",
    "\n",
    "                new_color_key = {k: color_key[i] for i, k in enumerate(unique_labels)}\n",
    "                data[\"color\"] = pd.Series(labels).map(new_color_key)\n",
    "\n",
    "        colors = \"color\"\n",
    "\n",
    "    elif values is not None:\n",
    "        data[\"value\"] = values\n",
    "        palette = _to_hex(plt.get_cmap(cmap)(np.linspace(0, 1, 256)))\n",
    "        colors = btr.linear_cmap(\n",
    "            \"value\", palette, low=np.min(values), high=np.max(values)\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        colors = matplotlib.colors.rgb2hex(plt.get_cmap(cmap)(0.5))\n",
    "\n",
    "    # print(data['color'].unique())\n",
    "    # print(colors)\n",
    "\n",
    "    if hover_data is not None:\n",
    "        tooltip_dict = {}\n",
    "        for col_name in hover_data:\n",
    "            data[col_name] = hover_data[col_name]\n",
    "            tooltip_dict[col_name] = \"@{\" + col_name + \"}\"\n",
    "        tooltips = list(tooltip_dict.items())\n",
    "    else:\n",
    "        tooltips = None\n",
    "\n",
    "    if alpha is not None:\n",
    "        data[\"alpha\"] = alpha\n",
    "    else:\n",
    "        alpha = 1\n",
    "        data[\"alpha\"] = alpha\n",
    "\n",
    "    data_source = bokeh.plotting.ColumnDataSource(data)\n",
    "    data_source.data['module'] = hover_data['module']\n",
    "    data_source.data['ID'] = hover_data['ID']\n",
    "    data_source.data['radius'] = np.ones_like(hover_data['ID']) * radius\n",
    "    data_source.data['alpha'] = np.ones_like(hover_data['ID']) * alpha\n",
    "    \n",
    "    # print(data_source.data['ID'][:5])\n",
    "\n",
    "    plot = bokeh.plotting.figure(\n",
    "        width=800,\n",
    "        height=500,\n",
    "        tooltips=tooltips,\n",
    "        tools=\"tap,box_select,pan,wheel_zoom,box_zoom,reset,save\",\n",
    "        background_fill_color=background,\n",
    "        title=title\n",
    "#             x_range=(np.floor(min(points[:,0])), np.ceil(max(points[:,0]))), # Get axes\n",
    "#             y_range=(np.floor(min(points[:,1])), np.ceil(max(points[:,1])))\n",
    "    )\n",
    "\n",
    "    if point_size is not None:\n",
    "\n",
    "        plot.circle(\n",
    "            x=\"x\",\n",
    "            y=\"y\",\n",
    "            source=data_source,\n",
    "            color=colors,\n",
    "            size=point_size,\n",
    "            alpha=\"alpha\",\n",
    "            line_color='black'\n",
    "        )\n",
    "\n",
    "    elif radius is not None:\n",
    "        plot.circle(\n",
    "            x=\"x\",\n",
    "            y=\"y\",\n",
    "            source=data_source,\n",
    "            color=colors,\n",
    "            radius=radius,\n",
    "            alpha=\"alpha\",\n",
    "            line_color='black'\n",
    "        )\n",
    "\n",
    "    plot.grid.visible = False\n",
    "    plot.axis.visible = False\n",
    "\n",
    "    \n",
    "    x_heatmap_profile = x\n",
    "    \n",
    "    # ['Ll', \n",
    "    #      'Lm', \n",
    "    #      'Lh', \n",
    "    #      'S0', \n",
    "    #      'S3', \n",
    "    #      'S6', \n",
    "    #      'S9', \n",
    "    #      # 'S12', \n",
    "    #      'S15', \n",
    "    #      'S24', \n",
    "    #      'C0', \n",
    "    #      # 'C2', \n",
    "    #      'C4', \n",
    "    #      'C6', \n",
    "    #      'C8', \n",
    "    #      'C10', \n",
    "    #      'C12', \n",
    "    #      'C14', \n",
    "    #      'C16', \n",
    "    #      'C18']\n",
    "    \n",
    "    if normalized:\n",
    "        hm_min = 0\n",
    "        hm_max = 1\n",
    "        \n",
    "    else:\n",
    "        hm_min = 2\n",
    "        hm_max = 16\n",
    "    \n",
    "    # For companion heatmap plot\n",
    "    ttherm_ids = embedding_df['TTHERM_ID'].values\n",
    "    hm_df = embedding_df[['TTHERM_ID'] + x_heatmap_profile]\n",
    "    hm_df['module'] = hover_data['module'].values\n",
    "    hm_df_tidy = hm_df.melt(id_vars=['TTHERM_ID', 'module'], var_name='phase', value_name='normalized_expression')\n",
    "    hm_cds = bokeh.plotting.ColumnDataSource(hm_df_tidy)\n",
    "    hm_cds.data['fill_alpha'] = [0.7]*len(hm_df_tidy)\n",
    "    hm_cds.data['line_alpha'] = [0.7]*len(hm_df_tidy)\n",
    "    # hm_cds.data['y_axis'] = ttherm_ids\n",
    "    \n",
    "    hm = heatmap(hm_cds, bokeh.palettes.Inferno256, hm_min, hm_max, x_heatmap_profile, ttherm_ids)\n",
    "    \n",
    "    \n",
    "\n",
    "    # For companion expression plot\n",
    "\n",
    "    expr_source = bokeh.plotting.ColumnDataSource(dict(\n",
    "        ID=['blah'], \n",
    "        expr_xs=[['Ll']], \n",
    "        expr_ys=[[0]],\n",
    "        alpha=[0],\n",
    "        color=['black']))\n",
    "    \n",
    "    if normalized:\n",
    "        y_axis_label = 'Geometric mean expression of normalized replicates'\n",
    "        y_range = (-0.01, 1.01)\n",
    "        \n",
    "    else:\n",
    "        y_axis_label = 'Geometric mean expression of replicates (log2-scale)'\n",
    "        y_range = (3.9, 16.1)\n",
    "    \n",
    "    expr_fig = bokeh.plotting.figure(width=800, \n",
    "                                     height=500,\n",
    "                                     background_fill_color=background,\n",
    "                                     # x_axis_label='Phase or condition',\n",
    "                                     y_axis_label=y_axis_label,\n",
    "                                     x_range=x_heatmap_profile, \n",
    "                                     y_range=y_range\n",
    "                                    )\n",
    "\n",
    "    expr_fig.multi_line('expr_xs', \n",
    "                        'expr_ys', \n",
    "                        source=expr_source, \n",
    "                        alpha='alpha', \n",
    "                        line_width=3, \n",
    "                        line_join='round',\n",
    "                        line_color=\"color\"\n",
    "                       )\n",
    "\n",
    "    expr_fig.xaxis.major_label_orientation = np.pi/4\n",
    "    expr_fig.xaxis.major_label_text_font_size = '12pt'\n",
    "    expr_fig.yaxis.major_label_text_font_size = '12pt'\n",
    "    expr_fig.yaxis.axis_label_text_font_size = '12pt'\n",
    "    expr_fig.xgrid.grid_line_color='whitesmoke'\n",
    "    expr_fig.xgrid.grid_line_alpha=0.2\n",
    "    expr_fig.ygrid.grid_line_color='whitesmoke'\n",
    "    expr_fig.ygrid.grid_line_alpha=0.2\n",
    "\n",
    "    # For data table\n",
    "    s2 = bokeh.plotting.ColumnDataSource(data=dict(ID=[]))\n",
    "\n",
    "    columns = [TableColumn(field=\"ID\",  title=\"TTHERM_ID\", width=160, formatter=HTMLTemplateFormatter(template='<a href=\"http://tet.ciliate.org/index.php/feature/details/feature_details.php?feature_name=<%= ID %>\"target=\"_blank\"><%= ID %></a>')),\n",
    "               TableColumn(field=\"module\",  title=\"Module\", width=160),\n",
    "               TableColumn(field='TGD2021_description', title='TGD2021_description', width=160),\n",
    "               TableColumn(field=\"Description\", title=\"eggNOG_description\", width=160),\n",
    "               TableColumn(field=\"Preferred_name\", title=\"eggNOG_preferred_name\", width=160),\n",
    "               TableColumn(field=\"max_annot_lvl\", title=\"max_annot_lvl\", width=160),\n",
    "               TableColumn(field=\"COG_category\", title=\"COG_category\", width=160),\n",
    "               TableColumn(field='EC', title='EC', width=160),\n",
    "               TableColumn(field='GOs', title='GOs', width=160),\n",
    "               TableColumn(field='KEGG_ko', title='KEGG_ko', width=160),\n",
    "               TableColumn(field='KEGG_Pathway', title='KEGG_Pathway', width=160),\n",
    "               TableColumn(field='KEGG_Module', title='KEGG_Module', width=160),\n",
    "               TableColumn(field='KEGG_Reaction', title='KEGG_Reaction', width=160),\n",
    "               TableColumn(field='KEGG_rclass', title='KEGG_rclass', width=160),\n",
    "               TableColumn(field='BRITE', title='BRITE', width=160),\n",
    "               TableColumn(field='KEGG_TC', title='KEGG_TC', width=160),\n",
    "               TableColumn(field='CAZy', title='CAZy', width=160),\n",
    "               TableColumn(field='BiGG_Reaction', title='BiGG_Reaction', width=160),\n",
    "#                    TableColumn(field=\"x\",  title=\"x\"),\n",
    "#                    TableColumn(field=\"y\",  title=\"y\")\n",
    "              ]\n",
    "    table = DataTable(source=s2, \n",
    "                      columns=columns, \n",
    "                      width=1600, \n",
    "                      height=500,\n",
    "                      editable=True,\n",
    "                      selectable=True,\n",
    "                      sortable=True,\n",
    "                      index_width=10,\n",
    "                      fit_columns=False,\n",
    "                     )\n",
    "    \n",
    "    heatmap_callback = CustomJS(\n",
    "        args=dict(\n",
    "            s1=data_source,\n",
    "            s_hm=hm_cds,\n",
    "            cols=x_heatmap_profile\n",
    "        ),\n",
    "        code=\"\"\"\n",
    "        var d1 = s1.data;\n",
    "        var d_hm = s_hm.data;\n",
    "        \n",
    "        var inds = s1.selected.indices;\n",
    "        const num_cols = cols.length;\n",
    "        \n",
    "        //d_hm['TTHERM_ID'] = []\n",
    "        //d_hm['normalized_expression'] = []\n",
    "        d_hm['fill_alpha'] = []\n",
    "        d_hm['line_alpha'] = []\n",
    "        \n",
    "        var selected_ttherm_ids = [];\n",
    "        \n",
    "        //Careful here! Number is hardcoded to match the number of genes in dataset EDIT: now it is dynamic)\n",
    "        var ttids = d_hm['TTHERM_ID'].slice(0, \"\"\"+str(num_genes)+\"\"\");\n",
    "        \n",
    "        if (inds.length == 0) {\n",
    "            d_hm['fill_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.7)\n",
    "            d_hm['line_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.7)\n",
    "        }else{\n",
    "        \n",
    "            // Start with everything deselected\n",
    "            d_hm['fill_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.01)\n",
    "            d_hm['line_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.01)\n",
    "        \n",
    "            // Get the selected indices\n",
    "            for (var i = 0; i < inds.length; i++) {\n",
    "                selected_ttherm_ids.push(d1['ID'][inds[i]])\n",
    "            }\n",
    "            console.log(selected_ttherm_ids);\n",
    "            \n",
    "            // iterate over the selected ttherm ids\n",
    "            for (var j = 0; j < selected_ttherm_ids.length; j++) {\n",
    "            \n",
    "                // var selected_gene = selected_ttherm_ids[j];\n",
    "                // console.log(selected_gene);\n",
    "\n",
    "                // ad hoc function to find if ttherm ids match\n",
    "                var match = (element) => element == selected_ttherm_ids[j];\n",
    "            \n",
    "                // get index of matching ttherm id in heatmap\n",
    "                var gene_index = ttids.findIndex(match);\n",
    "                console.log(gene_index);\n",
    "                \n",
    "                // loop over the columns and highlight the selected genes\n",
    "                for (var k = 0; k < num_cols; k++) {\n",
    "                \n",
    "                    d_hm['fill_alpha'][gene_index] = 0.7\n",
    "                    d_hm['line_alpha'][gene_index] = 0.7\n",
    "\n",
    "                    gene_index = gene_index + ttids.length\n",
    "                \n",
    "                }\n",
    "            \n",
    "            }\n",
    "            \n",
    "        }\n",
    "        \n",
    "        console.log(d_hm);\n",
    "        \n",
    "        s_hm.change.emit();\n",
    "        \n",
    "        \"\"\"\n",
    "    )\n",
    "\n",
    "    expression_callback = CustomJS(\n",
    "        args=dict(\n",
    "            s1=data_source,\n",
    "            s_expr=expr_source,\n",
    "            alpha=alpha,\n",
    "        ),\n",
    "        code=\"\"\"\n",
    "        var d1 = s1.data;\n",
    "        var d_expr = s_expr.data;\n",
    "\n",
    "        var inds = s1.selected.indices;\n",
    "        // console.log(inds)\n",
    "\n",
    "        // console.log(d1['ID'].length)\n",
    "\n",
    "        // d1['alpha'] = Array(d1['ID'].length).fill(0.2)\n",
    "\n",
    "        // console.log(d_expr['ID'].length, d_expr['expr_xs'].length, d_expr['expr_ys'].length)\n",
    "\n",
    "        d_expr['ID'] = [['blah']]\n",
    "        d_expr['expr_xs'] = [['Ll']]\n",
    "        d_expr['expr_ys'] = [[0]]\n",
    "        d_expr['alpha'] = [0]\n",
    "        d_expr['color'] = ['black']\n",
    "        // s_expr.change.emit();\n",
    "\n",
    "        // debugger;\n",
    "\n",
    "        for (var i = 0; i < inds.length; i++) {\n",
    "            // d_expr['alpha'][inds[i]] = 1/(inds.length * 2)\n",
    "            // console.log(inds[i], i)\n",
    "            d_expr['ID'].push(Array(18).fill(d1['ID'][inds[i]]))\n",
    "            d_expr['expr_xs'].push(d1['expr_xs'][inds[i]])\n",
    "            d_expr['expr_ys'].push(d1['expr_ys'][inds[i]])\n",
    "            d_expr['alpha'].push(Math.min(1, Math.max(7/(inds.length), 0.05)))\n",
    "            d_expr['color'].push(d1['color'][inds[i]])\n",
    "            // console.log(d_expr)\n",
    "            // console.log(i)\n",
    "            // console.log(\n",
    "            //     d_expr['ID'].length, \n",
    "            //     d_expr['expr_xs'].length, \n",
    "            //     d_expr['expr_ys'].length\n",
    "            // )\n",
    "        }\n",
    "\n",
    "        // s1.change.emit();\n",
    "        s_expr.change.emit();\n",
    "        // console.log(s_expr.data)\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "    )\n",
    "\n",
    "    selection_callback =  CustomJS(args=dict(\n",
    "                                          s1=data_source, \n",
    "                                          s2=s2,\n",
    "                                          table=table), \n",
    "                                               code=\"\"\"\n",
    "\n",
    "        var d1 = s1.data;\n",
    "        var d2 = s2.data;\n",
    "\n",
    "\n",
    "        var inds = s1.selected.indices;\n",
    "\n",
    "        d2['module'] = []\n",
    "        d2['ID'] = []\n",
    "        d2['TGD2021_description'] = []\n",
    "        d2['Description'] = []\n",
    "        d2['Preferred_name'] = []\n",
    "        d2['max_annot_lvl'] = []\n",
    "        d2['COG_category'] = []\n",
    "        d2['EC'] = []\n",
    "        d2['GOs'] = []\n",
    "        d2['KEGG_ko'] = []\n",
    "        d2['KEGG_Pathway'] = []\n",
    "        d2['KEGG_Module'] = []\n",
    "        d2['KEGG_Reaction'] = []\n",
    "        d2['KEGG_rclass'] = []\n",
    "        d2['BRITE'] = []\n",
    "        d2['KEGG_TC'] = []\n",
    "        d2['CAZy'] = []\n",
    "        d2['BiGG_Reaction'] = []\n",
    "\n",
    "        for (var i = 0; i < inds.length; i++) {\n",
    "            d2['module'].push(d1['module'][inds[i]])\n",
    "            d2['ID'].push(d1['ID'][inds[i]])\n",
    "            d2['TGD2021_description'].push(d1['TGD2021_description'][inds[i]])\n",
    "            d2['Description'].push(d1['Description'][inds[i]])\n",
    "            d2['Preferred_name'].push(d1['Preferred_name'][inds[i]])\n",
    "            d2['max_annot_lvl'].push(d1['max_annot_lvl'][inds[i]])\n",
    "            d2['COG_category'].push(d1['COG_category'][inds[i]])\n",
    "            d2['EC'].push(d1['EC'][inds[i]])\n",
    "            d2['GOs'].push(d1['GOs'][inds[i]])\n",
    "            d2['KEGG_ko'].push(d1['KEGG_ko'][inds[i]])\n",
    "            d2['KEGG_Pathway'].push(d1['KEGG_Pathway'][inds[i]])\n",
    "            d2['KEGG_Module'].push(d1['KEGG_Module'][inds[i]])\n",
    "            d2['KEGG_Reaction'].push(d1['KEGG_Reaction'][inds[i]])\n",
    "            d2['KEGG_rclass'].push(d1['KEGG_rclass'][inds[i]])\n",
    "            d2['BRITE'].push(d1['BRITE'][inds[i]])\n",
    "            d2['KEGG_TC'].push(d1['KEGG_TC'][inds[i]])\n",
    "            d2['CAZy'].push(d1['CAZy'][inds[i]])\n",
    "            d2['BiGG_Reaction'].push(d1['BiGG_Reaction'][inds[i]])\n",
    "        }\n",
    "        s2.change.emit();\n",
    "        table.change.emit();\n",
    "    \"\"\")\n",
    "\n",
    "    data_source.selected.js_on_change('indices', selection_callback, expression_callback, heatmap_callback)\n",
    "\n",
    "    if interactive_text_search:\n",
    "        text_input = TextInput(value=\"Search module(s) or TTHERM_ID(s), e.g. TTHERM_00321680, TTHERM_00313130...\", width=600)\n",
    "\n",
    "        if interactive_text_search_columns is None:\n",
    "            interactive_text_search_columns = []\n",
    "            if hover_data is not None:\n",
    "                interactive_text_search_columns.extend(hover_data.columns)\n",
    "            if labels is not None:\n",
    "                interactive_text_search_columns.append(\"label\")\n",
    "\n",
    "        if len(interactive_text_search_columns) == 0:\n",
    "            warn(\n",
    "                \"interactive_text_search_columns set to True, but no hover_data or labels provided.\"\n",
    "                \"Please provide hover_data or labels to use interactive text search.\"\n",
    "            )\n",
    "\n",
    "        else:\n",
    "            callback = CustomJS(\n",
    "                args=dict(\n",
    "                    source=data_source,\n",
    "                    s2=s2,\n",
    "                    table=table,\n",
    "                    matching_alpha=interactive_text_search_alpha_contrast,\n",
    "                    non_matching_alpha=1 - interactive_text_search_alpha_contrast,\n",
    "                    search_columns=interactive_text_search_columns,\n",
    "                    default_radius=radius,\n",
    "                    default_alpha=alpha\n",
    "                ),\n",
    "                code=\"\"\"\n",
    "                var data = source.data;\n",
    "                var text_search = cb_obj.value;\n",
    "                var d2 = s2.data;\n",
    "\n",
    "                // var ref_expr = ref_e_s.data;\n",
    "                // var d3 = sel_e_s.data;\n",
    "\n",
    "                var search_terms = text_search.split(',');\n",
    "\n",
    "                d2['module'] = []\n",
    "                d2['ID'] = []\n",
    "\n",
    "                // d3['xs'] = []\n",
    "                // d3['ys'] = []\n",
    "\n",
    "                var search_columns_dict = {}\n",
    "                for (var col in search_columns){\n",
    "                    search_columns_dict[col] = search_columns[col]\n",
    "                }\n",
    "\n",
    "                // First, clear the data table and selection\n",
    "                data['alpha'] = []\n",
    "                data['radius'] = []\n",
    "                source.selected.indices = []\n",
    "\n",
    "                // source.change.emit();\n",
    "                s2.change.emit();\n",
    "                // sel_e_s.change.emit();\n",
    "                table.change.emit();\n",
    "\n",
    "                // Run search\n",
    "                if (text_search.length > 0){\n",
    "                    // Loop over columns and values\n",
    "                    // If there is no match for any column for a given row, change the alpha value\n",
    "                    var string_match = false;\n",
    "                    for (var i = 0; i < data.x.length; i++) {\n",
    "                        string_match = false\n",
    "                        for (var j in search_columns_dict) {\n",
    "                            if (search_terms.some(t => String(data[search_columns_dict[j]][i]).includes(t.trim()))) {\n",
    "                                string_match = true\n",
    "                            }\n",
    "                        }\n",
    "                        if (string_match){\n",
    "                            data['alpha'][i] = matching_alpha\n",
    "                            data['radius'][i] = 1\n",
    "                            d2['module'].push(data['module'][i])\n",
    "                            d2['ID'].push(data['ID'][i])\n",
    "\n",
    "                            // d3['xs'].push(ref_expr['xs'][i])\n",
    "                            // d3['ys'].push(ref_expr['ys'][i])\n",
    "\n",
    "                            // So that these points are actually considered selected\n",
    "                            source.selected.indices.push(i)\n",
    "\n",
    "                        }else{\n",
    "                            data['alpha'][i] = non_matching_alpha\n",
    "                            data['radius'][i] = 0.01\n",
    "                        }\n",
    "                    }\n",
    "                    source.change.emit();\n",
    "                    s2.change.emit();\n",
    "                    // sel_e_s.change.emit();\n",
    "                    table.change.emit();\n",
    "\n",
    "                } else {\n",
    "\n",
    "                    // Loop over columns and values\n",
    "                    // If there is no match for any column for a given row, change the alpha value\n",
    "                    var string_match = false;\n",
    "                    for (var i = 0; i < data.x.length; i++) {\n",
    "                        string_match = false\n",
    "                        for (var j in search_columns_dict) {\n",
    "                            if (search_terms.some(t => String(data[search_columns_dict[j]][i]).includes(t.trim()))) {\n",
    "                                string_match = true\n",
    "                            }\n",
    "                        }\n",
    "                        if (string_match){\n",
    "                            data['alpha'][i] = default_alpha\n",
    "                            data['radius'][i] = default_radius\n",
    "                            d2['module'].push()\n",
    "                            d2['ID'].push()\n",
    "\n",
    "                            // d3['xs'].push()\n",
    "                            // d3['ys'].push()\n",
    "\n",
    "                        }else{\n",
    "                            data['alpha'][i] = non_matching_alpha\n",
    "                            data['radius'][i] = 0.01\n",
    "                        }\n",
    "                    }\n",
    "                    source.change.emit();\n",
    "                    s2.change.emit();\n",
    "                    // sel_e_s.change.emit();\n",
    "                    table.change.emit();\n",
    "\n",
    "                }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            \"\"\",\n",
    "            )\n",
    "\n",
    "            text_input.js_on_change(\"value\", callback, selection_callback, expression_callback, heatmap_callback)\n",
    "\n",
    "    # Lifted from https://stackoverflow.com/questions/31824124/is-there-a-way-to-save-bokeh-data-table-content\n",
    "    button1 = Button(label=\"Download Annotation Table\", button_type=\"success\", width=550)\n",
    "    button1.js_on_click(\n",
    "        CustomJS(\n",
    "            args=dict(source_data=data_source),\n",
    "            code=\"\"\"\n",
    "            var inds = source_data.selected.indices;\n",
    "            var data = source_data.data;\n",
    "            var out = \"TTHERM_ID\\tmodule\\tTGD2021_description\\teggNOG_description\\teggNOG_preferred_name\\tmax_annot_lvl\\tCOG_category\\tGOs\\tEC\\tKEGG_ko\\tKEGG_Pathway\\tKEGG_Module\\tKEGG_Reaction\\tKEGG_rclass\\tBRITE\\tKEGG_TC\\tCAZy\\tBiGG_Reaction\\\\n\";\n",
    "            for (var i = 0; i < inds.length; i++) {\n",
    "                out += data['ID'][inds[i]] + \"\\t\" + data['module'][inds[i]] + \"\\t\" + data['TGD2021_description'][inds[i]] + \"\\t\" + data['Description'][inds[i]] + \"\\t\" + data['Preferred_name'][inds[i]] + \"\\t\" + data['max_annot_lvl'][inds[i]] + \"\\t\" + data['COG_category'][inds[i]] + \"\\t\" + data['GOs'][inds[i]] + \"\\t\" + data['EC'][inds[i]] + \"\\t\" + data['KEGG_ko'][inds[i]] + \"\\t\" + data['KEGG_Pathway'][inds[i]] + \"\\t\" + data['KEGG_Module'][inds[i]] + \"\\t\" + data['KEGG_Reaction'][inds[i]] + \"\\t\" + data['KEGG_rclass'][inds[i]] + \"\\t\" + data['BRITE'][inds[i]] + \"\\t\" + data['KEGG_TC'][inds[i]] + \"\\t\" + data['CAZy'][inds[i]] + \"\\t\" + data['BiGG_Reaction'][inds[i]] + \"\\\\n\";\n",
    "            }\n",
    "            var file = new Blob([out], {type: 'text/plain'});\n",
    "            var elem = window.document.createElement('a');\n",
    "            elem.href = window.URL.createObjectURL(file);\n",
    "            elem.download = 'selected-annotation-data.tsv';\n",
    "            document.body.appendChild(elem);\n",
    "            elem.click();\n",
    "            document.body.removeChild(elem);\n",
    "            \"\"\"))        \n",
    "    \n",
    "    # NEED TO STOP HARDCODING THIS FILE\n",
    "    enrich_df = pd.read_csv('../enrichment/test_nn3_full_enrichment.csv')\n",
    "    colors = [color_key[int(m)] for m in enrich_df['module'].values]\n",
    "    enrich_df['color'] = colors\n",
    "    \n",
    "    enrich_cds = bokeh.models.ColumnDataSource(enrich_df)\n",
    "    enrich_p = plot_enrichment(enrich_cds)\n",
    "    \n",
    "    button2 = Button(label=\"Download Functional Enrichment Data\", button_type=\"success\", width=450)\n",
    "    button2.js_on_click(\n",
    "        CustomJS(\n",
    "            args=dict(source_data=enrich_cds),\n",
    "            code=\"\"\"\n",
    "            // var inds = source_data.selected.indices;\n",
    "            var data = source_data.data;\n",
    "            var out = \"module\\tterm\\tinfo\\tfold_change\\tbonferroni\\\\n\";\n",
    "            for (var i = 0; i < data['module'].length; i++) {\n",
    "                out += data['module'][i] + \"\\t\" + data['term'][i] + \"\\t\" + data['info'][i] + \"\\t\" + data['fold_change'][i] + \"\\t\" + data['bonferroni'][i] + \"\\\\n\";\n",
    "            }\n",
    "            var file = new Blob([out], {type: 'text/plain'});\n",
    "            var elem = window.document.createElement('a');\n",
    "            elem.href = window.URL.createObjectURL(file);\n",
    "            elem.download = 'enrichment-data.tsv';\n",
    "            document.body.appendChild(elem);\n",
    "            elem.click();\n",
    "            document.body.removeChild(elem);\n",
    "            \"\"\"))  \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    if interactive_text_search:\n",
    "        plot = column(row(column(plot, expr_fig), hm, enrich_p), row(text_input, button1, button2), table)\n",
    "    else:\n",
    "        plot = column(row(column(plot, expr_fig), hm, enrich_p), row(button1, button2), table)\n",
    "\n",
    "    return plot\n",
    "\n",
    "def get_centroid(module_df):\n",
    "    \n",
    "    # get rid of ttherm_ids\n",
    "    data_cols = [c for c in module_df.columns if ('TTHERM' not in c) and ('label' not in c)]\n",
    "    data = module_df[data_cols]\n",
    "    \n",
    "    centroid = data.apply(np.mean, axis=0).values\n",
    "    \n",
    "    return centroid\n",
    "\n",
    "def get_module_centroid_df(expr_df, cluster_label_df, alg, phases):\n",
    "    \n",
    "    merge = expr_df.merge(cluster_label_df, on='TTHERM_ID')\n",
    "    \n",
    "    grouped = merge.groupby(f'{alg}_label_{phases}')\n",
    "    \n",
    "    centroid_rows = []\n",
    "    \n",
    "    for label, grp_df in grouped:\n",
    "        # print(grp_df.head())\n",
    "        centroid = get_centroid(grp_df)\n",
    "        centroid_rows.append(centroid)\n",
    "    \n",
    "    data_cols = [c for c in merge.columns if ('TTHERM' not in c) and ('label' not in c)]\n",
    "    \n",
    "    centroid_df = pd.DataFrame(centroid_rows)\n",
    "    centroid_df.columns = data_cols\n",
    "    centroid_df.index.rename('module', inplace=True)\n",
    "        \n",
    "    return centroid_df\n",
    "\n",
    "def get_all_module_centroids(expr_df, cluster_label_df, alg, phases):\n",
    "    \n",
    "    merge = expr_df.merge(cluster_label_df, on='TTHERM_ID')\n",
    "    \n",
    "    grouped = merge.groupby(f'{alg}_label_{phases}')\n",
    "    \n",
    "    module_centroid_list = []\n",
    "    \n",
    "    for label, grp_df in grouped:\n",
    "        \n",
    "        centroid = get_centroid(grp_df)\n",
    "        module_centroid_list.append( (label, centroid) )\n",
    "        \n",
    "    return module_centroid_list\n",
    "\n",
    "def arrange_modules(expr_df, cluster_label_df, alg, phases):\n",
    "    \n",
    "    if phases == 'full':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12',\n",
    "             'S15', \n",
    "             'S24', \n",
    "             'C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10', \n",
    "             'C12', \n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "        \n",
    "    elif phases == 'veg':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12', \n",
    "             'S15', \n",
    "             'S24']\n",
    "        \n",
    "    elif phases == 'sex':\n",
    "        \n",
    "        x = ['C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10',\n",
    "             'C12',\n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "    cols = ['TTHERM_ID'] + [c for c in expr_df.columns if c.split('_')[0] in x]\n",
    "    \n",
    "    module_centroid_df = get_module_centroid_df(expr_df[cols], cluster_label_df, alg, phases)\n",
    "    \n",
    "    linkage = scipy.cluster.hierarchy.linkage(module_centroid_df, method='average', metric='correlation', optimal_ordering=True)\n",
    "    r_cophcorre, ar_copdist = scipy.cluster.hierarchy.cophenet(linkage, scipy.spatial.distance.pdist(module_centroid_df, metric='correlation'))\n",
    "    \n",
    "    # print(f'The Copheretic correlation is: {r_cophcorre}')\n",
    "    \n",
    "    d_dendro = scipy.cluster.hierarchy.dendrogram(linkage, no_plot=True)\n",
    "    cat_sorted = list(module_centroid_df.iloc[d_dendro['leaves'],:].index)\n",
    "    \n",
    "    sorter_index = dict(zip(cat_sorted, range(len(cat_sorted))))\n",
    "    \n",
    "    reassigned_df = cluster_label_df.copy(deep=True)\n",
    "    \n",
    "    \n",
    "    \n",
    "    reassigned_df[f'{alg}_label_{phases}'] = reassigned_df[f'{alg}_label_{phases}'].map(sorter_index)\n",
    "    print(len(reassigned_df))\n",
    "    \n",
    "    arranged_dfs = []\n",
    "    \n",
    "    for cat in cat_sorted:\n",
    "        \n",
    "        mini_df = reassigned_df.loc[reassigned_df[f'{alg}_label_{phases}'] == cat]\n",
    "        # gene_count += len(mini_df)\n",
    "\n",
    "        arranged_dfs.append(mini_df)\n",
    "        \n",
    "#     gene_count = 0\n",
    "    \n",
    "#     for mdf in arranged_dfs:\n",
    "#         gene_count += len(mdf)\n",
    "    \n",
    "#     print(gene_count)\n",
    "        \n",
    "    arranged_df = pd.concat(arranged_dfs)\n",
    "    \n",
    "    \n",
    "    return arranged_df\n",
    "\n",
    "def plot_embedding(expression_df, embedding_df, annotation_df, label_df, clust_alg, phases, palette, n_components=2, n_neighbors=15, title=None, random_state=42, radius=0.01, normalized=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    Function to plot the UMAP of expression data.\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # get new index for clustered heatmap\n",
    "    label_df = arrange_modules(expression_df, label_df, clust_alg, phases)\n",
    "    \n",
    "    # Weirdly, the heatmap looks better-arranged when I just sort by the modules, as\n",
    "    # given by the hierarchical clustering done by arrange_modules(), than if\n",
    "    # I stay with the order they were given automatically\n",
    "    label_df = label_df.sort_values(by=[f'{clust_alg}_label_{phases}', 'TTHERM_ID'], ascending=False)\n",
    "    new_index = label_df.index\n",
    "    \n",
    "    labels = label_df[f'{clust_alg}_label_{phases}'].values\n",
    "    \n",
    "    data = expression_df[list(expression_df.columns)[1:]].values\n",
    "    \n",
    "    embedding_df['TTHERM_ID'] = expression_df['TTHERM_ID'].values\n",
    "    \n",
    "    merge = expression_df.merge(embedding_df, on='TTHERM_ID')\n",
    "    \n",
    "    merge = merge.reindex(new_index)\n",
    "    \n",
    "    # take part of annotation df that shared TTHERM_IDs with expression df\n",
    "    relevant_annot = annotation_df.iloc[np.in1d(annotation_df['TTHERM_ID'].values, merge['TTHERM_ID'].values)]\n",
    "    merge = merge.merge(relevant_annot, on='TTHERM_ID')\n",
    "    \n",
    "\n",
    "    mean_expression_df = get_geom_mean_expression(merge)\n",
    "    \n",
    "    ttherm_ids = merge['TTHERM_ID'].values\n",
    "    merge = merge.merge(mean_expression_df, on='TTHERM_ID')\n",
    "    \n",
    "    if phases == 'full':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12',\n",
    "             'S15', \n",
    "             'S24', \n",
    "             'C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10', \n",
    "             'C12', \n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "        \n",
    "    elif phases == 'veg':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12', \n",
    "             'S15', \n",
    "             'S24']\n",
    "        \n",
    "    elif phases == 'sex':\n",
    "        \n",
    "        x = ['C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10',\n",
    "             'C12',\n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "    else:\n",
    "        print('Selected phases must be one of full, sex, or veg!')\n",
    "        return\n",
    "\n",
    "    xs = [x for ttid in ttherm_ids]\n",
    "    ys = [merge.loc[merge['TTHERM_ID'] == ttid, x].values[0] for ttid in ttherm_ids]\n",
    "\n",
    "    merge['expr_xs'] = xs\n",
    "    merge['expr_ys'] = ys\n",
    "    \n",
    "    # print(merge.head())\n",
    "\n",
    "    \n",
    "#     pdb.set_trace()\n",
    "    hover_data = pd.DataFrame({\n",
    "                               # 'index':np.arange(len(data)),\n",
    "                               'ID':merge['TTHERM_ID'].values,\n",
    "                               'module':[f'm{int(l):02d}' for l in labels]})\n",
    "    \n",
    "#     palette = [palette[l] for l in sorted(label_df[label_key].unique())]\n",
    "    \n",
    "    p = interactive(merge,\n",
    "                    x,\n",
    "                    # mean_expression_df,\n",
    "                    title=title,\n",
    "                    hover_data=hover_data, \n",
    "                    labels=labels, \n",
    "                    color_key=palette, \n",
    "#                     color_key_cmap='Paired',\n",
    "                    background='black', \n",
    "                    radius=radius,\n",
    "                    alpha=0.7,\n",
    "#                     width=600, \n",
    "#                     height=500,\n",
    "                    interactive_text_search=True,\n",
    "                    normalized=normalized\n",
    "                   )\n",
    "    \n",
    "    #p.children[1].title = title\n",
    "    \n",
    "    return p\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "id": "5ade6488",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list_test = ['TTHERM_000013409', 'TTHERM_01321550', 'TTHERM_00011710', 'TTHERM_00321680', 'TTHERM_00355700', 'TTHERM_00938950', 'TTHERM_01372820', 'TTHERM_00013410', 'TTHERM_00390080', 'TTHERM_00516380', 'TTHERM_00038880', 'TTHERM_00059370', 'TTHERM_00473020', 'TTHERM_00497590', 'TTHERM_00558350', 'TTHERM_00052190', 'TTHERM_00392790', 'TTHERM_00410180', 'TTHERM_00685980', 'TTHERM_00445920', 'TTHERM_00471040', 'TTHERM_00140780', 'TTHERM_00145480', 'TTHERM_00321720', 'TTHERM_00628650', 'TTHERM_00526730', 'TTHERM_01156770', 'TTHERM_00312200', 'TTHERM_01332070', 'TTHERM_00318900', 'TTHERM_00340180', 'TTHERM_00592740', 'TTHERM_00440600', 'TTHERM_01321570', 'TTHERM_00537380', 'TTHERM_00585170', 'TTHERM_01197130', 'TTHERM_00554390', 'TTHERM_00649180', 'TTHERM_00691410', ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "911d5a38-541a-4c3a-8288-a8afcf42d7c5",
   "metadata": {},
   "source": [
    "## Define palettes for plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0efd5123-23e9-423e-86b9-68e89a479481",
   "metadata": {},
   "source": [
    "These palettes are from the R package Polychrome. The first is just palette36 with the first color replaced by white. The second is the alphabet palette with white prepended."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b03ec5e-9520-4a4e-aadb-f7a16d7dd6e6",
   "metadata": {},
   "source": [
    "The R code for 64 colors:\n",
    "\n",
    "library(Polychrome);\n",
    "seed <- c(\"#000000\", \"#ff0000\", \"#00ff00\", \"#0000ff\");\n",
    "p64 <- createPalette(64, seed, range=c(40,100));\n",
    "paste(p64, sep=\"\\n\");\n",
    "\n",
    "Then, replace the first and last with ~white"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "id": "680b6c31-013d-4611-97e6-78699c22ff29",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette45 = \"\"\"\n",
    "#51635F\\n#FF1C16\\n#16FC26\\n#403DFC\\n#FE0DCE\\n#F9AA00\\n#00FFD5\\n#22BFFE\\n#BB3551\\n#E6FE97\\n#ECADFF\\n#FFBFBD\\n#CF00F5\\n#0D8B00\\n#D7FEFF\\n#8D7200\\n#F76C00\\n#AD3288\\n#5C5AB8\\n#FC0080\\n#B8FF16\\n#00AAB4\\n#FBE11C\\n#9AAAD9\\n#8BBB8C\\n#934B47\\n#6EFE99\\n#9C6D91\\n#FB9778\\n#9D32AF\\n#D40087\\n#FFDC9D\\n#FF8DB6\\n#A96AFC\\n#FDDDFB\\n#168CF7\\n#FD6CF9\\n#F64553\\n#4D6A00\\n#FAFEDB\\n#A7977D\\n#0DFBFF\\n#86B80D\\n#FD8AE4\\n#B7B126\n",
    "\"\"\".split()\n",
    "\n",
    "palette32 = \"\"\"\n",
    "white\\n#F91622\\n#16FC0D\\n#5138FB\\n#FD00CF\\n#FDD51C\\n#16FDD7\\n#FC8B8E\\n#16BFFF\\n#DF9BFD\\n#669C2A\\n#FEE7C4\\n#F31685\\n#DF16FD\\n#C1F1FE\\n#A23D7E\\n#D5FD0D\\n#8C5A0D\\n#FC790D\\n#4F5CBC\\n#FFCBEF\\n#168D72\\n#68FA93\\n#C4FDC9\\n#F7A449\\n#16789B\\n#AD0DAB\\n#C4262E\\n#0DF1FF\\n#EFF994\\n#B6C1FE\\n#8F22CD\n",
    "\"\"\".split()\n",
    "\n",
    "palette35 = \"\"\"\n",
    "#585F6A\\n#FE1626\\n#00FB0D\\n#2E40FC\\n#FD0DCE\\n#FCD200\\n#F7868C\\n#16FFDC\\n#22BEFB\\n#D28EF6\\n#609000\\n#FFE7C9\\n#F51683\\n#FF730D\\n#CAFE16\\n#AA3586\\n#BEEEFD\\n#BD00FA\\n#895D22\\n#FEC7F0\\n#495AA1\\n#73F995\\n#229270\\n#ED963B\\n#F6FE97\\n#C5FFD0\\n#C50DC8\\n#6993FF\\n#C22A35\\n#16ECFC\\n#AA707E\\n#7A3BCB\\n#7C845C\\n#358FAA\\n#BDBAF6\n",
    "\"\"\".split()\n",
    "\n",
    "palette38 = \"\"\"\n",
    "#636265\\n#F60D16\\n#00F90D\\n#3540FB\\n#FD0DD0\\n#FDDB0D\\n#00FFE2\\n#FA8884\\n#2ABEFE\\n#E5A3FF\\n#518F00\\n#FEFDD5\\n#D51CFF\\n#ED007F\\n#A33879\\n#96731C\\n#C8FB16\\n#C0ECFE\\n#FBC1DA\\n#5658BA\\n#F96900\\n#F69F1C\\n#58FA9C\\n#008E72\\n#BA22B9\\n#167D97\\n#794D8A\\n#CEFE9C\\n#BB222E\\n#954D45\\n#00DCEF\\n#FD66B0\\n#B2FDD3\\n#FDBD9F\\n#A9B4F1\\n#B371FE\\n#849566\\n#2A8EFF\n",
    "\"\"\".split()\n",
    "\n",
    "palette64 = \"\"\"\n",
    "white\\n#FA002E\\n#22FC22\\n#221CFA\\n#FF3DD6\\n#FFDA00\\n#00FEFB\\n#F48684\\n#CEB4FE\\n#FFFFE5\\n#0D933D\\n#CC00F8\\n#800D5D\\n#F10084\\n#22267A\\n#0DADFF\\n#CBFD71\\n#9A761C\\n#F96C00\\n#6399A6\\n#FFBCDA\\n#8D0DA3\\n#F79F26\\n#00FFBF\\n#A37CFB\\n#F68EEB\\n#720D0D\\n#F163AA\\n#7E926A\\n#826386\\n#B41C32\\n#9BEBCE\\n#E2DB83\\n#56D4FA\\n#E6E2FB\\n#925D58\\n#F7C3A7\\n#62E970\\n#220DBD\\n#5583BB\\n#7EA01C\\n#CDFDB6\\n#FD00FB\\n#B30D97\\n#F5FF00\\n#DD77FD\\n#4282FC\\n#BBA6A4\\n#0D8068\\n#AB5F26\\n#F7C26E\\n#9EFE00\\n#9B2EFD\\n#C56887\\n#FD3D68\\n#ABF2FD\\n#835FAC\\n#FF16B1\\n#325371\\n#CA16CA\\n#D26322\\n#AFCFFE\\n#91A1FA\\nfloralwhite\n",
    "\"\"\".split()\n",
    "\n",
    "palette65 = \"\"\"\n",
    "white\\ngainsboro\\n#FA002E\\n#22FC22\\n#221CFA\\n#FF3DD6\\n#FFDA00\\n#00FEFB\\n#F48684\\n#CEB4FE\\n#FFFFE5\\n#0D933D\\n#CC00F8\\n#800D5D\\n#F10084\\n#22267A\\n#0DADFF\\n#CBFD71\\n#9A761C\\n#F96C00\\n#6399A6\\n#FFBCDA\\n#8D0DA3\\n#F79F26\\n#00FFBF\\n#A37CFB\\n#F68EEB\\n#720D0D\\n#F163AA\\n#7E926A\\n#826386\\n#B41C32\\n#9BEBCE\\n#E2DB83\\n#56D4FA\\n#E6E2FB\\n#925D58\\n#F7C3A7\\n#62E970\\n#220DBD\\n#5583BB\\n#7EA01C\\n#CDFDB6\\n#FD00FB\\n#B30D97\\n#F5FF00\\n#DD77FD\\n#4282FC\\n#BBA6A4\\n#0D8068\\n#AB5F26\\n#F7C26E\\n#9EFE00\\n#9B2EFD\\n#C56887\\n#FD3D68\\n#ABF2FD\\n#835FAC\\n#FF16B1\\n#325371\\n#CA16CA\\n#D26322\\n#AFCFFE\\n#91A1FA\\nfloralwhite\n",
    "\"\"\".split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "id": "d8970fc7-8ade-402d-9566-9e16c5593f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette36 = [\"#FFFFFF\", \n",
    "             \"#E4E1E3\", \n",
    "             \"#F6222E\", \n",
    "             \"#FE00FA\", \n",
    "             \"#16FF32\", \n",
    "             \"#3283FE\", \n",
    "             \"#FEAF16\", \n",
    "             \"#B00068\", \n",
    "             \"#1CFFCE\",\n",
    "             \"#90AD1C\", \n",
    "             \"#2ED9FF\", \n",
    "             \"#DEA0FD\", \n",
    "             \"#AA0DFE\", \n",
    "             \"#F8A19F\", \n",
    "             \"#325A9B\", \n",
    "             \"#C4451C\", \n",
    "             \"#1C8356\", \n",
    "             \"#85660D\",\n",
    "             \"#B10DA1\", \n",
    "             \"#FBE426\", \n",
    "             \"#1CBE4F\", \n",
    "             \"#FA0087\", \n",
    "             \"#FC1CBF\", \n",
    "             \"#F7E1A0\", \n",
    "             \"#C075A6\", \n",
    "             \"#782AB6\", \n",
    "             \"#AAF400\",\n",
    "             \"#BDCDFF\", \n",
    "             \"#822E1C\", \n",
    "             \"#B5EFB5\", \n",
    "             \"#7ED7D1\", \n",
    "             \"#1C7F93\", \n",
    "             \"#D85FF7\", \n",
    "             \"#683B79\", \n",
    "             \"#66B0FF\", \n",
    "             \"#3B00FB\",\n",
    "             \"magenta\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6239a9e1-9b79-4369-8931-8b33f3d5b027",
   "metadata": {},
   "source": [
    "change index 5 to #778899. Change index 9 to #2F4F4F. Add #FF7F50. Change index 9 to #FFBCD9. Change index 14 to #DEA5A4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "id": "6c36e65a-4442-489e-a5ef-de35adb81ae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette27 = [\"#FFFFFF\", \n",
    "             \"#AA0DFE\", \n",
    "             \"#3283FE\", \n",
    "             \"#85660D\", \n",
    "             \"#782AB6\", \n",
    "             \"#778899\", \n",
    "             \"#1C8356\", \n",
    "             \"#16FF32\", \n",
    "             \"#F7E1A0\", \n",
    "#              \"#2F4F4F\",\n",
    "             \"#FFBCD9\", \n",
    "             \"#C4451C\", \n",
    "             \"#DEA0FD\", \n",
    "             \"#FE00FA\", \n",
    "#              \"#325A9B\", \n",
    "             \"#FEAF16\", \n",
    "             \"#DEA5A4\", \n",
    "             \"#90AD1C\", \n",
    "             \"#F6222E\",\n",
    "             \"#1CFFCE\", \n",
    "             \"#2ED9FF\", \n",
    "             \"#B10DA1\", \n",
    "#              \"#C075A6\", \n",
    "#              \"#FC1CBF\", \n",
    "#              \"#B00068\", \n",
    "             \"#FBE426\", \n",
    "             \"#FA0087\",\n",
    "             \"#FF7F50\"\n",
    "            ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00491da4-7bee-4501-a31c-a48783f1e4ee",
   "metadata": {},
   "source": [
    "## Get the expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "id": "b5a1bd1e-a09e-4bd2-9f4e-659799f7e99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# full_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/allgood_filt_agg_tidy_2021aligned_qc_rma_expression_full.csv')\n",
    "full_filtered_df = full_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# full_filtered_norm_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_full_norm_filt_agg_tidy_2021aligned_qc_rma_expression.csv')\n",
    "# full_filtered_norm_df = full_filtered_norm_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# veg_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_filt_agg_tidy_2021aligned_qc_rma_expression_veg.csv')\n",
    "# veg_filtered_df = veg_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# veg_filtered_norm_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_veg_norm_filt_agg_tidy_2021aligned_qc_rma_expression.csv')\n",
    "# veg_filtered_norm_df = veg_filtered_norm_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# sex_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_filt_agg_tidy_2021aligned_qc_rma_expression_sex.csv')\n",
    "# sex_filtered_df = sex_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# sex_filtered_norm_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_sex_norm_filt_agg_tidy_2021aligned_qc_rma_expression.csv')\n",
    "# sex_filtered_norm_df = sex_filtered_norm_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "id": "8a5f902a-b493-42f2-b166-441dd9bf3e5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20326"
      ]
     },
     "execution_count": 456,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(full_filtered_df['TTHERM_ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "id": "082eccb7-7c4b-4a6b-aab8-61b3d5a01c68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20326"
      ]
     },
     "execution_count": 457,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(full_filtered_df['TTHERM_ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "c0e0ac3b-882d-4fa3-8995-6b146c5d7073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TTHERM_ID',\n",
       " 'Ll_GSM283687',\n",
       " 'Ll_GSM284355',\n",
       " 'Ll_GSM284362',\n",
       " 'Lm_GSM283690',\n",
       " 'Lm_GSM284357',\n",
       " 'Lm_GSM284363',\n",
       " 'Lh_GSM283691',\n",
       " 'Lh_GSM284360',\n",
       " 'Lh_GSM284364',\n",
       " 'S0_GSM285363',\n",
       " 'S0_GSM285554',\n",
       " 'S0_GSM285561',\n",
       " 'S3_GSM285542',\n",
       " 'S3_GSM285555',\n",
       " 'S3_GSM285562',\n",
       " 'S6_GSM285543',\n",
       " 'S6_GSM285556',\n",
       " 'S6_GSM285563',\n",
       " 'S9_GSM285544',\n",
       " 'S9_GSM285564',\n",
       " 'S15_GSM285559',\n",
       " 'S15_GSM285566',\n",
       " 'S24_GSM285547',\n",
       " 'S24_GSM285560']"
      ]
     },
     "execution_count": 458,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[c for c in full_filtered_df.columns if c[0] != 'C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "id": "aee9c6d0-031e-4907-9cda-c9376e8e9b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_filtered_norm_df = normalize_expression_per_gene(full_filtered_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "id": "ce08f21f-a261-45c8-be31-7976bb8106cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TTHERM_ID', 'Ll_GSM283687', 'Ll_GSM284355', 'Ll_GSM284362',\n",
       "       'Lm_GSM283690', 'Lm_GSM284357', 'Lm_GSM284363', 'Lh_GSM283691',\n",
       "       'Lh_GSM284360', 'Lh_GSM284364', 'S0_GSM285363', 'S0_GSM285554',\n",
       "       'S0_GSM285561', 'S3_GSM285542', 'S3_GSM285555', 'S3_GSM285562',\n",
       "       'S6_GSM285543', 'S6_GSM285556', 'S6_GSM285563', 'S9_GSM285544',\n",
       "       'S9_GSM285564', 'S15_GSM285559', 'S15_GSM285566', 'S24_GSM285547',\n",
       "       'S24_GSM285560', 'C0_GSM285570', 'C0_GSM656230', 'C4_GSM285574',\n",
       "       'C4_GSM285588', 'C4_GSM656234', 'C6_GSM285575', 'C6_GSM656232',\n",
       "       'C8_GSM285576', 'C8_GSM285590', 'C10_GSM285578', 'C10_GSM285591',\n",
       "       'C12_GSM285579', 'C12_GSM285592', 'C12_GSM656237', 'C14_GSM285580',\n",
       "       'C14_GSM285593', 'C14_GSM656238', 'C16_GSM285582', 'C16_GSM285595',\n",
       "       'C16_GSM656239', 'C18_GSM285583', 'C18_GSM285596', 'C18_GSM656240'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 460,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_norm_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "id": "8d8d6dcc-7b5e-483c-9372-4c6096ac55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll_GSM283687</th>\n",
       "      <th>Ll_GSM284355</th>\n",
       "      <th>Ll_GSM284362</th>\n",
       "      <th>Lm_GSM283690</th>\n",
       "      <th>Lm_GSM284357</th>\n",
       "      <th>Lm_GSM284363</th>\n",
       "      <th>Lh_GSM283691</th>\n",
       "      <th>Lh_GSM284360</th>\n",
       "      <th>Lh_GSM284364</th>\n",
       "      <th>...</th>\n",
       "      <th>C12_GSM656237</th>\n",
       "      <th>C14_GSM285580</th>\n",
       "      <th>C14_GSM285593</th>\n",
       "      <th>C14_GSM656238</th>\n",
       "      <th>C16_GSM285582</th>\n",
       "      <th>C16_GSM285595</th>\n",
       "      <th>C16_GSM656239</th>\n",
       "      <th>C18_GSM285583</th>\n",
       "      <th>C18_GSM285596</th>\n",
       "      <th>C18_GSM656240</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>6.928782</td>\n",
       "      <td>7.264201</td>\n",
       "      <td>6.934214</td>\n",
       "      <td>6.732989</td>\n",
       "      <td>6.970612</td>\n",
       "      <td>7.150978</td>\n",
       "      <td>6.126826</td>\n",
       "      <td>6.868968</td>\n",
       "      <td>6.641119</td>\n",
       "      <td>...</td>\n",
       "      <td>6.450318</td>\n",
       "      <td>8.049750</td>\n",
       "      <td>7.788162</td>\n",
       "      <td>7.052154</td>\n",
       "      <td>6.517742</td>\n",
       "      <td>6.918501</td>\n",
       "      <td>6.048861</td>\n",
       "      <td>7.041619</td>\n",
       "      <td>6.757932</td>\n",
       "      <td>5.817246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>9.633489</td>\n",
       "      <td>9.977124</td>\n",
       "      <td>10.027529</td>\n",
       "      <td>9.720665</td>\n",
       "      <td>9.605762</td>\n",
       "      <td>10.225542</td>\n",
       "      <td>10.279608</td>\n",
       "      <td>10.459966</td>\n",
       "      <td>10.693337</td>\n",
       "      <td>...</td>\n",
       "      <td>11.130466</td>\n",
       "      <td>11.207738</td>\n",
       "      <td>11.009172</td>\n",
       "      <td>10.615417</td>\n",
       "      <td>11.038938</td>\n",
       "      <td>11.009222</td>\n",
       "      <td>10.216348</td>\n",
       "      <td>11.099187</td>\n",
       "      <td>11.172276</td>\n",
       "      <td>10.561021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>5.066343</td>\n",
       "      <td>4.767264</td>\n",
       "      <td>5.010981</td>\n",
       "      <td>6.139047</td>\n",
       "      <td>4.619361</td>\n",
       "      <td>4.751761</td>\n",
       "      <td>5.818550</td>\n",
       "      <td>5.342529</td>\n",
       "      <td>5.483750</td>\n",
       "      <td>...</td>\n",
       "      <td>6.314438</td>\n",
       "      <td>7.423571</td>\n",
       "      <td>7.507645</td>\n",
       "      <td>7.417087</td>\n",
       "      <td>7.147801</td>\n",
       "      <td>7.747930</td>\n",
       "      <td>7.093641</td>\n",
       "      <td>7.672685</td>\n",
       "      <td>7.511290</td>\n",
       "      <td>6.890117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>4.696881</td>\n",
       "      <td>4.638401</td>\n",
       "      <td>4.956299</td>\n",
       "      <td>6.942556</td>\n",
       "      <td>5.101252</td>\n",
       "      <td>4.730307</td>\n",
       "      <td>8.457690</td>\n",
       "      <td>4.526411</td>\n",
       "      <td>4.908300</td>\n",
       "      <td>...</td>\n",
       "      <td>5.250233</td>\n",
       "      <td>4.974993</td>\n",
       "      <td>5.747498</td>\n",
       "      <td>5.252167</td>\n",
       "      <td>5.210531</td>\n",
       "      <td>7.083187</td>\n",
       "      <td>5.252222</td>\n",
       "      <td>5.037613</td>\n",
       "      <td>5.495281</td>\n",
       "      <td>5.013987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>4.654278</td>\n",
       "      <td>4.537105</td>\n",
       "      <td>4.928739</td>\n",
       "      <td>5.063991</td>\n",
       "      <td>4.584168</td>\n",
       "      <td>4.911880</td>\n",
       "      <td>5.935311</td>\n",
       "      <td>4.519470</td>\n",
       "      <td>4.757861</td>\n",
       "      <td>...</td>\n",
       "      <td>4.651688</td>\n",
       "      <td>4.920573</td>\n",
       "      <td>4.636333</td>\n",
       "      <td>4.883712</td>\n",
       "      <td>4.779395</td>\n",
       "      <td>4.744335</td>\n",
       "      <td>4.513140</td>\n",
       "      <td>4.838428</td>\n",
       "      <td>4.961475</td>\n",
       "      <td>4.653340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TTHERM_ID  Ll_GSM283687  Ll_GSM284355  Ll_GSM284362  Lm_GSM283690  \\\n",
       "0  TTHERM_000000042      6.928782      7.264201      6.934214      6.732989   \n",
       "1  TTHERM_000000045      9.633489      9.977124     10.027529      9.720665   \n",
       "2   TTHERM_00000010      5.066343      4.767264      5.010981      6.139047   \n",
       "3   TTHERM_00000020      4.696881      4.638401      4.956299      6.942556   \n",
       "4   TTHERM_00000030      4.654278      4.537105      4.928739      5.063991   \n",
       "\n",
       "   Lm_GSM284357  Lm_GSM284363  Lh_GSM283691  Lh_GSM284360  Lh_GSM284364  ...  \\\n",
       "0      6.970612      7.150978      6.126826      6.868968      6.641119  ...   \n",
       "1      9.605762     10.225542     10.279608     10.459966     10.693337  ...   \n",
       "2      4.619361      4.751761      5.818550      5.342529      5.483750  ...   \n",
       "3      5.101252      4.730307      8.457690      4.526411      4.908300  ...   \n",
       "4      4.584168      4.911880      5.935311      4.519470      4.757861  ...   \n",
       "\n",
       "   C12_GSM656237  C14_GSM285580  C14_GSM285593  C14_GSM656238  C16_GSM285582  \\\n",
       "0       6.450318       8.049750       7.788162       7.052154       6.517742   \n",
       "1      11.130466      11.207738      11.009172      10.615417      11.038938   \n",
       "2       6.314438       7.423571       7.507645       7.417087       7.147801   \n",
       "3       5.250233       4.974993       5.747498       5.252167       5.210531   \n",
       "4       4.651688       4.920573       4.636333       4.883712       4.779395   \n",
       "\n",
       "   C16_GSM285595  C16_GSM656239  C18_GSM285583  C18_GSM285596  C18_GSM656240  \n",
       "0       6.918501       6.048861       7.041619       6.757932       5.817246  \n",
       "1      11.009222      10.216348      11.099187      11.172276      10.561021  \n",
       "2       7.747930       7.093641       7.672685       7.511290       6.890117  \n",
       "3       7.083187       5.252222       5.037613       5.495281       5.013987  \n",
       "4       4.744335       4.513140       4.838428       4.961475       4.653340  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 461,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "id": "67dc9e8a-2fa5-4b76-a944-ad21323254e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll_GSM283687</th>\n",
       "      <th>Ll_GSM284355</th>\n",
       "      <th>Ll_GSM284362</th>\n",
       "      <th>Lm_GSM283690</th>\n",
       "      <th>Lm_GSM284357</th>\n",
       "      <th>Lm_GSM284363</th>\n",
       "      <th>Lh_GSM283691</th>\n",
       "      <th>Lh_GSM284360</th>\n",
       "      <th>Lh_GSM284364</th>\n",
       "      <th>...</th>\n",
       "      <th>C12_GSM656237</th>\n",
       "      <th>C14_GSM285580</th>\n",
       "      <th>C14_GSM285593</th>\n",
       "      <th>C14_GSM656238</th>\n",
       "      <th>C16_GSM285582</th>\n",
       "      <th>C16_GSM285595</th>\n",
       "      <th>C16_GSM656239</th>\n",
       "      <th>C18_GSM285583</th>\n",
       "      <th>C18_GSM285596</th>\n",
       "      <th>C18_GSM656240</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>0.604560</td>\n",
       "      <td>0.712935</td>\n",
       "      <td>0.606315</td>\n",
       "      <td>0.541298</td>\n",
       "      <td>0.618075</td>\n",
       "      <td>0.676352</td>\n",
       "      <td>0.345444</td>\n",
       "      <td>0.585234</td>\n",
       "      <td>0.511614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.449966</td>\n",
       "      <td>0.966750</td>\n",
       "      <td>0.882229</td>\n",
       "      <td>0.644422</td>\n",
       "      <td>0.471751</td>\n",
       "      <td>0.601238</td>\n",
       "      <td>0.320253</td>\n",
       "      <td>0.641018</td>\n",
       "      <td>0.549357</td>\n",
       "      <td>0.245417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>0.015138</td>\n",
       "      <td>0.202748</td>\n",
       "      <td>0.230267</td>\n",
       "      <td>0.062732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.338374</td>\n",
       "      <td>0.367892</td>\n",
       "      <td>0.466359</td>\n",
       "      <td>0.593770</td>\n",
       "      <td>...</td>\n",
       "      <td>0.832424</td>\n",
       "      <td>0.874612</td>\n",
       "      <td>0.766203</td>\n",
       "      <td>0.551230</td>\n",
       "      <td>0.782454</td>\n",
       "      <td>0.766230</td>\n",
       "      <td>0.333354</td>\n",
       "      <td>0.815348</td>\n",
       "      <td>0.855251</td>\n",
       "      <td>0.521531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>0.071967</td>\n",
       "      <td>0.023813</td>\n",
       "      <td>0.063054</td>\n",
       "      <td>0.244680</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021317</td>\n",
       "      <td>0.193078</td>\n",
       "      <td>0.116435</td>\n",
       "      <td>0.139173</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272919</td>\n",
       "      <td>0.451498</td>\n",
       "      <td>0.465034</td>\n",
       "      <td>0.450454</td>\n",
       "      <td>0.407097</td>\n",
       "      <td>0.503722</td>\n",
       "      <td>0.398377</td>\n",
       "      <td>0.491607</td>\n",
       "      <td>0.465621</td>\n",
       "      <td>0.365608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>0.049259</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.105976</td>\n",
       "      <td>0.540233</td>\n",
       "      <td>0.137667</td>\n",
       "      <td>0.056567</td>\n",
       "      <td>0.871488</td>\n",
       "      <td>0.011989</td>\n",
       "      <td>0.095481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.170239</td>\n",
       "      <td>0.110063</td>\n",
       "      <td>0.278956</td>\n",
       "      <td>0.170661</td>\n",
       "      <td>0.161559</td>\n",
       "      <td>0.570979</td>\n",
       "      <td>0.170673</td>\n",
       "      <td>0.123753</td>\n",
       "      <td>0.223814</td>\n",
       "      <td>0.118588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>0.027748</td>\n",
       "      <td>0.004712</td>\n",
       "      <td>0.081708</td>\n",
       "      <td>0.108299</td>\n",
       "      <td>0.013964</td>\n",
       "      <td>0.078393</td>\n",
       "      <td>0.279603</td>\n",
       "      <td>0.001245</td>\n",
       "      <td>0.048113</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027239</td>\n",
       "      <td>0.080103</td>\n",
       "      <td>0.024220</td>\n",
       "      <td>0.072855</td>\n",
       "      <td>0.052347</td>\n",
       "      <td>0.045454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.063952</td>\n",
       "      <td>0.088144</td>\n",
       "      <td>0.027564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TTHERM_ID  Ll_GSM283687  Ll_GSM284355  Ll_GSM284362  Lm_GSM283690  \\\n",
       "0  TTHERM_000000042      0.604560      0.712935      0.606315      0.541298   \n",
       "1  TTHERM_000000045      0.015138      0.202748      0.230267      0.062732   \n",
       "2   TTHERM_00000010      0.071967      0.023813      0.063054      0.244680   \n",
       "3   TTHERM_00000020      0.049259      0.036473      0.105976      0.540233   \n",
       "4   TTHERM_00000030      0.027748      0.004712      0.081708      0.108299   \n",
       "\n",
       "   Lm_GSM284357  Lm_GSM284363  Lh_GSM283691  Lh_GSM284360  Lh_GSM284364  ...  \\\n",
       "0      0.618075      0.676352      0.345444      0.585234      0.511614  ...   \n",
       "1      0.000000      0.338374      0.367892      0.466359      0.593770  ...   \n",
       "2      0.000000      0.021317      0.193078      0.116435      0.139173  ...   \n",
       "3      0.137667      0.056567      0.871488      0.011989      0.095481  ...   \n",
       "4      0.013964      0.078393      0.279603      0.001245      0.048113  ...   \n",
       "\n",
       "   C12_GSM656237  C14_GSM285580  C14_GSM285593  C14_GSM656238  C16_GSM285582  \\\n",
       "0       0.449966       0.966750       0.882229       0.644422       0.471751   \n",
       "1       0.832424       0.874612       0.766203       0.551230       0.782454   \n",
       "2       0.272919       0.451498       0.465034       0.450454       0.407097   \n",
       "3       0.170239       0.110063       0.278956       0.170661       0.161559   \n",
       "4       0.027239       0.080103       0.024220       0.072855       0.052347   \n",
       "\n",
       "   C16_GSM285595  C16_GSM656239  C18_GSM285583  C18_GSM285596  C18_GSM656240  \n",
       "0       0.601238       0.320253       0.641018       0.549357       0.245417  \n",
       "1       0.766230       0.333354       0.815348       0.855251       0.521531  \n",
       "2       0.503722       0.398377       0.491607       0.465621       0.365608  \n",
       "3       0.570979       0.170673       0.123753       0.223814       0.118588  \n",
       "4       0.045454       0.000000       0.063952       0.088144       0.027564  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 462,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_norm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "id": "8ac61c80-4683-4d36-a891-20d3d8fa506f",
   "metadata": {},
   "outputs": [],
   "source": [
    "veg_filtered_norm_df = full_filtered_norm_df[[c for c in full_filtered_norm_df.columns if c[0] != 'C']]\n",
    "sex_filtered_norm_df = full_filtered_norm_df[[c for c in full_filtered_norm_df.columns if not c[0] in ['L', 'S']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 464,
   "id": "57bf2513-9422-4564-ab74-e8da57af1287",
   "metadata": {},
   "outputs": [],
   "source": [
    "# full_mean_norm = get_geom_mean_expression(full_filtered_norm_df)\n",
    "# full_mean = get_geom_mean_expression(full_filtered_df)\n",
    "# full_mean_norm.to_csv('./full_mean_norm_df.csv', index=False)\n",
    "\n",
    "# patrick_genes = \"\"\"\n",
    "# TTHERM_00658810, TTHERM_000193469, TTHERM_00047330, TTHERM_000486279, TTHERM_00141040, TTHERM_00227750, TTHERM_00317390, TTHERM_00670750, TTHERM_01122800, TTHERM_01213910, TTHERM_00527180, TTHERM_00473020, TTHERM_00624730, TTHERM_00624720, TTHERM_00378890, TTHERM_00321730, TTHERM_00522600, TTHERM_01055600, TTHERM_01018540, TTHERM_00221120, TTHERM_00221130\n",
    "# \"\"\".split(', ')\n",
    "\n",
    "# patrick_genes = [p.strip() for p in patrick_genes]\n",
    "\n",
    "# patrick_df_norm = full_mean_norm.loc[full_mean_norm['TTHERM_ID'].isin(patrick_genes)]\n",
    "# patrick_df = full_mean.loc[full_mean['TTHERM_ID'].isin(patrick_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "id": "af3a4878-5584-47bc-9c34-2eca64f7b234",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(patrick_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "id": "5c728107-b6c0-4d7e-9974-dc3a5e12d049",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(patrick_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 467,
   "id": "450d2acd-06f5-40b9-b263-c2cf83da023b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# patrick_df.to_csv('/Users/eukarya/Downloads/patrick_genes.csv', index=False)\n",
    "# patrick_df_norm.to_csv('/Users/eukarya/Downloads/patrick_genes_normalized', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "id": "527cab73-a0d7-4ead-ba51-0808cb1fe003",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll</th>\n",
       "      <th>Lm</th>\n",
       "      <th>Lh</th>\n",
       "      <th>S0</th>\n",
       "      <th>S3</th>\n",
       "      <th>S6</th>\n",
       "      <th>S9</th>\n",
       "      <th>S15</th>\n",
       "      <th>S24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>0.640498</td>\n",
       "      <td>0.610955</td>\n",
       "      <td>0.477293</td>\n",
       "      <td>0.351885</td>\n",
       "      <td>0.451853</td>\n",
       "      <td>0.371189</td>\n",
       "      <td>0.545109</td>\n",
       "      <td>0.538667</td>\n",
       "      <td>0.605428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>0.145249</td>\n",
       "      <td>0.124606</td>\n",
       "      <td>0.473124</td>\n",
       "      <td>0.583956</td>\n",
       "      <td>0.719232</td>\n",
       "      <td>0.661353</td>\n",
       "      <td>0.850265</td>\n",
       "      <td>0.781021</td>\n",
       "      <td>0.991903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>0.052735</td>\n",
       "      <td>0.083277</td>\n",
       "      <td>0.149116</td>\n",
       "      <td>0.632689</td>\n",
       "      <td>0.644945</td>\n",
       "      <td>0.756778</td>\n",
       "      <td>0.719076</td>\n",
       "      <td>0.331694</td>\n",
       "      <td>0.460232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>0.063478</td>\n",
       "      <td>0.227909</td>\n",
       "      <td>0.275428</td>\n",
       "      <td>0.955571</td>\n",
       "      <td>0.537915</td>\n",
       "      <td>0.527843</td>\n",
       "      <td>0.349297</td>\n",
       "      <td>0.151907</td>\n",
       "      <td>0.209777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>0.037559</td>\n",
       "      <td>0.066152</td>\n",
       "      <td>0.103251</td>\n",
       "      <td>0.829570</td>\n",
       "      <td>0.406981</td>\n",
       "      <td>0.611470</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>0.109470</td>\n",
       "      <td>0.072589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>0.111025</td>\n",
       "      <td>0.140537</td>\n",
       "      <td>0.375651</td>\n",
       "      <td>0.337526</td>\n",
       "      <td>0.561633</td>\n",
       "      <td>0.583339</td>\n",
       "      <td>0.550147</td>\n",
       "      <td>0.823484</td>\n",
       "      <td>0.689789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>0.640733</td>\n",
       "      <td>0.640657</td>\n",
       "      <td>0.642801</td>\n",
       "      <td>0.734546</td>\n",
       "      <td>0.741562</td>\n",
       "      <td>0.804534</td>\n",
       "      <td>0.728016</td>\n",
       "      <td>0.843385</td>\n",
       "      <td>0.893712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>0.304821</td>\n",
       "      <td>0.276959</td>\n",
       "      <td>0.519766</td>\n",
       "      <td>0.316289</td>\n",
       "      <td>0.478726</td>\n",
       "      <td>0.460461</td>\n",
       "      <td>0.462791</td>\n",
       "      <td>0.610656</td>\n",
       "      <td>0.587612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>0.375721</td>\n",
       "      <td>0.332330</td>\n",
       "      <td>0.334817</td>\n",
       "      <td>0.318067</td>\n",
       "      <td>0.411782</td>\n",
       "      <td>0.443272</td>\n",
       "      <td>0.427828</td>\n",
       "      <td>0.656553</td>\n",
       "      <td>0.621862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>0.372264</td>\n",
       "      <td>0.461628</td>\n",
       "      <td>0.831169</td>\n",
       "      <td>0.446549</td>\n",
       "      <td>0.596457</td>\n",
       "      <td>0.702927</td>\n",
       "      <td>0.761623</td>\n",
       "      <td>0.652697</td>\n",
       "      <td>0.531782</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID        Ll        Lm        Lh        S0        S3  \\\n",
       "0      TTHERM_000000042  0.640498  0.610955  0.477293  0.351885  0.451853   \n",
       "1      TTHERM_000000045  0.145249  0.124606  0.473124  0.583956  0.719232   \n",
       "2       TTHERM_00000010  0.052735  0.083277  0.149116  0.632689  0.644945   \n",
       "3       TTHERM_00000020  0.063478  0.227909  0.275428  0.955571  0.537915   \n",
       "4       TTHERM_00000030  0.037559  0.066152  0.103251  0.829570  0.406981   \n",
       "...                 ...       ...       ...       ...       ...       ...   \n",
       "20321   TTHERM_02293890  0.111025  0.140537  0.375651  0.337526  0.561633   \n",
       "20322   TTHERM_02385080  0.640733  0.640657  0.642801  0.734546  0.741562   \n",
       "20323   TTHERM_02555200  0.304821  0.276959  0.519766  0.316289  0.478726   \n",
       "20324   TTHERM_02607240  0.375721  0.332330  0.334817  0.318067  0.411782   \n",
       "20325   TTHERM_02653470  0.372264  0.461628  0.831169  0.446549  0.596457   \n",
       "\n",
       "             S6        S9       S15       S24  \n",
       "0      0.371189  0.545109  0.538667  0.605428  \n",
       "1      0.661353  0.850265  0.781021  0.991903  \n",
       "2      0.756778  0.719076  0.331694  0.460232  \n",
       "3      0.527843  0.349297  0.151907  0.209777  \n",
       "4      0.611470  0.461538  0.109470  0.072589  \n",
       "...         ...       ...       ...       ...  \n",
       "20321  0.583339  0.550147  0.823484  0.689789  \n",
       "20322  0.804534  0.728016  0.843385  0.893712  \n",
       "20323  0.460461  0.462791  0.610656  0.587612  \n",
       "20324  0.443272  0.427828  0.656553  0.621862  \n",
       "20325  0.702927  0.761623  0.652697  0.531782  \n",
       "\n",
       "[20326 rows x 10 columns]"
      ]
     },
     "execution_count": 468,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "veg_mean = get_geom_mean_expression(veg_filtered_norm_df)\n",
    "veg_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "id": "1508416d-d531-4f87-85cf-0db9fd0a0916",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>C0</th>\n",
       "      <th>C4</th>\n",
       "      <th>C6</th>\n",
       "      <th>C8</th>\n",
       "      <th>C10</th>\n",
       "      <th>C12</th>\n",
       "      <th>C14</th>\n",
       "      <th>C16</th>\n",
       "      <th>C18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>0.164791</td>\n",
       "      <td>0.723792</td>\n",
       "      <td>0.532334</td>\n",
       "      <td>0.961895</td>\n",
       "      <td>0.510857</td>\n",
       "      <td>0.700169</td>\n",
       "      <td>0.825905</td>\n",
       "      <td>0.459875</td>\n",
       "      <td>0.468452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>0.919002</td>\n",
       "      <td>0.703379</td>\n",
       "      <td>0.517879</td>\n",
       "      <td>0.499102</td>\n",
       "      <td>0.786583</td>\n",
       "      <td>0.716830</td>\n",
       "      <td>0.725345</td>\n",
       "      <td>0.613134</td>\n",
       "      <td>0.724042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>0.918367</td>\n",
       "      <td>0.263778</td>\n",
       "      <td>0.169134</td>\n",
       "      <td>0.238973</td>\n",
       "      <td>0.206974</td>\n",
       "      <td>0.262770</td>\n",
       "      <td>0.455647</td>\n",
       "      <td>0.435617</td>\n",
       "      <td>0.439906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>0.538202</td>\n",
       "      <td>0.302916</td>\n",
       "      <td>0.196585</td>\n",
       "      <td>0.171958</td>\n",
       "      <td>0.159344</td>\n",
       "      <td>0.178610</td>\n",
       "      <td>0.184527</td>\n",
       "      <td>0.287901</td>\n",
       "      <td>0.154389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>0.166954</td>\n",
       "      <td>0.214577</td>\n",
       "      <td>0.088220</td>\n",
       "      <td>0.037427</td>\n",
       "      <td>0.062772</td>\n",
       "      <td>0.075196</td>\n",
       "      <td>0.058766</td>\n",
       "      <td>0.032336</td>\n",
       "      <td>0.059593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>0.654662</td>\n",
       "      <td>0.199828</td>\n",
       "      <td>0.093582</td>\n",
       "      <td>0.248135</td>\n",
       "      <td>0.754035</td>\n",
       "      <td>0.775166</td>\n",
       "      <td>0.605974</td>\n",
       "      <td>0.582757</td>\n",
       "      <td>0.683368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>0.469278</td>\n",
       "      <td>0.474059</td>\n",
       "      <td>0.394062</td>\n",
       "      <td>0.592505</td>\n",
       "      <td>0.598329</td>\n",
       "      <td>0.244594</td>\n",
       "      <td>0.627762</td>\n",
       "      <td>0.733024</td>\n",
       "      <td>0.814996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>0.480840</td>\n",
       "      <td>0.591126</td>\n",
       "      <td>0.617040</td>\n",
       "      <td>0.622603</td>\n",
       "      <td>0.319449</td>\n",
       "      <td>0.351828</td>\n",
       "      <td>0.523217</td>\n",
       "      <td>0.796344</td>\n",
       "      <td>0.848257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>0.929297</td>\n",
       "      <td>0.739198</td>\n",
       "      <td>0.619243</td>\n",
       "      <td>0.588040</td>\n",
       "      <td>0.160268</td>\n",
       "      <td>0.612148</td>\n",
       "      <td>0.804887</td>\n",
       "      <td>0.799902</td>\n",
       "      <td>0.602232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>0.599931</td>\n",
       "      <td>0.291566</td>\n",
       "      <td>0.236501</td>\n",
       "      <td>0.318720</td>\n",
       "      <td>0.290800</td>\n",
       "      <td>0.524813</td>\n",
       "      <td>0.788421</td>\n",
       "      <td>0.850434</td>\n",
       "      <td>0.789467</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID        C0        C4        C6        C8       C10  \\\n",
       "0      TTHERM_000000042  0.164791  0.723792  0.532334  0.961895  0.510857   \n",
       "1      TTHERM_000000045  0.919002  0.703379  0.517879  0.499102  0.786583   \n",
       "2       TTHERM_00000010  0.918367  0.263778  0.169134  0.238973  0.206974   \n",
       "3       TTHERM_00000020  0.538202  0.302916  0.196585  0.171958  0.159344   \n",
       "4       TTHERM_00000030  0.166954  0.214577  0.088220  0.037427  0.062772   \n",
       "...                 ...       ...       ...       ...       ...       ...   \n",
       "20321   TTHERM_02293890  0.654662  0.199828  0.093582  0.248135  0.754035   \n",
       "20322   TTHERM_02385080  0.469278  0.474059  0.394062  0.592505  0.598329   \n",
       "20323   TTHERM_02555200  0.480840  0.591126  0.617040  0.622603  0.319449   \n",
       "20324   TTHERM_02607240  0.929297  0.739198  0.619243  0.588040  0.160268   \n",
       "20325   TTHERM_02653470  0.599931  0.291566  0.236501  0.318720  0.290800   \n",
       "\n",
       "            C12       C14       C16       C18  \n",
       "0      0.700169  0.825905  0.459875  0.468452  \n",
       "1      0.716830  0.725345  0.613134  0.724042  \n",
       "2      0.262770  0.455647  0.435617  0.439906  \n",
       "3      0.178610  0.184527  0.287901  0.154389  \n",
       "4      0.075196  0.058766  0.032336  0.059593  \n",
       "...         ...       ...       ...       ...  \n",
       "20321  0.775166  0.605974  0.582757  0.683368  \n",
       "20322  0.244594  0.627762  0.733024  0.814996  \n",
       "20323  0.351828  0.523217  0.796344  0.848257  \n",
       "20324  0.612148  0.804887  0.799902  0.602232  \n",
       "20325  0.524813  0.788421  0.850434  0.789467  \n",
       "\n",
       "[20326 rows x 10 columns]"
      ]
     },
     "execution_count": 469,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sex_mean = get_geom_mean_expression(sex_filtered_norm_df)\n",
    "sex_mean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d243d8b-6248-4f01-a443-215c5a01412d",
   "metadata": {},
   "source": [
    "Get the annnotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "id": "6dce5f9d-6cbb-4705-97be-33211caad9ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annot = pd.read_csv('../eggnog/complete_eggnog_annotation.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc78ff5-bc76-4957-bf55-9021c18e88c7",
   "metadata": {},
   "source": [
    "Add the TGD annotations for all the genes to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "id": "6175d66a-0d48-4b4e-a799-4018c4e00f13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_01528530</td>\n",
       "      <td>5911.EAS00195</td>\n",
       "      <td>2.2e-14</td>\n",
       "      <td>49.9</td>\n",
       "      <td>KOG0594@1|root,KOG0594@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>G</td>\n",
       "      <td>cyclin-dependent protein serine/threonine kina...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_01528510</td>\n",
       "      <td>5911.EAR81750</td>\n",
       "      <td>1.1e-194</td>\n",
       "      <td>633.4</td>\n",
       "      <td>2EI75@1|root,2SNPE@2759|Eukaryota,3ZBUW@5878|C...</td>\n",
       "      <td>5878|Ciliophora</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_01528500</td>\n",
       "      <td>5911.EAR81749</td>\n",
       "      <td>4e-141</td>\n",
       "      <td>462.0</td>\n",
       "      <td>COG1100@1|root,KOG0074@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>KLT</td>\n",
       "      <td>GTP binding</td>\n",
       "      <td>ARL13B</td>\n",
       "      <td>GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko04031</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ADP-ribosylation factor family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_0015284992</td>\n",
       "      <td>5911.EAR97791</td>\n",
       "      <td>8.4e-64</td>\n",
       "      <td>205.1</td>\n",
       "      <td>2ERUB@1|root,2SUIJ@2759|Eukaryota</td>\n",
       "      <td>5911.EAR97791|-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Leucine-rich repeat-containing protein 74A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00897120</td>\n",
       "      <td>5911.EAR86071</td>\n",
       "      <td>5.2e-77</td>\n",
       "      <td>248.8</td>\n",
       "      <td>2ERUB@1|root,2SUIJ@2759|Eukaryota</td>\n",
       "      <td>5911.EAR86071|-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>kinase domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26253</th>\n",
       "      <td>TTHERM_01076820</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26254</th>\n",
       "      <td>TTHERM_001076861</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26255</th>\n",
       "      <td>TTHERM_01076910</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26256</th>\n",
       "      <td>TTHERM_001076981</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26257</th>\n",
       "      <td>TTHERM_01077010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26258 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               TTHERM_ID  seed_ortholog    evalue  score  \\\n",
       "0        TTHERM_01528530  5911.EAS00195   2.2e-14   49.9   \n",
       "1        TTHERM_01528510  5911.EAR81750  1.1e-194  633.4   \n",
       "2        TTHERM_01528500  5911.EAR81749    4e-141  462.0   \n",
       "3      TTHERM_0015284992  5911.EAR97791   8.4e-64  205.1   \n",
       "4        TTHERM_00897120  5911.EAR86071   5.2e-77  248.8   \n",
       "...                  ...            ...       ...    ...   \n",
       "26253    TTHERM_01076820              -         -      -   \n",
       "26254   TTHERM_001076861              -         -      -   \n",
       "26255    TTHERM_01076910              -         -      -   \n",
       "26256   TTHERM_001076981              -         -      -   \n",
       "26257    TTHERM_01077010              -         -      -   \n",
       "\n",
       "                                              eggNOG_OGs    max_annot_lvl  \\\n",
       "0                  KOG0594@1|root,KOG0594@2759|Eukaryota   2759|Eukaryota   \n",
       "1      2EI75@1|root,2SNPE@2759|Eukaryota,3ZBUW@5878|C...  5878|Ciliophora   \n",
       "2                  COG1100@1|root,KOG0074@2759|Eukaryota   2759|Eukaryota   \n",
       "3                      2ERUB@1|root,2SUIJ@2759|Eukaryota  5911.EAR97791|-   \n",
       "4                      2ERUB@1|root,2SUIJ@2759|Eukaryota  5911.EAR86071|-   \n",
       "...                                                  ...              ...   \n",
       "26253                                                  -                -   \n",
       "26254                                                  -                -   \n",
       "26255                                                  -                -   \n",
       "26256                                                  -                -   \n",
       "26257                                                  -                -   \n",
       "\n",
       "      COG_category                                        Description  \\\n",
       "0                G  cyclin-dependent protein serine/threonine kina...   \n",
       "1                -                                                  -   \n",
       "2              KLT                                        GTP binding   \n",
       "3                -                                                  -   \n",
       "4                -                                                  -   \n",
       "...            ...                                                ...   \n",
       "26253            -                                                  -   \n",
       "26254            -                                                  -   \n",
       "26255            -                                                  -   \n",
       "26256            -                                                  -   \n",
       "26257            -                                                  -   \n",
       "\n",
       "      Preferred_name                                                GOs  ...  \\\n",
       "0                  -                                                  -  ...   \n",
       "1                  -                                                  -  ...   \n",
       "2             ARL13B  GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...  ...   \n",
       "3                  -                                                  -  ...   \n",
       "4                  -                                                  -  ...   \n",
       "...              ...                                                ...  ...   \n",
       "26253              -                                                  -  ...   \n",
       "26254              -                                                  -  ...   \n",
       "26255              -                                                  -  ...   \n",
       "26256              -                                                  -  ...   \n",
       "26257              -                                                  -  ...   \n",
       "\n",
       "      KEGG_Pathway KEGG_Module KEGG_Reaction KEGG_rclass            BRITE  \\\n",
       "0                -           -             -           -                -   \n",
       "1                -           -             -           -                -   \n",
       "2                -           -             -           -  ko00000,ko04031   \n",
       "3                -           -             -           -                -   \n",
       "4                -           -             -           -                -   \n",
       "...            ...         ...           ...         ...              ...   \n",
       "26253            -           -             -           -                -   \n",
       "26254            -           -             -           -                -   \n",
       "26255            -           -             -           -                -   \n",
       "26256            -           -             -           -                -   \n",
       "26257            -           -             -           -                -   \n",
       "\n",
       "      KEGG_TC CAZy BiGG_Reaction PFAMs  \\\n",
       "0           -    -             -     -   \n",
       "1           -    -             -     -   \n",
       "2           -    -             -     -   \n",
       "3           -    -             -     -   \n",
       "4           -    -             -     -   \n",
       "...       ...  ...           ...   ...   \n",
       "26253       -    -             -     -   \n",
       "26254       -    -             -     -   \n",
       "26255       -    -             -     -   \n",
       "26256       -    -             -     -   \n",
       "26257       -    -             -     -   \n",
       "\n",
       "                              TGD2021_description  \n",
       "0                                  protein kinase  \n",
       "1                            hypothetical protein  \n",
       "2          ADP-ribosylation factor family protein  \n",
       "3      Leucine-rich repeat-containing protein 74A  \n",
       "4                           kinase domain protein  \n",
       "...                                           ...  \n",
       "26253                             Endoglucanase 1  \n",
       "26254              transmembrane protein putative  \n",
       "26255                             Endoglucanase 1  \n",
       "26256                             Endoglucanase 1  \n",
       "26257                             Endoglucanase 1  \n",
       "\n",
       "[26258 rows x 22 columns]"
      ]
     },
     "execution_count": 471,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "id": "5a190458-1a72-44d0-9389-47bab21b1ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checkout = \"\"\"\n",
    "# TTHERM_00000030\\nTTHERM_000001459\\nTTHERM_000002658\\nTTHERM_000002709\\nTTHERM_000011269\\nTTHERM_000011798\\nTTHERM_00001210\\nTTHERM_00001310\\nTTHERM_000013569\\nTTHERM_000013649\\nTTHERM_000016298\\nTTHERM_000028468\\nTTHERM_000028809\\nTTHERM_000030519\\nTTHERM_000033909\\nTTHERM_000035399\\nTTHERM_000037449\\nTTHERM_000037782\\nTTHERM_000039088\\nTTHERM_000041632\\nTTHERM_000041635\\nTTHERM_000042736\\nTTHERM_000042742\\nTTHERM_000051719\\nTTHERM_000058888\\nTTHERM_000059119\\nTTHERM_000059499\\nTTHERM_00006050\\nTTHERM_00006170\\nTTHERM_00006290\\nTTHERM_000066709\\nTTHERM_000066939\\nTTHERM_000069529\\nTTHERM_000075950\\nTTHERM_000079989\\nTTHERM_000083289\\nTTHERM_000083329\\nTTHERM_000085199\\nTTHERM_000086777\\nTTHERM_000092856\\nTTHERM_000092919\\nTTHERM_000095648\\nTTHERM_000096738\\nTTHERM_000096767\\nTTHERM_000097819\\nTTHERM_000099988\\nTTHERM_000101249\\nTTHERM_000105399\\nTTHERM_000105638\\nTTHERM_000105664\\nTTHERM_000115543\\nTTHERM_000123578\\nTTHERM_000123759\\nTTHERM_000123799\\nTTHERM_000125189\\nTTHERM_000128699\\nTTHERM_000128738\\nTTHERM_000128899\\nTTHERM_000134819\\nTTHERM_000135979\\nTTHERM_000136019\\nTTHERM_000136238\\nTTHERM_00013650\\nTTHERM_000142399\\nTTHERM_000145039\\nTTHERM_000145409\\nTTHERM_00014930\\nTTHERM_000150028\\nTTHERM_000151655\\nTTHERM_000151657\\nTTHERM_000151659\\nTTHERM_000151707\\nTTHERM_000155257\\nTTHERM_000158289\\nTTHERM_00016230\\nTTHERM_00016450\\nTTHERM_000170478\\nTTHERM_000170479\\nTTHERM_000182182\\nTTHERM_000185789\\nTTHERM_000188551\\nTTHERM_000191328\\nTTHERM_000191709\\nTTHERM_000193177\\nTTHERM_00019600\\nTTHERM_00019650\\nTTHERM_000198388\\nTTHERM_000198509\\nTTHERM_000201618\\nTTHERM_000204119\\nTTHERM_000205206\\nTTHERM_000207258\\nTTHERM_000212534\\nTTHERM_000213644\\nTTHERM_000216168\\nTTHERM_000218239\\nTTHERM_000218619\\nTTHERM_000218939\\nTTHERM_000222369\\nTTHERM_000225898\\nTTHERM_000225988\\nTTHERM_000227538\\nTTHERM_000227669\\nTTHERM_000227689\\nTTHERM_000238739\\nTTHERM_000239299\\nTTHERM_00024030\\nTTHERM_00024149\\nTTHERM_000241779\\nTTHERM_000242459\\nTTHERM_000242629\\nTTHERM_000243879\\nTTHERM_000244008\\nTTHERM_000245699\\nTTHERM_000246949\\nTTHERM_000248239\\nTTHERM_000251049\\nTTHERM_000252389\\nTTHERM_000252419\\nTTHERM_000257099\\nTTHERM_000261919\\nTTHERM_000264819\\nTTHERM_000266261\\nTTHERM_000268259\\nTTHERM_000274479\\nTTHERM_000275799\\nTTHERM_000277338\\nTTHERM_000279719\\nTTHERM_000279858\\nTTHERM_000284240\\nTTHERM_00028800\\nTTHERM_000294649\\nTTHERM_000294809\\nTTHERM_00030000\\nTTHERM_000300069\\nTTHERM_000300369\\nTTHERM_000300459\\nTTHERM_000300489\\nTTHERM_000300549\\nTTHERM_00030070\\nTTHERM_00030150\\nTTHERM_00030220\\nTTHERM_00030420\\nTTHERM_000304243\\nTTHERM_000305499\\nTTHERM_000307767\\nTTHERM_000309879\\nTTHERM_000310179\\nTTHERM_000310809\\nTTHERM_000316229\\nTTHERM_000317059\\nTTHERM_000320389\\nTTHERM_000320439\\nTTHERM_00032870\\nTTHERM_000334255\\nTTHERM_000335848\\nTTHERM_000343739\\nTTHERM_000344109\\nTTHERM_000344329\\nTTHERM_000344368\\nTTHERM_00034920\\nTTHERM_00035270\\nTTHERM_000355149\\nTTHERM_000355799\\nTTHERM_00035670\\nTTHERM_000361561\\nTTHERM_000362979\\nTTHERM_000364298\\nTTHERM_000365339\\nTTHERM_000365478\\nTTHERM_000365519\\nTTHERM_00036990\\nTTHERM_000372548\\nTTHERM_00037610\\nTTHERM_00037730\\nTTHERM_000378449\\nTTHERM_000382129\\nTTHERM_000383469\\nTTHERM_000388289\\nTTHERM_000390049\\nTTHERM_00039090\\nTTHERM_000391477\\nTTHERM_000396989\\nTTHERM_000398140\\nTTHERM_000401919\\nTTHERM_000402210\\nTTHERM_000404309\\nTTHERM_000406649\\nTTHERM_000408795\\nTTHERM_000408949\\nTTHERM_000415569\\nTTHERM_000418349\\nTTHERM_000418498\\nTTHERM_000420499\\nTTHERM_000423265\\nTTHERM_000423321\\nTTHERM_000424459\\nTTHERM_000424558\\nTTHERM_00042640\\nTTHERM_00042740\\nTTHERM_000429641\\nTTHERM_000429979\\nTTHERM_000433738\\nTTHERM_000436229\\nTTHERM_000437619\\nTTHERM_000440509\\nTTHERM_000440549\\nTTHERM_000442267\\nTTHERM_000442799\\nTTHERM_000443079\\nTTHERM_000445939\\nTTHERM_000448823\\nTTHERM_000455598\\nTTHERM_000455680\\nTTHERM_000456708\\nTTHERM_000458197\\nTTHERM_000458218\\nTTHERM_00046100\\nTTHERM_00046110\\nTTHERM_000461810\\nTTHERM_00046300\\nTTHERM_000463588\\nTTHERM_000463589\\nTTHERM_000463838\\nTTHERM_00046460\\nTTHERM_000467909\\nTTHERM_000471259\\nTTHERM_000471329\\nTTHERM_000473239\\nTTHERM_00047350\\nTTHERM_000474539\\nTTHERM_000476461\\nTTHERM_000486249\\nTTHERM_000487049\\nTTHERM_00048760\\nTTHERM_00048780\\nTTHERM_000488179\\nTTHERM_000490799\\nTTHERM_000491079\\nTTHERM_00049180\\nTTHERM_000492408\\nTTHERM_00050530\\nTTHERM_000516450\\nTTHERM_000516461\\nTTHERM_000519809\\nTTHERM_000519829\\nTTHERM_00052070\\nTTHERM_00052120\\nTTHERM_000526429\\nTTHERM_000530108\\nTTHERM_000531819\\nTTHERM_000535351\\nTTHERM_000537309\\nTTHERM_00053940\\nTTHERM_000541549\\nTTHERM_000544739\\nTTHERM_000548089\\nTTHERM_000548129\\nTTHERM_000558438\\nTTHERM_000560099\\nTTHERM_000560142\\nTTHERM_000561598\\nTTHERM_000571589\\nTTHERM_000572032\\nTTHERM_000573209\\nTTHERM_000574269\\nTTHERM_000575519\\nTTHERM_000575579\\nTTHERM_000577049\\nTTHERM_000578441\\nTTHERM_000580367\\nTTHERM_000581749\\nTTHERM_000581869\\nTTHERM_000582139\\nTTHERM_00058360\\nTTHERM_00058400\\nTTHERM_000584531\\nTTHERM_000584749\\nTTHERM_000584759\\nTTHERM_000585019\\nTTHERM_000585149\\nTTHERM_00058580\\nTTHERM_000586861\\nTTHERM_00058950\\nTTHERM_000590479\\nTTHERM_000597629\\nTTHERM_000598670\\nTTHERM_000600499\\nTTHERM_000616089\\nTTHERM_000616459\\nTTHERM_000619890\\nTTHERM_000624648\\nTTHERM_000630099\\nTTHERM_000633271\\nTTHERM_000633479\\nTTHERM_000635749\\nTTHERM_000637208\\nTTHERM_000637209\\nTTHERM_000640080\\nTTHERM_000641369\\nTTHERM_000644779\\nTTHERM_000646835\\nTTHERM_000646840\\nTTHERM_000646899\\nTTHERM_000647419\\nTTHERM_000648878\\nTTHERM_000652549\\nTTHERM_000654259\\nTTHERM_000655619\\nTTHERM_000657361\\nTTHERM_000672289\\nTTHERM_000678048\\nTTHERM_000681909\\nTTHERM_000691225\\nTTHERM_00069190\\nTTHERM_000691929\\nTTHERM_000693107\\nTTHERM_000693288\\nTTHERM_000695779\\nTTHERM_00069600\\nTTHERM_000699759\\nTTHERM_000703757\\nTTHERM_00070870\\nTTHERM_00071130\\nTTHERM_000716179\\nTTHERM_000717709\\nTTHERM_000717999\\nTTHERM_000723051\\nTTHERM_000723389\\nTTHERM_000726079\\nTTHERM_000726339\\nTTHERM_000731450\\nTTHERM_000748947\\nTTHERM_000748978\\nTTHERM_000753493\\nTTHERM_000753495\\nTTHERM_000753599\\nTTHERM_00075500\\nTTHERM_00075540\\nTTHERM_000756121\\nTTHERM_00075660\\nTTHERM_00075690\\nTTHERM_00075730\\nTTHERM_00075750\\nTTHERM_000762999\\nTTHERM_000766429\\nTTHERM_000768570\\nTTHERM_000770619\\nTTHERM_000773079\\nTTHERM_00077360\\nTTHERM_000774831\\nTTHERM_00077530\\nTTHERM_00077560\\nTTHERM_00077600\\nTTHERM_00077810\\nTTHERM_000783219\\nTTHERM_00079290\\nTTHERM_00079550\\nTTHERM_00079580\\nTTHERM_000798148\\nTTHERM_000800229\\nTTHERM_000801323\\nTTHERM_000804799\\nTTHERM_000810475\\nTTHERM_000810562\\nTTHERM_00081100\\nTTHERM_000812603\\nTTHERM_000815140\\nTTHERM_000817290\\nTTHERM_000823383\\nTTHERM_000827189\\nTTHERM_00083400\\nTTHERM_00083780\\nTTHERM_00083790\\nTTHERM_00083800\\nTTHERM_00083930\\nTTHERM_000845809\\nTTHERM_00085250\\nTTHERM_000862770\\nTTHERM_000867570\\nTTHERM_00086820\\nTTHERM_000873623\\nTTHERM_000881459\\nTTHERM_000885885\\nTTHERM_000885891\\nTTHERM_000886917\\nTTHERM_000891240\\nTTHERM_00089200\\nTTHERM_000895865\\nTTHERM_000895969\\nTTHERM_000898161\\nTTHERM_000901634\\nTTHERM_000901749\\nTTHERM_00090250\\nTTHERM_00090320\\nTTHERM_00090340\\nTTHERM_000903858\\nTTHERM_00090430\\nTTHERM_00091520\\nTTHERM_00091580\\nTTHERM_00091632\\nTTHERM_000923209\\nTTHERM_000925559\\nTTHERM_00092860\\nTTHERM_00094020\\nTTHERM_000942839\\nTTHERM_000944157\\nTTHERM_000948690\\nTTHERM_000951962\\nTTHERM_000956521\\nTTHERM_000957531\\nTTHERM_000957669\\nTTHERM_000963272\\nTTHERM_000964350\\nTTHERM_000965394\\nTTHERM_000967548\\nTTHERM_00096769\\nTTHERM_00097820\\nTTHERM_000985179\\nTTHERM_000993039\\nTTHERM_000997598\\nTTHERM_001001378\\nTTHERM_001008626\\nTTHERM_001009941\\nTTHERM_00101250\\nTTHERM_001018451\\nTTHERM_001026363\\nTTHERM_001027719\\nTTHERM_001034480\\nTTHERM_001035719\\nTTHERM_001035762\\nTTHERM_001037789\\nTTHERM_001038803\\nTTHERM_001041997\\nTTHERM_001042011\\nTTHERM_001044673\\nTTHERM_00105400\\nTTHERM_001054129\\nTTHERM_001054139\\nTTHERM_001054349\\nTTHERM_001055501\\nTTHERM_00107080\\nTTHERM_00107110\\nTTHERM_001076814\\nTTHERM_001081695\\nTTHERM_00108220\\nTTHERM_001100419\\nTTHERM_001107439\\nTTHERM_00112480\\nTTHERM_001126428\\nTTHERM_00112940\\nTTHERM_001140389\\nTTHERM_001141651\\nTTHERM_001144931\\nTTHERM_001144951\\nTTHERM_001144953\\nTTHERM_001147129\\nTTHERM_001159889\\nTTHERM_001164069\\nTTHERM_001166292\\nTTHERM_001169451\\nTTHERM_001170538\\nTTHERM_00117550\\nTTHERM_001186268\\nTTHERM_001192479\\nTTHERM_001198169\\nTTHERM_001202231\\nTTHERM_001212889\\nTTHERM_001214996\\nTTHERM_00122120\\nTTHERM_00122450\\nTTHERM_00122460\\nTTHERM_00122470\\nTTHERM_00122549\\nTTHERM_001225669\\nTTHERM_001230168\\nTTHERM_001232265\\nTTHERM_00123640\\nTTHERM_00123900\\nTTHERM_00123960\\nTTHERM_001250151\\nTTHERM_00125220\\nTTHERM_00125660\\nTTHERM_00125750\\nTTHERM_001262899\\nTTHERM_001264989\\nTTHERM_00127060\\nTTHERM_00127090\\nTTHERM_001273284\\nTTHERM_001284761\\nTTHERM_00128590\\nTTHERM_001287954\\nTTHERM_00128880\\nTTHERM_00128980\\nTTHERM_00129020\\nTTHERM_001292214\\nTTHERM_001292250\\nTTHERM_00129270\\nTTHERM_001295314\\nTTHERM_00129570\\nTTHERM_00129640\\nTTHERM_00129790\\nTTHERM_00129930\\nTTHERM_001302821\\nTTHERM_001305848\\nTTHERM_001305849\\nTTHERM_00131130\\nTTHERM_00131210\\nTTHERM_00131320\\nTTHERM_001325754\\nTTHERM_001325773\\nTTHERM_001333220\\nTTHERM_00133610\\nTTHERM_001344745\\nTTHERM_00136160\\nTTHERM_001371771\\nTTHERM_001374862\\nTTHERM_001374871\\nTTHERM_00137530\\nTTHERM_00137830\\nTTHERM_00138000\\nTTHERM_00138110\\nTTHERM_00138300\\nTTHERM_001383000\\nTTHERM_00141150\\nTTHERM_001415159\\nTTHERM_001420339\\nTTHERM_001421352\\nTTHERM_00142240\\nTTHERM_001423407\\nTTHERM_00143560\\nTTHERM_00143570\\nTTHERM_00145060\\nTTHERM_001451032\\nTTHERM_00145630\\nTTHERM_001473460\\nTTHERM_001476529\\nTTHERM_00148810\\nTTHERM_00149080\\nTTHERM_00149990\\nTTHERM_001500990\\nTTHERM_00151410\\nTTHERM_00151620\\nTTHERM_00151760\\nTTHERM_001527452\\nTTHERM_001541732\\nTTHERM_00155240\\nTTHERM_00155510\\nTTHERM_00155530\\nTTHERM_00156762\\nTTHERM_00158530\\nTTHERM_00160820\\nTTHERM_00161150\\nTTHERM_00161680\\nTTHERM_00170200\\nTTHERM_00170580\\nTTHERM_00170590\\nTTHERM_00171650\\nTTHERM_00171680\\nTTHERM_00171690\\nTTHERM_00171700\\nTTHERM_00185130\\nTTHERM_00185150\\nTTHERM_00185240\\nTTHERM_00185380\\nTTHERM_00185530\\nTTHERM_00185590\\nTTHERM_00185620\\nTTHERM_00186020\\nTTHERM_00187060\\nTTHERM_00187200\\nTTHERM_00187210\\nTTHERM_00187230\\nTTHERM_00189240\\nTTHERM_00189410\\nTTHERM_00189480\\nTTHERM_00190650\\nTTHERM_00190700\\nTTHERM_00190850\\nTTHERM_00191360\\nTTHERM_00191470\\nTTHERM_00191650\\nTTHERM_00191990\\nTTHERM_00192070\\nTTHERM_00192100\\nTTHERM_00192140\\nTTHERM_00193240\\nTTHERM_00193280\\nTTHERM_00193290\\nTTHERM_00193730\\nTTHERM_00194130\\nTTHERM_00194430\\nTTHERM_00196340\\nTTHERM_00196500\\nTTHERM_00197685\\nTTHERM_00198200\\nTTHERM_00198240\\nTTHERM_00198350\\nTTHERM_00198510\\nTTHERM_00198520\\nTTHERM_00200560\\nTTHERM_00201640\\nTTHERM_00202830\\nTTHERM_00202860\\nTTHERM_00203030\\nTTHERM_00204160\\nTTHERM_00207259\\nTTHERM_00209330\\nTTHERM_00209390\\nTTHERM_00210490\\nTTHERM_00214670\\nTTHERM_00216110\\nTTHERM_00218710\\nTTHERM_00220880\\nTTHERM_00220900\\nTTHERM_00222350\\nTTHERM_00222370\\nTTHERM_00222400\\nTTHERM_00225670\\nTTHERM_00225750\\nTTHERM_00225930\\nTTHERM_00225980\\nTTHERM_00227350\\nTTHERM_00227430\\nTTHERM_00228920\\nTTHERM_00229940\\nTTHERM_00233020\\nTTHERM_00234140\\nTTHERM_00234180\\nTTHERM_00236260\\nTTHERM_00237340\\nTTHERM_00239010\\nTTHERM_00241480\\nTTHERM_00241920\\nTTHERM_00241930\\nTTHERM_00242420\\nTTHERM_00243700\\nTTHERM_00245240\\nTTHERM_00245610\\nTTHERM_00245810\\nTTHERM_00247160\\nTTHERM_00248270\\nTTHERM_00249590\\nTTHERM_00249670\\nTTHERM_00249690\\nTTHERM_00249790\\nTTHERM_00251020\\nTTHERM_00251050\\nTTHERM_00251110\\nTTHERM_00251150\\nTTHERM_00251240\\nTTHERM_00252260\\nTTHERM_00256930\\nTTHERM_00257030\\nTTHERM_00257040\\nTTHERM_00257130\\nTTHERM_00257180\\nTTHERM_00258260\\nTTHERM_00259290\\nTTHERM_00259339\\nTTHERM_00259540\\nTTHERM_00261920\\nTTHERM_00263000\\nTTHERM_00263080\\nTTHERM_00263120\\nTTHERM_00263200\\nTTHERM_00263320\\nTTHERM_00263409\\nTTHERM_00263420\\nTTHERM_00263620\\nTTHERM_002653450\\nTTHERM_002653530\\nTTHERM_00266280\\nTTHERM_00266540\\nTTHERM_00266660\\nTTHERM_00268020\\nTTHERM_00268080\\nTTHERM_00268250\\nTTHERM_00268300\\nTTHERM_00268320\\nTTHERM_00268340\\nTTHERM_00274510\\nTTHERM_00274530\\nTTHERM_00275880\\nTTHERM_00277170\\nTTHERM_00279660\\nTTHERM_00283160\\nTTHERM_00283320\\nTTHERM_00283500\\nTTHERM_00283570\\nTTHERM_00285260\\nTTHERM_00285330\\nTTHERM_00285479\\nTTHERM_00285650\\nTTHERM_00289110\\nTTHERM_00289160\\nTTHERM_00289490\\nTTHERM_00290620\\nTTHERM_00290650\\nTTHERM_00292110\\nTTHERM_00292120\\nTTHERM_00294600\\nTTHERM_00294630\\nTTHERM_00295350\\nTTHERM_00295680\\nTTHERM_00297100\\nTTHERM_00297120\\nTTHERM_00298490\\nTTHERM_00299790\\nTTHERM_00301870\\nTTHERM_00309930\\nTTHERM_00310010\\nTTHERM_00310020\\nTTHERM_00310460\\nTTHERM_00310690\\nTTHERM_00310780\\nTTHERM_00310810\\nTTHERM_00312020\\nTTHERM_00312030\\nTTHERM_00312682\\nTTHERM_00313170\\nTTHERM_00313200\\nTTHERM_00313230\\nTTHERM_00313390\\nTTHERM_00313580\\nTTHERM_00313660\\nTTHERM_00316130\\nTTHERM_00317250\\nTTHERM_00317380\\nTTHERM_00317420\\nTTHERM_00320039\\nTTHERM_00320160\\nTTHERM_00321540\\nTTHERM_00324460\\nTTHERM_00325630\\nTTHERM_00328520\\nTTHERM_00329780\\nTTHERM_00329910\\nTTHERM_00329980\\nTTHERM_00331050\\nTTHERM_00338160\\nTTHERM_00339740\\nTTHERM_00340130\\nTTHERM_00343300\\nTTHERM_00343640\\nTTHERM_00343740\\nTTHERM_00343840\\nTTHERM_00344080\\nTTHERM_00344110\\nTTHERM_00344130\\nTTHERM_00344391\\nTTHERM_00346910\\nTTHERM_00348640\\nTTHERM_00348890\\nTTHERM_00351140\\nTTHERM_00353430\\nTTHERM_00355150\\nTTHERM_00355169\\nTTHERM_00355440\\nTTHERM_00355480\\nTTHERM_00355650\\nTTHERM_00355850\\nTTHERM_00359160\\nTTHERM_00361610\\nTTHERM_00361700\\nTTHERM_00361750\\nTTHERM_00361800\\nTTHERM_00361900\\nTTHERM_00363110\\nTTHERM_00370950\\nTTHERM_00371170\\nTTHERM_00371290\\nTTHERM_00372630\\nTTHERM_00376203\\nTTHERM_00377210\\nTTHERM_00378420\\nTTHERM_00378650\\nTTHERM_00378980\\nTTHERM_00382400\\nTTHERM_00383520\\nTTHERM_00383590\\nTTHERM_00384650\\nTTHERM_00384980\\nTTHERM_00387150\\nTTHERM_00391260\\nTTHERM_00391470\\nTTHERM_00392890\\nTTHERM_00393110\\nTTHERM_00398090\\nTTHERM_00399150\\nTTHERM_00399340\\nTTHERM_00399660\\nTTHERM_00400760\\nTTHERM_00400810\\nTTHERM_00405580\\nTTHERM_00408920\\nTTHERM_00409090\\nTTHERM_00409121\\nTTHERM_00410291\\nTTHERM_00411440\\nTTHERM_00411450\\nTTHERM_00412000\\nTTHERM_00412040\\nTTHERM_00412090\\nTTHERM_00414390\\nTTHERM_00415690\\nTTHERM_00417880\\nTTHERM_00417950\\nTTHERM_00420590\\nTTHERM_00421190\\nTTHERM_00421200\\nTTHERM_00423380\\nTTHERM_00424473\\nTTHERM_00425890\\nTTHERM_00426050\\nTTHERM_00427580\\nTTHERM_00429710\\nTTHERM_00429729\\nTTHERM_00430130\\nTTHERM_00430180\\nTTHERM_00433440\\nTTHERM_00433760\\nTTHERM_00434220\\nTTHERM_00437300\\nTTHERM_00437490\\nTTHERM_00437520\\nTTHERM_00438810\\nTTHERM_00439260\\nTTHERM_00441700\\nTTHERM_00441800\\nTTHERM_00444810\\nTTHERM_00444860\\nTTHERM_00446060\\nTTHERM_00446100\\nTTHERM_00448770\\nTTHERM_00449230\\nTTHERM_00450830\\nTTHERM_00450860\\nTTHERM_00450900\\nTTHERM_00450910\\nTTHERM_00456700\\nTTHERM_00458160\\nTTHERM_00462820\\nTTHERM_00463220\\nTTHERM_00463460\\nTTHERM_00463700\\nTTHERM_00463840\\nTTHERM_00467400\\nTTHERM_00467440\\nTTHERM_00470780\\nTTHERM_00473230\\nTTHERM_00473240\\nTTHERM_00475350\\nTTHERM_00475380\\nTTHERM_00476690\\nTTHERM_00476890\\nTTHERM_00476920\\nTTHERM_00476952\\nTTHERM_00481170\\nTTHERM_00483410\\nTTHERM_00485850\\nTTHERM_00485960\\nTTHERM_00488380\\nTTHERM_00490860\\nTTHERM_00490960\\nTTHERM_00490970\\nTTHERM_00490980\\nTTHERM_00491000\\nTTHERM_00491060\\nTTHERM_00491190\\nTTHERM_00494300\\nTTHERM_00494730\\nTTHERM_00494780\\nTTHERM_00494848\\nTTHERM_00497690\\nTTHERM_00497720\\nTTHERM_00498080\\nTTHERM_00499390\\nTTHERM_00500730\\nTTHERM_00500940\\nTTHERM_00502350\\nTTHERM_00502450\\nTTHERM_00502530\\nTTHERM_00508990\\nTTHERM_00516410\\nTTHERM_00516430\\nTTHERM_00518470\\nTTHERM_00518480\\nTTHERM_00519840\\nTTHERM_00520930\\nTTHERM_00522400\\nTTHERM_00525120\\nTTHERM_00526300\\nTTHERM_00526340\\nTTHERM_00527290\\nTTHERM_00527420\\nTTHERM_00529490\\nTTHERM_00529520\\nTTHERM_00530039\\nTTHERM_00530050\\nTTHERM_00530410\\nTTHERM_00530659\\nTTHERM_00532140\\nTTHERM_00532300\\nTTHERM_00533900\\nTTHERM_00535200\\nTTHERM_00535390\\nTTHERM_00535510\\nTTHERM_00535650\\nTTHERM_00537310\\nTTHERM_00538999\\nTTHERM_00540020\\nTTHERM_00540040\\nTTHERM_00541510\\nTTHERM_00541520\\nTTHERM_00543570\\nTTHERM_00543640\\nTTHERM_00545880\\nTTHERM_00548380\\nTTHERM_00549450\\nTTHERM_00554440\\nTTHERM_00558700\\nTTHERM_00559850\\nTTHERM_00559960\\nTTHERM_00560130\\nTTHERM_00561150\\nTTHERM_00561210\\nTTHERM_00561430\\nTTHERM_00561540\\nTTHERM_00561610\\nTTHERM_00566910\\nTTHERM_00569070\\nTTHERM_00569230\\nTTHERM_00569340\\nTTHERM_00571750\\nTTHERM_00572180\\nTTHERM_00575460\\nTTHERM_00578740\\nTTHERM_00578970\\nTTHERM_00580400\\nTTHERM_00581460\\nTTHERM_00581909\\nTTHERM_00582039\\nTTHERM_00582040\\nTTHERM_00582190\\nTTHERM_00583400\\nTTHERM_00584770\\nTTHERM_00584820\\nTTHERM_00586560\\nTTHERM_00586690\\nTTHERM_00586840\\nTTHERM_00588860\\nTTHERM_00590300\\nTTHERM_00590310\\nTTHERM_00592870\\nTTHERM_00592970\\nTTHERM_00594180\\nTTHERM_00594240\\nTTHERM_00597620\\nTTHERM_00598660\\nTTHERM_00607120\\nTTHERM_00607170\\nTTHERM_00607260\\nTTHERM_00609390\\nTTHERM_00616020\\nTTHERM_00616490\\nTTHERM_00617800\\nTTHERM_00620900\\nTTHERM_00621040\\nTTHERM_00621570\\nTTHERM_00625970\\nTTHERM_00627080\\nTTHERM_00627280\\nTTHERM_00629770\\nTTHERM_00629890\\nTTHERM_00629940\\nTTHERM_00629970\\nTTHERM_00630180\\nTTHERM_00630350\\nTTHERM_00630370\\nTTHERM_00630530\\nTTHERM_00630570\\nTTHERM_00630620\\nTTHERM_00630710\\nTTHERM_00630720\\nTTHERM_00633200\\nTTHERM_00633210\\nTTHERM_00633339\\nTTHERM_00633640\\nTTHERM_00634672\\nTTHERM_00635690\\nTTHERM_00635760\\nTTHERM_00637190\\nTTHERM_00637430\\nTTHERM_00637460\\nTTHERM_00640060\\nTTHERM_00640070\\nTTHERM_00641260\\nTTHERM_00641340\\nTTHERM_00643410\\nTTHERM_00644660\\nTTHERM_00644670\\nTTHERM_00644760\\nTTHERM_00647450\\nTTHERM_00648820\\nTTHERM_00648830\\nTTHERM_00648850\\nTTHERM_00649510\\nTTHERM_00652530\\nTTHERM_00653700\\nTTHERM_00654000\\nTTHERM_00656070\\nTTHERM_00657540\\nTTHERM_00657590\\nTTHERM_00658720\\nTTHERM_00658990\\nTTHERM_00660350\\nTTHERM_00661510\\nTTHERM_00666940\\nTTHERM_00670130\\nTTHERM_00670150\\nTTHERM_00672050\\nTTHERM_00672110\\nTTHERM_00675720\\nTTHERM_00675750\\nTTHERM_00675770\\nTTHERM_00675780\\nTTHERM_00678340\\nTTHERM_00680640\\nTTHERM_00683040\\nTTHERM_00683320\\nTTHERM_00684670\\nTTHERM_00685830\\nTTHERM_00685840\\nTTHERM_00686130\\nTTHERM_00688530\\nTTHERM_00688640\\nTTHERM_00688740\\nTTHERM_00689830\\nTTHERM_00689870\\nTTHERM_00689910\\nTTHERM_00691620\\nTTHERM_00691910\\nTTHERM_00694310\\nTTHERM_00694320\\nTTHERM_00694530\\nTTHERM_00695840\\nTTHERM_00697030\\nTTHERM_00697050\\nTTHERM_00697120\\nTTHERM_00697122\\nTTHERM_00697570\\nTTHERM_00698720\\nTTHERM_00699750\\nTTHERM_00699830\\nTTHERM_00703350\\nTTHERM_00704030\\nTTHERM_00705090\\nTTHERM_00707470\\nTTHERM_00709560\\nTTHERM_00709630\\nTTHERM_00709650\\nTTHERM_00709670\\nTTHERM_00709700\\nTTHERM_00712000\\nTTHERM_00713000\\nTTHERM_00713260\\nTTHERM_00713430\\nTTHERM_00713540\\nTTHERM_00715940\\nTTHERM_00716160\\nTTHERM_00716170\\nTTHERM_00716330\\nTTHERM_00717530\\nTTHERM_00717600\\nTTHERM_00717650\\nTTHERM_00717880\\nTTHERM_00717950\\nTTHERM_00720100\\nTTHERM_00720102\\nTTHERM_00721190\\nTTHERM_00721470\\nTTHERM_00721560\\nTTHERM_00721870\\nTTHERM_00722970\\nTTHERM_00723520\\nTTHERM_00723650\\nTTHERM_00725950\\nTTHERM_00725960\\nTTHERM_00726030\\nTTHERM_00727530\\nTTHERM_00727670\\nTTHERM_00727740\\nTTHERM_00730250\\nTTHERM_00732680\\nTTHERM_00733980\\nTTHERM_00734070\\nTTHERM_00734140\\nTTHERM_00735190\\nTTHERM_00736530\\nTTHERM_00736532\\nTTHERM_00748940\\nTTHERM_00751032\\nTTHERM_00753610\\nTTHERM_00754680\\nTTHERM_00754690\\nTTHERM_00756180\\nTTHERM_00758900\\nTTHERM_00760440\\nTTHERM_00760550\\nTTHERM_00760760\\nTTHERM_00765350\\nTTHERM_00773180\\nTTHERM_00773200\\nTTHERM_00773440\\nTTHERM_00773550\\nTTHERM_00775870\\nTTHERM_00775910\\nTTHERM_00775949\\nTTHERM_00777040\\nTTHERM_00778400\\nTTHERM_00780520\\nTTHERM_00782150\\nTTHERM_00784250\\nTTHERM_00784730\\nTTHERM_00787130\\nTTHERM_00790550\\nTTHERM_00790850\\nTTHERM_00790870\\nTTHERM_00794010\\nTTHERM_00794220\\nTTHERM_00794360\\nTTHERM_00794460\\nTTHERM_00797800\\nTTHERM_00798050\\nTTHERM_00798120\\nTTHERM_00798140\\nTTHERM_00798170\\nTTHERM_00799182\\nTTHERM_00799210\\nTTHERM_00802360\\nTTHERM_00802530\\nTTHERM_00803590\\nTTHERM_00803690\\nTTHERM_00804780\\nTTHERM_00807900\\nTTHERM_00808050\\nTTHERM_00809110\\nTTHERM_00809130\\nTTHERM_00809220\\nTTHERM_00809250\\nTTHERM_00809320\\nTTHERM_00809420\\nTTHERM_00810530\\nTTHERM_00812910\\nTTHERM_00816240\\nTTHERM_00819450\\nTTHERM_00820750\\nTTHERM_00821880\\nTTHERM_00822150\\nTTHERM_00822280\\nTTHERM_00823420\\nTTHERM_00823840\\nTTHERM_00823890\\nTTHERM_00824010\\nTTHERM_00824050\\nTTHERM_00825140\\nTTHERM_00827200\\nTTHERM_00829330\\nTTHERM_00833810\\nTTHERM_00834900\\nTTHERM_00834920\\nTTHERM_00834960\\nTTHERM_00835170\\nTTHERM_00835210\\nTTHERM_00836710\\nTTHERM_00840040\\nTTHERM_00841290\\nTTHERM_00842570\\nTTHERM_00842650\\nTTHERM_00846960\\nTTHERM_00847040\\nTTHERM_00848080\\nTTHERM_00852770\\nTTHERM_00857790\\nTTHERM_00857890\\nTTHERM_00858020\\nTTHERM_00860520\\nTTHERM_00861620\\nTTHERM_00861650\\nTTHERM_00861660\\nTTHERM_00861670\\nTTHERM_00861680\\nTTHERM_00864910\\nTTHERM_00869580\\nTTHERM_00874740\\nTTHERM_00874760\\nTTHERM_00875870\\nTTHERM_00881320\\nTTHERM_00884680\\nTTHERM_00887990\\nTTHERM_00888000\\nTTHERM_00888002\\nTTHERM_00888010\\nTTHERM_00891270\\nTTHERM_00892330\\nTTHERM_00894430\\nTTHERM_00894440\\nTTHERM_00895720\\nTTHERM_00895770\\nTTHERM_00895790\\nTTHERM_00896080\\nTTHERM_00898140\\nTTHERM_00898160\\nTTHERM_00899540\\nTTHERM_00901690\\nTTHERM_00904010\\nTTHERM_00904050\\nTTHERM_00907070\\nTTHERM_00907090\\nTTHERM_00912240\\nTTHERM_00912280\\nTTHERM_00918470\\nTTHERM_00923200\\nTTHERM_00925370\\nTTHERM_00925430\\nTTHERM_00925440\\nTTHERM_00925680\\nTTHERM_00925730\\nTTHERM_00925790\\nTTHERM_00925830\\nTTHERM_00925850\\nTTHERM_00925870\\nTTHERM_00927330\\nTTHERM_00929490\\nTTHERM_00930710\\nTTHERM_00930880\\nTTHERM_00932030\\nTTHERM_00933110\\nTTHERM_00935490\\nTTHERM_00939010\\nTTHERM_00939070\\nTTHERM_00941470\\nTTHERM_00944110\\nTTHERM_00944150\\nTTHERM_00947340\\nTTHERM_00947370\\nTTHERM_00951750\\nTTHERM_00951810\\nTTHERM_00954080\\nTTHERM_00954260\\nTTHERM_00957600\\nTTHERM_00961900\\nTTHERM_00962000\\nTTHERM_00972990\\nTTHERM_00974040\\nTTHERM_00974150\\nTTHERM_00974260\\nTTHERM_00975330\\nTTHERM_00975400\\nTTHERM_00979890\\nTTHERM_00979940\\nTTHERM_00985050\\nTTHERM_00985130\\nTTHERM_00985190\\nTTHERM_00989310\\nTTHERM_00992650\\nTTHERM_00992670\\nTTHERM_00994190\\nTTHERM_00994280\\nTTHERM_00994370\\nTTHERM_00997600\\nTTHERM_00997730\\nTTHERM_00997770\\nTTHERM_00998920\\nTTHERM_00999110\\nTTHERM_01002570\\nTTHERM_01002700\\nTTHERM_01002820\\nTTHERM_01002900\\nTTHERM_01005040\\nTTHERM_01006602\\nTTHERM_01012090\\nTTHERM_01013120\\nTTHERM_01013370\\nTTHERM_01014500\\nTTHERM_01014610\\nTTHERM_01015950\\nTTHERM_01018430\\nTTHERM_01020670\\nTTHERM_01026400\\nTTHERM_01027580\\nTTHERM_01027600\\nTTHERM_01030020\\nTTHERM_01034450\\nTTHERM_01040850\\nTTHERM_01043192\\nTTHERM_01044560\\nTTHERM_01046990\\nTTHERM_01050450\\nTTHERM_01050550\\nTTHERM_01055730\\nTTHERM_01058762\\nTTHERM_01063940\\nTTHERM_01068020\\nTTHERM_01071450\\nTTHERM_01074550\\nTTHERM_01075750\\nTTHERM_01075770\\nTTHERM_01076820\\nTTHERM_01076930\\nTTHERM_01079070\\nTTHERM_01079230\\nTTHERM_01079330\\nTTHERM_01080370\\nTTHERM_01080430\\nTTHERM_01084310\\nTTHERM_01087890\\nTTHERM_01088000\\nTTHERM_01092360\\nTTHERM_01093760\\nTTHERM_01097910\\nTTHERM_01097920\\nTTHERM_01097970\\nTTHERM_01099080\\nTTHERM_01102790\\nTTHERM_01104870\\nTTHERM_01106170\\nTTHERM_01106210\\nTTHERM_01107460\\nTTHERM_01108590\\nTTHERM_01114210\\nTTHERM_01117280\\nTTHERM_01117320\\nTTHERM_01121640\\nTTHERM_01122810\\nTTHERM_01125220\\nTTHERM_01127460\\nTTHERM_01127470\\nTTHERM_01129750\\nTTHERM_01131820\\nTTHERM_01132860\\nTTHERM_01137233\\nTTHERM_01138283\\nTTHERM_01139330\\nTTHERM_01139360\\nTTHERM_01141460\\nTTHERM_01141630\\nTTHERM_01142700\\nTTHERM_01148180\\nTTHERM_01149250\\nTTHERM_01149310\\nTTHERM_01150350\\nTTHERM_01150420\\nTTHERM_01156800\\nTTHERM_01156890\\nTTHERM_01165250\\nTTHERM_01169420\\nTTHERM_01169510\\nTTHERM_01169520\\nTTHERM_01170540\\nTTHERM_01178670\\nTTHERM_01178750\\nTTHERM_01183080\\nTTHERM_01186250\\nTTHERM_01188330\\nTTHERM_01188340\\nTTHERM_01188390\\nTTHERM_01194810\\nTTHERM_01197010\\nTTHERM_01197150\\nTTHERM_01202212\\nTTHERM_01206400\\nTTHERM_01207610\\nTTHERM_01217230\\nTTHERM_01223622\\nTTHERM_01224650\\nTTHERM_01227690\\nTTHERM_01227790\\nTTHERM_01232200\\nTTHERM_01236340\\nTTHERM_01244600\\nTTHERM_01245650\\nTTHERM_01246720\\nTTHERM_01248940\\nTTHERM_01250150\\nTTHERM_01250170\\nTTHERM_01250180\\nTTHERM_01251200\\nTTHERM_01260670\\nTTHERM_01263970\\nTTHERM_012701404\\nTTHERM_01273260\\nTTHERM_01284820\\nTTHERM_01287960\\nTTHERM_01293260\\nTTHERM_01302830\\nTTHERM_01306850\\nTTHERM_01307940\\nTTHERM_01311230\\nTTHERM_01311270\\nTTHERM_01311280\\nTTHERM_01318430\\nTTHERM_01318450\\nTTHERM_01318470\\nTTHERM_01321580\\nTTHERM_01321640\\nTTHERM_01322650\\nTTHERM_01323720\\nTTHERM_01323730\\nTTHERM_01324740\\nTTHERM_01326860\\nTTHERM_01326870\\nTTHERM_01330050\\nTTHERM_01332100\\nTTHERM_01334240\\nTTHERM_01338460\\nTTHERM_01338530\\nTTHERM_01347950\\nTTHERM_01351080\\nTTHERM_01354280\\nTTHERM_01364620\\nTTHERM_01365632\\nTTHERM_01365640\\nTTHERM_01374860\\nTTHERM_01393300\\nTTHERM_01396430\\nTTHERM_01399520\\nTTHERM_01403830\\nTTHERM_01405890\\nTTHERM_01407940\\nTTHERM_01413110\\nTTHERM_01415160\\nTTHERM_01417300\\nTTHERM_01421350\\nTTHERM_01423410\\nTTHERM_01423430\\nTTHERM_01424460\\nTTHERM_01434610\\nTTHERM_01434620\\nTTHERM_01434650\\nTTHERM_01446960\\nTTHERM_01461140\\nTTHERM_01462160\\nTTHERM_01466260\\nTTHERM_01467270\\nTTHERM_01469340\\nTTHERM_01485662\\nTTHERM_01498890\\nTTHERM_01512220\\nTTHERM_01532560\\nTTHERM_01538670\\nTTHERM_01542800\\nTTHERM_01546850\\nTTHERM_01551970\\nTTHERM_01571180\\nTTHERM_01594610\\nTTHERM_01606700\\nTTHERM_01607710\\nTTHERM_01615782\\nTTHERM_016600704\\nTTHERM_01671170\\nTTHERM_01683270\\nTTHERM_01808870\n",
    "# \"\"\".split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "id": "6b6de3be-caa1-41d2-a058-efb28081050c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checkout_df = complete_annot.loc[complete_annot['TTHERM_ID'].isin(checkout)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "id": "ec95a4d8-b618-499a-a3ed-066491940bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(checkout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "id": "eafe60a7-e772-4724-8adf-188ad7ce59e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checkout_df.to_csv('../checkout_different_filtered_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "id": "6b88aa5a-83d9-4712-b900-7e22e93c9836",
   "metadata": {},
   "outputs": [],
   "source": [
    "# complete_annot.to_csv('/Users/eukarya/Downloads/tgn_complete_annot.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29defd25-5786-4bca-a782-59b158b7d176",
   "metadata": {},
   "source": [
    "## Round one of plotting and consensus "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2525b47b-7edf-4ed3-b6e7-4739c399b426",
   "metadata": {},
   "source": [
    "First, cluster accoding to UMAP/Leiden in full-dimensional space, using Manhattan distance and setting the number of neighbors to 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "id": "6788cd86-a9f4-48f4-b416-a8448c849d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "leiden_label_df_round_1, full_dists = build_leiden_label_df(full_filtered_norm_df, 'full', random_state=42, n_neighbors=3, leiden_type='modularity')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.0001, 0.0002, 0.0003, 0.0004, 0.0005, 0.0006, 0.0007, 0.0008, 0.0009]\n",
      "0.0001 generated 29 clusters.\n",
      "0.0002 generated 49 clusters.\n",
      "0.0003 generated 70 clusters.\n",
      "0.0004 generated 91 clusters.\n",
      "0.0005 generated 107 clusters.\n",
      "0.0006 generated 128 clusters.\n",
      "0.0007 generated 139 clusters.\n",
      "0.0008 generated 153 clusters.\n",
      "0.0009 generated 174 clusters.\n"
     ]
    }
   ],
   "source": [
    "leiden_label_df_cpms = {}\n",
    "\n",
    "la_res_params = [0.0001, 0.0002, 0.0003, 0.0004, 0.0005, 0.0006, 0.0007, 0.0008, 0.0009]\n",
    "\n",
    "print(la_res_params)\n",
    "\n",
    "for curr_la_res_param in la_res_params:\n",
    "    leiden_label_df_cpm, full_dists_cpm = build_leiden_label_df(full_filtered_norm_df, 'full', random_state=42, n_neighbors=3, leiden_type='cpm', la_res_param=curr_la_res_param)\n",
    "    leiden_label_df_cpm_sorted = leiden_label_df_cpm.sort_values(by='leiden_label_full')\n",
    "    print(curr_la_res_param, 'generated', max(list(leiden_label_df_cpm_sorted['leiden_label_full'].values))+1, 'clusters.')\n",
    "    leiden_label_df_cpms[curr_la_res_param] = leiden_label_df_cpm_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 479,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0.0001:               TTHERM_ID  leiden_label_full\n",
       " 8483    TTHERM_00324430                  0\n",
       " 8537    TTHERM_00327110                  0\n",
       " 15797   TTHERM_00734050                  0\n",
       " 8533    TTHERM_00327080                  0\n",
       " 1475    TTHERM_00038810                  0\n",
       " ...                 ...                ...\n",
       " 3703   TTHERM_000951751                 27\n",
       " 20285   TTHERM_01659050                 27\n",
       " 19901   TTHERM_01318440                 28\n",
       " 4691   TTHERM_001318492                 28\n",
       " 19903   TTHERM_01318460                 28\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0002:               TTHERM_ID  leiden_label_full\n",
       " 9828    TTHERM_00411350                  0\n",
       " 8858    TTHERM_00346810                  0\n",
       " 8854    TTHERM_00346760                  0\n",
       " 3264   TTHERM_000805891                  0\n",
       " 18777   TTHERM_01078050                  0\n",
       " ...                 ...                ...\n",
       " 3703   TTHERM_000951751                 47\n",
       " 3879   TTHERM_001018540                 47\n",
       " 19901   TTHERM_01318440                 48\n",
       " 4691   TTHERM_001318492                 48\n",
       " 19903   TTHERM_01318460                 48\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0003:               TTHERM_ID  leiden_label_full\n",
       " 14101   TTHERM_00637530                  0\n",
       " 10436   TTHERM_00442140                  0\n",
       " 1742   TTHERM_000455549                  0\n",
       " 2770    TTHERM_00069370                  0\n",
       " 13512   TTHERM_00599750                  0\n",
       " ...                 ...                ...\n",
       " 3703   TTHERM_000951751                 68\n",
       " 20285   TTHERM_01659050                 68\n",
       " 4691   TTHERM_001318492                 69\n",
       " 19901   TTHERM_01318440                 69\n",
       " 19903   TTHERM_01318460                 69\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0004:               TTHERM_ID  leiden_label_full\n",
       " 6682    TTHERM_00227640                  0\n",
       " 2538    TTHERM_00061550                  0\n",
       " 7626    TTHERM_00289270                  0\n",
       " 8180    TTHERM_00313110                  0\n",
       " 10823   TTHERM_00456980                  0\n",
       " ...                 ...                ...\n",
       " 3879   TTHERM_001018540                 89\n",
       " 20285   TTHERM_01659050                 89\n",
       " 19903   TTHERM_01318460                 90\n",
       " 4691   TTHERM_001318492                 90\n",
       " 19901   TTHERM_01318440                 90\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0005:               TTHERM_ID  leiden_label_full\n",
       " 5191    TTHERM_00146450                  0\n",
       " 14958   TTHERM_00688470                  0\n",
       " 17258   TTHERM_00884620                  0\n",
       " 4452    TTHERM_00125358                  0\n",
       " 9786    TTHERM_00408870                  0\n",
       " ...                 ...                ...\n",
       " 3879   TTHERM_001018540                105\n",
       " 20285   TTHERM_01659050                105\n",
       " 19901   TTHERM_01318440                106\n",
       " 4691   TTHERM_001318492                106\n",
       " 19903   TTHERM_01318460                106\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0006:               TTHERM_ID  leiden_label_full\n",
       " 18772   TTHERM_01077000                  0\n",
       " 11833   TTHERM_00499420                  0\n",
       " 19876   TTHERM_01308010                  0\n",
       " 3970    TTHERM_00105110                  0\n",
       " 11823   TTHERM_00499310                  0\n",
       " ...                 ...                ...\n",
       " 3879   TTHERM_001018540                126\n",
       " 20285   TTHERM_01659050                126\n",
       " 19901   TTHERM_01318440                127\n",
       " 19903   TTHERM_01318460                127\n",
       " 4691   TTHERM_001318492                127\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0007:               TTHERM_ID  leiden_label_full\n",
       " 11685   TTHERM_00494720                  0\n",
       " 917     TTHERM_00024380                  0\n",
       " 922     TTHERM_00024410                  0\n",
       " 3860    TTHERM_00101460                  0\n",
       " 10240   TTHERM_00431340                  0\n",
       " ...                 ...                ...\n",
       " 3879   TTHERM_001018540                137\n",
       " 20285   TTHERM_01659050                137\n",
       " 19903   TTHERM_01318460                138\n",
       " 19901   TTHERM_01318440                138\n",
       " 4691   TTHERM_001318492                138\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0008:               TTHERM_ID  leiden_label_full\n",
       " 13040   TTHERM_00571860                  0\n",
       " 9836    TTHERM_00411470                  0\n",
       " 9819    TTHERM_00410250                  0\n",
       " 950    TTHERM_000253490                  0\n",
       " 4964    TTHERM_00140820                  0\n",
       " ...                 ...                ...\n",
       " 20285   TTHERM_01659050                151\n",
       " 3703   TTHERM_000951751                151\n",
       " 4691   TTHERM_001318492                152\n",
       " 19903   TTHERM_01318460                152\n",
       " 19901   TTHERM_01318440                152\n",
       " \n",
       " [20326 rows x 2 columns],\n",
       " 0.0009:               TTHERM_ID  leiden_label_full\n",
       " 8650    TTHERM_00334542                  0\n",
       " 7251   TTHERM_002653451                  0\n",
       " 7250   TTHERM_002653450                  0\n",
       " 290    TTHERM_000085199                  0\n",
       " 2364   TTHERM_000575579                  0\n",
       " ...                 ...                ...\n",
       " 3703   TTHERM_000951751                172\n",
       " 3879   TTHERM_001018540                172\n",
       " 19901   TTHERM_01318440                173\n",
       " 19903   TTHERM_01318460                173\n",
       " 4691   TTHERM_001318492                173\n",
       " \n",
       " [20326 rows x 2 columns]}"
      ]
     },
     "execution_count": 479,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_cpms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 480,
   "id": "3e876499",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full\n",
       "0      TTHERM_000000042                 19\n",
       "1      TTHERM_000000045                 12\n",
       "2       TTHERM_00000010                  8\n",
       "3       TTHERM_00000020                 29\n",
       "4       TTHERM_00000030                 32\n",
       "...                 ...                ...\n",
       "20321   TTHERM_02293890                 33\n",
       "20322   TTHERM_02385080                 33\n",
       "20323   TTHERM_02555200                 43\n",
       "20324   TTHERM_02607240                 31\n",
       "20325   TTHERM_02653470                 44\n",
       "\n",
       "[20326 rows x 2 columns]"
      ]
     },
     "execution_count": 480,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 481,
   "id": "4a52f1d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60"
      ]
     },
     "execution_count": 481,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_full'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "id": "d2970071-e2c8-49d9-8931-9ce6e4f8af44",
   "metadata": {},
   "outputs": [],
   "source": [
    "leiden_label_df_round_1, veg_dists = build_leiden_label_df(veg_filtered_norm_df, 'veg', lldf=leiden_label_df_round_1, random_state=42, n_neighbors=3)\n",
    "leiden_label_df_round_1, sex_dists = build_leiden_label_df(sex_filtered_norm_df, 'sex', lldf=leiden_label_df_round_1, random_state=42, n_neighbors=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "id": "e13fb345-b665-46ff-adbb-98d76c952b8d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>33</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>33</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>43</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>44</td>\n",
       "      <td>38</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "0      TTHERM_000000042                 19                17                35\n",
       "1      TTHERM_000000045                 12                 6                25\n",
       "2       TTHERM_00000010                  8                32                 5\n",
       "3       TTHERM_00000020                 29                28                18\n",
       "4       TTHERM_00000030                 32                32                 0\n",
       "...                 ...                ...               ...               ...\n",
       "20321   TTHERM_02293890                 33                43                 3\n",
       "20322   TTHERM_02385080                 33                18                19\n",
       "20323   TTHERM_02555200                 43                18                19\n",
       "20324   TTHERM_02607240                 31                20                21\n",
       "20325   TTHERM_02653470                 44                38                30\n",
       "\n",
       "[20326 rows x 4 columns]"
      ]
     },
     "execution_count": 483,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "id": "382f7ad6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52"
      ]
     },
     "execution_count": 484,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_veg'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "id": "0cffc2fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56"
      ]
     },
     "execution_count": 485,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_sex'].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ffe51d0-614c-49f4-bf8c-41d48c1b75e6",
   "metadata": {},
   "source": [
    "Check out how many clusters we have"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9125a92-356f-49ad-88e1-c7da95ded20a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "c1879f2a-3e20-44ff-8d11-72dee85b765f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# leiden_label_df_round_1.to_csv('../clustergrammer/leiden_label_df_round_1_nn2.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "id": "db82b99c-9360-4a60-80bc-1b78e216b08e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60"
      ]
     },
     "execution_count": 487,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_full'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 488,
   "id": "736a3bba-f713-427b-aecb-e469e84ec26c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52"
      ]
     },
     "execution_count": 488,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_veg'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 489,
   "id": "94d2d630-4411-4777-9e38-cc7c77f90251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56"
      ]
     },
     "execution_count": 489,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_sex'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "96125df3-98dc-4b99-bb70-86ddb1997737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "333.21311475409834"
      ]
     },
     "execution_count": 490,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_size_list = []\n",
    "for l in leiden_label_df_round_1['leiden_label_full'].unique():\n",
    "    df = leiden_label_df_round_1.loc[leiden_label_df_round_1['leiden_label_full'] == l]\n",
    "    module_size_list.append(len(df))\n",
    "    \n",
    "np.mean(module_size_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 491,
   "id": "32eac343-9a2b-4a27-a9b2-785597ea65f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "331.0"
      ]
     },
     "execution_count": 491,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(module_size_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 492,
   "id": "333048c1-ea2a-4d02-97d9-af0f0450118e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "798"
      ]
     },
     "execution_count": 492,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(module_size_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "id": "8fd25633-9cfd-4c31-847a-c49fb989e4d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>35</td>\n",
       "      <td>5911.EAR87740</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1776.8</td>\n",
       "      <td>2CVX5@1|root,2RT23@2759|Eukaryota,3ZDPV@5878|C...</td>\n",
       "      <td>5878|Ciliophora</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>B-box zinc finger protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "      <td>5911.EAR87737</td>\n",
       "      <td>1.7e-272</td>\n",
       "      <td>898.9</td>\n",
       "      <td>KOG0575@1|root,KOG0575@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ko04068,ko04110,ko04114,ko04136,ko04138,ko0413...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko01001,ko03029,ko0303...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>5911.EAR87739</td>\n",
       "      <td>1.6e-77</td>\n",
       "      <td>260.4</td>\n",
       "      <td>KOG0032@1|root,KOG0032@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ko04010,ko04114,ko04136,ko04138,ko04139,ko0414...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko01001,ko03019,ko0302...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Serine/Threonine kinase domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>5888.CAK56476</td>\n",
       "      <td>5.8999999999999995e-52</td>\n",
       "      <td>172.6</td>\n",
       "      <td>KOG0615@1|root,KOG0615@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ko04140,ko04152,ko04211,ko04626,ko04920,ko0492...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko01001</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Serine/Threonine kinase domain protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex  \\\n",
       "0  TTHERM_000000042                 19                17                35   \n",
       "1  TTHERM_000000045                 12                 6                25   \n",
       "2   TTHERM_00000010                  8                32                 5   \n",
       "3   TTHERM_00000020                 29                28                18   \n",
       "4   TTHERM_00000030                 32                32                 0   \n",
       "\n",
       "   seed_ortholog                  evalue   score  \\\n",
       "0  5911.EAR87740                     0.0  1776.8   \n",
       "1              -                       -       -   \n",
       "2  5911.EAR87737                1.7e-272   898.9   \n",
       "3  5911.EAR87739                 1.6e-77   260.4   \n",
       "4  5888.CAK56476  5.8999999999999995e-52   172.6   \n",
       "\n",
       "                                          eggNOG_OGs    max_annot_lvl  \\\n",
       "0  2CVX5@1|root,2RT23@2759|Eukaryota,3ZDPV@5878|C...  5878|Ciliophora   \n",
       "1                                                  -                -   \n",
       "2              KOG0575@1|root,KOG0575@2759|Eukaryota   2759|Eukaryota   \n",
       "3              KOG0032@1|root,KOG0032@2759|Eukaryota   2759|Eukaryota   \n",
       "4              KOG0615@1|root,KOG0615@2759|Eukaryota   2759|Eukaryota   \n",
       "\n",
       "  COG_category  ...                                       KEGG_Pathway  \\\n",
       "0            -  ...                                                  -   \n",
       "1            -  ...                                                  -   \n",
       "2            T  ...  ko04068,ko04110,ko04114,ko04136,ko04138,ko0413...   \n",
       "3            T  ...  ko04010,ko04114,ko04136,ko04138,ko04139,ko0414...   \n",
       "4            T  ...  ko04140,ko04152,ko04211,ko04626,ko04920,ko0492...   \n",
       "\n",
       "  KEGG_Module KEGG_Reaction KEGG_rclass  \\\n",
       "0           -             -           -   \n",
       "1           -             -           -   \n",
       "2           -             -           -   \n",
       "3           -             -           -   \n",
       "4           -             -           -   \n",
       "\n",
       "                                               BRITE KEGG_TC CAZy  \\\n",
       "0                                                  -       -    -   \n",
       "1                                                  -       -    -   \n",
       "2  ko00000,ko00001,ko01000,ko01001,ko03029,ko0303...       -    -   \n",
       "3  ko00000,ko00001,ko01000,ko01001,ko03019,ko0302...       -    -   \n",
       "4                    ko00000,ko00001,ko01000,ko01001       -    -   \n",
       "\n",
       "  BiGG_Reaction PFAMs                     TGD2021_description  \n",
       "0             -     -               B-box zinc finger protein  \n",
       "1             -     -          transmembrane protein putative  \n",
       "2             -     -                          protein kinase  \n",
       "3             -     -  Serine/Threonine kinase domain protein  \n",
       "4             -     -  Serine/Threonine kinase domain protein  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 493,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labeled_annot = leiden_label_df_round_1.merge(complete_annot, on='TTHERM_ID')\n",
    "labeled_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "id": "23d1e313-7738-4a3a-bf16-9b36210d0be1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>TTHERM_00001200</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>TTHERM_000046379</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>TTHERM_00006010</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>TTHERM_00006110</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>TTHERM_00006120</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19878</th>\n",
       "      <td>TTHERM_01308050</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19890</th>\n",
       "      <td>TTHERM_01310180</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19956</th>\n",
       "      <td>TTHERM_01337380</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19972</th>\n",
       "      <td>TTHERM_01344690</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20131</th>\n",
       "      <td>TTHERM_01443820</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>798 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "41      TTHERM_00001200                  0                 5                 2\n",
       "143    TTHERM_000046379                  0                23                 5\n",
       "197     TTHERM_00006010                  0                 5                 2\n",
       "205     TTHERM_00006110                  0                14                17\n",
       "207     TTHERM_00006120                  0                 5                 0\n",
       "...                 ...                ...               ...               ...\n",
       "19878   TTHERM_01308050                  0                14                 2\n",
       "19890   TTHERM_01310180                  0                 5                28\n",
       "19956   TTHERM_01337380                  0                23                 2\n",
       "19972   TTHERM_01344690                  0                 5                 0\n",
       "20131   TTHERM_01443820                  0                35                 2\n",
       "\n",
       "[798 rows x 4 columns]"
      ]
     },
     "execution_count": 494,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['leiden_label_full']==0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e03aeddd-8c49-4136-9c15-6befa0d61418",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "id": "28e85629-61e7-4d5d-b1ba-5da9a1832724",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12397</th>\n",
       "      <td>TTHERM_00532700</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "12397  TTHERM_00532700                 10                 3                21"
      ]
     },
     "execution_count": 495,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['TTHERM_ID']=='TTHERM_00532700']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "id": "b670c766-4a46-4403-9591-512df86365cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>TTHERM_00001180</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>TTHERM_00006380</td>\n",
       "      <td>in StAR and phosphatidylcholine transfer protein</td>\n",
       "      <td>START domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>TTHERM_00006450</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>TTHERM_000077009</td>\n",
       "      <td>Zinc finger LSD1 subclass family protein</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>TTHERM_000078939</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19842</th>\n",
       "      <td>TTHERM_01298480</td>\n",
       "      <td>hyaluronan metabolic process</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19982</th>\n",
       "      <td>TTHERM_01345810</td>\n",
       "      <td>Serine threonine-protein phosphatase</td>\n",
       "      <td>PPN1 serine/threonine protein phosphatase 2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20109</th>\n",
       "      <td>TTHERM_01433590</td>\n",
       "      <td>Ion channel</td>\n",
       "      <td>cation channel family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20202</th>\n",
       "      <td>TTHERM_01513250</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20264</th>\n",
       "      <td>TTHERM_01609750</td>\n",
       "      <td>WD domain, G-beta repeat protein</td>\n",
       "      <td>WD domain G-beta repeat protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>533 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID                                       Description  \\\n",
       "37      TTHERM_00001180                                                 -   \n",
       "231     TTHERM_00006380  in StAR and phosphatidylcholine transfer protein   \n",
       "237     TTHERM_00006450                                                 -   \n",
       "261    TTHERM_000077009          Zinc finger LSD1 subclass family protein   \n",
       "267    TTHERM_000078939                                                 -   \n",
       "...                 ...                                               ...   \n",
       "19842   TTHERM_01298480                      hyaluronan metabolic process   \n",
       "19982   TTHERM_01345810              Serine threonine-protein phosphatase   \n",
       "20109   TTHERM_01433590                                       Ion channel   \n",
       "20202   TTHERM_01513250                                                 -   \n",
       "20264   TTHERM_01609750                  WD domain, G-beta repeat protein   \n",
       "\n",
       "                                TGD2021_description  \n",
       "37                             hypothetical protein  \n",
       "231                            START domain protein  \n",
       "237                            hypothetical protein  \n",
       "261                  transmembrane protein putative  \n",
       "267                            hypothetical protein  \n",
       "...                                             ...  \n",
       "19842                          hypothetical protein  \n",
       "19982  PPN1 serine/threonine protein phosphatase 2A  \n",
       "20109                 cation channel family protein  \n",
       "20202                          hypothetical protein  \n",
       "20264               WD domain G-beta repeat protein  \n",
       "\n",
       "[533 rows x 3 columns]"
      ]
     },
     "execution_count": 496,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labeled_annot.loc[labeled_annot['leiden_label_full'] == 10][['TTHERM_ID', 'Description', 'TGD2021_description']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "id": "8d82230e-f85b-42ad-b9c2-75a69c56ef89",
   "metadata": {},
   "outputs": [],
   "source": [
    "leiden_label_df_round_1.to_csv('./test_nn3_leiden_label_df_round_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "id": "de9bf727",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>29</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>33</td>\n",
       "      <td>43</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>33</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>43</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>31</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>44</td>\n",
       "      <td>38</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "0      TTHERM_000000042                 19                17                35\n",
       "1      TTHERM_000000045                 12                 6                25\n",
       "2       TTHERM_00000010                  8                32                 5\n",
       "3       TTHERM_00000020                 29                28                18\n",
       "4       TTHERM_00000030                 32                32                 0\n",
       "...                 ...                ...               ...               ...\n",
       "20321   TTHERM_02293890                 33                43                 3\n",
       "20322   TTHERM_02385080                 33                18                19\n",
       "20323   TTHERM_02555200                 43                18                19\n",
       "20324   TTHERM_02607240                 31                20                21\n",
       "20325   TTHERM_02653470                 44                38                30\n",
       "\n",
       "[20326 rows x 4 columns]"
      ]
     },
     "execution_count": 498,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "e0d1a05d-c704-48d5-952d-7cf01f75bb99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "56"
      ]
     },
     "execution_count": 499,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(np.array(module_size_list) >= 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "0e2ef7d6-cf4f-452e-b183-4cdda16fcf3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "20049  TTHERM_01386050                 30                 3                27"
      ]
     },
     "execution_count": 500,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "id": "83198b81-f8ff-4230-8047-741d63a685af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# full_pymde_data = full_filtered_norm_df[list(full_filtered_norm_df.columns)[1:]].values\n",
    "\n",
    "# pymde.seed(42)\n",
    "# torch.seed(42)\n",
    "\n",
    "# full_embedding = pymde.preserve_neighbors(\n",
    "#     full_pymde_data, \n",
    "#     verbose=True,\n",
    "#     n_neighbors=5,\n",
    "#     constraint=pymde.Standardized()\n",
    "# ).embed()\n",
    "\n",
    "# pymde_full_df = pd.DataFrame(np.array(full_embedding), columns=('x', 'y'))\n",
    "# bokeh.plotting.output_file(filename='./plots/full_norm_pymde_leiden_round_1.2.html', mode='inline')\n",
    "\n",
    "# p = plot_embedding(full_filtered_norm_df, pymde_full_df, complete_annot, leiden_label_df_round_1, 'full', palette32, n_neighbors=5, radius=0.005)\n",
    "# bokeh.plotting.save(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f4677a3-55c9-49d8-9941-fde9c264c9fc",
   "metadata": {},
   "source": [
    "For ordering the leiden clusters in the heatmap, try computing the cityblock distance between centroids and then ordering by smallest distance, starting with largest cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "id": "c41737b2-bfb7-43cc-a5dc-c95cfbb53512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  b\n",
       "0  1  0\n",
       "1  2  2\n",
       "2  3  0\n",
       "3  4  2"
      ]
     },
     "execution_count": 502,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_centroid = pd.DataFrame({'a': np.array([1, 2, 3, 4]), 'b': np.array([0, 2, 0, 2])})\n",
    "test_centroid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "id": "1cedb93e-ab46-4c19-a492-353247f834a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([2.5, 1. ])"
      ]
     },
     "execution_count": 503,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_centroid.mean(axis=0).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2acf27e-4961-4ed6-9a79-f6c146f27bd1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "id": "b5da096c-f282-4050-9d3e-862ee1ce6a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "l = get_all_module_centroids(full_filtered_norm_df, leiden_label_df_round_1, 'leiden', 'full')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdcf9180-1fcc-4ab1-895a-8fed06c73c09",
   "metadata": {},
   "source": [
    "Module 0 is always the largest for leiden and WGCNA; need to check about CLR/MCL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "id": "9dd13886-0d8e-40fe-9bce-587fe2756aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdf = get_module_centroid_df(full_filtered_norm_df, leiden_label_df_round_1, 'leiden', 'full')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "id": "27798f08-c9d5-42f5-a0c4-0b0188191475",
   "metadata": {},
   "outputs": [],
   "source": [
    "# arr_df = arrange_modules(full_filtered_norm_df, leiden_label_df_round_1, 'leiden', 'full') # FIXME\n",
    "arr_df = leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "id": "8b0d4e56-7832-4310-9605-efb90db185ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20326\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3821</th>\n",
       "      <td>TTHERM_001009942</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3822</th>\n",
       "      <td>TTHERM_001009943</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18827</th>\n",
       "      <td>TTHERM_01081690</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>TTHERM_000037389</td>\n",
       "      <td>44</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>TTHERM_000077669</td>\n",
       "      <td>44</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19961</th>\n",
       "      <td>TTHERM_01338500</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20013</th>\n",
       "      <td>TTHERM_01358410</td>\n",
       "      <td>55</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20036</th>\n",
       "      <td>TTHERM_01372810</td>\n",
       "      <td>55</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20124</th>\n",
       "      <td>TTHERM_01436760</td>\n",
       "      <td>55</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20317</th>\n",
       "      <td>TTHERM_02094560</td>\n",
       "      <td>55</td>\n",
       "      <td>17</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "3821   TTHERM_001009942                 35                21                14\n",
       "3822   TTHERM_001009943                 35                21                14\n",
       "18827   TTHERM_01081690                 35                21                19\n",
       "118    TTHERM_000037389                 44                13                 1\n",
       "263    TTHERM_000077669                 44                 2                 1\n",
       "...                 ...                ...               ...               ...\n",
       "19961   TTHERM_01338500                 55                 1                30\n",
       "20013   TTHERM_01358410                 55                 2                20\n",
       "20036   TTHERM_01372810                 55                 1                 6\n",
       "20124   TTHERM_01436760                 55                 2                 1\n",
       "20317   TTHERM_02094560                 55                17                33\n",
       "\n",
       "[20326 rows x 4 columns]"
      ]
     },
     "execution_count": 507,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arrange_modules(full_filtered_df, leiden_label_df_round_1, 'leiden', 'full')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "id": "5cacc481-ad08-41ca-898d-6d1706091849",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>TTHERM_00001200</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>TTHERM_000046379</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>TTHERM_00006010</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>TTHERM_00006110</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>TTHERM_00006120</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19878</th>\n",
       "      <td>TTHERM_01308050</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19890</th>\n",
       "      <td>TTHERM_01310180</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19956</th>\n",
       "      <td>TTHERM_01337380</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19972</th>\n",
       "      <td>TTHERM_01344690</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20131</th>\n",
       "      <td>TTHERM_01443820</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>798 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "41      TTHERM_00001200                  0                 5                 2\n",
       "143    TTHERM_000046379                  0                23                 5\n",
       "197     TTHERM_00006010                  0                 5                 2\n",
       "205     TTHERM_00006110                  0                14                17\n",
       "207     TTHERM_00006120                  0                 5                 0\n",
       "...                 ...                ...               ...               ...\n",
       "19878   TTHERM_01308050                  0                14                 2\n",
       "19890   TTHERM_01310180                  0                 5                28\n",
       "19956   TTHERM_01337380                  0                23                 2\n",
       "19972   TTHERM_01344690                  0                 5                 0\n",
       "20131   TTHERM_01443820                  0                35                 2\n",
       "\n",
       "[798 rows x 4 columns]"
      ]
     },
     "execution_count": 508,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr_df.loc[arr_df['leiden_label_full'] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "id": "478eef64-9c36-4e5e-b146-2844b08a0e3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "20049  TTHERM_01386050                 30                 3                27"
      ]
     },
     "execution_count": 509,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr_df.loc[arr_df['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "id": "191fb6c9-2d12-4a9d-bfbe-b6b9d9485c40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>30</td>\n",
       "      <td>3</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "20049  TTHERM_01386050                 30                 3                27"
      ]
     },
     "execution_count": 510,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19808</th>\n",
       "      <td>TTHERM_01289060</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16925</th>\n",
       "      <td>TTHERM_00842410</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12328</th>\n",
       "      <td>TTHERM_00530750</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18813</th>\n",
       "      <td>TTHERM_01080470</td>\n",
       "      <td>0</td>\n",
       "      <td>37</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4025</th>\n",
       "      <td>TTHERM_00106800</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "19808  TTHERM_01289060                  0                37                11\n",
       "16925  TTHERM_00842410                  0                23                22\n",
       "12328  TTHERM_00530750                  0                14                17\n",
       "18813  TTHERM_01080470                  0                37                23\n",
       "4025   TTHERM_00106800                  0                14                17"
      ]
     },
     "execution_count": 511,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1_sorted = leiden_label_df_round_1.sort_values(by='leiden_label_full')\n",
    "leiden_label_df_round_1_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "metadata": {},
   "outputs": [],
   "source": [
    "def leiden_clustering_df_to_csv(df, export_file_path):\n",
    "    current_module = list(df['leiden_label_full'].values)[0]\n",
    "\n",
    "    new_cluster = True\n",
    "\n",
    "    with open(export_file_path, 'w') as f:\n",
    "        for idx, row in df.iterrows():\n",
    "            if current_module != row['leiden_label_full']:\n",
    "                current_module = row['leiden_label_full']\n",
    "                f.write('\\n')\n",
    "                new_cluster =True\n",
    "\n",
    "            elif not new_cluster:\n",
    "                f.write('\\t')\n",
    "            \n",
    "            f.write(row['TTHERM_ID'])\n",
    "            new_cluster = False\n",
    "\n",
    "        f.write('\\n')\n",
    "    \n",
    "    print(export_file_path, 'successfully exported.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./leiden_clustering.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0001.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0002.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0003.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0004.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0005.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0006.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0007.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0008.labels successfully exported.\n",
      "./leiden_clustering_cpm_0.0009.labels successfully exported.\n"
     ]
    }
   ],
   "source": [
    "leiden_clustering_df_to_csv(leiden_label_df_round_1_sorted, './leiden_clustering.labels')\n",
    "\n",
    "for la_param, df in leiden_label_df_cpms.items():\n",
    "    leiden_clustering_df_to_csv(df, f'./leiden_clustering_cpm_{str(la_param)}.labels')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n",
      "61\n"
     ]
    }
   ],
   "source": [
    "with open('./leiden_clustering.labels', 'r') as f:\n",
    "    num_leiden_clusters = f.readlines()\n",
    "    print(len(num_leiden_clusters))\n",
    "\n",
    "print(max(list(leiden_label_df_round_1_sorted['leiden_label_full'].values)) + 1) # module numbering starts at 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f0f4ee4d-b47e-4e5d-883b-f30c9a90dd91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rcl_mcl_df.to_csv('./test_nn3_leiden_label_df_round_1_rearranged.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cf55445",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "55189b03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_00000040</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000637219</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00139660</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00473080</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00543660</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_000332209</td>\n",
       "      <td>898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_00036950</td>\n",
       "      <td>899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_01084350</td>\n",
       "      <td>899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_00664030</td>\n",
       "      <td>900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_00429850</td>\n",
       "      <td>900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full\n",
       "0       TTHERM_00000040                  0\n",
       "1      TTHERM_000637219                  0\n",
       "2       TTHERM_00139660                  0\n",
       "3       TTHERM_00473080                  0\n",
       "4       TTHERM_00543660                  0\n",
       "...                 ...                ...\n",
       "20321  TTHERM_000332209                898\n",
       "20322   TTHERM_00036950                899\n",
       "20323   TTHERM_01084350                899\n",
       "20324   TTHERM_00664030                900\n",
       "20325   TTHERM_00429850                900\n",
       "\n",
       "[20326 rows x 2 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rcl_mcl_df = pd.read_csv('./mcl_rcl/mcl_rcl_label_df_round_1.csv')\n",
    "rcl_mcl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "25666f03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30\n",
      "30\n",
      "30\n",
      "30\n",
      "30\n",
      "30\n",
      "30\n",
      "30\n",
      "128\n",
      "30\n",
      "128\n",
      "128\n",
      "128\n",
      "128\n",
      "128\n",
      "30\n",
      "128\n",
      "30\n",
      "128\n",
      "128\n",
      "128\n",
      "128\n",
      "535\n",
      "128\n",
      "128\n",
      "535\n",
      "535\n",
      "30\n",
      "30\n",
      "128\n",
      "128\n",
      "535\n",
      "30\n",
      "30\n",
      "128\n",
      "128\n",
      "30\n",
      "128\n",
      "128\n",
      "128\n"
     ]
    }
   ],
   "source": [
    "for gene in gene_list_test:\n",
    "    print((rcl_mcl_df.loc[rcl_mcl_df['TTHERM_ID'] == gene])['leiden_label_full'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20326\n"
     ]
    }
   ],
   "source": [
    "clust_alg = 'leiden'\n",
    "phases = 'full'\n",
    "\n",
    "arranged_rcl_mcl_df = arrange_modules(full_filtered_norm_df, rcl_mcl_df, clust_alg, phases) \n",
    "sorted_rcl_mcl_df = arranged_rcl_mcl_df.sort_values(by=[f'{clust_alg}_label_{phases}', 'TTHERM_ID'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_00000040</td>\n",
       "      <td>675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         TTHERM_ID  leiden_label_full\n",
       "0  TTHERM_00000040                675"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arranged_rcl_mcl_df.loc[arranged_rcl_mcl_df['TTHERM_ID'] == 'TTHERM_00000040']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_00000040</td>\n",
       "      <td>675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         TTHERM_ID  leiden_label_full\n",
       "0  TTHERM_00000040                675"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_rcl_mcl_df.loc[sorted_rcl_mcl_df['TTHERM_ID'] == 'TTHERM_00000040']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "91b739eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_rcl_mcl_df.to_csv('./mcl_rcl_label_df_round_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "73c2c36a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17649\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([299, 560, 516, 539, 111, 265, 653, 374, 548, 136, 584, 479, 262,\n",
       "       356, 240, 511, 505,  87, 640, 401, 415, 360, 603, 695, 646, 687,\n",
       "       725, 755, 497, 138, 306, 253, 518, 242, 484, 218, 684, 503, 232,\n",
       "       679,  55, 455,  33, 772, 787, 606,   8,  82, 592, 373, 366,  68,\n",
       "       661, 223, 166, 581, 631,  35, 544, 133, 572, 100, 142, 792, 517,\n",
       "       154,  85,  88, 388, 444, 351,  99, 364, 209, 248, 325, 336, 163,\n",
       "       294, 281,  48, 650, 231, 323, 609,   0, 504, 526, 666, 191, 419,\n",
       "       692, 728,  14,   1, 633, 565, 732, 260,  36, 194,  98, 309, 553,\n",
       "       405, 247, 143, 140,   7, 408, 465,  25, 427, 651, 753, 793, 165,\n",
       "       487, 496, 120, 318, 739, 604, 414, 774, 586, 284, 507,  53, 164,\n",
       "       317, 387, 208, 514,  89, 447, 489, 706, 264, 445, 160, 483, 721,\n",
       "       638, 577, 741,  61, 202, 398, 190, 363, 224, 395, 315,  37, 574,\n",
       "       480, 345, 520, 113, 474,  95, 460, 756,  18, 256, 665, 241, 132,\n",
       "       214, 383, 672, 438, 456,  29, 188, 681, 304, 275, 103, 781, 699,\n",
       "       686,  50, 185, 635, 641, 675, 534, 107, 737, 482, 349, 450, 749,\n",
       "         6, 558, 357, 311,  56, 754, 773, 591, 525, 478, 235,  73, 676,\n",
       "       671, 125, 118, 186, 127, 221, 179, 610, 342, 341,  83, 451, 249,\n",
       "        91, 618, 771, 543, 277,  66,  28, 715, 183, 673, 648, 141,   5,\n",
       "       258, 421, 367, 499, 600, 800, 719, 752, 244, 238, 210,  49, 759,\n",
       "       760, 149, 219, 375, 134, 717, 806, 512, 524, 712, 161, 261, 274,\n",
       "       636, 121, 502, 430, 169, 594, 429, 649, 658,  12,  13, 361,  97,\n",
       "       698, 369, 583,  86, 297, 196, 104, 605, 271, 173, 593, 216, 225,\n",
       "       130,  17, 382, 255,  80, 660, 776, 174, 557, 314, 236, 595, 320,\n",
       "       308, 533, 637, 257, 420, 770,  19, 233,  39, 327, 189, 678, 617,\n",
       "       688, 171, 239, 285, 463, 352, 378, 403, 562, 767, 805, 498,  40,\n",
       "       808, 439, 180, 555, 713, 580, 691, 784, 298, 269, 348, 807, 425,\n",
       "       799, 696, 559,  34, 263,  72, 276, 380, 322, 446, 488, 708, 563,\n",
       "       168, 243, 390, 705, 201, 494,   4, 193, 454, 321, 355, 226, 782,\n",
       "       411, 733, 731, 270, 757, 182, 371, 324, 714, 657,  77, 738, 643,\n",
       "       453, 766, 735, 747, 632, 677, 612, 423, 135, 114, 710, 146, 159,\n",
       "       564,  52, 162, 796, 720, 150,  62,  94, 289, 353, 621, 293, 647,\n",
       "       147, 615, 551, 428, 319, 795, 582, 212, 400, 222, 307, 619,  44,\n",
       "       470, 522, 230, 213, 326, 339, 519, 246, 443, 170,  57, 547, 722,\n",
       "       310, 397, 569, 585, 181, 251, 779,  32, 412, 108, 152,  58, 106,\n",
       "       466, 347, 778, 486, 481, 590, 529, 286, 495, 368, 601, 528, 370,\n",
       "        79, 521, 804, 769, 350, 426,  46,  92, 628, 101, 393, 625, 237,\n",
       "       570, 332, 701, 302, 279, 742, 407, 682, 123, 126, 664, 122, 744,\n",
       "       513, 803, 329, 156, 746, 652, 587, 785, 611, 417, 404, 192, 259,\n",
       "       200, 145, 515,  45, 131, 440, 627, 477, 435,   2, 416, 442, 704,\n",
       "       139, 312, 588, 630, 469, 129,  38, 468, 358, 578, 334, 464, 680,\n",
       "       663, 187, 158, 227, 490,  74,  30, 184, 254,  54, 780, 441, 296,\n",
       "       461, 422,  90, 413, 622, 178, 768, 313, 536, 758, 396, 694, 331,\n",
       "       552, 654, 399, 473, 775, 523, 267, 501, 250, 217, 614, 626, 102,\n",
       "       418, 437, 197, 707, 700, 448,  67, 452, 245, 790, 215, 763, 268,\n",
       "       554, 748, 493, 199, 764,  76, 157, 459, 278, 153,  63, 198, 608,\n",
       "       540, 801,  81, 655, 148,  78, 549,  10, 573, 761, 634, 697, 492,\n",
       "       295, 124,  43, 556, 394, 693, 282, 668, 291, 662, 431, 527, 475,\n",
       "        31, 724,  75, 266, 316,  15, 458, 624, 362, 471, 433,  23, 689,\n",
       "       579,  60, 110, 550, 384, 391, 424,  70, 607, 116, 305,  24, 195,\n",
       "       151, 228, 389, 726, 620, 500, 359, 599, 467, 462, 381, 354, 377,\n",
       "       385, 645, 745, 656, 730, 485, 750, 788, 723, 344, 683, 105, 740,\n",
       "       685,  84,   9, 117,  16,  27,  20, 432, 119, 406,  47, 333, 338,\n",
       "       288, 508, 449, 802, 537, 109,  59, 798, 629, 576, 434, 491, 789,\n",
       "       172, 229, 220, 669, 379, 597, 623,  26,  51, 535, 337, 613,  42,\n",
       "       292, 206, 602, 530, 568, 659, 734, 301, 175, 137, 144, 155,  11,\n",
       "       330, 303,  22, 644, 207, 328, 567, 410, 797, 571, 765, 743, 532,\n",
       "       167, 616, 510,   3, 542, 176, 711, 670, 211, 566,  69,  65, 541,\n",
       "       538, 392, 300,  96, 272, 794, 729, 709, 791, 386, 343,  93,  64,\n",
       "       736, 283, 205, 575, 762, 409, 674, 777, 667, 365, 531, 639, 340,\n",
       "       509, 335, 751, 716, 115,  21, 372, 436, 376, 702, 589,  71, 596,\n",
       "       177, 112, 290, 506, 476, 203, 346, 273, 561, 727, 598, 718, 234,\n",
       "       287, 252,  41, 280, 642, 786, 472, 783, 690, 703, 128, 402, 204,\n",
       "       457, 546, 545])"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arrange_modules(full_filtered_norm_df, rcl_mcl_df, 'leiden', 'full')['leiden_label_full'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d79a58-e289-445f-a3f5-abeb9b6b87cd",
   "metadata": {},
   "source": [
    "#### Go to ../enrichment/enrichment.ipynb to generate the functional annotation enrichment data from this clustering. Proceed once done."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
