{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2e305761-a971-4232-8a21-5907cfeedfad",
   "metadata": {},
   "source": [
    "## Import all packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5d9163ea-8dc4-46e8-9542-ba27cc741b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pdb\n",
    "from sklearn.metrics import silhouette_score, pairwise_distances, silhouette_samples\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.datasets import make_blobs\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "\n",
    "import bokeh\n",
    "from bokeh.plotting import show as show_interactive\n",
    "from bokeh.plotting import output_file, output_notebook\n",
    "from bokeh.layouts import column, row\n",
    "from bokeh.models import CustomJS, TextInput, LassoSelectTool, Select, MultiSelect, ColorBar, Legend, LegendItem\n",
    "from bokeh.models.widgets import DataTable, DateFormatter, TableColumn, Button, HTMLTemplateFormatter\n",
    "from bokeh.events import SelectionGeometry\n",
    "from bokeh.transform import linear_cmap, jitter\n",
    "from matplotlib.pyplot import show as show_static\n",
    "# from clustergrammer2 import net, Network, CGM2\n",
    "\n",
    "import igraph as ig\n",
    "import leidenalg as la\n",
    "\n",
    "import scipy.stats as st\n",
    "import scipy.spatial\n",
    "import scipy.cluster.hierarchy\n",
    "\n",
    "import glob\n",
    "import json\n",
    "import re\n",
    "import copy\n",
    "\n",
    "import requests\n",
    "import bs4\n",
    "import tqdm\n",
    "import os\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "import umap\n",
    "import pymde\n",
    "\n",
    "import torch\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "from pynndescent import NNDescent\n",
    "\n",
    "# bokeh.io.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "20cf4cee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20326"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/allgood_filt_agg_tidy_2021aligned_qc_rma_expression_full.csv')\n",
    "full_filtered_df = full_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "num_genes = full_filtered_df.shape[0]\n",
    "num_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "d2bc3de5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll_GSM283687</th>\n",
       "      <th>Ll_GSM284355</th>\n",
       "      <th>Ll_GSM284362</th>\n",
       "      <th>Lm_GSM283690</th>\n",
       "      <th>Lm_GSM284357</th>\n",
       "      <th>Lm_GSM284363</th>\n",
       "      <th>Lh_GSM283691</th>\n",
       "      <th>Lh_GSM284360</th>\n",
       "      <th>Lh_GSM284364</th>\n",
       "      <th>...</th>\n",
       "      <th>C12_GSM656237</th>\n",
       "      <th>C14_GSM285580</th>\n",
       "      <th>C14_GSM285593</th>\n",
       "      <th>C14_GSM656238</th>\n",
       "      <th>C16_GSM285582</th>\n",
       "      <th>C16_GSM285595</th>\n",
       "      <th>C16_GSM656239</th>\n",
       "      <th>C18_GSM285583</th>\n",
       "      <th>C18_GSM285596</th>\n",
       "      <th>C18_GSM656240</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>6.928782</td>\n",
       "      <td>7.264201</td>\n",
       "      <td>6.934214</td>\n",
       "      <td>6.732989</td>\n",
       "      <td>6.970612</td>\n",
       "      <td>7.150978</td>\n",
       "      <td>6.126826</td>\n",
       "      <td>6.868968</td>\n",
       "      <td>6.641119</td>\n",
       "      <td>...</td>\n",
       "      <td>6.450318</td>\n",
       "      <td>8.049750</td>\n",
       "      <td>7.788162</td>\n",
       "      <td>7.052154</td>\n",
       "      <td>6.517742</td>\n",
       "      <td>6.918501</td>\n",
       "      <td>6.048861</td>\n",
       "      <td>7.041619</td>\n",
       "      <td>6.757932</td>\n",
       "      <td>5.817246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>9.633489</td>\n",
       "      <td>9.977124</td>\n",
       "      <td>10.027529</td>\n",
       "      <td>9.720665</td>\n",
       "      <td>9.605762</td>\n",
       "      <td>10.225542</td>\n",
       "      <td>10.279608</td>\n",
       "      <td>10.459966</td>\n",
       "      <td>10.693337</td>\n",
       "      <td>...</td>\n",
       "      <td>11.130466</td>\n",
       "      <td>11.207738</td>\n",
       "      <td>11.009172</td>\n",
       "      <td>10.615417</td>\n",
       "      <td>11.038938</td>\n",
       "      <td>11.009222</td>\n",
       "      <td>10.216348</td>\n",
       "      <td>11.099187</td>\n",
       "      <td>11.172276</td>\n",
       "      <td>10.561021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>5.066343</td>\n",
       "      <td>4.767264</td>\n",
       "      <td>5.010981</td>\n",
       "      <td>6.139047</td>\n",
       "      <td>4.619361</td>\n",
       "      <td>4.751761</td>\n",
       "      <td>5.818550</td>\n",
       "      <td>5.342529</td>\n",
       "      <td>5.483750</td>\n",
       "      <td>...</td>\n",
       "      <td>6.314438</td>\n",
       "      <td>7.423571</td>\n",
       "      <td>7.507645</td>\n",
       "      <td>7.417087</td>\n",
       "      <td>7.147801</td>\n",
       "      <td>7.747930</td>\n",
       "      <td>7.093641</td>\n",
       "      <td>7.672685</td>\n",
       "      <td>7.511290</td>\n",
       "      <td>6.890117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>4.696881</td>\n",
       "      <td>4.638401</td>\n",
       "      <td>4.956299</td>\n",
       "      <td>6.942556</td>\n",
       "      <td>5.101252</td>\n",
       "      <td>4.730307</td>\n",
       "      <td>8.457690</td>\n",
       "      <td>4.526411</td>\n",
       "      <td>4.908300</td>\n",
       "      <td>...</td>\n",
       "      <td>5.250233</td>\n",
       "      <td>4.974993</td>\n",
       "      <td>5.747498</td>\n",
       "      <td>5.252167</td>\n",
       "      <td>5.210531</td>\n",
       "      <td>7.083187</td>\n",
       "      <td>5.252222</td>\n",
       "      <td>5.037613</td>\n",
       "      <td>5.495281</td>\n",
       "      <td>5.013987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>4.654278</td>\n",
       "      <td>4.537105</td>\n",
       "      <td>4.928739</td>\n",
       "      <td>5.063991</td>\n",
       "      <td>4.584168</td>\n",
       "      <td>4.911880</td>\n",
       "      <td>5.935311</td>\n",
       "      <td>4.519470</td>\n",
       "      <td>4.757861</td>\n",
       "      <td>...</td>\n",
       "      <td>4.651688</td>\n",
       "      <td>4.920573</td>\n",
       "      <td>4.636333</td>\n",
       "      <td>4.883712</td>\n",
       "      <td>4.779395</td>\n",
       "      <td>4.744335</td>\n",
       "      <td>4.513140</td>\n",
       "      <td>4.838428</td>\n",
       "      <td>4.961475</td>\n",
       "      <td>4.653340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>8.292906</td>\n",
       "      <td>8.105281</td>\n",
       "      <td>7.937988</td>\n",
       "      <td>7.999612</td>\n",
       "      <td>8.855890</td>\n",
       "      <td>7.806027</td>\n",
       "      <td>9.758197</td>\n",
       "      <td>8.501278</td>\n",
       "      <td>8.374183</td>\n",
       "      <td>...</td>\n",
       "      <td>9.448039</td>\n",
       "      <td>9.630978</td>\n",
       "      <td>9.521447</td>\n",
       "      <td>9.229861</td>\n",
       "      <td>9.490282</td>\n",
       "      <td>9.283237</td>\n",
       "      <td>9.411544</td>\n",
       "      <td>9.215158</td>\n",
       "      <td>9.712400</td>\n",
       "      <td>10.122851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>14.742986</td>\n",
       "      <td>13.773021</td>\n",
       "      <td>14.830606</td>\n",
       "      <td>14.312092</td>\n",
       "      <td>14.348803</td>\n",
       "      <td>14.587798</td>\n",
       "      <td>14.803615</td>\n",
       "      <td>13.853572</td>\n",
       "      <td>14.678446</td>\n",
       "      <td>...</td>\n",
       "      <td>13.036664</td>\n",
       "      <td>14.959828</td>\n",
       "      <td>14.772800</td>\n",
       "      <td>13.594750</td>\n",
       "      <td>14.728186</td>\n",
       "      <td>15.033645</td>\n",
       "      <td>14.133952</td>\n",
       "      <td>15.138900</td>\n",
       "      <td>14.903656</td>\n",
       "      <td>14.365655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>10.300151</td>\n",
       "      <td>8.699775</td>\n",
       "      <td>10.506309</td>\n",
       "      <td>9.353966</td>\n",
       "      <td>9.394486</td>\n",
       "      <td>10.295839</td>\n",
       "      <td>10.878299</td>\n",
       "      <td>10.507599</td>\n",
       "      <td>10.150730</td>\n",
       "      <td>...</td>\n",
       "      <td>9.802110</td>\n",
       "      <td>10.171732</td>\n",
       "      <td>11.301930</td>\n",
       "      <td>10.151894</td>\n",
       "      <td>11.540889</td>\n",
       "      <td>11.962657</td>\n",
       "      <td>10.931498</td>\n",
       "      <td>11.211766</td>\n",
       "      <td>12.170715</td>\n",
       "      <td>11.585618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>7.752533</td>\n",
       "      <td>7.176984</td>\n",
       "      <td>5.453495</td>\n",
       "      <td>6.332388</td>\n",
       "      <td>7.268512</td>\n",
       "      <td>6.065287</td>\n",
       "      <td>8.836045</td>\n",
       "      <td>6.155319</td>\n",
       "      <td>5.236456</td>\n",
       "      <td>...</td>\n",
       "      <td>6.947031</td>\n",
       "      <td>8.447849</td>\n",
       "      <td>8.537948</td>\n",
       "      <td>8.133894</td>\n",
       "      <td>7.981674</td>\n",
       "      <td>8.265161</td>\n",
       "      <td>8.834731</td>\n",
       "      <td>7.083681</td>\n",
       "      <td>6.774539</td>\n",
       "      <td>9.130942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>8.975957</td>\n",
       "      <td>9.137541</td>\n",
       "      <td>8.921606</td>\n",
       "      <td>9.264407</td>\n",
       "      <td>9.572660</td>\n",
       "      <td>9.127379</td>\n",
       "      <td>10.663918</td>\n",
       "      <td>10.925996</td>\n",
       "      <td>10.197815</td>\n",
       "      <td>...</td>\n",
       "      <td>9.278051</td>\n",
       "      <td>11.120612</td>\n",
       "      <td>10.751137</td>\n",
       "      <td>9.564864</td>\n",
       "      <td>11.169112</td>\n",
       "      <td>11.102387</td>\n",
       "      <td>9.792246</td>\n",
       "      <td>10.604000</td>\n",
       "      <td>10.873286</td>\n",
       "      <td>9.900377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  Ll_GSM283687  Ll_GSM284355  Ll_GSM284362  \\\n",
       "0      TTHERM_000000042      6.928782      7.264201      6.934214   \n",
       "1      TTHERM_000000045      9.633489      9.977124     10.027529   \n",
       "2       TTHERM_00000010      5.066343      4.767264      5.010981   \n",
       "3       TTHERM_00000020      4.696881      4.638401      4.956299   \n",
       "4       TTHERM_00000030      4.654278      4.537105      4.928739   \n",
       "...                 ...           ...           ...           ...   \n",
       "20321   TTHERM_02293890      8.292906      8.105281      7.937988   \n",
       "20322   TTHERM_02385080     14.742986     13.773021     14.830606   \n",
       "20323   TTHERM_02555200     10.300151      8.699775     10.506309   \n",
       "20324   TTHERM_02607240      7.752533      7.176984      5.453495   \n",
       "20325   TTHERM_02653470      8.975957      9.137541      8.921606   \n",
       "\n",
       "       Lm_GSM283690  Lm_GSM284357  Lm_GSM284363  Lh_GSM283691  Lh_GSM284360  \\\n",
       "0          6.732989      6.970612      7.150978      6.126826      6.868968   \n",
       "1          9.720665      9.605762     10.225542     10.279608     10.459966   \n",
       "2          6.139047      4.619361      4.751761      5.818550      5.342529   \n",
       "3          6.942556      5.101252      4.730307      8.457690      4.526411   \n",
       "4          5.063991      4.584168      4.911880      5.935311      4.519470   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "20321      7.999612      8.855890      7.806027      9.758197      8.501278   \n",
       "20322     14.312092     14.348803     14.587798     14.803615     13.853572   \n",
       "20323      9.353966      9.394486     10.295839     10.878299     10.507599   \n",
       "20324      6.332388      7.268512      6.065287      8.836045      6.155319   \n",
       "20325      9.264407      9.572660      9.127379     10.663918     10.925996   \n",
       "\n",
       "       Lh_GSM284364  ...  C12_GSM656237  C14_GSM285580  C14_GSM285593  \\\n",
       "0          6.641119  ...       6.450318       8.049750       7.788162   \n",
       "1         10.693337  ...      11.130466      11.207738      11.009172   \n",
       "2          5.483750  ...       6.314438       7.423571       7.507645   \n",
       "3          4.908300  ...       5.250233       4.974993       5.747498   \n",
       "4          4.757861  ...       4.651688       4.920573       4.636333   \n",
       "...             ...  ...            ...            ...            ...   \n",
       "20321      8.374183  ...       9.448039       9.630978       9.521447   \n",
       "20322     14.678446  ...      13.036664      14.959828      14.772800   \n",
       "20323     10.150730  ...       9.802110      10.171732      11.301930   \n",
       "20324      5.236456  ...       6.947031       8.447849       8.537948   \n",
       "20325     10.197815  ...       9.278051      11.120612      10.751137   \n",
       "\n",
       "       C14_GSM656238  C16_GSM285582  C16_GSM285595  C16_GSM656239  \\\n",
       "0           7.052154       6.517742       6.918501       6.048861   \n",
       "1          10.615417      11.038938      11.009222      10.216348   \n",
       "2           7.417087       7.147801       7.747930       7.093641   \n",
       "3           5.252167       5.210531       7.083187       5.252222   \n",
       "4           4.883712       4.779395       4.744335       4.513140   \n",
       "...              ...            ...            ...            ...   \n",
       "20321       9.229861       9.490282       9.283237       9.411544   \n",
       "20322      13.594750      14.728186      15.033645      14.133952   \n",
       "20323      10.151894      11.540889      11.962657      10.931498   \n",
       "20324       8.133894       7.981674       8.265161       8.834731   \n",
       "20325       9.564864      11.169112      11.102387       9.792246   \n",
       "\n",
       "       C18_GSM285583  C18_GSM285596  C18_GSM656240  \n",
       "0           7.041619       6.757932       5.817246  \n",
       "1          11.099187      11.172276      10.561021  \n",
       "2           7.672685       7.511290       6.890117  \n",
       "3           5.037613       5.495281       5.013987  \n",
       "4           4.838428       4.961475       4.653340  \n",
       "...              ...            ...            ...  \n",
       "20321       9.215158       9.712400      10.122851  \n",
       "20322      15.138900      14.903656      14.365655  \n",
       "20323      11.211766      12.170715      11.585618  \n",
       "20324       7.083681       6.774539       9.130942  \n",
       "20325      10.604000      10.873286       9.900377  \n",
       "\n",
       "[20326 rows x 48 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "9c9ac460",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_first_3_veg_phases = [\n",
    "            #  'Ll', \n",
    "            #  'Lm', \n",
    "            #  'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "            #  'S9', \n",
    "            #  # 'S12', \n",
    "            #  'S15', \n",
    "            #  'S24'\n",
    "             ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "01961773",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['S0_GSM285363',\n",
       " 'S0_GSM285554',\n",
       " 'S0_GSM285561',\n",
       " 'S3_GSM285542',\n",
       " 'S3_GSM285555',\n",
       " 'S3_GSM285562',\n",
       " 'S6_GSM285543',\n",
       " 'S6_GSM285556',\n",
       " 'S6_GSM285563']"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_3_veg_phases_cols = [col for col in list(full_filtered_df.columns) if col.split('_')[0] in list_first_3_veg_phases]\n",
    "first_3_veg_phases_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "31c6b3e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 10)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "((full_filtered_df[['TTHERM_ID']+first_3_veg_phases_cols])[(full_filtered_df[first_3_veg_phases_cols].sum(axis=1) < 30)]).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7ace6be-7982-4461-bc6e-f49a60b1e56c",
   "metadata": {},
   "source": [
    "## Define functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "bb82d41d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_nns(data_df, nn, metric, random_state=42, n_jobs=-1, p_minkowski=1, distance_matrix=None):\n",
    "    \n",
    "    # if metric == 'clr':\n",
    "    num_neighbors = NearestNeighbors(n_neighbors=nn, metric='precomputed', n_jobs=-1).fit(distance_matrix)\n",
    "    nn_dists, nn_idxs = num_neighbors.kneighbors(return_distance=True)\n",
    "    return nn_idxs, nn_dists\n",
    "\n",
    "    # n_trees = min(64, 5 + int(round((data_df.shape[0]) ** 0.5 / 20.0)))\n",
    "    # n_iters = max(5, int(round(np.log2(data_df.shape[0]))))\n",
    "\n",
    "    # if metric == 'minkowski':\n",
    "    #     knn_search_index = NNDescent(\n",
    "    #             data_df,\n",
    "    #             n_neighbors=nn,\n",
    "    #             metric=metric,\n",
    "    #             metric_kwds={'p': p_minkowski},\n",
    "    #             random_state=random_state,\n",
    "    #             n_trees=n_trees,\n",
    "    #             n_iters=n_iters,\n",
    "    #             max_candidates=60,\n",
    "    #             # low_memory=low_memory,\n",
    "    #             n_jobs=n_jobs,\n",
    "    #             verbose=False,\n",
    "    #             compressed=False,\n",
    "    #         )\n",
    "    # else:\n",
    "    #     knn_search_index = NNDescent(\n",
    "    #                 data_df,\n",
    "    #                 n_neighbors=nn,\n",
    "    #                 metric=metric,\n",
    "    #                 # metric_kwds=metric_kwds,\n",
    "    #                 random_state=random_state,\n",
    "    #                 n_trees=n_trees,\n",
    "    #                 n_iters=n_iters,\n",
    "    #                 max_candidates=60,\n",
    "    #                 # low_memory=low_memory,\n",
    "    #                 n_jobs=n_jobs,\n",
    "    #                 verbose=False,\n",
    "    #                 compressed=False,\n",
    "    #             )\n",
    "    # nn_idxs, nn_dists = knn_search_index.neighbor_graph\n",
    "\n",
    "    # return nn_idxs, nn_dists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "674c6371-69ed-44cd-891b-f9bdcb1ee41b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_geom_mean_expression(expression_df):\n",
    "    \"\"\"\n",
    "    \n",
    "    Function to take an expression dataframe from the microarrays and collapse it into the means of\n",
    "    all replicate chips.\n",
    "    \"\"\"\n",
    "    # C2 and S12 got removed during quality control\n",
    "    x = [\n",
    "        'Ll', \n",
    "        'Lm', \n",
    "        'Lh', \n",
    "        'S0', \n",
    "        'S3', \n",
    "        'S6', \n",
    "        'S9', \n",
    "        # 'S12', \n",
    "        'S15', \n",
    "        'S24', \n",
    "        'C0', \n",
    "        # 'C2', \n",
    "        'C4', \n",
    "        'C6', \n",
    "        'C8', \n",
    "        'C10', \n",
    "        'C12', \n",
    "        'C14', \n",
    "        'C16', \n",
    "        'C18']\n",
    "    \n",
    "    # cols = expression_df.columns[1:]\n",
    "    # x = [c for c in x if c in cols]\n",
    "    \n",
    "    condition_expr_dict = {c.split(\"_\")[0]: [] for c in expression_df.columns[1:]}\n",
    "    \n",
    "    for c in list(expression_df.columns)[1:]:\n",
    "        \n",
    "        cond = c.split('_')[0]\n",
    "        if cond in condition_expr_dict.keys():\n",
    "            expr_list = condition_expr_dict.get(cond, [])\n",
    "\n",
    "            # Need to avoid true zeros\n",
    "            expr_list.append(expression_df[c].values)\n",
    "            condition_expr_dict[cond] = expr_list\n",
    "        \n",
    "    condition_mean_dict = {c: (st.mstats.gmean(np.array(condition_expr_dict[c]) + 1, 0) - 1) for c in condition_expr_dict.keys() if c in x}\n",
    "    \n",
    "    mean_expr_df = pd.DataFrame(condition_mean_dict)\n",
    "    mean_expr_df['TTHERM_ID'] = expression_df['TTHERM_ID'].values\n",
    "    cols = list(mean_expr_df.columns)\n",
    "    reorder = cols[-1:] + cols[:-1]\n",
    "    mean_expr_df = mean_expr_df[reorder]\n",
    "    \n",
    "    return mean_expr_df\n",
    "\n",
    "def normalizer(array):\n",
    "    \"\"\"\n",
    "    Normalizes the values of an array to range from zero to one\n",
    "    \"\"\"\n",
    "    \n",
    "    a = np.array(array)\n",
    "    \n",
    "    normalized = (array - np.min(array)) / (np.max(array) - np.min(array))\n",
    "    \n",
    "    return normalized\n",
    "\n",
    "def normalize_expression_per_gene(expression_df):\n",
    "    \"\"\"\n",
    "    Function to normalize all gene expression to range from zero to one.\n",
    "    \"\"\"\n",
    "    if 'TTHERM_ID' in expression_df.columns:\n",
    "        ttids = expression_df['TTHERM_ID'].values\n",
    "        data = expression_df[list(expression_df.columns)[1:]]\n",
    "        \n",
    "        norm_expression_df = data.apply(lambda row: normalizer(row), axis=1)\n",
    "        norm_expression_df['TTHERM_ID'] = ttids\n",
    "        \n",
    "        columns = norm_expression_df.columns.tolist()\n",
    "        \n",
    "        rearrangment = columns[-1:] + columns[:-1]\n",
    "        \n",
    "        norm_expression_df = norm_expression_df[rearrangment]\n",
    "        \n",
    "    else:\n",
    "        norm_expression_df = expression_df.apply(lambda row: normalizer(row), axis=1)\n",
    "    \n",
    "    return norm_expression_df\n",
    "    \n",
    "\n",
    "\n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in order of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "    \n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    \n",
    "def run_leiden(df, n_components=2, n_neighbors=3, random_state=42, metric='manhattan', leiden_type='modularity', la_res_param=1.0, return_dists=True):\n",
    "    \"\"\"\n",
    "    Function to compute the simplicial sets for coexpression using UMAP and to then apply\n",
    "    the Leiden algorithm to cluster the resulting graph.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas dataframe\n",
    "        the expression data\n",
    "    n_components : int (default 2)\n",
    "        the number of dimensions onto which the data should be projected\n",
    "    n_neighbors : int (default 15)\n",
    "        a parameter for the UMAP algorithm. I think it has to do with balancing\n",
    "        local vs. global topology in the data\n",
    "    random_state : float (default 42)\n",
    "        Constraining this parameter makes the output reproducible\n",
    "    metric : str (default \"euclidean\")\n",
    "        The distance function\n",
    "    return_dists : Bool (default True)\n",
    "        Whether the function should return the computed distances\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    leiden_modules : np array\n",
    "        An array of ints, each corresponding to the module (or cluster) to which a gene belongs,\n",
    "        listed in ortder of the input dataframe\n",
    "    \"\"\"\n",
    "    \n",
    "    data = df[list(df.columns)[1:]].values\n",
    "\n",
    "    labels_idxs = list(range(data.shape[0]))  \n",
    "\n",
    "#     mapper = umap.UMAP(random_state=random_state, n_components=n_components, n_neighbors=n_neighbors).fit(data)\n",
    "    \n",
    "    manhattan_distance_matrix = pairwise_distances(data, metric='manhattan', n_jobs=-1)\n",
    "    # nn = NearestNeighbors(n_neighbors=n_neighbors, metric='precomputed', n_jobs=-1).fit(manhattan_distance_matrix)\n",
    "    # nn_dists, nn_idxs = nn.kneighbors(return_distance=True)\n",
    "\n",
    "    nn_idxs, nn_dists = compute_nns(data, n_neighbors, metric, random_state, distance_matrix=manhattan_distance_matrix)\n",
    "    \n",
    "    result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, knn_indices=nn_idxs, knn_dists=nn_dists, return_dists=return_dists)\n",
    "    # result, sigmas, rhos, dists = umap.umap_.fuzzy_simplicial_set(data, n_neighbors, random_state, metric, return_dists=return_dists)\n",
    "\n",
    "\n",
    "    \n",
    "    sources, targets = result.nonzero()\n",
    "    edge_list = zip(sources, targets)\n",
    "    weights = result.data\n",
    "    \n",
    "    g = ig.Graph(edges=edge_list, edge_attrs={'weight': weights})\n",
    "    nx_g = nx.Graph(g.get_edgelist())\n",
    "    \n",
    "    if leiden_type == 'modularity':\n",
    "        partition = la.find_partition(g, la.ModularityVertexPartition, seed=random_state, weights='weight')\n",
    "    elif leiden_type == 'cpm':\n",
    "        partition = la.find_partition(g, la.CPMVertexPartition, resolution_parameter = la_res_param, seed=random_state, weights='weight')\n",
    "    else:\n",
    "        raise ValueError('Invalid value for leiden_type parameter')\n",
    "    leiden_modules = np.array(partition.membership)\n",
    "\n",
    "    communities = {}\n",
    "\n",
    "    for idx, membership in enumerate(leiden_modules):\n",
    "        if membership not in communities:\n",
    "            communities[membership] = []\n",
    "        communities[membership].append(labels_idxs[idx])\n",
    "\n",
    "    sscore = silhouette_score(manhattan_distance_matrix, leiden_modules, metric='precomputed')\n",
    "    modularity = nx.community.quality.modularity(nx_g, communities.values(), weight='weight')\n",
    "    \n",
    "    return leiden_modules, dists, sscore, modularity\n",
    "\n",
    "def build_leiden_label_df(data_df, phases, random_state=42, n_neighbors=3, metric='manhattan', leiden_type='modularity', la_res_param=1.0, lldf=None):\n",
    "    \"\"\"\n",
    "    Function to build a dataframe of genes labeled according to their UMAP/Leiden modules\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    data_df : pandas DataFrame\n",
    "        The expression data\n",
    "    phases : str ('full', 'veg', or 'sex')\n",
    "        The physiological phases for which expression data is being provided\n",
    "    lldf : pandas DataFrame (default None)\n",
    "        Another leiden label df (lldf) to which to add a column\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    lldf : pandas DataFrame\n",
    "        Leiden Label DataFrame. Gene IDs and their corresponding UMAP/Leiden module\n",
    "        computed for a specific physiological regime (full set (full), vegetative only\n",
    "        (veg), or sexual only (sex))\n",
    "    \"\"\"\n",
    "    \n",
    "    if type(lldf) == type(None):\n",
    "        lldf = pd.DataFrame.from_dict({'TTHERM_ID': []})\n",
    "    \n",
    "    leiden_modules, dists, sscore, modularity = run_leiden(data_df, random_state=random_state, n_neighbors=n_neighbors, metric=metric, leiden_type=leiden_type, la_res_param=la_res_param)\n",
    "    lldf['TTHERM_ID'] = data_df['TTHERM_ID'].values\n",
    "    \n",
    "    lldf[f'leiden_label_{phases}'] = leiden_modules\n",
    "    \n",
    "    return lldf, dists, sscore, modularity\n",
    "\n",
    "\n",
    "# The two functions below are taken and adapted from the UMAP package\n",
    "def _get_umap_embedding(umap_object):\n",
    "    if hasattr(umap_object, \"embedding_\"):\n",
    "        return umap_object.embedding_\n",
    "    elif hasattr(umap_object, \"embedding\"):\n",
    "        return umap_object.embedding\n",
    "    else:\n",
    "        raise ValueError(\"Could not find embedding attribute of umap_object\")\n",
    "        \n",
    "def plot_enrichment(enrich_column_data_source):\n",
    "    \n",
    "    # pdb.set_trace()\n",
    "    \n",
    "    # y_range = FactorRange(factors=[str(y) for y in enrich_df['module'].unique()])\n",
    "    \n",
    "    # grouped = enrich_df.groupby('module')\n",
    "    \n",
    "    hover = [\n",
    "        ('module', '@module'),\n",
    "        ('term', '@term'),\n",
    "        ('info', '@info'),\n",
    "        ('fold-change', '@fold_change'),\n",
    "        ('bonferroni', '@bonferroni')\n",
    "    ]\n",
    "    \n",
    "    p = bokeh.plotting.figure(\n",
    "        height=1000,\n",
    "        width=400,\n",
    "        # y_range=y_range,\n",
    "        title='Functional term enrichment in modules',\n",
    "        x_axis_label='fold-change',\n",
    "        y_axis_label='module',\n",
    "        x_axis_type='log',\n",
    "        tooltips=hover,\n",
    "        # background_fill_color='black'\n",
    "    )\n",
    "    \n",
    "    # cds = bokeh.models.ColumnDataSource(enrich_df)\n",
    "    # print(enrich_df.head())\n",
    "    \n",
    "    p.circle(y=jitter('module', width=0.4), x='fold_change', source=enrich_column_data_source, alpha=0.3, size=7, color='color', line_color='black')\n",
    "    # p.xaxis.major_label_orientation = 45\n",
    "    p.ygrid.minor_grid_line_color = 'navy'\n",
    "    p.ygrid.minor_grid_line_alpha = 0.1\n",
    "    # p.xgrid.band_fill_alpha = 0.1\n",
    "    # p.xgrid.band_fill_color = \"navy\"\n",
    "    ticker = []\n",
    "    for m in enrich_column_data_source.data['module']:\n",
    "        if m not in ticker:\n",
    "            ticker.append(m)\n",
    "    p.yaxis.ticker = ticker\n",
    "    p.y_range.flipped = True\n",
    "    p.xaxis.major_label_text_font_size = '12pt'\n",
    "    p.yaxis.major_label_text_font_size = '12pt'\n",
    "    p.yaxis.axis_label_text_font_size = '12pt'\n",
    "    p.xaxis.axis_label_text_font_size = '12pt'\n",
    "    \n",
    "    return p\n",
    "\n",
    "def heatmap(column_data_source, ls_color_palette, r_low, r_high, x_axis_factors, y_axis_factors, s_z=\"normalized_expression\", index_name='TTHERM_ID', col_name='phase'):\n",
    "    # adapted from https://gitlab.com/biotransistor/bokehheat/-/blob/master/bokehheat/heat.py\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df_matrx: a dataframe in same xy orientation as the final heatmap.\n",
    "          the index should cary the y axis label.\n",
    "          the column should cary the x axis label.\n",
    "          the matrix as such should only cary the z axis values.\n",
    "\n",
    "        ls_color_palette: a list color strings to specify the color spectrum.\n",
    "            this variable is compatible with the ordinary bokeh palettes:\n",
    "            https://bokeh.pydata.org/en/latest/docs/reference/palettes.html\n",
    "\n",
    "        r_low: quantitative minimum value. the dataset can contain lower values,\n",
    "            but for color labeling they will be mapped to this minimum value.\n",
    "            e.g.: -8.\n",
    "\n",
    "        r_high: quantitative maximum value. the dataset can contain lower values,\n",
    "            but for color labeling they will be mapped to this maximum value.\n",
    "            e.g.: 8.\n",
    "\n",
    "        s_z: string. label that specifies what the values in the matrix actually\n",
    "            are. e.g.: 'gene expression [log2]'\n",
    "\n",
    "    output:\n",
    "        p: bokeh plot object.\n",
    "\n",
    "    description:\n",
    "        this function will return a bokeh based interactive heatmap plot.\n",
    "        the color are representing the z value.\n",
    "    \"\"\"\n",
    "    # index as string\n",
    "#     df_matrix.index = df_matrix.index.astype(str)\n",
    "#     df_matrix.columns = df_matrix.columns.astype(str)\n",
    "\n",
    "#     # handle y and x axis name\n",
    "#     if (df_matrix.index.name == None):\n",
    "#         df_matrix.index.name = \"y_axis\"\n",
    "#     if (df_matrix.columns.name == None):\n",
    "#         df_matrix.columns.name = \"x_axis\"\n",
    "    # pdb.set_trace()\n",
    "    s_y = index_name\n",
    "    \n",
    "    # df_matrix.columns.name = 'phase'\n",
    "    s_x = col_name\n",
    "    \n",
    "    \n",
    "    # print(df_matrix.head())\n",
    "    \n",
    "    # melt dataframe\n",
    "    # df_tidy = df_matrix.reset_index().melt(\n",
    "    #     id_vars=[df_matrix.index.name],\n",
    "    #     value_name=s_z\n",
    "    # )\n",
    "    # print(df_tidy.head())\n",
    "    # color declaration\n",
    "    d_zcolormapper = linear_cmap(\n",
    "        field_name=s_z,\n",
    "        palette=ls_color_palette,\n",
    "        low=r_low,\n",
    "        high=r_high\n",
    "    )\n",
    "    # tooltip declaration\n",
    "    lt_tooltip = [\n",
    "        (s_y, f\"@{s_y}\"),\n",
    "        (s_x, f\"@{s_x}\"),\n",
    "        (s_z, f\"@{s_z}\"),\n",
    "        ('module', f'@module')\n",
    "    ]\n",
    "    # generate figure\n",
    "    o_colorbar = ColorBar(color_mapper=d_zcolormapper['transform'])\n",
    "    p = bokeh.plotting.figure(\n",
    "        y_range=y_axis_factors,\n",
    "        x_range=x_axis_factors,\n",
    "        width=400,\n",
    "        height=1000,\n",
    "        tools = \"box_zoom,hover,pan,reset,wheel_zoom,save\",  # have to be set hardcoded\n",
    "        active_drag = \"box_zoom\",  # have to be set hardcoded\n",
    "        tooltips=lt_tooltip,\n",
    "        title=s_z,\n",
    "        toolbar_location='right',\n",
    "        \n",
    "    )\n",
    "    \n",
    "    p.rect(\n",
    "        source=column_data_source,\n",
    "        x=s_x,\n",
    "        y=s_y,\n",
    "        color=d_zcolormapper,\n",
    "        width=1,\n",
    "        height=1,\n",
    "        fill_alpha='fill_alpha',\n",
    "        line_alpha='line_alpha',\n",
    "        # line_color='white',\n",
    "        nonselection_fill_alpha=0.01,\n",
    "        nonselection_line_alpha=0.01,\n",
    "        # nonselection_line_color=\"white\"\n",
    "    )\n",
    "    p.add_layout(o_colorbar, place='left')\n",
    "    # p.yaxis.major_label_orientation = \"horizontal\"\n",
    "    p.xaxis.major_label_orientation = 45\n",
    "    # p.yaxis.major_label_text_font_size = '0pt'\n",
    "    p.yaxis.visible = False\n",
    "    p.xaxis.major_label_text_font_size = '12pt'\n",
    "\n",
    "    # out\n",
    "    return(p)\n",
    "        \n",
    "def interactive(\n",
    "    embedding_df,\n",
    "    x,\n",
    "    # mean_expression_df,\n",
    "    title=None,\n",
    "    labels=None,\n",
    "    values=None,\n",
    "    hover_data=None,\n",
    "    theme=None,\n",
    "    cmap=\"Blues\",\n",
    "    color_key=None,\n",
    "    color_key_cmap=\"Spectral\",\n",
    "    background=\"white\",\n",
    "#     width=800,\n",
    "#     height=800,\n",
    "    point_size=None,\n",
    "    radius=None, # My contribution\n",
    "#     subset_points=None,\n",
    "    interactive_text_search=False,\n",
    "    interactive_text_search_columns=None,\n",
    "    interactive_text_search_alpha_contrast=0.999,\n",
    "    alpha=None,\n",
    "    normalized=True\n",
    "):\n",
    "    \"\"\"Create an interactive bokeh plot of a UMAP embedding.\n",
    "    While static plots are useful, sometimes a plot that\n",
    "    supports interactive zooming, and hover tooltips for\n",
    "    individual points is much more desireable. This function\n",
    "    provides a simple interface for creating such plots. The\n",
    "    result is a bokeh plot that will be displayed in a notebook.\n",
    "    Note that more complex tooltips etc. will require custom\n",
    "    code -- this is merely meant to provide fast and easy\n",
    "    access to interactive plotting.\n",
    "    Parameters\n",
    "    ----------\n",
    "    embedding_df: pandas DataFrame\n",
    "        A expression dataframe with columns x and y, which are the\n",
    "        2D embedding of a model (e.g., UMAP or pyMDE) on the expression data, and all the\n",
    "        annotations, geometric means of expression, etc.\n",
    "    x: list\n",
    "        The categories for the x-axes of the heatmap and expression profiles\n",
    "    labels: array, shape (n_samples,) (optional, default None)\n",
    "        An array of labels (assumed integer or categorical),\n",
    "        one for each data sample.\n",
    "        This will be used for coloring the points in\n",
    "        the plot according to their label. Note that\n",
    "        this option is mutually exclusive to the ``values``\n",
    "        option.\n",
    "    values: array, shape (n_samples,) (optional, default None)\n",
    "        An array of values (assumed float or continuous),\n",
    "        one for each sample.\n",
    "        This will be used for coloring the points in\n",
    "        the plot according to a colorscale associated\n",
    "        to the total range of values. Note that this\n",
    "        option is mutually exclusive to the ``labels``\n",
    "        option.\n",
    "    hover_data: DataFrame, shape (n_samples, n_tooltip_features)\n",
    "    (optional, default None)\n",
    "        A dataframe of tooltip data. Each column of the dataframe\n",
    "        should be a Series of length ``n_samples`` providing a value\n",
    "        for each data point. Column names will be used for\n",
    "        identifying information within the tooltip.\n",
    "    theme: string (optional, default None)\n",
    "        A color theme to use for plotting. A small set of\n",
    "        predefined themes are provided which have relatively\n",
    "        good aesthetics. Available themes are:\n",
    "           * 'blue'\n",
    "           * 'red'\n",
    "           * 'green'\n",
    "           * 'inferno'\n",
    "           * 'fire'\n",
    "           * 'viridis'\n",
    "           * 'darkblue'\n",
    "           * 'darkred'\n",
    "           * 'darkgreen'\n",
    "    cmap: string (optional, default 'Blues')\n",
    "        The name of a matplotlib colormap to use for coloring\n",
    "        or shading points. If no labels or values are passed\n",
    "        this will be used for shading points according to\n",
    "        density (largely only of relevance for very large\n",
    "        datasets). If values are passed this will be used for\n",
    "        shading according the value. Note that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    color_key: dict or array, shape (n_categories) (optional, default None)\n",
    "        A way to assign colors to categoricals. This can either be\n",
    "        an explicit dict mapping labels to colors (as strings of form\n",
    "        '#RRGGBB'), or an array like object providing one color for\n",
    "        each distinct category being provided in ``labels``. Either\n",
    "        way this mapping will be used to color points according to\n",
    "        the label. Note that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    color_key_cmap: string (optional, default 'Spectral')\n",
    "        The name of a matplotlib colormap to use for categorical coloring.\n",
    "        If an explicit ``color_key`` is not given a color mapping for\n",
    "        categories can be generated from the label list and selecting\n",
    "        a matching list of colors from the given colormap. Note\n",
    "        that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    background: string (optional, default 'white')\n",
    "        The color of the background. Usually this will be either\n",
    "        'white' or 'black', but any color name will work. Ideally\n",
    "        one wants to match this appropriately to the colors being\n",
    "        used for points etc. This is one of the things that themes\n",
    "        handle for you. Note that if theme\n",
    "        is passed then this value will be overridden by the\n",
    "        corresponding option of the theme.\n",
    "    width: int (optional, default 800)\n",
    "        The desired width of the plot in pixels.\n",
    "    height: int (optional, default 800)\n",
    "        The desired height of the plot in pixels\n",
    "    point_size: int (optional, default None)\n",
    "        The size of each point marker\n",
    "    radius: int (optional, default None)\n",
    "        The radius of each point marker (adjusts the point size while zooming)\n",
    "    subset_points: array, shape (n_samples,) (optional, default None)\n",
    "        A way to select a subset of points based on an array of boolean\n",
    "        values.\n",
    "    interactive_text_search: bool (optional, default False)\n",
    "        Whether to include a text search widget above the interactive plot\n",
    "    interactive_text_search_columns: list (optional, default None)\n",
    "        Columns of data source to search. Searches labels and hover_data by default.\n",
    "    interactive_text_search_alpha_contrast: float (optional, default 0.95)\n",
    "        Alpha value for points matching text search. Alpha value for points\n",
    "        not matching text search will be 1 - interactive_text_search_alpha_contrast\n",
    "    alpha: float (optional, default: None)\n",
    "        The alpha blending value, between 0 (transparent) and 1 (opaque).\n",
    "    Returns\n",
    "    -------\n",
    "    \"\"\"\n",
    "    if theme is not None:\n",
    "        cmap = _themes[theme][\"cmap\"]\n",
    "        color_key_cmap = _themes[theme][\"color_key_cmap\"]\n",
    "        background = _themes[theme][\"background\"]\n",
    "\n",
    "    if labels is not None and values is not None:\n",
    "        raise ValueError(\n",
    "            \"Conflicting options; only one of labels or values should be set\"\n",
    "        )\n",
    "        \n",
    "    if alpha is not None:\n",
    "        if not 0.0 <= alpha <= 1.0:\n",
    "            raise ValueError(\"Alpha must be between 0 and 1 inclusive\")\n",
    "\n",
    "    if point_size is None and radius is None:\n",
    "        point_size = 100.0 / np.sqrt(points.shape[0])\n",
    "        \n",
    "    data = embedding_df\n",
    "    # data = data.set_index('TTHERM_ID')\n",
    "    # pdb.set_trace()\n",
    "    if radius is not None:\n",
    "        data['radius'] = radius\n",
    "\n",
    "    if labels is not None:\n",
    "        data[\"label\"] = labels\n",
    "\n",
    "        if color_key is None:\n",
    "            unique_labels = np.unique(labels)\n",
    "            num_labels = unique_labels.shape[0]\n",
    "            color_key = _to_hex(\n",
    "                plt.get_cmap(color_key_cmap)(np.linspace(0, 1, num_labels))\n",
    "            )\n",
    "\n",
    "        if isinstance(color_key, dict):\n",
    "            data[\"color\"] = pd.Series(labels).map(color_key)\n",
    "        else:\n",
    "            # print('here')\n",
    "            unique_labels = np.unique(labels)\n",
    "            if len(color_key) < unique_labels.shape[0]:\n",
    "                # raise ValueError(\n",
    "                #     \"Color key must have enough colors for the number of labels\"\n",
    "                # )\n",
    "                \n",
    "                print('Color key has fewer colors than labels. Making all white')\n",
    "                data['color'] = ['white']*len(labels)\n",
    "            else:\n",
    "\n",
    "                new_color_key = {k: color_key[i] for i, k in enumerate(unique_labels)}\n",
    "                data[\"color\"] = pd.Series(labels).map(new_color_key)\n",
    "\n",
    "        colors = \"color\"\n",
    "\n",
    "    elif values is not None:\n",
    "        data[\"value\"] = values\n",
    "        palette = _to_hex(plt.get_cmap(cmap)(np.linspace(0, 1, 256)))\n",
    "        colors = btr.linear_cmap(\n",
    "            \"value\", palette, low=np.min(values), high=np.max(values)\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        colors = matplotlib.colors.rgb2hex(plt.get_cmap(cmap)(0.5))\n",
    "\n",
    "    # print(data['color'].unique())\n",
    "    # print(colors)\n",
    "\n",
    "    if hover_data is not None:\n",
    "        tooltip_dict = {}\n",
    "        for col_name in hover_data:\n",
    "            data[col_name] = hover_data[col_name]\n",
    "            tooltip_dict[col_name] = \"@{\" + col_name + \"}\"\n",
    "        tooltips = list(tooltip_dict.items())\n",
    "    else:\n",
    "        tooltips = None\n",
    "\n",
    "    if alpha is not None:\n",
    "        data[\"alpha\"] = alpha\n",
    "    else:\n",
    "        alpha = 1\n",
    "        data[\"alpha\"] = alpha\n",
    "\n",
    "    data_source = bokeh.plotting.ColumnDataSource(data)\n",
    "    data_source.data['module'] = hover_data['module']\n",
    "    data_source.data['ID'] = hover_data['ID']\n",
    "    data_source.data['radius'] = np.ones_like(hover_data['ID']) * radius\n",
    "    data_source.data['alpha'] = np.ones_like(hover_data['ID']) * alpha\n",
    "    \n",
    "    # print(data_source.data['ID'][:5])\n",
    "\n",
    "    plot = bokeh.plotting.figure(\n",
    "        width=800,\n",
    "        height=500,\n",
    "        tooltips=tooltips,\n",
    "        tools=\"tap,box_select,pan,wheel_zoom,box_zoom,reset,save\",\n",
    "        background_fill_color=background,\n",
    "        title=title\n",
    "#             x_range=(np.floor(min(points[:,0])), np.ceil(max(points[:,0]))), # Get axes\n",
    "#             y_range=(np.floor(min(points[:,1])), np.ceil(max(points[:,1])))\n",
    "    )\n",
    "\n",
    "    if point_size is not None:\n",
    "\n",
    "        plot.circle(\n",
    "            x=\"x\",\n",
    "            y=\"y\",\n",
    "            source=data_source,\n",
    "            color=colors,\n",
    "            size=point_size,\n",
    "            alpha=\"alpha\",\n",
    "            line_color='black'\n",
    "        )\n",
    "\n",
    "    elif radius is not None:\n",
    "        plot.circle(\n",
    "            x=\"x\",\n",
    "            y=\"y\",\n",
    "            source=data_source,\n",
    "            color=colors,\n",
    "            radius=radius,\n",
    "            alpha=\"alpha\",\n",
    "            line_color='black'\n",
    "        )\n",
    "\n",
    "    plot.grid.visible = False\n",
    "    plot.axis.visible = False\n",
    "\n",
    "    \n",
    "    x_heatmap_profile = x\n",
    "    \n",
    "    # ['Ll', \n",
    "    #      'Lm', \n",
    "    #      'Lh', \n",
    "    #      'S0', \n",
    "    #      'S3', \n",
    "    #      'S6', \n",
    "    #      'S9', \n",
    "    #      # 'S12', \n",
    "    #      'S15', \n",
    "    #      'S24', \n",
    "    #      'C0', \n",
    "    #      # 'C2', \n",
    "    #      'C4', \n",
    "    #      'C6', \n",
    "    #      'C8', \n",
    "    #      'C10', \n",
    "    #      'C12', \n",
    "    #      'C14', \n",
    "    #      'C16', \n",
    "    #      'C18']\n",
    "    \n",
    "    if normalized:\n",
    "        hm_min = 0\n",
    "        hm_max = 1\n",
    "        \n",
    "    else:\n",
    "        hm_min = 2\n",
    "        hm_max = 16\n",
    "    \n",
    "    # For companion heatmap plot\n",
    "    ttherm_ids = embedding_df['TTHERM_ID'].values\n",
    "    hm_df = embedding_df[['TTHERM_ID'] + x_heatmap_profile]\n",
    "    hm_df['module'] = hover_data['module'].values\n",
    "    hm_df_tidy = hm_df.melt(id_vars=['TTHERM_ID', 'module'], var_name='phase', value_name='normalized_expression')\n",
    "    hm_cds = bokeh.plotting.ColumnDataSource(hm_df_tidy)\n",
    "    hm_cds.data['fill_alpha'] = [0.7]*len(hm_df_tidy)\n",
    "    hm_cds.data['line_alpha'] = [0.7]*len(hm_df_tidy)\n",
    "    # hm_cds.data['y_axis'] = ttherm_ids\n",
    "    \n",
    "    hm = heatmap(hm_cds, bokeh.palettes.Inferno256, hm_min, hm_max, x_heatmap_profile, ttherm_ids)\n",
    "    \n",
    "    \n",
    "\n",
    "    # For companion expression plot\n",
    "\n",
    "    expr_source = bokeh.plotting.ColumnDataSource(dict(\n",
    "        ID=['blah'], \n",
    "        expr_xs=[['Ll']], \n",
    "        expr_ys=[[0]],\n",
    "        alpha=[0],\n",
    "        color=['black']))\n",
    "    \n",
    "    if normalized:\n",
    "        y_axis_label = 'Geometric mean expression of normalized replicates'\n",
    "        y_range = (-0.01, 1.01)\n",
    "        \n",
    "    else:\n",
    "        y_axis_label = 'Geometric mean expression of replicates (log2-scale)'\n",
    "        y_range = (3.9, 16.1)\n",
    "    \n",
    "    expr_fig = bokeh.plotting.figure(width=800, \n",
    "                                     height=500,\n",
    "                                     background_fill_color=background,\n",
    "                                     # x_axis_label='Phase or condition',\n",
    "                                     y_axis_label=y_axis_label,\n",
    "                                     x_range=x_heatmap_profile, \n",
    "                                     y_range=y_range\n",
    "                                    )\n",
    "\n",
    "    expr_fig.multi_line('expr_xs', \n",
    "                        'expr_ys', \n",
    "                        source=expr_source, \n",
    "                        alpha='alpha', \n",
    "                        line_width=3, \n",
    "                        line_join='round',\n",
    "                        line_color=\"color\"\n",
    "                       )\n",
    "\n",
    "    expr_fig.xaxis.major_label_orientation = np.pi/4\n",
    "    expr_fig.xaxis.major_label_text_font_size = '12pt'\n",
    "    expr_fig.yaxis.major_label_text_font_size = '12pt'\n",
    "    expr_fig.yaxis.axis_label_text_font_size = '12pt'\n",
    "    expr_fig.xgrid.grid_line_color='whitesmoke'\n",
    "    expr_fig.xgrid.grid_line_alpha=0.2\n",
    "    expr_fig.ygrid.grid_line_color='whitesmoke'\n",
    "    expr_fig.ygrid.grid_line_alpha=0.2\n",
    "\n",
    "    # For data table\n",
    "    s2 = bokeh.plotting.ColumnDataSource(data=dict(ID=[]))\n",
    "\n",
    "    columns = [TableColumn(field=\"ID\",  title=\"TTHERM_ID\", width=160, formatter=HTMLTemplateFormatter(template='<a href=\"http://tet.ciliate.org/index.php/feature/details/feature_details.php?feature_name=<%= ID %>\"target=\"_blank\"><%= ID %></a>')),\n",
    "               TableColumn(field=\"module\",  title=\"Module\", width=160),\n",
    "               TableColumn(field='TGD2021_description', title='TGD2021_description', width=160),\n",
    "               TableColumn(field=\"Description\", title=\"eggNOG_description\", width=160),\n",
    "               TableColumn(field=\"Preferred_name\", title=\"eggNOG_preferred_name\", width=160),\n",
    "               TableColumn(field=\"max_annot_lvl\", title=\"max_annot_lvl\", width=160),\n",
    "               TableColumn(field=\"COG_category\", title=\"COG_category\", width=160),\n",
    "               TableColumn(field='EC', title='EC', width=160),\n",
    "               TableColumn(field='GOs', title='GOs', width=160),\n",
    "               TableColumn(field='KEGG_ko', title='KEGG_ko', width=160),\n",
    "               TableColumn(field='KEGG_Pathway', title='KEGG_Pathway', width=160),\n",
    "               TableColumn(field='KEGG_Module', title='KEGG_Module', width=160),\n",
    "               TableColumn(field='KEGG_Reaction', title='KEGG_Reaction', width=160),\n",
    "               TableColumn(field='KEGG_rclass', title='KEGG_rclass', width=160),\n",
    "               TableColumn(field='BRITE', title='BRITE', width=160),\n",
    "               TableColumn(field='KEGG_TC', title='KEGG_TC', width=160),\n",
    "               TableColumn(field='CAZy', title='CAZy', width=160),\n",
    "               TableColumn(field='BiGG_Reaction', title='BiGG_Reaction', width=160),\n",
    "#                    TableColumn(field=\"x\",  title=\"x\"),\n",
    "#                    TableColumn(field=\"y\",  title=\"y\")\n",
    "              ]\n",
    "    table = DataTable(source=s2, \n",
    "                      columns=columns, \n",
    "                      width=1600, \n",
    "                      height=500,\n",
    "                      editable=True,\n",
    "                      selectable=True,\n",
    "                      sortable=True,\n",
    "                      index_width=10,\n",
    "                      fit_columns=False,\n",
    "                     )\n",
    "    \n",
    "    heatmap_callback = CustomJS(\n",
    "        args=dict(\n",
    "            s1=data_source,\n",
    "            s_hm=hm_cds,\n",
    "            cols=x_heatmap_profile\n",
    "        ),\n",
    "        code=\"\"\"\n",
    "        var d1 = s1.data;\n",
    "        var d_hm = s_hm.data;\n",
    "        \n",
    "        var inds = s1.selected.indices;\n",
    "        const num_cols = cols.length;\n",
    "        \n",
    "        //d_hm['TTHERM_ID'] = []\n",
    "        //d_hm['normalized_expression'] = []\n",
    "        d_hm['fill_alpha'] = []\n",
    "        d_hm['line_alpha'] = []\n",
    "        \n",
    "        var selected_ttherm_ids = [];\n",
    "        \n",
    "        //Careful here! Number is hardcoded to match the number of genes in dataset EDIT: now it is dynamic)\n",
    "        var ttids = d_hm['TTHERM_ID'].slice(0, \"\"\"+str(num_genes)+\"\"\");\n",
    "        \n",
    "        if (inds.length == 0) {\n",
    "            d_hm['fill_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.7)\n",
    "            d_hm['line_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.7)\n",
    "        }else{\n",
    "        \n",
    "            // Start with everything deselected\n",
    "            d_hm['fill_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.01)\n",
    "            d_hm['line_alpha'] = Array(d_hm['TTHERM_ID'].length).fill(0.01)\n",
    "        \n",
    "            // Get the selected indices\n",
    "            for (var i = 0; i < inds.length; i++) {\n",
    "                selected_ttherm_ids.push(d1['ID'][inds[i]])\n",
    "            }\n",
    "            console.log(selected_ttherm_ids);\n",
    "            \n",
    "            // iterate over the selected ttherm ids\n",
    "            for (var j = 0; j < selected_ttherm_ids.length; j++) {\n",
    "            \n",
    "                // var selected_gene = selected_ttherm_ids[j];\n",
    "                // console.log(selected_gene);\n",
    "\n",
    "                // ad hoc function to find if ttherm ids match\n",
    "                var match = (element) => element == selected_ttherm_ids[j];\n",
    "            \n",
    "                // get index of matching ttherm id in heatmap\n",
    "                var gene_index = ttids.findIndex(match);\n",
    "                console.log(gene_index);\n",
    "                \n",
    "                // loop over the columns and highlight the selected genes\n",
    "                for (var k = 0; k < num_cols; k++) {\n",
    "                \n",
    "                    d_hm['fill_alpha'][gene_index] = 0.7\n",
    "                    d_hm['line_alpha'][gene_index] = 0.7\n",
    "\n",
    "                    gene_index = gene_index + ttids.length\n",
    "                \n",
    "                }\n",
    "            \n",
    "            }\n",
    "            \n",
    "        }\n",
    "        \n",
    "        console.log(d_hm);\n",
    "        \n",
    "        s_hm.change.emit();\n",
    "        \n",
    "        \"\"\"\n",
    "    )\n",
    "\n",
    "    expression_callback = CustomJS(\n",
    "        args=dict(\n",
    "            s1=data_source,\n",
    "            s_expr=expr_source,\n",
    "            alpha=alpha,\n",
    "        ),\n",
    "        code=\"\"\"\n",
    "        var d1 = s1.data;\n",
    "        var d_expr = s_expr.data;\n",
    "\n",
    "        var inds = s1.selected.indices;\n",
    "        // console.log(inds)\n",
    "\n",
    "        // console.log(d1['ID'].length)\n",
    "\n",
    "        // d1['alpha'] = Array(d1['ID'].length).fill(0.2)\n",
    "\n",
    "        // console.log(d_expr['ID'].length, d_expr['expr_xs'].length, d_expr['expr_ys'].length)\n",
    "\n",
    "        d_expr['ID'] = [['blah']]\n",
    "        d_expr['expr_xs'] = [['Ll']]\n",
    "        d_expr['expr_ys'] = [[0]]\n",
    "        d_expr['alpha'] = [0]\n",
    "        d_expr['color'] = ['black']\n",
    "        // s_expr.change.emit();\n",
    "\n",
    "        // debugger;\n",
    "\n",
    "        for (var i = 0; i < inds.length; i++) {\n",
    "            // d_expr['alpha'][inds[i]] = 1/(inds.length * 2)\n",
    "            // console.log(inds[i], i)\n",
    "            d_expr['ID'].push(Array(18).fill(d1['ID'][inds[i]]))\n",
    "            d_expr['expr_xs'].push(d1['expr_xs'][inds[i]])\n",
    "            d_expr['expr_ys'].push(d1['expr_ys'][inds[i]])\n",
    "            d_expr['alpha'].push(Math.min(1, Math.max(7/(inds.length), 0.05)))\n",
    "            d_expr['color'].push(d1['color'][inds[i]])\n",
    "            // console.log(d_expr)\n",
    "            // console.log(i)\n",
    "            // console.log(\n",
    "            //     d_expr['ID'].length, \n",
    "            //     d_expr['expr_xs'].length, \n",
    "            //     d_expr['expr_ys'].length\n",
    "            // )\n",
    "        }\n",
    "\n",
    "        // s1.change.emit();\n",
    "        s_expr.change.emit();\n",
    "        // console.log(s_expr.data)\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "    )\n",
    "\n",
    "    selection_callback =  CustomJS(args=dict(\n",
    "                                          s1=data_source, \n",
    "                                          s2=s2,\n",
    "                                          table=table), \n",
    "                                               code=\"\"\"\n",
    "\n",
    "        var d1 = s1.data;\n",
    "        var d2 = s2.data;\n",
    "\n",
    "\n",
    "        var inds = s1.selected.indices;\n",
    "\n",
    "        d2['module'] = []\n",
    "        d2['ID'] = []\n",
    "        d2['TGD2021_description'] = []\n",
    "        d2['Description'] = []\n",
    "        d2['Preferred_name'] = []\n",
    "        d2['max_annot_lvl'] = []\n",
    "        d2['COG_category'] = []\n",
    "        d2['EC'] = []\n",
    "        d2['GOs'] = []\n",
    "        d2['KEGG_ko'] = []\n",
    "        d2['KEGG_Pathway'] = []\n",
    "        d2['KEGG_Module'] = []\n",
    "        d2['KEGG_Reaction'] = []\n",
    "        d2['KEGG_rclass'] = []\n",
    "        d2['BRITE'] = []\n",
    "        d2['KEGG_TC'] = []\n",
    "        d2['CAZy'] = []\n",
    "        d2['BiGG_Reaction'] = []\n",
    "\n",
    "        for (var i = 0; i < inds.length; i++) {\n",
    "            d2['module'].push(d1['module'][inds[i]])\n",
    "            d2['ID'].push(d1['ID'][inds[i]])\n",
    "            d2['TGD2021_description'].push(d1['TGD2021_description'][inds[i]])\n",
    "            d2['Description'].push(d1['Description'][inds[i]])\n",
    "            d2['Preferred_name'].push(d1['Preferred_name'][inds[i]])\n",
    "            d2['max_annot_lvl'].push(d1['max_annot_lvl'][inds[i]])\n",
    "            d2['COG_category'].push(d1['COG_category'][inds[i]])\n",
    "            d2['EC'].push(d1['EC'][inds[i]])\n",
    "            d2['GOs'].push(d1['GOs'][inds[i]])\n",
    "            d2['KEGG_ko'].push(d1['KEGG_ko'][inds[i]])\n",
    "            d2['KEGG_Pathway'].push(d1['KEGG_Pathway'][inds[i]])\n",
    "            d2['KEGG_Module'].push(d1['KEGG_Module'][inds[i]])\n",
    "            d2['KEGG_Reaction'].push(d1['KEGG_Reaction'][inds[i]])\n",
    "            d2['KEGG_rclass'].push(d1['KEGG_rclass'][inds[i]])\n",
    "            d2['BRITE'].push(d1['BRITE'][inds[i]])\n",
    "            d2['KEGG_TC'].push(d1['KEGG_TC'][inds[i]])\n",
    "            d2['CAZy'].push(d1['CAZy'][inds[i]])\n",
    "            d2['BiGG_Reaction'].push(d1['BiGG_Reaction'][inds[i]])\n",
    "        }\n",
    "        s2.change.emit();\n",
    "        table.change.emit();\n",
    "    \"\"\")\n",
    "\n",
    "    data_source.selected.js_on_change('indices', selection_callback, expression_callback, heatmap_callback)\n",
    "\n",
    "    if interactive_text_search:\n",
    "        text_input = TextInput(value=\"Search module(s) or TTHERM_ID(s), e.g. TTHERM_00321680, TTHERM_00313130...\", width=600)\n",
    "\n",
    "        if interactive_text_search_columns is None:\n",
    "            interactive_text_search_columns = []\n",
    "            if hover_data is not None:\n",
    "                interactive_text_search_columns.extend(hover_data.columns)\n",
    "            if labels is not None:\n",
    "                interactive_text_search_columns.append(\"label\")\n",
    "\n",
    "        if len(interactive_text_search_columns) == 0:\n",
    "            warn(\n",
    "                \"interactive_text_search_columns set to True, but no hover_data or labels provided.\"\n",
    "                \"Please provide hover_data or labels to use interactive text search.\"\n",
    "            )\n",
    "\n",
    "        else:\n",
    "            callback = CustomJS(\n",
    "                args=dict(\n",
    "                    source=data_source,\n",
    "                    s2=s2,\n",
    "                    table=table,\n",
    "                    matching_alpha=interactive_text_search_alpha_contrast,\n",
    "                    non_matching_alpha=1 - interactive_text_search_alpha_contrast,\n",
    "                    search_columns=interactive_text_search_columns,\n",
    "                    default_radius=radius,\n",
    "                    default_alpha=alpha\n",
    "                ),\n",
    "                code=\"\"\"\n",
    "                var data = source.data;\n",
    "                var text_search = cb_obj.value;\n",
    "                var d2 = s2.data;\n",
    "\n",
    "                // var ref_expr = ref_e_s.data;\n",
    "                // var d3 = sel_e_s.data;\n",
    "\n",
    "                var search_terms = text_search.split(',');\n",
    "\n",
    "                d2['module'] = []\n",
    "                d2['ID'] = []\n",
    "\n",
    "                // d3['xs'] = []\n",
    "                // d3['ys'] = []\n",
    "\n",
    "                var search_columns_dict = {}\n",
    "                for (var col in search_columns){\n",
    "                    search_columns_dict[col] = search_columns[col]\n",
    "                }\n",
    "\n",
    "                // First, clear the data table and selection\n",
    "                data['alpha'] = []\n",
    "                data['radius'] = []\n",
    "                source.selected.indices = []\n",
    "\n",
    "                // source.change.emit();\n",
    "                s2.change.emit();\n",
    "                // sel_e_s.change.emit();\n",
    "                table.change.emit();\n",
    "\n",
    "                // Run search\n",
    "                if (text_search.length > 0){\n",
    "                    // Loop over columns and values\n",
    "                    // If there is no match for any column for a given row, change the alpha value\n",
    "                    var string_match = false;\n",
    "                    for (var i = 0; i < data.x.length; i++) {\n",
    "                        string_match = false\n",
    "                        for (var j in search_columns_dict) {\n",
    "                            if (search_terms.some(t => String(data[search_columns_dict[j]][i]).includes(t.trim()))) {\n",
    "                                string_match = true\n",
    "                            }\n",
    "                        }\n",
    "                        if (string_match){\n",
    "                            data['alpha'][i] = matching_alpha\n",
    "                            data['radius'][i] = 1\n",
    "                            d2['module'].push(data['module'][i])\n",
    "                            d2['ID'].push(data['ID'][i])\n",
    "\n",
    "                            // d3['xs'].push(ref_expr['xs'][i])\n",
    "                            // d3['ys'].push(ref_expr['ys'][i])\n",
    "\n",
    "                            // So that these points are actually considered selected\n",
    "                            source.selected.indices.push(i)\n",
    "\n",
    "                        }else{\n",
    "                            data['alpha'][i] = non_matching_alpha\n",
    "                            data['radius'][i] = 0.01\n",
    "                        }\n",
    "                    }\n",
    "                    source.change.emit();\n",
    "                    s2.change.emit();\n",
    "                    // sel_e_s.change.emit();\n",
    "                    table.change.emit();\n",
    "\n",
    "                } else {\n",
    "\n",
    "                    // Loop over columns and values\n",
    "                    // If there is no match for any column for a given row, change the alpha value\n",
    "                    var string_match = false;\n",
    "                    for (var i = 0; i < data.x.length; i++) {\n",
    "                        string_match = false\n",
    "                        for (var j in search_columns_dict) {\n",
    "                            if (search_terms.some(t => String(data[search_columns_dict[j]][i]).includes(t.trim()))) {\n",
    "                                string_match = true\n",
    "                            }\n",
    "                        }\n",
    "                        if (string_match){\n",
    "                            data['alpha'][i] = default_alpha\n",
    "                            data['radius'][i] = default_radius\n",
    "                            d2['module'].push()\n",
    "                            d2['ID'].push()\n",
    "\n",
    "                            // d3['xs'].push()\n",
    "                            // d3['ys'].push()\n",
    "\n",
    "                        }else{\n",
    "                            data['alpha'][i] = non_matching_alpha\n",
    "                            data['radius'][i] = 0.01\n",
    "                        }\n",
    "                    }\n",
    "                    source.change.emit();\n",
    "                    s2.change.emit();\n",
    "                    // sel_e_s.change.emit();\n",
    "                    table.change.emit();\n",
    "\n",
    "                }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "            \"\"\",\n",
    "            )\n",
    "\n",
    "            text_input.js_on_change(\"value\", callback, selection_callback, expression_callback, heatmap_callback)\n",
    "\n",
    "    # Lifted from https://stackoverflow.com/questions/31824124/is-there-a-way-to-save-bokeh-data-table-content\n",
    "    button1 = Button(label=\"Download Annotation Table\", button_type=\"success\", width=550)\n",
    "    button1.js_on_click(\n",
    "        CustomJS(\n",
    "            args=dict(source_data=data_source),\n",
    "            code=\"\"\"\n",
    "            var inds = source_data.selected.indices;\n",
    "            var data = source_data.data;\n",
    "            var out = \"TTHERM_ID\\tmodule\\tTGD2021_description\\teggNOG_description\\teggNOG_preferred_name\\tmax_annot_lvl\\tCOG_category\\tGOs\\tEC\\tKEGG_ko\\tKEGG_Pathway\\tKEGG_Module\\tKEGG_Reaction\\tKEGG_rclass\\tBRITE\\tKEGG_TC\\tCAZy\\tBiGG_Reaction\\\\n\";\n",
    "            for (var i = 0; i < inds.length; i++) {\n",
    "                out += data['ID'][inds[i]] + \"\\t\" + data['module'][inds[i]] + \"\\t\" + data['TGD2021_description'][inds[i]] + \"\\t\" + data['Description'][inds[i]] + \"\\t\" + data['Preferred_name'][inds[i]] + \"\\t\" + data['max_annot_lvl'][inds[i]] + \"\\t\" + data['COG_category'][inds[i]] + \"\\t\" + data['GOs'][inds[i]] + \"\\t\" + data['EC'][inds[i]] + \"\\t\" + data['KEGG_ko'][inds[i]] + \"\\t\" + data['KEGG_Pathway'][inds[i]] + \"\\t\" + data['KEGG_Module'][inds[i]] + \"\\t\" + data['KEGG_Reaction'][inds[i]] + \"\\t\" + data['KEGG_rclass'][inds[i]] + \"\\t\" + data['BRITE'][inds[i]] + \"\\t\" + data['KEGG_TC'][inds[i]] + \"\\t\" + data['CAZy'][inds[i]] + \"\\t\" + data['BiGG_Reaction'][inds[i]] + \"\\\\n\";\n",
    "            }\n",
    "            var file = new Blob([out], {type: 'text/plain'});\n",
    "            var elem = window.document.createElement('a');\n",
    "            elem.href = window.URL.createObjectURL(file);\n",
    "            elem.download = 'selected-annotation-data.tsv';\n",
    "            document.body.appendChild(elem);\n",
    "            elem.click();\n",
    "            document.body.removeChild(elem);\n",
    "            \"\"\"))        \n",
    "    \n",
    "    # NEED TO STOP HARDCODING THIS FILE\n",
    "    enrich_df = pd.read_csv('../enrichment/test_nn3_full_enrichment.csv')\n",
    "    colors = [color_key[int(m)] for m in enrich_df['module'].values]\n",
    "    enrich_df['color'] = colors\n",
    "    \n",
    "    enrich_cds = bokeh.models.ColumnDataSource(enrich_df)\n",
    "    enrich_p = plot_enrichment(enrich_cds)\n",
    "    \n",
    "    button2 = Button(label=\"Download Functional Enrichment Data\", button_type=\"success\", width=450)\n",
    "    button2.js_on_click(\n",
    "        CustomJS(\n",
    "            args=dict(source_data=enrich_cds),\n",
    "            code=\"\"\"\n",
    "            // var inds = source_data.selected.indices;\n",
    "            var data = source_data.data;\n",
    "            var out = \"module\\tterm\\tinfo\\tfold_change\\tbonferroni\\\\n\";\n",
    "            for (var i = 0; i < data['module'].length; i++) {\n",
    "                out += data['module'][i] + \"\\t\" + data['term'][i] + \"\\t\" + data['info'][i] + \"\\t\" + data['fold_change'][i] + \"\\t\" + data['bonferroni'][i] + \"\\\\n\";\n",
    "            }\n",
    "            var file = new Blob([out], {type: 'text/plain'});\n",
    "            var elem = window.document.createElement('a');\n",
    "            elem.href = window.URL.createObjectURL(file);\n",
    "            elem.download = 'enrichment-data.tsv';\n",
    "            document.body.appendChild(elem);\n",
    "            elem.click();\n",
    "            document.body.removeChild(elem);\n",
    "            \"\"\"))  \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    if interactive_text_search:\n",
    "        plot = column(row(column(plot, expr_fig), hm, enrich_p), row(text_input, button1, button2), table)\n",
    "    else:\n",
    "        plot = column(row(column(plot, expr_fig), hm, enrich_p), row(button1, button2), table)\n",
    "\n",
    "    return plot\n",
    "\n",
    "def get_centroid(module_df):\n",
    "    \n",
    "    # get rid of ttherm_ids\n",
    "    data_cols = [c for c in module_df.columns if ('TTHERM' not in c) and ('label' not in c)]\n",
    "    data = module_df[data_cols]\n",
    "    \n",
    "    centroid = data.apply(np.mean, axis=0).values\n",
    "    \n",
    "    return centroid\n",
    "\n",
    "def get_module_centroid_df(expr_df, cluster_label_df, alg, phases):\n",
    "    \n",
    "    merge = expr_df.merge(cluster_label_df, on='TTHERM_ID')\n",
    "    \n",
    "    grouped = merge.groupby(f'{alg}_label_{phases}')\n",
    "    \n",
    "    centroid_rows = []\n",
    "    \n",
    "    for label, grp_df in grouped:\n",
    "        # print(grp_df.head())\n",
    "        centroid = get_centroid(grp_df)\n",
    "        centroid_rows.append(centroid)\n",
    "    \n",
    "    data_cols = [c for c in merge.columns if ('TTHERM' not in c) and ('label' not in c)]\n",
    "    \n",
    "    centroid_df = pd.DataFrame(centroid_rows)\n",
    "    centroid_df.columns = data_cols\n",
    "    centroid_df.index.rename('module', inplace=True)\n",
    "        \n",
    "    return centroid_df\n",
    "\n",
    "def get_all_module_centroids(expr_df, cluster_label_df, alg, phases):\n",
    "    \n",
    "    merge = expr_df.merge(cluster_label_df, on='TTHERM_ID')\n",
    "    \n",
    "    grouped = merge.groupby(f'{alg}_label_{phases}')\n",
    "    \n",
    "    module_centroid_list = []\n",
    "    \n",
    "    for label, grp_df in grouped:\n",
    "        \n",
    "        centroid = get_centroid(grp_df)\n",
    "        module_centroid_list.append( (label, centroid) )\n",
    "        \n",
    "    return module_centroid_list\n",
    "\n",
    "def arrange_modules(expr_df, cluster_label_df, alg, phases):\n",
    "    \n",
    "    if phases == 'full':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12',\n",
    "             'S15', \n",
    "             'S24', \n",
    "             'C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10', \n",
    "             'C12', \n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "        \n",
    "    elif phases == 'veg':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12', \n",
    "             'S15', \n",
    "             'S24']\n",
    "        \n",
    "    elif phases == 'sex':\n",
    "        \n",
    "        x = ['C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10',\n",
    "             'C12',\n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "    cols = ['TTHERM_ID'] + [c for c in expr_df.columns if c.split('_')[0] in x]\n",
    "    \n",
    "    module_centroid_df = get_module_centroid_df(expr_df[cols], cluster_label_df, alg, phases)\n",
    "    \n",
    "    linkage = scipy.cluster.hierarchy.linkage(module_centroid_df, method='average', metric='correlation', optimal_ordering=True)\n",
    "    r_cophcorre, ar_copdist = scipy.cluster.hierarchy.cophenet(linkage, scipy.spatial.distance.pdist(module_centroid_df, metric='correlation'))\n",
    "    \n",
    "    # print(f'The Copheretic correlation is: {r_cophcorre}')\n",
    "    \n",
    "    d_dendro = scipy.cluster.hierarchy.dendrogram(linkage, no_plot=True)\n",
    "    cat_sorted = list(module_centroid_df.iloc[d_dendro['leaves'],:].index)\n",
    "    \n",
    "    sorter_index = dict(zip(cat_sorted, range(len(cat_sorted))))\n",
    "    \n",
    "    reassigned_df = cluster_label_df.copy(deep=True)\n",
    "    \n",
    "    \n",
    "    \n",
    "    reassigned_df[f'{alg}_label_{phases}'] = reassigned_df[f'{alg}_label_{phases}'].map(sorter_index)\n",
    "    print(len(reassigned_df))\n",
    "    \n",
    "    arranged_dfs = []\n",
    "    \n",
    "    for cat in cat_sorted:\n",
    "        \n",
    "        mini_df = reassigned_df.loc[reassigned_df[f'{alg}_label_{phases}'] == cat]\n",
    "        # gene_count += len(mini_df)\n",
    "\n",
    "        arranged_dfs.append(mini_df)\n",
    "        \n",
    "#     gene_count = 0\n",
    "    \n",
    "#     for mdf in arranged_dfs:\n",
    "#         gene_count += len(mdf)\n",
    "    \n",
    "#     print(gene_count)\n",
    "        \n",
    "    arranged_df = pd.concat(arranged_dfs)\n",
    "    \n",
    "    \n",
    "    return arranged_df\n",
    "\n",
    "def plot_embedding(expression_df, embedding_df, annotation_df, label_df, clust_alg, phases, palette, n_components=2, n_neighbors=15, title=None, random_state=42, radius=0.01, normalized=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    Function to plot the UMAP of expression data.\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # get new index for clustered heatmap\n",
    "    label_df = arrange_modules(expression_df, label_df, clust_alg, phases)\n",
    "    \n",
    "    # Weirdly, the heatmap looks better-arranged when I just sort by the modules, as\n",
    "    # given by the hierarchical clustering done by arrange_modules(), than if\n",
    "    # I stay with the order they were given automatically\n",
    "    label_df = label_df.sort_values(by=[f'{clust_alg}_label_{phases}', 'TTHERM_ID'], ascending=False)\n",
    "    new_index = label_df.index\n",
    "    \n",
    "    labels = label_df[f'{clust_alg}_label_{phases}'].values\n",
    "    \n",
    "    data = expression_df[list(expression_df.columns)[1:]].values\n",
    "    \n",
    "    embedding_df['TTHERM_ID'] = expression_df['TTHERM_ID'].values\n",
    "    \n",
    "    merge = expression_df.merge(embedding_df, on='TTHERM_ID')\n",
    "    \n",
    "    merge = merge.reindex(new_index)\n",
    "    \n",
    "    # take part of annotation df that shared TTHERM_IDs with expression df\n",
    "    relevant_annot = annotation_df.iloc[np.in1d(annotation_df['TTHERM_ID'].values, merge['TTHERM_ID'].values)]\n",
    "    merge = merge.merge(relevant_annot, on='TTHERM_ID')\n",
    "    \n",
    "\n",
    "    mean_expression_df = get_geom_mean_expression(merge)\n",
    "    \n",
    "    ttherm_ids = merge['TTHERM_ID'].values\n",
    "    merge = merge.merge(mean_expression_df, on='TTHERM_ID')\n",
    "    \n",
    "    if phases == 'full':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12',\n",
    "             'S15', \n",
    "             'S24', \n",
    "             'C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10', \n",
    "             'C12', \n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "        \n",
    "    elif phases == 'veg':\n",
    "        \n",
    "        x = ['Ll', \n",
    "             'Lm', \n",
    "             'Lh', \n",
    "             'S0', \n",
    "             'S3', \n",
    "             'S6', \n",
    "             'S9', \n",
    "             # 'S12', \n",
    "             'S15', \n",
    "             'S24']\n",
    "        \n",
    "    elif phases == 'sex':\n",
    "        \n",
    "        x = ['C0', \n",
    "             # 'C2', \n",
    "             'C4', \n",
    "             'C6', \n",
    "             'C8', \n",
    "             'C10',\n",
    "             'C12',\n",
    "             'C14', \n",
    "             'C16', \n",
    "             'C18']\n",
    "        \n",
    "    else:\n",
    "        print('Selected phases must be one of full, sex, or veg!')\n",
    "        return\n",
    "\n",
    "    xs = [x for ttid in ttherm_ids]\n",
    "    ys = [merge.loc[merge['TTHERM_ID'] == ttid, x].values[0] for ttid in ttherm_ids]\n",
    "\n",
    "    merge['expr_xs'] = xs\n",
    "    merge['expr_ys'] = ys\n",
    "    \n",
    "    # print(merge.head())\n",
    "\n",
    "    \n",
    "#     pdb.set_trace()\n",
    "    hover_data = pd.DataFrame({\n",
    "                               # 'index':np.arange(len(data)),\n",
    "                               'ID':merge['TTHERM_ID'].values,\n",
    "                               'module':[f'm{int(l):02d}' for l in labels]})\n",
    "    \n",
    "#     palette = [palette[l] for l in sorted(label_df[label_key].unique())]\n",
    "    \n",
    "    p = interactive(merge,\n",
    "                    x,\n",
    "                    # mean_expression_df,\n",
    "                    title=title,\n",
    "                    hover_data=hover_data, \n",
    "                    labels=labels, \n",
    "                    color_key=palette, \n",
    "#                     color_key_cmap='Paired',\n",
    "                    background='black', \n",
    "                    radius=radius,\n",
    "                    alpha=0.7,\n",
    "#                     width=600, \n",
    "#                     height=500,\n",
    "                    interactive_text_search=True,\n",
    "                    normalized=normalized\n",
    "                   )\n",
    "    \n",
    "    #p.children[1].title = title\n",
    "    \n",
    "    return p\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5ade6488",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list_test = ['TTHERM_000013409', 'TTHERM_01321550', 'TTHERM_00011710', 'TTHERM_00321680', 'TTHERM_00355700', 'TTHERM_00938950', 'TTHERM_01372820', 'TTHERM_00013410', 'TTHERM_00390080', 'TTHERM_00516380', 'TTHERM_00038880', 'TTHERM_00059370', 'TTHERM_00473020', 'TTHERM_00497590', 'TTHERM_00558350', 'TTHERM_00052190', 'TTHERM_00392790', 'TTHERM_00410180', 'TTHERM_00685980', 'TTHERM_00445920', 'TTHERM_00471040', 'TTHERM_00140780', 'TTHERM_00145480', 'TTHERM_00321720', 'TTHERM_00628650', 'TTHERM_00526730', 'TTHERM_01156770', 'TTHERM_00312200', 'TTHERM_01332070', 'TTHERM_00318900', 'TTHERM_00340180', 'TTHERM_00592740', 'TTHERM_00440600', 'TTHERM_01321570', 'TTHERM_00537380', 'TTHERM_00585170', 'TTHERM_01197130', 'TTHERM_00554390', 'TTHERM_00649180', 'TTHERM_00691410', ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "911d5a38-541a-4c3a-8288-a8afcf42d7c5",
   "metadata": {},
   "source": [
    "## Define palettes for plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0efd5123-23e9-423e-86b9-68e89a479481",
   "metadata": {},
   "source": [
    "These palettes are from the R package Polychrome. The first is just palette36 with the first color replaced by white. The second is the alphabet palette with white prepended."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b03ec5e-9520-4a4e-aadb-f7a16d7dd6e6",
   "metadata": {},
   "source": [
    "The R code for 64 colors:\n",
    "\n",
    "library(Polychrome);\n",
    "seed <- c(\"#000000\", \"#ff0000\", \"#00ff00\", \"#0000ff\");\n",
    "p64 <- createPalette(64, seed, range=c(40,100));\n",
    "paste(p64, sep=\"\\n\");\n",
    "\n",
    "Then, replace the first and last with ~white"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "680b6c31-013d-4611-97e6-78699c22ff29",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette45 = \"\"\"\n",
    "#51635F\\n#FF1C16\\n#16FC26\\n#403DFC\\n#FE0DCE\\n#F9AA00\\n#00FFD5\\n#22BFFE\\n#BB3551\\n#E6FE97\\n#ECADFF\\n#FFBFBD\\n#CF00F5\\n#0D8B00\\n#D7FEFF\\n#8D7200\\n#F76C00\\n#AD3288\\n#5C5AB8\\n#FC0080\\n#B8FF16\\n#00AAB4\\n#FBE11C\\n#9AAAD9\\n#8BBB8C\\n#934B47\\n#6EFE99\\n#9C6D91\\n#FB9778\\n#9D32AF\\n#D40087\\n#FFDC9D\\n#FF8DB6\\n#A96AFC\\n#FDDDFB\\n#168CF7\\n#FD6CF9\\n#F64553\\n#4D6A00\\n#FAFEDB\\n#A7977D\\n#0DFBFF\\n#86B80D\\n#FD8AE4\\n#B7B126\n",
    "\"\"\".split()\n",
    "\n",
    "palette32 = \"\"\"\n",
    "white\\n#F91622\\n#16FC0D\\n#5138FB\\n#FD00CF\\n#FDD51C\\n#16FDD7\\n#FC8B8E\\n#16BFFF\\n#DF9BFD\\n#669C2A\\n#FEE7C4\\n#F31685\\n#DF16FD\\n#C1F1FE\\n#A23D7E\\n#D5FD0D\\n#8C5A0D\\n#FC790D\\n#4F5CBC\\n#FFCBEF\\n#168D72\\n#68FA93\\n#C4FDC9\\n#F7A449\\n#16789B\\n#AD0DAB\\n#C4262E\\n#0DF1FF\\n#EFF994\\n#B6C1FE\\n#8F22CD\n",
    "\"\"\".split()\n",
    "\n",
    "palette35 = \"\"\"\n",
    "#585F6A\\n#FE1626\\n#00FB0D\\n#2E40FC\\n#FD0DCE\\n#FCD200\\n#F7868C\\n#16FFDC\\n#22BEFB\\n#D28EF6\\n#609000\\n#FFE7C9\\n#F51683\\n#FF730D\\n#CAFE16\\n#AA3586\\n#BEEEFD\\n#BD00FA\\n#895D22\\n#FEC7F0\\n#495AA1\\n#73F995\\n#229270\\n#ED963B\\n#F6FE97\\n#C5FFD0\\n#C50DC8\\n#6993FF\\n#C22A35\\n#16ECFC\\n#AA707E\\n#7A3BCB\\n#7C845C\\n#358FAA\\n#BDBAF6\n",
    "\"\"\".split()\n",
    "\n",
    "palette38 = \"\"\"\n",
    "#636265\\n#F60D16\\n#00F90D\\n#3540FB\\n#FD0DD0\\n#FDDB0D\\n#00FFE2\\n#FA8884\\n#2ABEFE\\n#E5A3FF\\n#518F00\\n#FEFDD5\\n#D51CFF\\n#ED007F\\n#A33879\\n#96731C\\n#C8FB16\\n#C0ECFE\\n#FBC1DA\\n#5658BA\\n#F96900\\n#F69F1C\\n#58FA9C\\n#008E72\\n#BA22B9\\n#167D97\\n#794D8A\\n#CEFE9C\\n#BB222E\\n#954D45\\n#00DCEF\\n#FD66B0\\n#B2FDD3\\n#FDBD9F\\n#A9B4F1\\n#B371FE\\n#849566\\n#2A8EFF\n",
    "\"\"\".split()\n",
    "\n",
    "palette64 = \"\"\"\n",
    "white\\n#FA002E\\n#22FC22\\n#221CFA\\n#FF3DD6\\n#FFDA00\\n#00FEFB\\n#F48684\\n#CEB4FE\\n#FFFFE5\\n#0D933D\\n#CC00F8\\n#800D5D\\n#F10084\\n#22267A\\n#0DADFF\\n#CBFD71\\n#9A761C\\n#F96C00\\n#6399A6\\n#FFBCDA\\n#8D0DA3\\n#F79F26\\n#00FFBF\\n#A37CFB\\n#F68EEB\\n#720D0D\\n#F163AA\\n#7E926A\\n#826386\\n#B41C32\\n#9BEBCE\\n#E2DB83\\n#56D4FA\\n#E6E2FB\\n#925D58\\n#F7C3A7\\n#62E970\\n#220DBD\\n#5583BB\\n#7EA01C\\n#CDFDB6\\n#FD00FB\\n#B30D97\\n#F5FF00\\n#DD77FD\\n#4282FC\\n#BBA6A4\\n#0D8068\\n#AB5F26\\n#F7C26E\\n#9EFE00\\n#9B2EFD\\n#C56887\\n#FD3D68\\n#ABF2FD\\n#835FAC\\n#FF16B1\\n#325371\\n#CA16CA\\n#D26322\\n#AFCFFE\\n#91A1FA\\nfloralwhite\n",
    "\"\"\".split()\n",
    "\n",
    "palette65 = \"\"\"\n",
    "white\\ngainsboro\\n#FA002E\\n#22FC22\\n#221CFA\\n#FF3DD6\\n#FFDA00\\n#00FEFB\\n#F48684\\n#CEB4FE\\n#FFFFE5\\n#0D933D\\n#CC00F8\\n#800D5D\\n#F10084\\n#22267A\\n#0DADFF\\n#CBFD71\\n#9A761C\\n#F96C00\\n#6399A6\\n#FFBCDA\\n#8D0DA3\\n#F79F26\\n#00FFBF\\n#A37CFB\\n#F68EEB\\n#720D0D\\n#F163AA\\n#7E926A\\n#826386\\n#B41C32\\n#9BEBCE\\n#E2DB83\\n#56D4FA\\n#E6E2FB\\n#925D58\\n#F7C3A7\\n#62E970\\n#220DBD\\n#5583BB\\n#7EA01C\\n#CDFDB6\\n#FD00FB\\n#B30D97\\n#F5FF00\\n#DD77FD\\n#4282FC\\n#BBA6A4\\n#0D8068\\n#AB5F26\\n#F7C26E\\n#9EFE00\\n#9B2EFD\\n#C56887\\n#FD3D68\\n#ABF2FD\\n#835FAC\\n#FF16B1\\n#325371\\n#CA16CA\\n#D26322\\n#AFCFFE\\n#91A1FA\\nfloralwhite\n",
    "\"\"\".split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "d8970fc7-8ade-402d-9566-9e16c5593f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette36 = [\"#FFFFFF\", \n",
    "             \"#E4E1E3\", \n",
    "             \"#F6222E\", \n",
    "             \"#FE00FA\", \n",
    "             \"#16FF32\", \n",
    "             \"#3283FE\", \n",
    "             \"#FEAF16\", \n",
    "             \"#B00068\", \n",
    "             \"#1CFFCE\",\n",
    "             \"#90AD1C\", \n",
    "             \"#2ED9FF\", \n",
    "             \"#DEA0FD\", \n",
    "             \"#AA0DFE\", \n",
    "             \"#F8A19F\", \n",
    "             \"#325A9B\", \n",
    "             \"#C4451C\", \n",
    "             \"#1C8356\", \n",
    "             \"#85660D\",\n",
    "             \"#B10DA1\", \n",
    "             \"#FBE426\", \n",
    "             \"#1CBE4F\", \n",
    "             \"#FA0087\", \n",
    "             \"#FC1CBF\", \n",
    "             \"#F7E1A0\", \n",
    "             \"#C075A6\", \n",
    "             \"#782AB6\", \n",
    "             \"#AAF400\",\n",
    "             \"#BDCDFF\", \n",
    "             \"#822E1C\", \n",
    "             \"#B5EFB5\", \n",
    "             \"#7ED7D1\", \n",
    "             \"#1C7F93\", \n",
    "             \"#D85FF7\", \n",
    "             \"#683B79\", \n",
    "             \"#66B0FF\", \n",
    "             \"#3B00FB\",\n",
    "             \"magenta\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6239a9e1-9b79-4369-8931-8b33f3d5b027",
   "metadata": {},
   "source": [
    "change index 5 to #778899. Change index 9 to #2F4F4F. Add #FF7F50. Change index 9 to #FFBCD9. Change index 14 to #DEA5A4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6c36e65a-4442-489e-a5ef-de35adb81ae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "palette27 = [\"#FFFFFF\", \n",
    "             \"#AA0DFE\", \n",
    "             \"#3283FE\", \n",
    "             \"#85660D\", \n",
    "             \"#782AB6\", \n",
    "             \"#778899\", \n",
    "             \"#1C8356\", \n",
    "             \"#16FF32\", \n",
    "             \"#F7E1A0\", \n",
    "#              \"#2F4F4F\",\n",
    "             \"#FFBCD9\", \n",
    "             \"#C4451C\", \n",
    "             \"#DEA0FD\", \n",
    "             \"#FE00FA\", \n",
    "#              \"#325A9B\", \n",
    "             \"#FEAF16\", \n",
    "             \"#DEA5A4\", \n",
    "             \"#90AD1C\", \n",
    "             \"#F6222E\",\n",
    "             \"#1CFFCE\", \n",
    "             \"#2ED9FF\", \n",
    "             \"#B10DA1\", \n",
    "#              \"#C075A6\", \n",
    "#              \"#FC1CBF\", \n",
    "#              \"#B00068\", \n",
    "             \"#FBE426\", \n",
    "             \"#FA0087\",\n",
    "             \"#FF7F50\"\n",
    "            ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00491da4-7bee-4501-a31c-a48783f1e4ee",
   "metadata": {},
   "source": [
    "## Get the expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "b5a1bd1e-a09e-4bd2-9f4e-659799f7e99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# full_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/allgood_filt_agg_tidy_2021aligned_qc_rma_expression_full.csv')\n",
    "# full_filtered_df = full_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# full_filtered_norm_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_full_norm_filt_agg_tidy_2021aligned_qc_rma_expression.csv')\n",
    "# full_filtered_norm_df = full_filtered_norm_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# veg_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_filt_agg_tidy_2021aligned_qc_rma_expression_veg.csv')\n",
    "# veg_filtered_df = veg_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# veg_filtered_norm_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_veg_norm_filt_agg_tidy_2021aligned_qc_rma_expression.csv')\n",
    "# veg_filtered_norm_df = veg_filtered_norm_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# sex_filtered_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_filt_agg_tidy_2021aligned_qc_rma_expression_sex.csv')\n",
    "# sex_filtered_df = sex_filtered_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})\n",
    "\n",
    "# sex_filtered_norm_df = pd.read_csv('../microarray_probe_alignment_and_filtering/greedy_sex_norm_filt_agg_tidy_2021aligned_qc_rma_expression.csv')\n",
    "# sex_filtered_norm_df = sex_filtered_norm_df.rename(columns={'Unnamed: 0': 'TTHERM_ID'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "8a5f902a-b493-42f2-b166-441dd9bf3e5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20326"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(full_filtered_df['TTHERM_ID'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "082eccb7-7c4b-4a6b-aab8-61b3d5a01c68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20326"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(full_filtered_df['TTHERM_ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "c0e0ac3b-882d-4fa3-8995-6b146c5d7073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TTHERM_ID',\n",
       " 'Ll_GSM283687',\n",
       " 'Ll_GSM284355',\n",
       " 'Ll_GSM284362',\n",
       " 'Lm_GSM283690',\n",
       " 'Lm_GSM284357',\n",
       " 'Lm_GSM284363',\n",
       " 'Lh_GSM283691',\n",
       " 'Lh_GSM284360',\n",
       " 'Lh_GSM284364',\n",
       " 'S0_GSM285363',\n",
       " 'S0_GSM285554',\n",
       " 'S0_GSM285561',\n",
       " 'S3_GSM285542',\n",
       " 'S3_GSM285555',\n",
       " 'S3_GSM285562',\n",
       " 'S6_GSM285543',\n",
       " 'S6_GSM285556',\n",
       " 'S6_GSM285563',\n",
       " 'S9_GSM285544',\n",
       " 'S9_GSM285564',\n",
       " 'S15_GSM285559',\n",
       " 'S15_GSM285566',\n",
       " 'S24_GSM285547',\n",
       " 'S24_GSM285560']"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[c for c in full_filtered_df.columns if c[0] != 'C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "aee9c6d0-031e-4907-9cda-c9376e8e9b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_filtered_norm_df = normalize_expression_per_gene(full_filtered_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "ce08f21f-a261-45c8-be31-7976bb8106cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TTHERM_ID', 'Ll_GSM283687', 'Ll_GSM284355', 'Ll_GSM284362',\n",
       "       'Lm_GSM283690', 'Lm_GSM284357', 'Lm_GSM284363', 'Lh_GSM283691',\n",
       "       'Lh_GSM284360', 'Lh_GSM284364', 'S0_GSM285363', 'S0_GSM285554',\n",
       "       'S0_GSM285561', 'S3_GSM285542', 'S3_GSM285555', 'S3_GSM285562',\n",
       "       'S6_GSM285543', 'S6_GSM285556', 'S6_GSM285563', 'S9_GSM285544',\n",
       "       'S9_GSM285564', 'S15_GSM285559', 'S15_GSM285566', 'S24_GSM285547',\n",
       "       'S24_GSM285560', 'C0_GSM285570', 'C0_GSM656230', 'C4_GSM285574',\n",
       "       'C4_GSM285588', 'C4_GSM656234', 'C6_GSM285575', 'C6_GSM656232',\n",
       "       'C8_GSM285576', 'C8_GSM285590', 'C10_GSM285578', 'C10_GSM285591',\n",
       "       'C12_GSM285579', 'C12_GSM285592', 'C12_GSM656237', 'C14_GSM285580',\n",
       "       'C14_GSM285593', 'C14_GSM656238', 'C16_GSM285582', 'C16_GSM285595',\n",
       "       'C16_GSM656239', 'C18_GSM285583', 'C18_GSM285596', 'C18_GSM656240'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_norm_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "8d8d6dcc-7b5e-483c-9372-4c6096ac55f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll_GSM283687</th>\n",
       "      <th>Ll_GSM284355</th>\n",
       "      <th>Ll_GSM284362</th>\n",
       "      <th>Lm_GSM283690</th>\n",
       "      <th>Lm_GSM284357</th>\n",
       "      <th>Lm_GSM284363</th>\n",
       "      <th>Lh_GSM283691</th>\n",
       "      <th>Lh_GSM284360</th>\n",
       "      <th>Lh_GSM284364</th>\n",
       "      <th>...</th>\n",
       "      <th>C12_GSM656237</th>\n",
       "      <th>C14_GSM285580</th>\n",
       "      <th>C14_GSM285593</th>\n",
       "      <th>C14_GSM656238</th>\n",
       "      <th>C16_GSM285582</th>\n",
       "      <th>C16_GSM285595</th>\n",
       "      <th>C16_GSM656239</th>\n",
       "      <th>C18_GSM285583</th>\n",
       "      <th>C18_GSM285596</th>\n",
       "      <th>C18_GSM656240</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>6.928782</td>\n",
       "      <td>7.264201</td>\n",
       "      <td>6.934214</td>\n",
       "      <td>6.732989</td>\n",
       "      <td>6.970612</td>\n",
       "      <td>7.150978</td>\n",
       "      <td>6.126826</td>\n",
       "      <td>6.868968</td>\n",
       "      <td>6.641119</td>\n",
       "      <td>...</td>\n",
       "      <td>6.450318</td>\n",
       "      <td>8.049750</td>\n",
       "      <td>7.788162</td>\n",
       "      <td>7.052154</td>\n",
       "      <td>6.517742</td>\n",
       "      <td>6.918501</td>\n",
       "      <td>6.048861</td>\n",
       "      <td>7.041619</td>\n",
       "      <td>6.757932</td>\n",
       "      <td>5.817246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>9.633489</td>\n",
       "      <td>9.977124</td>\n",
       "      <td>10.027529</td>\n",
       "      <td>9.720665</td>\n",
       "      <td>9.605762</td>\n",
       "      <td>10.225542</td>\n",
       "      <td>10.279608</td>\n",
       "      <td>10.459966</td>\n",
       "      <td>10.693337</td>\n",
       "      <td>...</td>\n",
       "      <td>11.130466</td>\n",
       "      <td>11.207738</td>\n",
       "      <td>11.009172</td>\n",
       "      <td>10.615417</td>\n",
       "      <td>11.038938</td>\n",
       "      <td>11.009222</td>\n",
       "      <td>10.216348</td>\n",
       "      <td>11.099187</td>\n",
       "      <td>11.172276</td>\n",
       "      <td>10.561021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>5.066343</td>\n",
       "      <td>4.767264</td>\n",
       "      <td>5.010981</td>\n",
       "      <td>6.139047</td>\n",
       "      <td>4.619361</td>\n",
       "      <td>4.751761</td>\n",
       "      <td>5.818550</td>\n",
       "      <td>5.342529</td>\n",
       "      <td>5.483750</td>\n",
       "      <td>...</td>\n",
       "      <td>6.314438</td>\n",
       "      <td>7.423571</td>\n",
       "      <td>7.507645</td>\n",
       "      <td>7.417087</td>\n",
       "      <td>7.147801</td>\n",
       "      <td>7.747930</td>\n",
       "      <td>7.093641</td>\n",
       "      <td>7.672685</td>\n",
       "      <td>7.511290</td>\n",
       "      <td>6.890117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>4.696881</td>\n",
       "      <td>4.638401</td>\n",
       "      <td>4.956299</td>\n",
       "      <td>6.942556</td>\n",
       "      <td>5.101252</td>\n",
       "      <td>4.730307</td>\n",
       "      <td>8.457690</td>\n",
       "      <td>4.526411</td>\n",
       "      <td>4.908300</td>\n",
       "      <td>...</td>\n",
       "      <td>5.250233</td>\n",
       "      <td>4.974993</td>\n",
       "      <td>5.747498</td>\n",
       "      <td>5.252167</td>\n",
       "      <td>5.210531</td>\n",
       "      <td>7.083187</td>\n",
       "      <td>5.252222</td>\n",
       "      <td>5.037613</td>\n",
       "      <td>5.495281</td>\n",
       "      <td>5.013987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>4.654278</td>\n",
       "      <td>4.537105</td>\n",
       "      <td>4.928739</td>\n",
       "      <td>5.063991</td>\n",
       "      <td>4.584168</td>\n",
       "      <td>4.911880</td>\n",
       "      <td>5.935311</td>\n",
       "      <td>4.519470</td>\n",
       "      <td>4.757861</td>\n",
       "      <td>...</td>\n",
       "      <td>4.651688</td>\n",
       "      <td>4.920573</td>\n",
       "      <td>4.636333</td>\n",
       "      <td>4.883712</td>\n",
       "      <td>4.779395</td>\n",
       "      <td>4.744335</td>\n",
       "      <td>4.513140</td>\n",
       "      <td>4.838428</td>\n",
       "      <td>4.961475</td>\n",
       "      <td>4.653340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TTHERM_ID  Ll_GSM283687  Ll_GSM284355  Ll_GSM284362  Lm_GSM283690  \\\n",
       "0  TTHERM_000000042      6.928782      7.264201      6.934214      6.732989   \n",
       "1  TTHERM_000000045      9.633489      9.977124     10.027529      9.720665   \n",
       "2   TTHERM_00000010      5.066343      4.767264      5.010981      6.139047   \n",
       "3   TTHERM_00000020      4.696881      4.638401      4.956299      6.942556   \n",
       "4   TTHERM_00000030      4.654278      4.537105      4.928739      5.063991   \n",
       "\n",
       "   Lm_GSM284357  Lm_GSM284363  Lh_GSM283691  Lh_GSM284360  Lh_GSM284364  ...  \\\n",
       "0      6.970612      7.150978      6.126826      6.868968      6.641119  ...   \n",
       "1      9.605762     10.225542     10.279608     10.459966     10.693337  ...   \n",
       "2      4.619361      4.751761      5.818550      5.342529      5.483750  ...   \n",
       "3      5.101252      4.730307      8.457690      4.526411      4.908300  ...   \n",
       "4      4.584168      4.911880      5.935311      4.519470      4.757861  ...   \n",
       "\n",
       "   C12_GSM656237  C14_GSM285580  C14_GSM285593  C14_GSM656238  C16_GSM285582  \\\n",
       "0       6.450318       8.049750       7.788162       7.052154       6.517742   \n",
       "1      11.130466      11.207738      11.009172      10.615417      11.038938   \n",
       "2       6.314438       7.423571       7.507645       7.417087       7.147801   \n",
       "3       5.250233       4.974993       5.747498       5.252167       5.210531   \n",
       "4       4.651688       4.920573       4.636333       4.883712       4.779395   \n",
       "\n",
       "   C16_GSM285595  C16_GSM656239  C18_GSM285583  C18_GSM285596  C18_GSM656240  \n",
       "0       6.918501       6.048861       7.041619       6.757932       5.817246  \n",
       "1      11.009222      10.216348      11.099187      11.172276      10.561021  \n",
       "2       7.747930       7.093641       7.672685       7.511290       6.890117  \n",
       "3       7.083187       5.252222       5.037613       5.495281       5.013987  \n",
       "4       4.744335       4.513140       4.838428       4.961475       4.653340  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "67dc9e8a-2fa5-4b76-a944-ad21323254e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll_GSM283687</th>\n",
       "      <th>Ll_GSM284355</th>\n",
       "      <th>Ll_GSM284362</th>\n",
       "      <th>Lm_GSM283690</th>\n",
       "      <th>Lm_GSM284357</th>\n",
       "      <th>Lm_GSM284363</th>\n",
       "      <th>Lh_GSM283691</th>\n",
       "      <th>Lh_GSM284360</th>\n",
       "      <th>Lh_GSM284364</th>\n",
       "      <th>...</th>\n",
       "      <th>C12_GSM656237</th>\n",
       "      <th>C14_GSM285580</th>\n",
       "      <th>C14_GSM285593</th>\n",
       "      <th>C14_GSM656238</th>\n",
       "      <th>C16_GSM285582</th>\n",
       "      <th>C16_GSM285595</th>\n",
       "      <th>C16_GSM656239</th>\n",
       "      <th>C18_GSM285583</th>\n",
       "      <th>C18_GSM285596</th>\n",
       "      <th>C18_GSM656240</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>0.604560</td>\n",
       "      <td>0.712935</td>\n",
       "      <td>0.606315</td>\n",
       "      <td>0.541298</td>\n",
       "      <td>0.618075</td>\n",
       "      <td>0.676352</td>\n",
       "      <td>0.345444</td>\n",
       "      <td>0.585234</td>\n",
       "      <td>0.511614</td>\n",
       "      <td>...</td>\n",
       "      <td>0.449966</td>\n",
       "      <td>0.966750</td>\n",
       "      <td>0.882229</td>\n",
       "      <td>0.644422</td>\n",
       "      <td>0.471751</td>\n",
       "      <td>0.601238</td>\n",
       "      <td>0.320253</td>\n",
       "      <td>0.641018</td>\n",
       "      <td>0.549357</td>\n",
       "      <td>0.245417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>0.015138</td>\n",
       "      <td>0.202748</td>\n",
       "      <td>0.230267</td>\n",
       "      <td>0.062732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.338374</td>\n",
       "      <td>0.367892</td>\n",
       "      <td>0.466359</td>\n",
       "      <td>0.593770</td>\n",
       "      <td>...</td>\n",
       "      <td>0.832424</td>\n",
       "      <td>0.874612</td>\n",
       "      <td>0.766203</td>\n",
       "      <td>0.551230</td>\n",
       "      <td>0.782454</td>\n",
       "      <td>0.766230</td>\n",
       "      <td>0.333354</td>\n",
       "      <td>0.815348</td>\n",
       "      <td>0.855251</td>\n",
       "      <td>0.521531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>0.071967</td>\n",
       "      <td>0.023813</td>\n",
       "      <td>0.063054</td>\n",
       "      <td>0.244680</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021317</td>\n",
       "      <td>0.193078</td>\n",
       "      <td>0.116435</td>\n",
       "      <td>0.139173</td>\n",
       "      <td>...</td>\n",
       "      <td>0.272919</td>\n",
       "      <td>0.451498</td>\n",
       "      <td>0.465034</td>\n",
       "      <td>0.450454</td>\n",
       "      <td>0.407097</td>\n",
       "      <td>0.503722</td>\n",
       "      <td>0.398377</td>\n",
       "      <td>0.491607</td>\n",
       "      <td>0.465621</td>\n",
       "      <td>0.365608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>0.049259</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.105976</td>\n",
       "      <td>0.540233</td>\n",
       "      <td>0.137667</td>\n",
       "      <td>0.056567</td>\n",
       "      <td>0.871488</td>\n",
       "      <td>0.011989</td>\n",
       "      <td>0.095481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.170239</td>\n",
       "      <td>0.110063</td>\n",
       "      <td>0.278956</td>\n",
       "      <td>0.170661</td>\n",
       "      <td>0.161559</td>\n",
       "      <td>0.570979</td>\n",
       "      <td>0.170673</td>\n",
       "      <td>0.123753</td>\n",
       "      <td>0.223814</td>\n",
       "      <td>0.118588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>0.027748</td>\n",
       "      <td>0.004712</td>\n",
       "      <td>0.081708</td>\n",
       "      <td>0.108299</td>\n",
       "      <td>0.013964</td>\n",
       "      <td>0.078393</td>\n",
       "      <td>0.279603</td>\n",
       "      <td>0.001245</td>\n",
       "      <td>0.048113</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027239</td>\n",
       "      <td>0.080103</td>\n",
       "      <td>0.024220</td>\n",
       "      <td>0.072855</td>\n",
       "      <td>0.052347</td>\n",
       "      <td>0.045454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.063952</td>\n",
       "      <td>0.088144</td>\n",
       "      <td>0.027564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TTHERM_ID  Ll_GSM283687  Ll_GSM284355  Ll_GSM284362  Lm_GSM283690  \\\n",
       "0  TTHERM_000000042      0.604560      0.712935      0.606315      0.541298   \n",
       "1  TTHERM_000000045      0.015138      0.202748      0.230267      0.062732   \n",
       "2   TTHERM_00000010      0.071967      0.023813      0.063054      0.244680   \n",
       "3   TTHERM_00000020      0.049259      0.036473      0.105976      0.540233   \n",
       "4   TTHERM_00000030      0.027748      0.004712      0.081708      0.108299   \n",
       "\n",
       "   Lm_GSM284357  Lm_GSM284363  Lh_GSM283691  Lh_GSM284360  Lh_GSM284364  ...  \\\n",
       "0      0.618075      0.676352      0.345444      0.585234      0.511614  ...   \n",
       "1      0.000000      0.338374      0.367892      0.466359      0.593770  ...   \n",
       "2      0.000000      0.021317      0.193078      0.116435      0.139173  ...   \n",
       "3      0.137667      0.056567      0.871488      0.011989      0.095481  ...   \n",
       "4      0.013964      0.078393      0.279603      0.001245      0.048113  ...   \n",
       "\n",
       "   C12_GSM656237  C14_GSM285580  C14_GSM285593  C14_GSM656238  C16_GSM285582  \\\n",
       "0       0.449966       0.966750       0.882229       0.644422       0.471751   \n",
       "1       0.832424       0.874612       0.766203       0.551230       0.782454   \n",
       "2       0.272919       0.451498       0.465034       0.450454       0.407097   \n",
       "3       0.170239       0.110063       0.278956       0.170661       0.161559   \n",
       "4       0.027239       0.080103       0.024220       0.072855       0.052347   \n",
       "\n",
       "   C16_GSM285595  C16_GSM656239  C18_GSM285583  C18_GSM285596  C18_GSM656240  \n",
       "0       0.601238       0.320253       0.641018       0.549357       0.245417  \n",
       "1       0.766230       0.333354       0.815348       0.855251       0.521531  \n",
       "2       0.503722       0.398377       0.491607       0.465621       0.365608  \n",
       "3       0.570979       0.170673       0.123753       0.223814       0.118588  \n",
       "4       0.045454       0.000000       0.063952       0.088144       0.027564  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_filtered_norm_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "8ac61c80-4683-4d36-a891-20d3d8fa506f",
   "metadata": {},
   "outputs": [],
   "source": [
    "veg_filtered_norm_df = full_filtered_norm_df[[c for c in full_filtered_norm_df.columns if c[0] != 'C']]\n",
    "sex_filtered_norm_df = full_filtered_norm_df[[c for c in full_filtered_norm_df.columns if not c[0] in ['L', 'S']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "57bf2513-9422-4564-ab74-e8da57af1287",
   "metadata": {},
   "outputs": [],
   "source": [
    "# full_mean_norm = get_geom_mean_expression(full_filtered_norm_df)\n",
    "# full_mean = get_geom_mean_expression(full_filtered_df)\n",
    "# full_mean_norm.to_csv('./full_mean_norm_df.csv', index=False)\n",
    "\n",
    "# patrick_genes = \"\"\"\n",
    "# TTHERM_00658810, TTHERM_000193469, TTHERM_00047330, TTHERM_000486279, TTHERM_00141040, TTHERM_00227750, TTHERM_00317390, TTHERM_00670750, TTHERM_01122800, TTHERM_01213910, TTHERM_00527180, TTHERM_00473020, TTHERM_00624730, TTHERM_00624720, TTHERM_00378890, TTHERM_00321730, TTHERM_00522600, TTHERM_01055600, TTHERM_01018540, TTHERM_00221120, TTHERM_00221130\n",
    "# \"\"\".split(', ')\n",
    "\n",
    "# patrick_genes = [p.strip() for p in patrick_genes]\n",
    "\n",
    "# patrick_df_norm = full_mean_norm.loc[full_mean_norm['TTHERM_ID'].isin(patrick_genes)]\n",
    "# patrick_df = full_mean.loc[full_mean['TTHERM_ID'].isin(patrick_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "af3a4878-5584-47bc-9c34-2eca64f7b234",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(patrick_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "5c728107-b6c0-4d7e-9974-dc3a5e12d049",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(patrick_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "450d2acd-06f5-40b9-b263-c2cf83da023b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# patrick_df.to_csv('/Users/eukarya/Downloads/patrick_genes.csv', index=False)\n",
    "# patrick_df_norm.to_csv('/Users/eukarya/Downloads/patrick_genes_normalized', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "527cab73-a0d7-4ead-ba51-0808cb1fe003",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Ll</th>\n",
       "      <th>Lm</th>\n",
       "      <th>Lh</th>\n",
       "      <th>S0</th>\n",
       "      <th>S3</th>\n",
       "      <th>S6</th>\n",
       "      <th>S9</th>\n",
       "      <th>S15</th>\n",
       "      <th>S24</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>0.640498</td>\n",
       "      <td>0.610955</td>\n",
       "      <td>0.477293</td>\n",
       "      <td>0.351885</td>\n",
       "      <td>0.451853</td>\n",
       "      <td>0.371189</td>\n",
       "      <td>0.545109</td>\n",
       "      <td>0.538667</td>\n",
       "      <td>0.605428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>0.145249</td>\n",
       "      <td>0.124606</td>\n",
       "      <td>0.473124</td>\n",
       "      <td>0.583956</td>\n",
       "      <td>0.719232</td>\n",
       "      <td>0.661353</td>\n",
       "      <td>0.850265</td>\n",
       "      <td>0.781021</td>\n",
       "      <td>0.991903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>0.052735</td>\n",
       "      <td>0.083277</td>\n",
       "      <td>0.149116</td>\n",
       "      <td>0.632689</td>\n",
       "      <td>0.644945</td>\n",
       "      <td>0.756778</td>\n",
       "      <td>0.719076</td>\n",
       "      <td>0.331694</td>\n",
       "      <td>0.460232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>0.063478</td>\n",
       "      <td>0.227909</td>\n",
       "      <td>0.275428</td>\n",
       "      <td>0.955571</td>\n",
       "      <td>0.537915</td>\n",
       "      <td>0.527843</td>\n",
       "      <td>0.349297</td>\n",
       "      <td>0.151907</td>\n",
       "      <td>0.209777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>0.037559</td>\n",
       "      <td>0.066152</td>\n",
       "      <td>0.103251</td>\n",
       "      <td>0.829570</td>\n",
       "      <td>0.406981</td>\n",
       "      <td>0.611470</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>0.109470</td>\n",
       "      <td>0.072589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>0.111025</td>\n",
       "      <td>0.140537</td>\n",
       "      <td>0.375651</td>\n",
       "      <td>0.337526</td>\n",
       "      <td>0.561633</td>\n",
       "      <td>0.583339</td>\n",
       "      <td>0.550147</td>\n",
       "      <td>0.823484</td>\n",
       "      <td>0.689789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>0.640733</td>\n",
       "      <td>0.640657</td>\n",
       "      <td>0.642801</td>\n",
       "      <td>0.734546</td>\n",
       "      <td>0.741562</td>\n",
       "      <td>0.804534</td>\n",
       "      <td>0.728016</td>\n",
       "      <td>0.843385</td>\n",
       "      <td>0.893712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>0.304821</td>\n",
       "      <td>0.276959</td>\n",
       "      <td>0.519766</td>\n",
       "      <td>0.316289</td>\n",
       "      <td>0.478726</td>\n",
       "      <td>0.460461</td>\n",
       "      <td>0.462791</td>\n",
       "      <td>0.610656</td>\n",
       "      <td>0.587612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>0.375721</td>\n",
       "      <td>0.332330</td>\n",
       "      <td>0.334817</td>\n",
       "      <td>0.318067</td>\n",
       "      <td>0.411782</td>\n",
       "      <td>0.443272</td>\n",
       "      <td>0.427828</td>\n",
       "      <td>0.656553</td>\n",
       "      <td>0.621862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>0.372264</td>\n",
       "      <td>0.461628</td>\n",
       "      <td>0.831169</td>\n",
       "      <td>0.446549</td>\n",
       "      <td>0.596457</td>\n",
       "      <td>0.702927</td>\n",
       "      <td>0.761623</td>\n",
       "      <td>0.652697</td>\n",
       "      <td>0.531782</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID        Ll        Lm        Lh        S0        S3  \\\n",
       "0      TTHERM_000000042  0.640498  0.610955  0.477293  0.351885  0.451853   \n",
       "1      TTHERM_000000045  0.145249  0.124606  0.473124  0.583956  0.719232   \n",
       "2       TTHERM_00000010  0.052735  0.083277  0.149116  0.632689  0.644945   \n",
       "3       TTHERM_00000020  0.063478  0.227909  0.275428  0.955571  0.537915   \n",
       "4       TTHERM_00000030  0.037559  0.066152  0.103251  0.829570  0.406981   \n",
       "...                 ...       ...       ...       ...       ...       ...   \n",
       "20321   TTHERM_02293890  0.111025  0.140537  0.375651  0.337526  0.561633   \n",
       "20322   TTHERM_02385080  0.640733  0.640657  0.642801  0.734546  0.741562   \n",
       "20323   TTHERM_02555200  0.304821  0.276959  0.519766  0.316289  0.478726   \n",
       "20324   TTHERM_02607240  0.375721  0.332330  0.334817  0.318067  0.411782   \n",
       "20325   TTHERM_02653470  0.372264  0.461628  0.831169  0.446549  0.596457   \n",
       "\n",
       "             S6        S9       S15       S24  \n",
       "0      0.371189  0.545109  0.538667  0.605428  \n",
       "1      0.661353  0.850265  0.781021  0.991903  \n",
       "2      0.756778  0.719076  0.331694  0.460232  \n",
       "3      0.527843  0.349297  0.151907  0.209777  \n",
       "4      0.611470  0.461538  0.109470  0.072589  \n",
       "...         ...       ...       ...       ...  \n",
       "20321  0.583339  0.550147  0.823484  0.689789  \n",
       "20322  0.804534  0.728016  0.843385  0.893712  \n",
       "20323  0.460461  0.462791  0.610656  0.587612  \n",
       "20324  0.443272  0.427828  0.656553  0.621862  \n",
       "20325  0.702927  0.761623  0.652697  0.531782  \n",
       "\n",
       "[20326 rows x 10 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "veg_mean = get_geom_mean_expression(veg_filtered_norm_df)\n",
    "veg_mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "1508416d-d531-4f87-85cf-0db9fd0a0916",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>C0</th>\n",
       "      <th>C4</th>\n",
       "      <th>C6</th>\n",
       "      <th>C8</th>\n",
       "      <th>C10</th>\n",
       "      <th>C12</th>\n",
       "      <th>C14</th>\n",
       "      <th>C16</th>\n",
       "      <th>C18</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>0.164791</td>\n",
       "      <td>0.723792</td>\n",
       "      <td>0.532334</td>\n",
       "      <td>0.961895</td>\n",
       "      <td>0.510857</td>\n",
       "      <td>0.700169</td>\n",
       "      <td>0.825905</td>\n",
       "      <td>0.459875</td>\n",
       "      <td>0.468452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>0.919002</td>\n",
       "      <td>0.703379</td>\n",
       "      <td>0.517879</td>\n",
       "      <td>0.499102</td>\n",
       "      <td>0.786583</td>\n",
       "      <td>0.716830</td>\n",
       "      <td>0.725345</td>\n",
       "      <td>0.613134</td>\n",
       "      <td>0.724042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>0.918367</td>\n",
       "      <td>0.263778</td>\n",
       "      <td>0.169134</td>\n",
       "      <td>0.238973</td>\n",
       "      <td>0.206974</td>\n",
       "      <td>0.262770</td>\n",
       "      <td>0.455647</td>\n",
       "      <td>0.435617</td>\n",
       "      <td>0.439906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>0.538202</td>\n",
       "      <td>0.302916</td>\n",
       "      <td>0.196585</td>\n",
       "      <td>0.171958</td>\n",
       "      <td>0.159344</td>\n",
       "      <td>0.178610</td>\n",
       "      <td>0.184527</td>\n",
       "      <td>0.287901</td>\n",
       "      <td>0.154389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>0.166954</td>\n",
       "      <td>0.214577</td>\n",
       "      <td>0.088220</td>\n",
       "      <td>0.037427</td>\n",
       "      <td>0.062772</td>\n",
       "      <td>0.075196</td>\n",
       "      <td>0.058766</td>\n",
       "      <td>0.032336</td>\n",
       "      <td>0.059593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>0.654662</td>\n",
       "      <td>0.199828</td>\n",
       "      <td>0.093582</td>\n",
       "      <td>0.248135</td>\n",
       "      <td>0.754035</td>\n",
       "      <td>0.775166</td>\n",
       "      <td>0.605974</td>\n",
       "      <td>0.582757</td>\n",
       "      <td>0.683368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>0.469278</td>\n",
       "      <td>0.474059</td>\n",
       "      <td>0.394062</td>\n",
       "      <td>0.592505</td>\n",
       "      <td>0.598329</td>\n",
       "      <td>0.244594</td>\n",
       "      <td>0.627762</td>\n",
       "      <td>0.733024</td>\n",
       "      <td>0.814996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>0.480840</td>\n",
       "      <td>0.591126</td>\n",
       "      <td>0.617040</td>\n",
       "      <td>0.622603</td>\n",
       "      <td>0.319449</td>\n",
       "      <td>0.351828</td>\n",
       "      <td>0.523217</td>\n",
       "      <td>0.796344</td>\n",
       "      <td>0.848257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>0.929297</td>\n",
       "      <td>0.739198</td>\n",
       "      <td>0.619243</td>\n",
       "      <td>0.588040</td>\n",
       "      <td>0.160268</td>\n",
       "      <td>0.612148</td>\n",
       "      <td>0.804887</td>\n",
       "      <td>0.799902</td>\n",
       "      <td>0.602232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>0.599931</td>\n",
       "      <td>0.291566</td>\n",
       "      <td>0.236501</td>\n",
       "      <td>0.318720</td>\n",
       "      <td>0.290800</td>\n",
       "      <td>0.524813</td>\n",
       "      <td>0.788421</td>\n",
       "      <td>0.850434</td>\n",
       "      <td>0.789467</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID        C0        C4        C6        C8       C10  \\\n",
       "0      TTHERM_000000042  0.164791  0.723792  0.532334  0.961895  0.510857   \n",
       "1      TTHERM_000000045  0.919002  0.703379  0.517879  0.499102  0.786583   \n",
       "2       TTHERM_00000010  0.918367  0.263778  0.169134  0.238973  0.206974   \n",
       "3       TTHERM_00000020  0.538202  0.302916  0.196585  0.171958  0.159344   \n",
       "4       TTHERM_00000030  0.166954  0.214577  0.088220  0.037427  0.062772   \n",
       "...                 ...       ...       ...       ...       ...       ...   \n",
       "20321   TTHERM_02293890  0.654662  0.199828  0.093582  0.248135  0.754035   \n",
       "20322   TTHERM_02385080  0.469278  0.474059  0.394062  0.592505  0.598329   \n",
       "20323   TTHERM_02555200  0.480840  0.591126  0.617040  0.622603  0.319449   \n",
       "20324   TTHERM_02607240  0.929297  0.739198  0.619243  0.588040  0.160268   \n",
       "20325   TTHERM_02653470  0.599931  0.291566  0.236501  0.318720  0.290800   \n",
       "\n",
       "            C12       C14       C16       C18  \n",
       "0      0.700169  0.825905  0.459875  0.468452  \n",
       "1      0.716830  0.725345  0.613134  0.724042  \n",
       "2      0.262770  0.455647  0.435617  0.439906  \n",
       "3      0.178610  0.184527  0.287901  0.154389  \n",
       "4      0.075196  0.058766  0.032336  0.059593  \n",
       "...         ...       ...       ...       ...  \n",
       "20321  0.775166  0.605974  0.582757  0.683368  \n",
       "20322  0.244594  0.627762  0.733024  0.814996  \n",
       "20323  0.351828  0.523217  0.796344  0.848257  \n",
       "20324  0.612148  0.804887  0.799902  0.602232  \n",
       "20325  0.524813  0.788421  0.850434  0.789467  \n",
       "\n",
       "[20326 rows x 10 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sex_mean = get_geom_mean_expression(sex_filtered_norm_df)\n",
    "sex_mean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d243d8b-6248-4f01-a443-215c5a01412d",
   "metadata": {},
   "source": [
    "Get the annnotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "6dce5f9d-6cbb-4705-97be-33211caad9ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_annot = pd.read_csv('../eggnog/complete_eggnog_annotation.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc78ff5-bc76-4957-bf55-9021c18e88c7",
   "metadata": {},
   "source": [
    "Add the TGD annotations for all the genes to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "6175d66a-0d48-4b4e-a799-4018c4e00f13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_01528530</td>\n",
       "      <td>5911.EAS00195</td>\n",
       "      <td>2.2e-14</td>\n",
       "      <td>49.9</td>\n",
       "      <td>KOG0594@1|root,KOG0594@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>G</td>\n",
       "      <td>cyclin-dependent protein serine/threonine kina...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_01528510</td>\n",
       "      <td>5911.EAR81750</td>\n",
       "      <td>1.1e-194</td>\n",
       "      <td>633.4</td>\n",
       "      <td>2EI75@1|root,2SNPE@2759|Eukaryota,3ZBUW@5878|C...</td>\n",
       "      <td>5878|Ciliophora</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_01528500</td>\n",
       "      <td>5911.EAR81749</td>\n",
       "      <td>4e-141</td>\n",
       "      <td>462.0</td>\n",
       "      <td>COG1100@1|root,KOG0074@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>KLT</td>\n",
       "      <td>GTP binding</td>\n",
       "      <td>ARL13B</td>\n",
       "      <td>GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko04031</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ADP-ribosylation factor family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_0015284992</td>\n",
       "      <td>5911.EAR97791</td>\n",
       "      <td>8.4e-64</td>\n",
       "      <td>205.1</td>\n",
       "      <td>2ERUB@1|root,2SUIJ@2759|Eukaryota</td>\n",
       "      <td>5911.EAR97791|-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Leucine-rich repeat-containing protein 74A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00897120</td>\n",
       "      <td>5911.EAR86071</td>\n",
       "      <td>5.2e-77</td>\n",
       "      <td>248.8</td>\n",
       "      <td>2ERUB@1|root,2SUIJ@2759|Eukaryota</td>\n",
       "      <td>5911.EAR86071|-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>kinase domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26253</th>\n",
       "      <td>TTHERM_01076820</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26254</th>\n",
       "      <td>TTHERM_001076861</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26255</th>\n",
       "      <td>TTHERM_01076910</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26256</th>\n",
       "      <td>TTHERM_001076981</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26257</th>\n",
       "      <td>TTHERM_01077010</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Endoglucanase 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26258 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               TTHERM_ID  seed_ortholog    evalue  score  \\\n",
       "0        TTHERM_01528530  5911.EAS00195   2.2e-14   49.9   \n",
       "1        TTHERM_01528510  5911.EAR81750  1.1e-194  633.4   \n",
       "2        TTHERM_01528500  5911.EAR81749    4e-141  462.0   \n",
       "3      TTHERM_0015284992  5911.EAR97791   8.4e-64  205.1   \n",
       "4        TTHERM_00897120  5911.EAR86071   5.2e-77  248.8   \n",
       "...                  ...            ...       ...    ...   \n",
       "26253    TTHERM_01076820              -         -      -   \n",
       "26254   TTHERM_001076861              -         -      -   \n",
       "26255    TTHERM_01076910              -         -      -   \n",
       "26256   TTHERM_001076981              -         -      -   \n",
       "26257    TTHERM_01077010              -         -      -   \n",
       "\n",
       "                                              eggNOG_OGs    max_annot_lvl  \\\n",
       "0                  KOG0594@1|root,KOG0594@2759|Eukaryota   2759|Eukaryota   \n",
       "1      2EI75@1|root,2SNPE@2759|Eukaryota,3ZBUW@5878|C...  5878|Ciliophora   \n",
       "2                  COG1100@1|root,KOG0074@2759|Eukaryota   2759|Eukaryota   \n",
       "3                      2ERUB@1|root,2SUIJ@2759|Eukaryota  5911.EAR97791|-   \n",
       "4                      2ERUB@1|root,2SUIJ@2759|Eukaryota  5911.EAR86071|-   \n",
       "...                                                  ...              ...   \n",
       "26253                                                  -                -   \n",
       "26254                                                  -                -   \n",
       "26255                                                  -                -   \n",
       "26256                                                  -                -   \n",
       "26257                                                  -                -   \n",
       "\n",
       "      COG_category                                        Description  \\\n",
       "0                G  cyclin-dependent protein serine/threonine kina...   \n",
       "1                -                                                  -   \n",
       "2              KLT                                        GTP binding   \n",
       "3                -                                                  -   \n",
       "4                -                                                  -   \n",
       "...            ...                                                ...   \n",
       "26253            -                                                  -   \n",
       "26254            -                                                  -   \n",
       "26255            -                                                  -   \n",
       "26256            -                                                  -   \n",
       "26257            -                                                  -   \n",
       "\n",
       "      Preferred_name                                                GOs  ...  \\\n",
       "0                  -                                                  -  ...   \n",
       "1                  -                                                  -  ...   \n",
       "2             ARL13B  GO:0000902,GO:0000904,GO:0001947,GO:0002009,GO...  ...   \n",
       "3                  -                                                  -  ...   \n",
       "4                  -                                                  -  ...   \n",
       "...              ...                                                ...  ...   \n",
       "26253              -                                                  -  ...   \n",
       "26254              -                                                  -  ...   \n",
       "26255              -                                                  -  ...   \n",
       "26256              -                                                  -  ...   \n",
       "26257              -                                                  -  ...   \n",
       "\n",
       "      KEGG_Pathway KEGG_Module KEGG_Reaction KEGG_rclass            BRITE  \\\n",
       "0                -           -             -           -                -   \n",
       "1                -           -             -           -                -   \n",
       "2                -           -             -           -  ko00000,ko04031   \n",
       "3                -           -             -           -                -   \n",
       "4                -           -             -           -                -   \n",
       "...            ...         ...           ...         ...              ...   \n",
       "26253            -           -             -           -                -   \n",
       "26254            -           -             -           -                -   \n",
       "26255            -           -             -           -                -   \n",
       "26256            -           -             -           -                -   \n",
       "26257            -           -             -           -                -   \n",
       "\n",
       "      KEGG_TC CAZy BiGG_Reaction PFAMs  \\\n",
       "0           -    -             -     -   \n",
       "1           -    -             -     -   \n",
       "2           -    -             -     -   \n",
       "3           -    -             -     -   \n",
       "4           -    -             -     -   \n",
       "...       ...  ...           ...   ...   \n",
       "26253       -    -             -     -   \n",
       "26254       -    -             -     -   \n",
       "26255       -    -             -     -   \n",
       "26256       -    -             -     -   \n",
       "26257       -    -             -     -   \n",
       "\n",
       "                              TGD2021_description  \n",
       "0                                  protein kinase  \n",
       "1                            hypothetical protein  \n",
       "2          ADP-ribosylation factor family protein  \n",
       "3      Leucine-rich repeat-containing protein 74A  \n",
       "4                           kinase domain protein  \n",
       "...                                           ...  \n",
       "26253                             Endoglucanase 1  \n",
       "26254              transmembrane protein putative  \n",
       "26255                             Endoglucanase 1  \n",
       "26256                             Endoglucanase 1  \n",
       "26257                             Endoglucanase 1  \n",
       "\n",
       "[26258 rows x 22 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "6b6de3be-caa1-41d2-a058-efb28081050c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checkout_df = complete_annot.loc[complete_annot['TTHERM_ID'].isin(checkout)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "ec95a4d8-b618-499a-a3ed-066491940bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(checkout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "eafe60a7-e772-4724-8adf-188ad7ce59e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checkout_df.to_csv('../checkout_different_filtered_genes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "6b88aa5a-83d9-4712-b900-7e22e93c9836",
   "metadata": {},
   "outputs": [],
   "source": [
    "# complete_annot.to_csv('/Users/eukarya/Downloads/tgn_complete_annot.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29defd25-5786-4bca-a782-59b158b7d176",
   "metadata": {},
   "source": [
    "## Round one of plotting and consensus "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2525b47b-7edf-4ed3-b6e7-4739c399b426",
   "metadata": {},
   "source": [
    "First, cluster accoding to UMAP/Leiden in full-dimensional space, using Manhattan distance and setting the number of neighbors to 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d03955e5",
   "metadata": {},
   "source": [
    "### LEIDEN MODULARITY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "92c456b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# leiden_label_df_round_1, full_dists, sscore, modularity = build_leiden_label_df(full_filtered_norm_df, 'full', random_state=42, n_neighbors=3, leiden_type='modularity')\n",
    "leiden_label_df_round_1, full_dists, sscore, modularity =  build_leiden_label_df(full_filtered_norm_df, 'full', random_state=42, n_neighbors=6, leiden_type='cpm', la_res_param=0.030)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "6b8d5fd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20326\n"
     ]
    }
   ],
   "source": [
    "clust_alg = 'leiden'\n",
    "phases = 'full'\n",
    "\n",
    "leiden_label_df_round_1_arranged = arrange_modules(full_filtered_norm_df, leiden_label_df_round_1, clust_alg, phases) \n",
    "leiden_label_df_round_1_arranged_sorted = leiden_label_df_round_1_arranged.sort_values(by=[f'{clust_alg}_label_{phases}', 'TTHERM_ID'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "30ff03ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.021391190803871626, 0.500338049198614)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sscore, modularity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "5d980050",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19792</th>\n",
       "      <td>TTHERM_01284740</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19234</th>\n",
       "      <td>TTHERM_01126420</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18748</th>\n",
       "      <td>TTHERM_01075650</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18496</th>\n",
       "      <td>TTHERM_01035720</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17764</th>\n",
       "      <td>TTHERM_00947510</td>\n",
       "      <td>778</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full\n",
       "19792  TTHERM_01284740                778\n",
       "19234  TTHERM_01126420                778\n",
       "18748  TTHERM_01075650                778\n",
       "18496  TTHERM_01035720                778\n",
       "17764  TTHERM_00947510                778"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1_arranged_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "3b50c719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    525\n",
       "dtype: int64"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1_arranged_sorted['leiden_label_full'].mode()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f25e4781",
   "metadata": {},
   "source": [
    "### LEIDEN CPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# leiden_label_df_cpms = {}\n",
    "\n",
    "# la_res_params = [0.0001, 0.0002, 0.0003, 0.0004, 0.0005, 0.0006, 0.0007, 0.0008, 0.0009, 1]\n",
    "\n",
    "# print(la_res_params)\n",
    "\n",
    "# for curr_la_res_param in la_res_params:\n",
    "#     leiden_label_df_cpm, full_dists_cpm, cpm_sscore, cpm_modularity = build_leiden_label_df(full_filtered_norm_df, 'full', random_state=42, n_neighbors=5, leiden_type='cpm', la_res_param=curr_la_res_param)\n",
    "#     leiden_label_df_cpm_sorted = leiden_label_df_cpm.sort_values(by='leiden_label_full')\n",
    "#     print(curr_la_res_param, 'generated', max(list(leiden_label_df_cpm_sorted['leiden_label_full'].values))+1, 'clusters.')\n",
    "#     leiden_label_df_cpms[curr_la_res_param] = leiden_label_df_cpm_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "# leiden_label_df_cpms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "3e876499",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>372</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full\n",
       "0      TTHERM_000000042                228\n",
       "1      TTHERM_000000045                190\n",
       "2       TTHERM_00000010                475\n",
       "3       TTHERM_00000020                391\n",
       "4       TTHERM_00000030                  6\n",
       "...                 ...                ...\n",
       "20321   TTHERM_02293890                416\n",
       "20322   TTHERM_02385080                628\n",
       "20323   TTHERM_02555200                450\n",
       "20324   TTHERM_02607240                301\n",
       "20325   TTHERM_02653470                372\n",
       "\n",
       "[20326 rows x 2 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "4a52f1d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "778"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_full'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "d2970071-e2c8-49d9-8931-9ce6e4f8af44",
   "metadata": {},
   "outputs": [],
   "source": [
    "leiden_label_df_round_1, veg_dists, veg_sscore, veg_modularity = build_leiden_label_df(veg_filtered_norm_df, 'veg', lldf=leiden_label_df_round_1, random_state=42, n_neighbors=3)\n",
    "leiden_label_df_round_1, sex_dists, sex_sscore, sex_modularity = build_leiden_label_df(sex_filtered_norm_df, 'sex', lldf=leiden_label_df_round_1, random_state=42, n_neighbors=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "e13fb345-b665-46ff-adbb-98d76c952b8d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>228</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>190</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>475</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>391</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>416</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>628</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>450</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>301</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>372</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "0      TTHERM_000000042                228                 6                17\n",
       "1      TTHERM_000000045                190                 0                10\n",
       "2       TTHERM_00000010                475                19                 4\n",
       "3       TTHERM_00000020                391                19                 4\n",
       "4       TTHERM_00000030                  6                19                 0\n",
       "...                 ...                ...               ...               ...\n",
       "20321   TTHERM_02293890                416                12                 8\n",
       "20322   TTHERM_02385080                628                16                15\n",
       "20323   TTHERM_02555200                450                18                15\n",
       "20324   TTHERM_02607240                301                17                 3\n",
       "20325   TTHERM_02653470                372                10                11\n",
       "\n",
       "[20326 rows x 4 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "382f7ad6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_veg'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "0cffc2fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_sex'].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ffe51d0-614c-49f4-bf8c-41d48c1b75e6",
   "metadata": {},
   "source": [
    "Check out how many clusters we have"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9125a92-356f-49ad-88e1-c7da95ded20a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c1879f2a-3e20-44ff-8d11-72dee85b765f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# leiden_label_df_round_1.to_csv('../clustergrammer/leiden_label_df_round_1_nn2.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "db82b99c-9360-4a60-80bc-1b78e216b08e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "778"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_full'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "736a3bba-f713-427b-aecb-e469e84ec26c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_veg'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "94d2d630-4411-4777-9e38-cc7c77f90251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(leiden_label_df_round_1['leiden_label_sex'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "96125df3-98dc-4b99-bb70-86ddb1997737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26.09242618741977"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_size_list = []\n",
    "for l in leiden_label_df_round_1['leiden_label_full'].unique():\n",
    "    df = leiden_label_df_round_1.loc[leiden_label_df_round_1['leiden_label_full'] == l]\n",
    "    module_size_list.append(len(df))\n",
    "    \n",
    "np.mean(module_size_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "32eac343-9a2b-4a27-a9b2-785597ea65f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24.0"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(module_size_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "333048c1-ea2a-4d02-97d9-af0f0450118e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "74"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(module_size_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "8fd25633-9cfd-4c31-847a-c49fb989e4d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "      <th>seed_ortholog</th>\n",
       "      <th>evalue</th>\n",
       "      <th>score</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG_category</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>228</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>5911.EAR87740</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1776.8</td>\n",
       "      <td>2CVX5@1|root,2RT23@2759|Eukaryota,3ZDPV@5878|C...</td>\n",
       "      <td>5878|Ciliophora</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>B-box zinc finger protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>190</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>475</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "      <td>5911.EAR87737</td>\n",
       "      <td>1.7e-272</td>\n",
       "      <td>898.9</td>\n",
       "      <td>KOG0575@1|root,KOG0575@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ko04068,ko04110,ko04114,ko04136,ko04138,ko0413...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko01001,ko03029,ko0303...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>protein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>391</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "      <td>5911.EAR87739</td>\n",
       "      <td>1.6e-77</td>\n",
       "      <td>260.4</td>\n",
       "      <td>KOG0032@1|root,KOG0032@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ko04010,ko04114,ko04136,ko04138,ko04139,ko0414...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko01001,ko03019,ko0302...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Serine/Threonine kinase domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>5888.CAK56476</td>\n",
       "      <td>5.8999999999999995e-52</td>\n",
       "      <td>172.6</td>\n",
       "      <td>KOG0615@1|root,KOG0615@2759|Eukaryota</td>\n",
       "      <td>2759|Eukaryota</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>ko04140,ko04152,ko04211,ko04626,ko04920,ko0492...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko00001,ko01000,ko01001</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Serine/Threonine kinase domain protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex  \\\n",
       "0  TTHERM_000000042                228                 6                17   \n",
       "1  TTHERM_000000045                190                 0                10   \n",
       "2   TTHERM_00000010                475                19                 4   \n",
       "3   TTHERM_00000020                391                19                 4   \n",
       "4   TTHERM_00000030                  6                19                 0   \n",
       "\n",
       "   seed_ortholog                  evalue   score  \\\n",
       "0  5911.EAR87740                     0.0  1776.8   \n",
       "1              -                       -       -   \n",
       "2  5911.EAR87737                1.7e-272   898.9   \n",
       "3  5911.EAR87739                 1.6e-77   260.4   \n",
       "4  5888.CAK56476  5.8999999999999995e-52   172.6   \n",
       "\n",
       "                                          eggNOG_OGs    max_annot_lvl  \\\n",
       "0  2CVX5@1|root,2RT23@2759|Eukaryota,3ZDPV@5878|C...  5878|Ciliophora   \n",
       "1                                                  -                -   \n",
       "2              KOG0575@1|root,KOG0575@2759|Eukaryota   2759|Eukaryota   \n",
       "3              KOG0032@1|root,KOG0032@2759|Eukaryota   2759|Eukaryota   \n",
       "4              KOG0615@1|root,KOG0615@2759|Eukaryota   2759|Eukaryota   \n",
       "\n",
       "  COG_category  ...                                       KEGG_Pathway  \\\n",
       "0            -  ...                                                  -   \n",
       "1            -  ...                                                  -   \n",
       "2            T  ...  ko04068,ko04110,ko04114,ko04136,ko04138,ko0413...   \n",
       "3            T  ...  ko04010,ko04114,ko04136,ko04138,ko04139,ko0414...   \n",
       "4            T  ...  ko04140,ko04152,ko04211,ko04626,ko04920,ko0492...   \n",
       "\n",
       "  KEGG_Module KEGG_Reaction KEGG_rclass  \\\n",
       "0           -             -           -   \n",
       "1           -             -           -   \n",
       "2           -             -           -   \n",
       "3           -             -           -   \n",
       "4           -             -           -   \n",
       "\n",
       "                                               BRITE KEGG_TC CAZy  \\\n",
       "0                                                  -       -    -   \n",
       "1                                                  -       -    -   \n",
       "2  ko00000,ko00001,ko01000,ko01001,ko03029,ko0303...       -    -   \n",
       "3  ko00000,ko00001,ko01000,ko01001,ko03019,ko0302...       -    -   \n",
       "4                    ko00000,ko00001,ko01000,ko01001       -    -   \n",
       "\n",
       "  BiGG_Reaction PFAMs                     TGD2021_description  \n",
       "0             -     -               B-box zinc finger protein  \n",
       "1             -     -          transmembrane protein putative  \n",
       "2             -     -                          protein kinase  \n",
       "3             -     -  Serine/Threonine kinase domain protein  \n",
       "4             -     -  Serine/Threonine kinase domain protein  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labeled_annot = leiden_label_df_round_1.merge(complete_annot, on='TTHERM_ID')\n",
    "labeled_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "23d1e313-7738-4a3a-bf16-9b36210d0be1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>TTHERM_000011798</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>TTHERM_000158529</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>TTHERM_000188553</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892</th>\n",
       "      <td>TTHERM_000241859</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>TTHERM_000294809</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18541</th>\n",
       "      <td>TTHERM_01044390</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18816</th>\n",
       "      <td>TTHERM_01080530</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18871</th>\n",
       "      <td>TTHERM_01084310</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18979</th>\n",
       "      <td>TTHERM_01097910</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19041</th>\n",
       "      <td>TTHERM_01104840</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "36     TTHERM_000011798                  0                 6                 0\n",
       "628    TTHERM_000158529                  0                 2                 0\n",
       "723    TTHERM_000188553                  0                 3                 4\n",
       "892    TTHERM_000241859                  0                 2                 0\n",
       "1057   TTHERM_000294809                  0                 3                 0\n",
       "...                 ...                ...               ...               ...\n",
       "18541   TTHERM_01044390                  0                19                20\n",
       "18816   TTHERM_01080530                  0                19                 0\n",
       "18871   TTHERM_01084310                  0                 2                 0\n",
       "18979   TTHERM_01097910                  0                20                 0\n",
       "19041   TTHERM_01104840                  0                 3                 0\n",
       "\n",
       "[74 rows x 4 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['leiden_label_full']==0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e03aeddd-8c49-4136-9c15-6befa0d61418",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "28e85629-61e7-4d5d-b1ba-5da9a1832724",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12397</th>\n",
       "      <td>TTHERM_00532700</td>\n",
       "      <td>334</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "12397  TTHERM_00532700                334                15                 3"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['TTHERM_ID']=='TTHERM_00532700']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "b670c766-4a46-4403-9591-512df86365cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>Description</th>\n",
       "      <th>TGD2021_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>TTHERM_00016520</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>754</th>\n",
       "      <td>TTHERM_000196029</td>\n",
       "      <td>Possible plasma membrane-binding motif in junc...</td>\n",
       "      <td>radial spoke head-like protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>TTHERM_000372529</td>\n",
       "      <td>Essential component of the vacuolar proton pum...</td>\n",
       "      <td>nucleoside diphosphate kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>TTHERM_00041580</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1722</th>\n",
       "      <td>TTHERM_00044960</td>\n",
       "      <td>-</td>\n",
       "      <td>spindle pole protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>TTHERM_00047660</td>\n",
       "      <td>Thioredoxin and glutathione reductase family p...</td>\n",
       "      <td>thioredoxin glutathione reductase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2176</th>\n",
       "      <td>TTHERM_00052270</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4261</th>\n",
       "      <td>TTHERM_001164064</td>\n",
       "      <td>Trichohyalin-plectin-homology domain</td>\n",
       "      <td>tumor suppressor mitostatin protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4287</th>\n",
       "      <td>TTHERM_00118600</td>\n",
       "      <td>pyruvate kinase activity</td>\n",
       "      <td>PYK1 pyruvate kinase complex alpha subunit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4845</th>\n",
       "      <td>TTHERM_00137550</td>\n",
       "      <td>Calmodulin-binding</td>\n",
       "      <td>DNA polymerase delta C4-type zinc-finger protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4945</th>\n",
       "      <td>TTHERM_00139700</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4979</th>\n",
       "      <td>TTHERM_00141000</td>\n",
       "      <td>Protein tyrosine kinase</td>\n",
       "      <td>cyclin-dependent kinase-like Serine/Threonine ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5452</th>\n",
       "      <td>TTHERM_00155380</td>\n",
       "      <td>-</td>\n",
       "      <td>BBC57 EF hand protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5491</th>\n",
       "      <td>TTHERM_00157880</td>\n",
       "      <td>Leucine-rich repeat</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5598</th>\n",
       "      <td>TTHERM_00161270</td>\n",
       "      <td>CAMSAP CH domain</td>\n",
       "      <td>BBC31 sperm flagellar protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5936</th>\n",
       "      <td>TTHERM_00190730</td>\n",
       "      <td>Spfh domain band 7 family protein</td>\n",
       "      <td>SPFH domain/band 7 family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6215</th>\n",
       "      <td>TTHERM_00196150</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6225</th>\n",
       "      <td>TTHERM_00196280</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499</th>\n",
       "      <td>TTHERM_00219250</td>\n",
       "      <td>Zinc knuckle</td>\n",
       "      <td>RRM5 RNA recognition motif 1 in vertebrate ser...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7222</th>\n",
       "      <td>TTHERM_00264850</td>\n",
       "      <td>Domain of unknown function (DUF4586)</td>\n",
       "      <td>Potassium channel AKT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7441</th>\n",
       "      <td>TTHERM_00279780</td>\n",
       "      <td>-</td>\n",
       "      <td>dual specificity phosphatase domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7903</th>\n",
       "      <td>TTHERM_00299910</td>\n",
       "      <td>-</td>\n",
       "      <td>Zinc finger CCCH domain-containing protein 9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8168</th>\n",
       "      <td>TTHERM_00312960</td>\n",
       "      <td>U2 snRNA binding</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8242</th>\n",
       "      <td>TTHERM_00314970</td>\n",
       "      <td>-</td>\n",
       "      <td>kelch motif protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8312</th>\n",
       "      <td>TTHERM_00316930</td>\n",
       "      <td>smoothened signaling pathway</td>\n",
       "      <td>IQ and ubiquitin-like domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8399</th>\n",
       "      <td>TTHERM_00319990</td>\n",
       "      <td>Axonemal dynein light chain</td>\n",
       "      <td>p28B 33 kDa inner dynein arm light chain axone...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8627</th>\n",
       "      <td>TTHERM_00334290</td>\n",
       "      <td>Leucine-rich repeat</td>\n",
       "      <td>LC1 dynein light chain 1 axonemal protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8900</th>\n",
       "      <td>TTHERM_00348390</td>\n",
       "      <td>Domain of unknown function (DUF4496)</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9090</th>\n",
       "      <td>TTHERM_00355760</td>\n",
       "      <td>FHA domain</td>\n",
       "      <td>FHA domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9255</th>\n",
       "      <td>TTHERM_00372420</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9768</th>\n",
       "      <td>TTHERM_00406590</td>\n",
       "      <td>-</td>\n",
       "      <td>BBC118 EF-hand pair protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9968</th>\n",
       "      <td>TTHERM_00418580</td>\n",
       "      <td>Protein of unknown function (DUF667)</td>\n",
       "      <td>transcription factor iib protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10086</th>\n",
       "      <td>TTHERM_00421090</td>\n",
       "      <td>Leucine Rich repeat</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10218</th>\n",
       "      <td>TTHERM_00430020</td>\n",
       "      <td>Radial spokehead-like protein</td>\n",
       "      <td>radial spokehead protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10519</th>\n",
       "      <td>TTHERM_00444180</td>\n",
       "      <td>Radial spoke head</td>\n",
       "      <td>radial spokehead protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10631</th>\n",
       "      <td>TTHERM_00446510</td>\n",
       "      <td>It is involved in the biological process descr...</td>\n",
       "      <td>protein phosphatase 2C containing protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10784</th>\n",
       "      <td>TTHERM_00455540</td>\n",
       "      <td>Adenylate kinase</td>\n",
       "      <td>adenylate kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10843</th>\n",
       "      <td>TTHERM_00459310</td>\n",
       "      <td>Universal stress protein family</td>\n",
       "      <td>universal stress family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11138</th>\n",
       "      <td>TTHERM_00471260</td>\n",
       "      <td>DnaJ protein homolog</td>\n",
       "      <td>DnaJ (Hsp40) family B protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11369</th>\n",
       "      <td>TTHERM_00481300</td>\n",
       "      <td>Protein of unknown function (DUF2816)</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12429</th>\n",
       "      <td>TTHERM_00535160</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12496</th>\n",
       "      <td>TTHERM_00535929</td>\n",
       "      <td>-</td>\n",
       "      <td>DNA double-strand break repair rad50 ATPase pu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12906</th>\n",
       "      <td>TTHERM_00564090</td>\n",
       "      <td>Protein kinase C conserved region 2 (CalB)</td>\n",
       "      <td>Dpy-30 motif protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13843</th>\n",
       "      <td>TTHERM_00624660</td>\n",
       "      <td>RIB43A</td>\n",
       "      <td>protofilament ribbon protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14661</th>\n",
       "      <td>TTHERM_00670590</td>\n",
       "      <td>Ubiquitin-conjugating enzyme E2, catalytic dom...</td>\n",
       "      <td>UCV2 ubiquitin-conjugating enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14664</th>\n",
       "      <td>TTHERM_00670610</td>\n",
       "      <td>Regulatory subunit of type ii pka r-subunit fa...</td>\n",
       "      <td>regulatory subunit of type II PKA R-subunit fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14963</th>\n",
       "      <td>TTHERM_00688540</td>\n",
       "      <td>C2 domain of PTEN tumour-suppressor protein</td>\n",
       "      <td>PTEN tumor-suppressor protein C2 domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15128</th>\n",
       "      <td>TTHERM_00695730</td>\n",
       "      <td>Mortierella verticillata NRRL 6337</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15633</th>\n",
       "      <td>TTHERM_00725990</td>\n",
       "      <td>-</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16176</th>\n",
       "      <td>TTHERM_00780460</td>\n",
       "      <td>GMP-PDE, delta subunit</td>\n",
       "      <td>GMP-PDE delta subunit family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16197</th>\n",
       "      <td>TTHERM_00780680</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16722</th>\n",
       "      <td>TTHERM_00825340</td>\n",
       "      <td>-</td>\n",
       "      <td>transmembrane protein putative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16870</th>\n",
       "      <td>TTHERM_00836640</td>\n",
       "      <td>WD repeat domain 90</td>\n",
       "      <td>WD repeat protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17645</th>\n",
       "      <td>TTHERM_00938880</td>\n",
       "      <td>peptidyl-serine phosphorylation</td>\n",
       "      <td>casein kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17671</th>\n",
       "      <td>TTHERM_00939240</td>\n",
       "      <td>EF hand domain-containing protein</td>\n",
       "      <td>EF hand protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17925</th>\n",
       "      <td>TTHERM_00974270</td>\n",
       "      <td>Domain of unknown function (DUF4483)</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18153</th>\n",
       "      <td>TTHERM_01001540</td>\n",
       "      <td>Rab GTPase binding</td>\n",
       "      <td>Yip1 domain protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18600</th>\n",
       "      <td>TTHERM_01049330</td>\n",
       "      <td>Domain of unknown function (DUF4490)</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18642</th>\n",
       "      <td>TTHERM_01052990</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID                                        Description  \\\n",
       "685     TTHERM_00016520                                                  -   \n",
       "754    TTHERM_000196029  Possible plasma membrane-binding motif in junc...   \n",
       "1413   TTHERM_000372529  Essential component of the vacuolar proton pum...   \n",
       "1587    TTHERM_00041580                                                  -   \n",
       "1722    TTHERM_00044960                                                  -   \n",
       "1952    TTHERM_00047660  Thioredoxin and glutathione reductase family p...   \n",
       "2176    TTHERM_00052270                                                  -   \n",
       "4261   TTHERM_001164064               Trichohyalin-plectin-homology domain   \n",
       "4287    TTHERM_00118600                           pyruvate kinase activity   \n",
       "4845    TTHERM_00137550                                 Calmodulin-binding   \n",
       "4945    TTHERM_00139700                                                  -   \n",
       "4979    TTHERM_00141000                            Protein tyrosine kinase   \n",
       "5452    TTHERM_00155380                                                  -   \n",
       "5491    TTHERM_00157880                                Leucine-rich repeat   \n",
       "5598    TTHERM_00161270                                   CAMSAP CH domain   \n",
       "5936    TTHERM_00190730                  Spfh domain band 7 family protein   \n",
       "6215    TTHERM_00196150                                                  -   \n",
       "6225    TTHERM_00196280                                                  -   \n",
       "6499    TTHERM_00219250                                       Zinc knuckle   \n",
       "7222    TTHERM_00264850               Domain of unknown function (DUF4586)   \n",
       "7441    TTHERM_00279780                                                  -   \n",
       "7903    TTHERM_00299910                                                  -   \n",
       "8168    TTHERM_00312960                                   U2 snRNA binding   \n",
       "8242    TTHERM_00314970                                                  -   \n",
       "8312    TTHERM_00316930                       smoothened signaling pathway   \n",
       "8399    TTHERM_00319990                        Axonemal dynein light chain   \n",
       "8627    TTHERM_00334290                                Leucine-rich repeat   \n",
       "8900    TTHERM_00348390               Domain of unknown function (DUF4496)   \n",
       "9090    TTHERM_00355760                                         FHA domain   \n",
       "9255    TTHERM_00372420                                                  -   \n",
       "9768    TTHERM_00406590                                                  -   \n",
       "9968    TTHERM_00418580               Protein of unknown function (DUF667)   \n",
       "10086   TTHERM_00421090                                Leucine Rich repeat   \n",
       "10218   TTHERM_00430020                      Radial spokehead-like protein   \n",
       "10519   TTHERM_00444180                                  Radial spoke head   \n",
       "10631   TTHERM_00446510  It is involved in the biological process descr...   \n",
       "10784   TTHERM_00455540                                   Adenylate kinase   \n",
       "10843   TTHERM_00459310                    Universal stress protein family   \n",
       "11138   TTHERM_00471260                               DnaJ protein homolog   \n",
       "11369   TTHERM_00481300              Protein of unknown function (DUF2816)   \n",
       "12429   TTHERM_00535160                                                  -   \n",
       "12496   TTHERM_00535929                                                  -   \n",
       "12906   TTHERM_00564090         Protein kinase C conserved region 2 (CalB)   \n",
       "13843   TTHERM_00624660                                             RIB43A   \n",
       "14661   TTHERM_00670590  Ubiquitin-conjugating enzyme E2, catalytic dom...   \n",
       "14664   TTHERM_00670610  Regulatory subunit of type ii pka r-subunit fa...   \n",
       "14963   TTHERM_00688540        C2 domain of PTEN tumour-suppressor protein   \n",
       "15128   TTHERM_00695730                 Mortierella verticillata NRRL 6337   \n",
       "15633   TTHERM_00725990                                                  -   \n",
       "16176   TTHERM_00780460                             GMP-PDE, delta subunit   \n",
       "16197   TTHERM_00780680                                                  -   \n",
       "16722   TTHERM_00825340                                                  -   \n",
       "16870   TTHERM_00836640                                WD repeat domain 90   \n",
       "17645   TTHERM_00938880                    peptidyl-serine phosphorylation   \n",
       "17671   TTHERM_00939240                  EF hand domain-containing protein   \n",
       "17925   TTHERM_00974270               Domain of unknown function (DUF4483)   \n",
       "18153   TTHERM_01001540                                 Rab GTPase binding   \n",
       "18600   TTHERM_01049330               Domain of unknown function (DUF4490)   \n",
       "18642   TTHERM_01052990                                                  -   \n",
       "\n",
       "                                     TGD2021_description  \n",
       "685                                 hypothetical protein  \n",
       "754                       radial spoke head-like protein  \n",
       "1413                       nucleoside diphosphate kinase  \n",
       "1587                                hypothetical protein  \n",
       "1722                       spindle pole protein putative  \n",
       "1952                   thioredoxin glutathione reductase  \n",
       "2176                                hypothetical protein  \n",
       "4261                 tumor suppressor mitostatin protein  \n",
       "4287          PYK1 pyruvate kinase complex alpha subunit  \n",
       "4845    DNA polymerase delta C4-type zinc-finger protein  \n",
       "4945                                hypothetical protein  \n",
       "4979   cyclin-dependent kinase-like Serine/Threonine ...  \n",
       "5452                      BBC57 EF hand protein putative  \n",
       "5491                                hypothetical protein  \n",
       "5598                       BBC31 sperm flagellar protein  \n",
       "5936                   SPFH domain/band 7 family protein  \n",
       "6215                                hypothetical protein  \n",
       "6225                                hypothetical protein  \n",
       "6499   RRM5 RNA recognition motif 1 in vertebrate ser...  \n",
       "7222                              Potassium channel AKT1  \n",
       "7441         dual specificity phosphatase domain protein  \n",
       "7903        Zinc finger CCCH domain-containing protein 9  \n",
       "8168                                hypothetical protein  \n",
       "8242                                 kelch motif protein  \n",
       "8312                IQ and ubiquitin-like domain protein  \n",
       "8399   p28B 33 kDa inner dynein arm light chain axone...  \n",
       "8627           LC1 dynein light chain 1 axonemal protein  \n",
       "8900                                hypothetical protein  \n",
       "9090                                  FHA domain protein  \n",
       "9255                                hypothetical protein  \n",
       "9768                         BBC118 EF-hand pair protein  \n",
       "9968           transcription factor iib protein putative  \n",
       "10086                               hypothetical protein  \n",
       "10218                           radial spokehead protein  \n",
       "10519                           radial spokehead protein  \n",
       "10631          protein phosphatase 2C containing protein  \n",
       "10784                                   adenylate kinase  \n",
       "10843                    universal stress family protein  \n",
       "11138                      DnaJ (Hsp40) family B protein  \n",
       "11369                               hypothetical protein  \n",
       "12429                               hypothetical protein  \n",
       "12496  DNA double-strand break repair rad50 ATPase pu...  \n",
       "12906                               Dpy-30 motif protein  \n",
       "13843                       protofilament ribbon protein  \n",
       "14661                  UCV2 ubiquitin-conjugating enzyme  \n",
       "14664  regulatory subunit of type II PKA R-subunit fa...  \n",
       "14963    PTEN tumor-suppressor protein C2 domain protein  \n",
       "15128                               hypothetical protein  \n",
       "15633                     transmembrane protein putative  \n",
       "16176               GMP-PDE delta subunit family protein  \n",
       "16197                               hypothetical protein  \n",
       "16722                     transmembrane protein putative  \n",
       "16870                                  WD repeat protein  \n",
       "17645                                      casein kinase  \n",
       "17671                                    EF hand protein  \n",
       "17925                               hypothetical protein  \n",
       "18153                                Yip1 domain protein  \n",
       "18600                               hypothetical protein  \n",
       "18642                               hypothetical protein  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labeled_annot.loc[labeled_annot['leiden_label_full'] == 10][['TTHERM_ID', 'Description', 'TGD2021_description']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "8d82230e-f85b-42ad-b9c2-75a69c56ef89",
   "metadata": {},
   "outputs": [],
   "source": [
    "leiden_label_df_round_1_arranged_sorted.to_csv('./test_nn3_leiden_label_df_round_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "de9bf727",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTHERM_000000042</td>\n",
       "      <td>228</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TTHERM_000000045</td>\n",
       "      <td>190</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTHERM_00000010</td>\n",
       "      <td>475</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TTHERM_00000020</td>\n",
       "      <td>391</td>\n",
       "      <td>19</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTHERM_00000030</td>\n",
       "      <td>6</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20321</th>\n",
       "      <td>TTHERM_02293890</td>\n",
       "      <td>416</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20322</th>\n",
       "      <td>TTHERM_02385080</td>\n",
       "      <td>628</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20323</th>\n",
       "      <td>TTHERM_02555200</td>\n",
       "      <td>450</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20324</th>\n",
       "      <td>TTHERM_02607240</td>\n",
       "      <td>301</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20325</th>\n",
       "      <td>TTHERM_02653470</td>\n",
       "      <td>372</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "0      TTHERM_000000042                228                 6                17\n",
       "1      TTHERM_000000045                190                 0                10\n",
       "2       TTHERM_00000010                475                19                 4\n",
       "3       TTHERM_00000020                391                19                 4\n",
       "4       TTHERM_00000030                  6                19                 0\n",
       "...                 ...                ...               ...               ...\n",
       "20321   TTHERM_02293890                416                12                 8\n",
       "20322   TTHERM_02385080                628                16                15\n",
       "20323   TTHERM_02555200                450                18                15\n",
       "20324   TTHERM_02607240                301                17                 3\n",
       "20325   TTHERM_02653470                372                10                11\n",
       "\n",
       "[20326 rows x 4 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "e0d1a05d-c704-48d5-952d-7cf01f75bb99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "771"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(np.array(module_size_list) >= 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "0e2ef7d6-cf4f-452e-b183-4cdda16fcf3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>70</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "20049  TTHERM_01386050                 70                15                13"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "83198b81-f8ff-4230-8047-741d63a685af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# full_pymde_data = full_filtered_norm_df[list(full_filtered_norm_df.columns)[1:]].values\n",
    "\n",
    "# pymde.seed(42)\n",
    "# torch.seed(42)\n",
    "\n",
    "# full_embedding = pymde.preserve_neighbors(\n",
    "#     full_pymde_data, \n",
    "#     verbose=True,\n",
    "#     n_neighbors=5,\n",
    "#     constraint=pymde.Standardized()\n",
    "# ).embed()\n",
    "\n",
    "# pymde_full_df = pd.DataFrame(np.array(full_embedding), columns=('x', 'y'))\n",
    "# bokeh.plotting.output_file(filename='./plots/full_norm_pymde_leiden_round_1.2.html', mode='inline')\n",
    "\n",
    "# p = plot_embedding(full_filtered_norm_df, pymde_full_df, complete_annot, leiden_label_df_round_1, 'full', palette32, n_neighbors=5, radius=0.005)\n",
    "# bokeh.plotting.save(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f4677a3-55c9-49d8-9941-fde9c264c9fc",
   "metadata": {},
   "source": [
    "For ordering the leiden clusters in the heatmap, try computing the cityblock distance between centroids and then ordering by smallest distance, starting with largest cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "c41737b2-bfb7-43cc-a5dc-c95cfbb53512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   a  b\n",
       "0  1  0\n",
       "1  2  2\n",
       "2  3  0\n",
       "3  4  2"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_centroid = pd.DataFrame({'a': np.array([1, 2, 3, 4]), 'b': np.array([0, 2, 0, 2])})\n",
    "test_centroid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "1cedb93e-ab46-4c19-a492-353247f834a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([2.5, 1. ])"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_centroid.mean(axis=0).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2acf27e-4961-4ed6-9a79-f6c146f27bd1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "b5da096c-f282-4050-9d3e-862ee1ce6a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "l = get_all_module_centroids(full_filtered_norm_df, leiden_label_df_round_1, 'leiden', 'full')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdcf9180-1fcc-4ab1-895a-8fed06c73c09",
   "metadata": {},
   "source": [
    "Module 0 is always the largest for leiden and WGCNA; need to check about CLR/MCL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "9dd13886-0d8e-40fe-9bce-587fe2756aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdf = get_module_centroid_df(full_filtered_norm_df, leiden_label_df_round_1, 'leiden', 'full')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "27798f08-c9d5-42f5-a0c4-0b0188191475",
   "metadata": {},
   "outputs": [],
   "source": [
    "# arr_df = arrange_modules(full_filtered_norm_df, leiden_label_df_round_1, 'leiden', 'full') # FIXME\n",
    "arr_df = leiden_label_df_round_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "8b0d4e56-7832-4310-9605-efb90db185ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20326\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1638</th>\n",
       "      <td>TTHERM_000429640</td>\n",
       "      <td>617</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2821</th>\n",
       "      <td>TTHERM_00070750</td>\n",
       "      <td>617</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2879</th>\n",
       "      <td>TTHERM_000718079</td>\n",
       "      <td>617</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3192</th>\n",
       "      <td>TTHERM_00079410</td>\n",
       "      <td>617</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3273</th>\n",
       "      <td>TTHERM_000810475</td>\n",
       "      <td>617</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17501</th>\n",
       "      <td>TTHERM_00925730</td>\n",
       "      <td>718</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17607</th>\n",
       "      <td>TTHERM_00933260</td>\n",
       "      <td>718</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18447</th>\n",
       "      <td>TTHERM_01028810</td>\n",
       "      <td>718</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19182</th>\n",
       "      <td>TTHERM_01121640</td>\n",
       "      <td>718</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20300</th>\n",
       "      <td>TTHERM_01683270</td>\n",
       "      <td>718</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20326 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "1638   TTHERM_000429640                617                19                15\n",
       "2821    TTHERM_00070750                617                13                13\n",
       "2879   TTHERM_000718079                617                 3                 7\n",
       "3192    TTHERM_00079410                617                 0                 7\n",
       "3273   TTHERM_000810475                617                 0                18\n",
       "...                 ...                ...               ...               ...\n",
       "17501   TTHERM_00925730                718                 7                14\n",
       "17607   TTHERM_00933260                718                 7                 7\n",
       "18447   TTHERM_01028810                718                 3                14\n",
       "19182   TTHERM_01121640                718                 7                 0\n",
       "20300   TTHERM_01683270                718                 6                14\n",
       "\n",
       "[20326 rows x 4 columns]"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arrange_modules(full_filtered_df, leiden_label_df_round_1, 'leiden', 'full')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "5cacc481-ad08-41ca-898d-6d1706091849",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>TTHERM_000011798</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>TTHERM_000158529</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>TTHERM_000188553</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892</th>\n",
       "      <td>TTHERM_000241859</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1057</th>\n",
       "      <td>TTHERM_000294809</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18541</th>\n",
       "      <td>TTHERM_01044390</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18816</th>\n",
       "      <td>TTHERM_01080530</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18871</th>\n",
       "      <td>TTHERM_01084310</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18979</th>\n",
       "      <td>TTHERM_01097910</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19041</th>\n",
       "      <td>TTHERM_01104840</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "36     TTHERM_000011798                  0                 6                 0\n",
       "628    TTHERM_000158529                  0                 2                 0\n",
       "723    TTHERM_000188553                  0                 3                 4\n",
       "892    TTHERM_000241859                  0                 2                 0\n",
       "1057   TTHERM_000294809                  0                 3                 0\n",
       "...                 ...                ...               ...               ...\n",
       "18541   TTHERM_01044390                  0                19                20\n",
       "18816   TTHERM_01080530                  0                19                 0\n",
       "18871   TTHERM_01084310                  0                 2                 0\n",
       "18979   TTHERM_01097910                  0                20                 0\n",
       "19041   TTHERM_01104840                  0                 3                 0\n",
       "\n",
       "[74 rows x 4 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr_df.loc[arr_df['leiden_label_full'] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "478eef64-9c36-4e5e-b146-2844b08a0e3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>70</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "20049  TTHERM_01386050                 70                15                13"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr_df.loc[arr_df['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "191fb6c9-2d12-4a9d-bfbe-b6b9d9485c40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20049</th>\n",
       "      <td>TTHERM_01386050</td>\n",
       "      <td>70</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "20049  TTHERM_01386050                 70                15                13"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1.loc[leiden_label_df_round_1['TTHERM_ID'] == 'TTHERM_01386050']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTHERM_ID</th>\n",
       "      <th>leiden_label_full</th>\n",
       "      <th>leiden_label_veg</th>\n",
       "      <th>leiden_label_sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7178</th>\n",
       "      <td>TTHERM_00263160</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14187</th>\n",
       "      <td>TTHERM_00644760</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9067</th>\n",
       "      <td>TTHERM_00355480</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13228</th>\n",
       "      <td>TTHERM_00581570</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4373</th>\n",
       "      <td>TTHERM_00122450</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TTHERM_ID  leiden_label_full  leiden_label_veg  leiden_label_sex\n",
       "7178   TTHERM_00263160                  0                 3                 0\n",
       "14187  TTHERM_00644760                  0                 2                14\n",
       "9067   TTHERM_00355480                  0                 2                 0\n",
       "13228  TTHERM_00581570                  0                 3                 0\n",
       "4373   TTHERM_00122450                  0                 2                 0"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leiden_label_df_round_1_sorted = leiden_label_df_round_1.sort_values(by='leiden_label_full')\n",
    "leiden_label_df_round_1_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "def leiden_clustering_df_to_csv(df, export_file_path):\n",
    "    current_module = list(df['leiden_label_full'].values)[0]\n",
    "\n",
    "    new_cluster = True\n",
    "\n",
    "    with open(export_file_path, 'w') as f:\n",
    "        for idx, row in df.iterrows():\n",
    "            if current_module != row['leiden_label_full']:\n",
    "                current_module = row['leiden_label_full']\n",
    "                f.write('\\n')\n",
    "                new_cluster =True\n",
    "\n",
    "            elif not new_cluster:\n",
    "                f.write('\\t')\n",
    "            \n",
    "            f.write(row['TTHERM_ID'])\n",
    "            new_cluster = False\n",
    "\n",
    "        f.write('\\n')\n",
    "    \n",
    "    print(export_file_path, 'successfully exported.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "# leiden_clustering_df_to_csv(leiden_label_df_round_1_sorted, './leiden_clustering.labels')\n",
    "\n",
    "# for la_param, df in leiden_label_df_cpms.items():\n",
    "#     leiden_clustering_df_to_csv(df, f'./leiden_clustering_cpm_{str(la_param)}.labels')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('./leiden_clustering.labels', 'r') as f:\n",
    "#     num_leiden_clusters = f.readlines()\n",
    "#     print(len(num_leiden_clusters))\n",
    "\n",
    "# print(max(list(leiden_label_df_round_1_sorted['leiden_label_full'].values)) + 1) # module numbering starts at 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "f0f4ee4d-b47e-4e5d-883b-f30c9a90dd91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rcl_mcl_df.to_csv('./test_nn3_leiden_label_df_round_1_rearranged.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cf55445",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "55189b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rcl_mcl_df = pd.read_csv('./mcl_rcl/mcl_rcl_label_df_round_1.csv')\n",
    "# rcl_mcl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "25666f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for gene in gene_list_test:\n",
    "#     print((rcl_mcl_df.loc[rcl_mcl_df['TTHERM_ID'] == gene])['leiden_label_full'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clust_alg = 'leiden'\n",
    "# phases = 'full'\n",
    "\n",
    "# arranged_rcl_mcl_df = arrange_modules(full_filtered_norm_df, rcl_mcl_df, clust_alg, phases) \n",
    "# sorted_rcl_mcl_df = arranged_rcl_mcl_df.sort_values(by=[f'{clust_alg}_label_{phases}', 'TTHERM_ID'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "# arranged_rcl_mcl_df.loc[arranged_rcl_mcl_df['TTHERM_ID'] == 'TTHERM_00000040']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sorted_rcl_mcl_df.loc[sorted_rcl_mcl_df['TTHERM_ID'] == 'TTHERM_00000040']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "91b739eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sorted_rcl_mcl_df.to_csv('./mcl_rcl_label_df_round_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "73c2c36a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# arrange_modules(full_filtered_norm_df, rcl_mcl_df, 'leiden', 'full')['leiden_label_full'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d79a58-e289-445f-a3f5-abeb9b6b87cd",
   "metadata": {},
   "source": [
    "#### Go to ../enrichment/enrichment.ipynb to generate the functional annotation enrichment data from this clustering. Proceed once done."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
